From 215ba93b9260286b916c1c73f73750f9279645e8 Mon Sep 17 00:00:00 2001
From: Amine Khaldi <amine.khaldi@reactos.org>
Date: Mon, 23 Feb 2015 20:14:03 +0100
Subject: dinput: Skip Wine specific __wine_check_for_events calls in ReactOS.

---
 dlls/dinput/keyboard.c | 2 ++
 dlls/dinput/mouse.c    | 2 ++
 2 files changed, 4 insertions(+)

diff --git a/dlls/dinput/keyboard.c b/dlls/dinput/keyboard.c
index 9e78257..3019552 100644
--- a/dlls/dinput/keyboard.c
+++ b/dlls/dinput/keyboard.c
@@ -335,7 +335,9 @@ static HRESULT WINAPI SysKeyboardWImpl_GetDeviceState(LPDIRECTINPUTDEVICE8W ifac
     if (len != This->base.data_format.user_df->dwDataSize )
         return DIERR_INVALIDPARAM;
 
+#ifndef __REACTOS__
     __wine_check_for_events( QS_ALLINPUT );
+#endif
 
     EnterCriticalSection(&This->base.crit);
 
diff --git a/dlls/dinput/mouse.c b/dlls/dinput/mouse.c
index 44e1d57..8249513 100644
--- a/dlls/dinput/mouse.c
+++ b/dlls/dinput/mouse.c
@@ -552,7 +552,9 @@ static HRESULT WINAPI SysMouseWImpl_GetDeviceState(LPDIRECTINPUTDEVICE8W iface,
 
     if(This->base.acquired == 0) return DIERR_NOTACQUIRED;
 
+#ifndef __REACTOS__
     __wine_check_for_events( QS_ALLINPUT );
+#endif
 
     TRACE("(this=%p,0x%08x,%p):\n", This, len, ptr);
     _dump_mouse_state(&This->m_state);
-- 
2.3.0

