From d98a585a7d809889d84a45546d7d259a9ab35356 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 27 Oct 2016 01:33:34 +0200
Subject: cabinet: Set index of folder in FDICopy callback.

---
 dlls/cabinet/fdi.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/cabinet/fdi.c b/dlls/cabinet/fdi.c
index fdff0bf..758133e 100644
--- a/dlls/cabinet/fdi.c
+++ b/dlls/cabinet/fdi.c
@@ -2640,6 +2640,7 @@ BOOL __cdecl FDICopy(
       fdin.date = file->date;
       fdin.time = file->time;
       fdin.attribs = file->attribs;
+      fdin.iFolder = file->index;
       if ((filehf = ((*pfnfdin)(fdintCOPY_FILE, &fdin))) == -1) {
         set_error( fdi, FDIERROR_USER_ABORT, 0 );
         filehf = 0;
@@ -2765,6 +2766,7 @@ BOOL __cdecl FDICopy(
       fdin.date = file->date;
       fdin.time = file->time;
       fdin.attribs = file->attribs; /* FIXME: filter _A_EXEC? */
+      fdin.iFolder = file->index;
       ((*pfnfdin)(fdintCLOSE_FILE_INFO, &fdin));
       filehf = 0;
 
-- 
2.9.0

