From 2ba7980e47f0c593a17f13e73a19dec6fa4f8955 Mon Sep 17 00:00:00 2001
From: Jarkko Korpi <jarkko_korpi@hotmail.com>
Date: Mon, 24 Aug 2015 21:44:20 +0300
Subject: winspool.drv Add case 8 for SetPrinterW.

---
 dlls/winspool.drv/info.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/dlls/winspool.drv/info.c b/dlls/winspool.drv/info.c
index ca2c34d..e39509d 100644
--- a/dlls/winspool.drv/info.c
+++ b/dlls/winspool.drv/info.c
@@ -3452,6 +3452,14 @@ static void set_printer_2( HKEY key, const PRINTER_INFO_2W *pi )
     set_reg_DWORD( key, UntilTimeW, pi->UntilTime );
 }
 
+static BOOL set_printer_8( HKEY key, const PRINTER_INFO_8W *pi )
+{
+    if (!pi->pDevMode) return FALSE;
+
+    set_reg_devmode( key, Default_DevModeW, pi->pDevMode );
+    return TRUE;
+}
+
 static BOOL set_printer_9( HKEY key, const PRINTER_INFO_9W *pi )
 {
     if (!pi->pDevMode) return FALSE;
@@ -3484,7 +3492,12 @@ BOOL WINAPI SetPrinterW( HANDLE printer, DWORD level, LPBYTE data, DWORD command
         ret = TRUE;
         break;
     }
-
+    case 8:
+    {
+        PRINTER_INFO_8W *pi = (PRINTER_INFO_8W *)data;
+        ret = set_printer_8( key, pi );
+        break;
+    }
     case 9:
     {
         PRINTER_INFO_9W *pi = (PRINTER_INFO_9W *)data;
-- 
2.5.0

