From 938e979d889737ed2dcd4197aff9734891f06449 Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Thu, 13 Feb 2014 16:07:26 -0700
Subject: shell32: Set the default security attributes for user shell folders.

---
 dlls/shell32/shellpath.c |   94 +++++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 93 insertions(+), 1 deletion(-)

diff --git a/dlls/shell32/shellpath.c b/dlls/shell32/shellpath.c
index 875be38..3476a88 100644
--- a/dlls/shell32/shellpath.c
+++ b/dlls/shell32/shellpath.c
@@ -2169,6 +2169,70 @@ cleanup:
     return hr;
 }
 
+PSECURITY_DESCRIPTOR _SHGetUserSecurityDescriptor( void )
+{
+    PSECURITY_DESCRIPTOR sd = HeapAlloc(GetProcessHeap(), 0, SECURITY_DESCRIPTOR_MIN_LENGTH);
+    PACL dacl = HeapAlloc(GetProcessHeap(), 0, 100);
+    PSID admin_sid = NULL, user_sid;
+    TOKEN_USER *user = NULL;
+    BOOL ret = FALSE;
+    DWORD sid_size;
+    HANDLE token;
+
+    if(!sd || !dacl) goto cleanup;
+
+    /* find the user SID */
+    ret = TRUE;
+    if (!OpenThreadToken(GetCurrentThread(), TOKEN_READ, TRUE, &token))
+    {
+        if (GetLastError() != ERROR_NO_TOKEN) ret = FALSE;
+        else if (!OpenProcessToken(GetCurrentProcess(), TOKEN_READ, &token)) ret = FALSE;
+    }
+    if (!ret) goto cleanup;
+    sid_size = 0;
+    ret = GetTokenInformation(token, TokenUser, NULL, 0, &sid_size);
+    if (GetLastError() != ERROR_INSUFFICIENT_BUFFER) goto cleanup;
+    user = HeapAlloc(GetProcessHeap(), 0, sid_size);
+    if (!user) goto cleanup;
+    ret = GetTokenInformation(token, TokenUser, user, sid_size, &sid_size);
+    if (!ret) goto cleanup;
+    CloseHandle(token);
+    user_sid = user->User.Sid;
+
+    /* find the administrator group SID */
+    sid_size = 0;
+    ret = CreateWellKnownSid(WinBuiltinAdministratorsSid, NULL, NULL, &sid_size);
+    if(GetLastError() != ERROR_INSUFFICIENT_BUFFER) goto cleanup;
+    admin_sid = HeapAlloc(GetProcessHeap(), 0, sid_size);
+    if(!admin_sid) goto cleanup;
+    ret = CreateWellKnownSid(WinBuiltinAdministratorsSid, NULL, admin_sid, &sid_size);
+    if(!ret) goto cleanup;
+
+    /* build the DACL */
+    ret = InitializeSecurityDescriptor(sd, SECURITY_DESCRIPTOR_REVISION);
+    if(!ret) goto cleanup;
+    ret = InitializeAcl(dacl, 100, ACL_REVISION);
+    if(!ret) goto cleanup;
+    ret = AddAccessAllowedAceEx(dacl, ACL_REVISION, OBJECT_INHERIT_ACE|CONTAINER_INHERIT_ACE,
+                                GENERIC_ALL, user_sid);
+    if(!ret) goto cleanup;
+    ret = AddAccessAllowedAceEx(dacl, ACL_REVISION, OBJECT_INHERIT_ACE|CONTAINER_INHERIT_ACE,
+                                GENERIC_ALL, admin_sid);
+    if(!ret) goto cleanup;
+    ret = SetSecurityDescriptorDacl(sd, TRUE, dacl, FALSE);
+
+cleanup:
+    HeapFree(GetProcessHeap(), 0, user);
+    HeapFree(GetProcessHeap(), 0, admin_sid);
+    if(!ret)
+    {
+        HeapFree(GetProcessHeap(), 0, dacl);
+        HeapFree(GetProcessHeap(), 0, sd);
+        sd = NULL;
+    }
+    return sd;
+}
+
 /*************************************************************************
  * SHGetFolderPathAndSubDirW		[SHELL32.@]
  */
@@ -2180,6 +2244,8 @@ HRESULT WINAPI SHGetFolderPathAndSubDirW(
 	LPCWSTR pszSubPath,/* [I] sub directory of the specified folder */
 	LPWSTR pszPath)    /* [O] converted path */
 {
+    SECURITY_ATTRIBUTES sa, *sec = NULL;
+    PSECURITY_DESCRIPTOR sd = NULL;
     HRESULT    hr;
     WCHAR      szBuildPath[MAX_PATH], szTemp[MAX_PATH];
     DWORD      folder = nFolder & CSIDL_FOLDER_MASK;
@@ -2292,8 +2358,25 @@ HRESULT WINAPI SHGetFolderPathAndSubDirW(
         goto end;
     }
 
+    /* build the appropriate security attributes for the directory */
+    switch (type)
+    {
+        case CSIDL_Type_User:
+            sd = _SHGetUserSecurityDescriptor();
+            break;
+        default:
+            break;
+    }
+    if (sd)
+    {
+        sa.nLength = sizeof(sa);
+        sa.lpSecurityDescriptor = sd;
+        sa.bInheritHandle = TRUE;
+        sec = &sa;
+    }
+
     /* create directory/directories */
-    ret = SHCreateDirectoryExW(hwndOwner, szBuildPath, NULL);
+    ret = SHCreateDirectoryExW(hwndOwner, szBuildPath, sec);
     if (ret && ret != ERROR_ALREADY_EXISTS)
     {
         ERR("Failed to create directory %s.\n", debugstr_w(szBuildPath));
@@ -2303,6 +2386,15 @@ HRESULT WINAPI SHGetFolderPathAndSubDirW(
 
     TRACE("Created missing system directory %s\n", debugstr_w(szBuildPath));
 end:
+    if (sd)
+    {
+        BOOL present, defaulted;
+        PACL dacl = NULL;
+
+        GetSecurityDescriptorDacl(sd, &present, &dacl, &defaulted);
+        HeapFree(GetProcessHeap(), 0, dacl);
+        HeapFree(GetProcessHeap(), 0, sd);
+    }
     TRACE("returning 0x%08x (final path is %s)\n", hr, debugstr_w(szBuildPath));
     return hr;
 }
-- 
1.7.9.5

