From d4d2a048787580327efdeaad7baac703a1993b63 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 10 May 2017 02:13:01 +0200
Subject: mfplat: Fix invalid memory access.

---
 dlls/mfplat/main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/mfplat/main.c b/dlls/mfplat/main.c
index c1c700368f..56bc2840f0 100644
--- a/dlls/mfplat/main.c
+++ b/dlls/mfplat/main.c
@@ -262,7 +262,7 @@ static BOOL match_type(const WCHAR *clsid_str, const WCHAR *type_str, MFT_REGIST
     if (RegQueryValueExW(hfilter, type_str, NULL, &reg_type, (LPBYTE)info, &size) != ERROR_SUCCESS)
         goto out;
 
-    for (i = 0; i < size / sizeof(MFT_REGISTER_TYPE_INFO) + 1; i++)
+    for (i = 0; i < size / sizeof(MFT_REGISTER_TYPE_INFO); i++)
     {
         if (IsEqualGUID(&info[i].guidMajorType, &type->guidMajorType) &&
             IsEqualGUID(&info[i].guidSubtype,   &type->guidSubtype))
-- 
2.12.2

