From 322b917bad2ab43eb5181a429f1360ce9aca384b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sat, 26 Mar 2016 06:03:56 +0100
Subject: include/rpcndr.h: Fix definition of EXTERN_GUID.

---
 include/rpcndr.h | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/include/rpcndr.h b/include/rpcndr.h
index f795566..9029950 100644
--- a/include/rpcndr.h
+++ b/include/rpcndr.h
@@ -45,9 +45,7 @@ extern "C" {
 	{ l, w1, w2, { b1, b2,  b3,  b4,  b5,  b6,  b7,  b8 } }
 #else
 #define EXTERN_GUID(name, l, w1, w2, b1, b2, b3, b4, b5, b6, b7, b8) \
-        const GUID DECLSPEC_SELECTANY name DECLSPEC_HIDDEN; \
-        const GUID DECLSPEC_SELECTANY name = \
-	{ l, w1, w2, { b1, b2,  b3,  b4,  b5,  b6,  b7,  b8 } }
+        EXTERN_C const GUID name
 #endif
 #endif
 
-- 
2.7.1

