From 317b8941e2c36fd568799915c8023332b5054331 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <z.figura12@gmail.com>
Date: Sat, 6 Apr 2019 21:31:55 -0500
Subject: [PATCH] ntdll: Add stub for
 NtQuerySystemInformation(SystemExtendedProcessInformation).

---
 dlls/ntdll/nt.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/ntdll/nt.c b/dlls/ntdll/nt.c
index f42934969..c019a9c29 100644
--- a/dlls/ntdll/nt.c
+++ b/dlls/ntdll/nt.c
@@ -2768,6 +2768,11 @@ NTSTATUS WINAPI NtQuerySystemInformation(
             }
         }
         break;
+    case SystemExtendedProcessInformation:
+        FIXME("SystemExtendedProcessInformation, len %u, buffer %p, stub!\n", Length, SystemInformation);
+        memset(SystemInformation, 0, Length);
+        ret = STATUS_SUCCESS;
+        break;
     default:
 	FIXME("(0x%08x,%p,0x%08x,%p) stub\n",
 	      SystemInformationClass,SystemInformation,Length,ResultLength);
-- 
2.20.1

