From 7fa97568c8aadc5b2d27dace09946efe403f0367 Mon Sep 17 00:00:00 2001
From: Jarkko Korpi <jarkko_korpi@hotmail.com>
Date: Mon, 14 Dec 2015 10:49:25 +0200
Subject: user32: Remove dead code from SCROLL_TrackScrollBar.

---
 dlls/user32/scroll.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/dlls/user32/scroll.c b/dlls/user32/scroll.c
index be205ef..b28ee44 100644
--- a/dlls/user32/scroll.c
+++ b/dlls/user32/scroll.c
@@ -1096,7 +1096,6 @@ static void SCROLL_HandleScrollEvent( HWND hwnd, INT nBar, UINT msg, POINT pt)
 void SCROLL_TrackScrollBar( HWND hwnd, INT scrollbar, POINT pt )
 {
     MSG msg;
-    INT xoffset = 0, yoffset = 0;
 
     if (scrollbar != SB_CTL)
     {
@@ -1117,8 +1116,8 @@ void SCROLL_TrackScrollBar( HWND hwnd, INT scrollbar, POINT pt )
             msg.message == WM_MOUSEMOVE ||
             (msg.message == WM_SYSTIMER && msg.wParam == SCROLL_TIMER))
         {
-            pt.x = (short)LOWORD(msg.lParam) + xoffset;
-            pt.y = (short)HIWORD(msg.lParam) + yoffset;
+            pt.x = (short)LOWORD(msg.lParam);
+            pt.y = (short)HIWORD(msg.lParam);
             SCROLL_HandleScrollEvent( hwnd, scrollbar, msg.message, pt );
         }
         else
-- 
2.6.4

