From 4b299b286d73648250fedd699f2ef98e73a5e1a8 Mon Sep 17 00:00:00 2001
From: Alexandre Julliard <julliard@winehq.org>
Date: Mon, 2 Sep 2019 20:44:56 +0200
Subject: [PATCH] configure: Don't link to libunwind if not necessary.

Signed-off-by: Alexandre Julliard <julliard@winehq.org>
---
 configure    | 2 +-
 configure.ac | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/configure b/configure
index 0377509c60..e374edfdc5 100755
--- a/configure
+++ b/configure
@@ -14904,9 +14904,9 @@ rm -f core conftest.err conftest.$ac_objext \
 fi
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $wine_cv_have_libunwind_unw_getcontext" >&5
 $as_echo "$wine_cv_have_libunwind_unw_getcontext" >&6; }
-            test "$wine_cv_have_libunwind_unw_getcontext" = yes || UNWIND_LIBS=""
             LIBS=$save_libs
         fi
+        test "$wine_cv_have_libunwind_unw_getcontext" = yes || UNWIND_LIBS=""
         if test "x$wine_cv_have_unw_getcontext$wine_cv_have_libunwind_unw_getcontext" != xnono
         then
 
diff --git a/configure.ac b/configure.ac
index d0af7ef4a1..ff81a172f9 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1697,9 +1697,9 @@ then
                     [[unw_context_t context; unw_getcontext( &context );]])],
                     [wine_cv_have_libunwind_unw_getcontext="yes"],
                     [wine_cv_have_libunwind_unw_getcontext="no"])])
-            test "$wine_cv_have_libunwind_unw_getcontext" = yes || UNWIND_LIBS=""
             LIBS=$save_libs
         fi
+        test "$wine_cv_have_libunwind_unw_getcontext" = yes || UNWIND_LIBS=""
         if test "x$wine_cv_have_unw_getcontext$wine_cv_have_libunwind_unw_getcontext" != xnono
         then
             AC_DEFINE(HAVE_LIBUNWIND,1,[Define to 1 if you have the `unwind' library (-lunwind).])
-- 
2.23.0.rc1

