From 24ba61d54e9af4927f0e3be4c1789f3007d20b8b Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 18 Apr 2015 15:55:20 +0200
Subject: wine.inf: Add a ProfileList\<UserSID> registry subkey.

---
 loader/wine.inf.in | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index d901251..68a1b69 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -492,6 +492,8 @@ HKLM,%CurrentVersionNT%\Perflib,,16
 HKLM,%CurrentVersionNT%\Ports,,16
 HKLM,%CurrentVersionNT%\Print,,16
 HKLM,%CurrentVersionNT%\ProfileList,,16
+; Update/Replace if local_user_sid in server/token.c changes
+HKLM,%CurrentVersionNT%\ProfileList\S-1-5-21-0-0-0-1000,,16
 
 [CurrentVersionWow64]
 HKLM,%CurrentVersion%,"ProgramFilesDir (x86)",,"%16426%"
-- 
2.3.5

