From 211c1988706a03296da84d5b2607de3ec9e5ca68 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 29 Apr 2016 17:11:26 +0200
Subject: ntdll: Initialize Reserved_0 bit in NtQueryInformationThread.

---
 dlls/ntdll/thread.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/ntdll/thread.c b/dlls/ntdll/thread.c
index 40747d7..f6a5fbe 100644
--- a/dlls/ntdll/thread.c
+++ b/dlls/ntdll/thread.c
@@ -1045,6 +1045,7 @@ NTSTATUS WINAPI NtQueryInformationThread( HANDLE handle, THREADINFOCLASS class,
                     tdi->Entry.HighWord.Bits.Granularity = 1;
                     tdi->Entry.HighWord.Bits.Default_Big = 1;
                     tdi->Entry.HighWord.Bits.Type        = 0x12;
+                    tdi->Entry.HighWord.Bits.Reserved_0  = 0;
                     /* it has to be one of the system GDT selectors */
                     if (sel != (wine_get_ds() & ~3) && sel != (wine_get_ss() & ~3))
                     {
-- 
2.8.0

