From 87680099180ae97cdd4ab0060929e77a93ca453c Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Thu, 21 Aug 2014 22:31:53 -0600
Subject: ntdll: Make nt_to_unix_file_name_attr return always safe to free.

---
 dlls/ntdll/directory.c |    1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/ntdll/directory.c b/dlls/ntdll/directory.c
index a2796b2..aac7522 100644
--- a/dlls/ntdll/directory.c
+++ b/dlls/ntdll/directory.c
@@ -2941,6 +2941,7 @@ NTSTATUS nt_to_unix_file_name_attr( const OBJECT_ATTRIBUTES *attr, ANSI_STRING *
     NTSTATUS status;
     BOOLEAN check_case = !(attr->Attributes & OBJ_CASE_INSENSITIVE);
 
+    unix_name_ret->Buffer = NULL;
     if (!attr->RootDirectory)  /* without root dir fall back to normal lookup */
         return wine_nt_to_unix_file_name( attr->ObjectName, unix_name_ret, disposition, check_case );
 
-- 
1.7.9.5

