From 3fdd706cddf8cf517f38488f445ca4ab30bba088 Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Thu, 21 Aug 2014 22:35:55 -0600
Subject: ntdll: Make wine_nt_to_unix_file_name return always safe to free.

---
 dlls/ntdll/directory.c |    1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/ntdll/directory.c b/dlls/ntdll/directory.c
index aac7522..01f35ab 100644
--- a/dlls/ntdll/directory.c
+++ b/dlls/ntdll/directory.c
@@ -3026,6 +3026,7 @@ NTSTATUS CDECL wine_nt_to_unix_file_name( const UNICODE_STRING *nameW, ANSI_STRI
 
     name     = nameW->Buffer;
     name_len = nameW->Length / sizeof(WCHAR);
+    unix_name_ret->Buffer = NULL;
 
     if (!name_len || !IS_SEPARATOR(name[0])) return STATUS_OBJECT_PATH_SYNTAX_BAD;
 
-- 
1.7.9.5

