From f9af5e33f532b5bcc4c9e4195efe54bbb29ed826 Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Thu, 21 Aug 2014 22:45:34 -0600
Subject: ntdll: Fix unlikely leak on STATUS_NO_SUCH_FILE in init_redirects.

---
 dlls/ntdll/directory.c |    1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/ntdll/directory.c b/dlls/ntdll/directory.c
index 9e5dbae..4fb89e3 100644
--- a/dlls/ntdll/directory.c
+++ b/dlls/ntdll/directory.c
@@ -2466,6 +2466,7 @@ static void init_redirects(void)
     if (status)
     {
         ERR( "cannot open %s (%x)\n", debugstr_w(user_shared_data->NtSystemRoot), status );
+        RtlFreeAnsiString( &unix_name );
         return;
     }
     if (!stat( unix_name.Buffer, &st ))
-- 
1.7.9.5

