From c8c9fa7bab2da87b6a14380265cf88bc53eea6de Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Thu, 21 Aug 2014 22:57:06 -0600
Subject: ntdll: Fix leak on STATUS_NO_SUCH_FILE in load_builtin_dll.

---
 dlls/ntdll/loader.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index 79aa341..c9c2848 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -1712,6 +1712,7 @@ static NTSTATUS load_builtin_dll( LPCWSTR load_path, LPCWSTR path, HANDLE file,
         if (wine_nt_to_unix_file_name( &nt_name, &unix_name, FILE_OPEN, FALSE ))
         {
             RtlFreeUnicodeString( &nt_name );
+            RtlFreeAnsiString( &unix_name );
             return STATUS_DLL_NOT_FOUND;
         }
         prev_info = builtin_load_info;
@@ -1720,7 +1721,7 @@ static NTSTATUS load_builtin_dll( LPCWSTR load_path, LPCWSTR path, HANDLE file,
         handle = wine_dlopen( unix_name.Buffer, RTLD_NOW, error, sizeof(error) );
         builtin_load_info = prev_info;
         RtlFreeUnicodeString( &nt_name );
-        RtlFreeHeap( GetProcessHeap(), 0, unix_name.Buffer );
+        RtlFreeAnsiString( &unix_name );
         if (!handle)
         {
             WARN( "failed to load .so lib for builtin %s: %s\n", debugstr_w(path), error );
-- 
1.7.9.5

