From a2c8ef31f29e7e6473c6c0c35a7b9681bc99f1da Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Thu, 21 Aug 2014 22:50:19 -0600
Subject: kernel32: Fix leak on STATUS_NO_SUCH_FILE in RemoveDirectoryW.

---
 dlls/kernel32/path.c |    2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/kernel32/path.c b/dlls/kernel32/path.c
index 09fb04b..4a41ab8 100644
--- a/dlls/kernel32/path.c
+++ b/dlls/kernel32/path.c
@@ -1602,6 +1602,7 @@ BOOL WINAPI RemoveDirectoryW( LPCWSTR path )
         SetLastError( ERROR_PATH_NOT_FOUND );
         return FALSE;
     }
+    unix_name.Buffer = NULL;
     attr.Length = sizeof(attr);
     attr.RootDirectory = 0;
     attr.Attributes = OBJ_CASE_INSENSITIVE;
@@ -1619,6 +1620,7 @@ BOOL WINAPI RemoveDirectoryW( LPCWSTR path )
     if (status != STATUS_SUCCESS)
     {
         SetLastError( RtlNtStatusToDosError(status) );
+        RtlFreeAnsiString( &unix_name );
         return FALSE;
     }
 
-- 
1.7.9.5

