From d52a094e2e5831a52fac98016fc71879edc8193e Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Thu, 21 Aug 2014 22:50:19 -0600
Subject: kernel32: Fix leak on STATUS_NO_SUCH_FILE in RemoveDirectoryW.

---
 dlls/kernel32/path.c |    1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/kernel32/path.c b/dlls/kernel32/path.c
index 09fb04b..eeba48a 100644
--- a/dlls/kernel32/path.c
+++ b/dlls/kernel32/path.c
@@ -1619,6 +1619,7 @@ BOOL WINAPI RemoveDirectoryW( LPCWSTR path )
     if (status != STATUS_SUCCESS)
     {
         SetLastError( RtlNtStatusToDosError(status) );
+        RtlFreeAnsiString( &unix_name );
         return FALSE;
     }
 
-- 
1.7.9.5

