From 921d9d1c4f4a140ae8d0057079b62a0b8adb6205 Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Thu, 21 Aug 2014 22:54:09 -0600
Subject: kernel32: Fix leak on STATUS_NO_SUCH_FILE in QueryDosDeviceW.

---
 dlls/kernel32/volume.c |    6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/kernel32/volume.c b/dlls/kernel32/volume.c
index d396764..d580f1d 100644
--- a/dlls/kernel32/volume.c
+++ b/dlls/kernel32/volume.c
@@ -1354,10 +1354,8 @@ DWORD WINAPI QueryDosDeviceW( LPCWSTR devname, LPWSTR target, DWORD bufsize )
             status = wine_nt_to_unix_file_name( &nt_name, &unix_name, FILE_OPEN, TRUE );
             if (status) SetLastError( RtlNtStatusToDosError(status) );
             else
-            {
                 ret = MultiByteToWideChar( CP_UNIXCP, 0, unix_name.Buffer, -1, target, bufsize );
-                RtlFreeAnsiString( &unix_name );
-            }
+            RtlFreeAnsiString( &unix_name );
         }
     done:
         if (ret)
@@ -1409,6 +1407,7 @@ DWORD WINAPI QueryDosDeviceW( LPCWSTR devname, LPWSTR target, DWORD bufsize )
                 p[4] = 0;
                 p += 5;
             }
+            RtlFreeAnsiString( &unix_name );
         }
         strcpyW( nt_buffer + 4, lptW );
         for (i = 1; i <= 9; i++)
@@ -1427,6 +1426,7 @@ DWORD WINAPI QueryDosDeviceW( LPCWSTR devname, LPWSTR target, DWORD bufsize )
                 p[4] = 0;
                 p += 5;
             }
+            RtlFreeAnsiString( &unix_name );
         }
 
         RtlInitUnicodeString( &nt_name, dosdevW );
-- 
1.7.9.5

