From 4039f0b985055de28b5dd300b1634a5079b4ea32 Mon Sep 17 00:00:00 2001
From: Axel Davy <axel.davy@ens.fr>
Date: Wed, 7 Jan 2015 13:48:04 +0100
Subject: [PATCH 20/21] d3dadapter: don't initialize dri2 fallback if we use
 dri3

Signed-off-by: Axel Davy <axel.davy@ens.fr>
---
 dlls/winex11.drv/d3dadapter.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/dlls/winex11.drv/d3dadapter.c b/dlls/winex11.drv/d3dadapter.c
index c2355a5..994b453 100644
--- a/dlls/winex11.drv/d3dadapter.c
+++ b/dlls/winex11.drv/d3dadapter.c
@@ -219,7 +219,8 @@ DRI3Present_Release( struct DRI3Present *This )
         ChangeDisplaySettingsExW(This->devname, &(This->initial_mode), 0, CDS_FULLSCREEN, NULL);
         PRESENTDestroy(gdi_display, This->present_priv);
 #ifdef D3DADAPTER9_DRI2
-        DRI2FallbackDestroy(This->dri2_priv);
+        if (is_dri2_fallback)
+            DRI2FallbackDestroy(This->dri2_priv);
 #endif
         HeapFree(GetProcessHeap(), 0, This);
     }
@@ -647,7 +648,8 @@ DRI3Present_new( Display *dpy,
 
     PRESENTInit(dpy, &(This->present_priv));
 #ifdef D3DADAPTER9_DRI2
-    DRI2FallbackInit(dpy, &(This->dri2_priv));
+    if (is_dri2_fallback)
+        DRI2FallbackInit(dpy, &(This->dri2_priv));
 #endif
     *out = This;
 
-- 
2.2.2

