From 26dd77e71e29a2465efd33cb67f793111fe36c68 Mon Sep 17 00:00:00 2001
From: Tiziano Bacocco <tizbac2@gmail.com>
Date: Mon, 26 Jan 2015 23:26:22 +0100
Subject: [PATCH 21/21] d3dadapter: Fix uninitialized dmSize field of the
 DEVMODEW structure

Signed-off-by: Tiziano Bacocco <tizbac2@gmail.com>
---
 dlls/winex11.drv/d3dadapter.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/winex11.drv/d3dadapter.c b/dlls/winex11.drv/d3dadapter.c
index 994b453..2918297 100644
--- a/dlls/winex11.drv/d3dadapter.c
+++ b/dlls/winex11.drv/d3dadapter.c
@@ -597,6 +597,7 @@ DRI3Present_ChangePresentParameters( struct DRI3Present *This,
         newMode.dmPelsWidth = params->BackBufferWidth;
         newMode.dmPelsHeight = params->BackBufferHeight;
         newMode.dmFields = DM_PELSWIDTH | DM_PELSHEIGHT;
+        newMode.dmSize = sizeof(DEVMODEW);
         ChangeDisplaySettingsExW(This->devname, &newMode, 0, CDS_FULLSCREEN, NULL);
 
         style = fullscreen_style(0);
-- 
2.2.2

