From 2bec406f8c9a80d8bb7beef2a9cb39455be6b125 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sat, 14 Jan 2017 23:06:28 +0100
Subject: dsound: Respect return value of callback in DirectSoundEnumerateA/W.

---
 dlls/dsound/dsound_main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/dsound/dsound_main.c b/dlls/dsound/dsound_main.c
index db2ca506bda..2e12fe7a40d 100644
--- a/dlls/dsound/dsound_main.c
+++ b/dlls/dsound/dsound_main.c
@@ -522,7 +522,7 @@ HRESULT enumerate_mmdevices(EDataFlow flow, GUID *guids,
         }
 
         if(device != defdev){
-            send_device(device, &guids[n], cb, user);
+            keep_going = send_device(device, &guids[n], cb, user);
             ++n;
         }
 
-- 
2.11.0

