From a83bb75f363f799616d7b39cb4829c15638b0fff Mon Sep 17 00:00:00 2001
From: Christian Costa <titan.costa@gmail.com>
Date: Mon, 9 Nov 2015 10:20:21 +0100
Subject: ntdll: Add dll override default rule for purist mode.

---
 dlls/ntdll/loadorder.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/ntdll/loadorder.c b/dlls/ntdll/loadorder.c
index eaeba37..5068812 100644
--- a/dlls/ntdll/loadorder.c
+++ b/dlls/ntdll/loadorder.c
@@ -557,6 +557,7 @@ enum loadorder get_load_order( const WCHAR *app_name, const WCHAR *path )
     enum loadorder ret = LO_INVALID;
     HANDLE std_key, app_key = 0;
     WCHAR *module, *basename;
+    WCHAR default_rule[] = {'*',0};
 
     if (!init_done) init_load_order();
     std_key = get_override_standard_key();
@@ -588,6 +589,10 @@ enum loadorder get_load_order( const WCHAR *app_name, const WCHAR *path )
         goto done;
     }
 
+    /* default rule '*' (used for purist mode) */
+    if ((ret = get_load_order_value( std_key, app_key, default_rule )) != LO_INVALID)
+        goto done;
+
     /* and last the hard-coded default */
     ret = LO_DEFAULT;
     TRACE( "got hardcoded %s for %s\n", debugstr_loadorder(ret), debugstr_w(path) );
-- 
2.6.2

