From 7d9df59da745fe72a232617a340b795d23fa4dc8 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Fri, 9 Dec 2016 12:47:24 +0800
Subject: include: Fix the GUID_WICPixelFormat32bppCMYK definition.

---
 include/wincodec.idl | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/wincodec.idl b/include/wincodec.idl
index e8c7023..04765b4 100644
--- a/include/wincodec.idl
+++ b/include/wincodec.idl
@@ -221,7 +221,7 @@ cpp_quote("DEFINE_GUID(GUID_WICPixelFormat64bppPRGBA, 0x6fddc324,0x4e03,0x4bfe,0
 
 cpp_quote("DEFINE_GUID(GUID_WICPixelFormat128bppRGBAFloat, 0x6fddc324,0x4e03,0x4bfe,0xb1,0x85,0x3d,0x77,0x76,0x8d,0xc9,0x19);")
 
-cpp_quote("DEFINE_GUID(GUID_WICPixelFormat32bppCMYK, 0x6fddc324,0x4e03,0x4fbe,0xb1,0x85,0x3d,0x77,0x76,0x8d,0xc9,0x1c);")
+cpp_quote("DEFINE_GUID(GUID_WICPixelFormat32bppCMYK, 0x6fddc324,0x4e03,0x4bfe,0xb1,0x85,0x3d,0x77,0x76,0x8d,0xc9,0x1c);")
 
 typedef struct WICRect {
     INT X;
-- 
2.9.0

