From 6844c121a7cd8a7f90afc0c033cf1c3519fd8d9a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 28 Jul 2017 02:41:27 +0200
Subject: server: Avoid crash when async_terminate destroys async object in
 free_async_queue.

---
 server/async.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/server/async.c b/server/async.c
index 926c121f6f9..9e91f9eadf3 100644
--- a/server/async.c
+++ b/server/async.c
@@ -194,10 +194,12 @@ void free_async_queue( struct async_queue *queue )
 
     LIST_FOR_EACH_ENTRY_SAFE( async, next, &queue->queue, struct async, queue_entry )
     {
+        grab_object( &async->obj );
         if (!async->completion) async->completion = fd_get_completion( async->fd, &async->comp_key );
         async->fd = NULL;
         async_terminate( async, STATUS_HANDLES_CLOSED );
         async->queue = NULL;
+        release_object( &async->obj );
     }
 }
 
-- 
2.13.1

