From f873a7b1be1f76c5dba0198353ed1256c8eaabd8 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 12 Feb 2017 05:20:04 +0100
Subject: wined3d: Create buffers on default pool in wined3d_buffer_create.

---
 dlls/wined3d/buffer.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/wined3d/buffer.c b/dlls/wined3d/buffer.c
index 386e40c1409..7fbccd7668b 100644
--- a/dlls/wined3d/buffer.c
+++ b/dlls/wined3d/buffer.c
@@ -1500,7 +1500,7 @@ HRESULT CDECL wined3d_buffer_create(struct wined3d_device *device, const struct
     FIXME("Ignoring access flags (pool).\n");
 
     hr = buffer_init(object, device, desc->byte_width, desc->usage, WINED3DFMT_UNKNOWN,
-            WINED3D_POOL_MANAGED, desc->bind_flags, data, parent, parent_ops);
+            WINED3D_POOL_DEFAULT, desc->bind_flags, data, parent, parent_ops);
     if (FAILED(hr))
     {
         WARN("Failed to initialize buffer, hr %#x.\n", hr);
-- 
2.11.0

