From fd6c5551002854ec585eca40e6889d3461c96b24 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 31 Aug 2017 01:47:49 +0200
Subject: [PATCH] d3d9/tests: Check MaxVertexBlendMatrixIndex capability.

---
 dlls/d3d9/tests/visual.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/dlls/d3d9/tests/visual.c b/dlls/d3d9/tests/visual.c
index 8e01445970f..1f0d4a85d1d 100644
--- a/dlls/d3d9/tests/visual.c
+++ b/dlls/d3d9/tests/visual.c
@@ -21971,7 +21971,9 @@ static void test_indexed_vertex_blending(void)
     IDirect3DDevice9 *device;
     IDirect3D9 *d3d;
     ULONG refcount;
+    D3DCAPS9 caps;
     HWND window;
+    HRESULT hr;
 
     window = CreateWindowA("static", "d3d9_test", WS_OVERLAPPEDWINDOW | WS_VISIBLE,
             0, 0, 640, 480, NULL, NULL, NULL, NULL);
@@ -22000,7 +22002,14 @@ static void test_indexed_vertex_blending(void)
     if (SUCCEEDED(IDirect3D9_CreateDevice(d3d, D3DADAPTER_DEFAULT, D3DDEVTYPE_HAL, window,
             D3DCREATE_SOFTWARE_VERTEXPROCESSING, &present_parameters, &device)))
     {
+        memset(&caps, 0, sizeof(caps));
+        hr = IDirect3DDevice9_GetDeviceCaps(device, &caps);
+        ok(SUCCEEDED(hr), "Failed to get device caps, hr %#x.\n", hr);
+        todo_wine ok(caps.MaxVertexBlendMatrixIndex == 255, "Expected 255 as maximum blend matrix index, got %u.\n",
+                caps.MaxVertexBlendMatrixIndex);
+
         do_test_indexed_vertex_blending(device,"IVB software");
+
         refcount = IDirect3DDevice9_Release(device);
         ok(!refcount, "Device has %u references left.\n", refcount);
     }
@@ -22010,7 +22019,26 @@ static void test_indexed_vertex_blending(void)
     if (SUCCEEDED(IDirect3D9_CreateDevice(d3d, D3DADAPTER_DEFAULT, D3DDEVTYPE_HAL, window,
             D3DCREATE_MIXED_VERTEXPROCESSING, &present_parameters, &device)))
     {
+        memset(&caps, 0, sizeof(caps));
+        hr = IDirect3DDevice9_GetDeviceCaps(device, &caps);
+        ok(SUCCEEDED(hr), "Failed to get device caps, hr %#x.\n", hr);
+        ok(caps.MaxVertexBlendMatrixIndex < 255, "Expected less than 255 as maximum blend matrix index, got %u.\n",
+                caps.MaxVertexBlendMatrixIndex);
+
+        hr = IDirect3DDevice9_SetSoftwareVertexProcessing(device, TRUE);
+        ok(SUCCEEDED(hr), "Failed to enable software processing, hr %#x.\n", hr);
+
+        memset(&caps, 0, sizeof(caps));
+        hr = IDirect3DDevice9_GetDeviceCaps(device, &caps);
+        ok(SUCCEEDED(hr), "Failed to get device caps, hr %#x.\n", hr);
+        todo_wine ok(caps.MaxVertexBlendMatrixIndex == 255, "Expected 255 as maximum blend matrix index, got %u.\n",
+                caps.MaxVertexBlendMatrixIndex);
+
+        hr = IDirect3DDevice9_SetSoftwareVertexProcessing(device, FALSE);
+        ok(SUCCEEDED(hr), "Failed to disable software processing, hr %#x.\n", hr);
+
         do_test_indexed_vertex_blending(device,"IVB mixed");
+
         refcount = IDirect3DDevice9_Release(device);
         ok(!refcount, "Device has %u references left.\n", refcount);
     }
-- 
2.26.2

