From a535cf73bc68b02cfd5fb8d03e86a442f472db4b Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Thu, 20 Oct 2016 16:50:20 +0800
Subject: windowscodecs: Improve stub for
 IWICMetadataQueryReader::GetMetadataByName.

---
 dlls/windowscodecs/metadataquery.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/windowscodecs/metadataquery.c b/dlls/windowscodecs/metadataquery.c
index 10bf295..24de2f8 100644
--- a/dlls/windowscodecs/metadataquery.c
+++ b/dlls/windowscodecs/metadataquery.c
@@ -109,7 +109,7 @@ static HRESULT WINAPI mqr_GetMetadataByName(IWICMetadataQueryReader *iface,
 {
     QueryReader *This = impl_from_IWICMetadataQueryReader(iface);
     FIXME("(%p,%s,%p)\n", This, wine_dbgstr_w(wzName), pvarValue);
-    return E_NOTIMPL;
+    return WINCODEC_ERR_PROPERTYNOTFOUND;
 }
 
 static HRESULT WINAPI mqr_GetEnumerator(IWICMetadataQueryReader *iface,
-- 
2.9.0

