From 82ead58d02badf4e84124277868942b089d9b7ba Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Thu, 27 Oct 2016 14:48:45 +0800
Subject: windowscodecs: Add registration of the GIF encoder.

---
 dlls/windowscodecs/windowscodecs_wincodec.idl | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/dlls/windowscodecs/windowscodecs_wincodec.idl b/dlls/windowscodecs/windowscodecs_wincodec.idl
index fd7ff01..1681b1d 100644
--- a/dlls/windowscodecs/windowscodecs_wincodec.idl
+++ b/dlls/windowscodecs/windowscodecs_wincodec.idl
@@ -70,6 +70,13 @@ coclass WICBmpEncoder { interface IWICBitmapEncoder; }
 coclass WICGifDecoder { interface IWICBitmapDecoder; }
 
 [
+    helpstring("WIC GIF Encoder"),
+    threading(both),
+    uuid(114f5598-0b22-40a0-86a1-c83ea495adbd)
+]
+coclass WICGifEncoder { interface IWICBitmapEncoder; }
+
+[
     helpstring("WIC ICO Decoder"),
     threading(both),
     uuid(c61bfcdf-2e0f-4aad-a8d7-e06bafebcdfe)
-- 
2.9.0

