From 6e645b3f441ca310a17a06487b0a55fa629aa0d1 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 2 May 2015 18:42:09 +0200
Subject: kernel32: Do not omit mandatory argument for VirtualProtect.

---
 dlls/kernel32/except.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/kernel32/except.c b/dlls/kernel32/except.c
index bd50e88..2a7dbaa 100644
--- a/dlls/kernel32/except.c
+++ b/dlls/kernel32/except.c
@@ -401,6 +401,7 @@ static BOOL start_debugger_atomic(PEXCEPTION_POINTERS epointers)
  */
 static inline BOOL check_resource_write( void *addr )
 {
+    DWORD old_prot;
     void *rsrc;
     DWORD size;
     MEMORY_BASIC_INFORMATION info;
@@ -412,7 +413,7 @@ static inline BOOL check_resource_write( void *addr )
         return FALSE;
     if (addr < rsrc || (char *)addr >= (char *)rsrc + size) return FALSE;
     TRACE( "Broken app is writing to the resource data, enabling work-around\n" );
-    VirtualProtect( rsrc, size, PAGE_READWRITE, NULL );
+    VirtualProtect( rsrc, size, PAGE_READWRITE, &old_prot );
     return TRUE;
 }
 
-- 
2.3.5

