From 10c4be44f7321c54dd01785ba31799f89530ae1b Mon Sep 17 00:00:00 2001
From: Qian Hong <qhong@codeweavers.com>
Date: Fri, 28 Aug 2015 12:02:36 +0800
Subject: server: Map EXDEV to STATUS_NOT_SAME_DEVICE.

---
 server/file.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/server/file.c b/server/file.c
index a07ca16..4481b8c 100644
--- a/server/file.c
+++ b/server/file.c
@@ -661,6 +661,7 @@ void file_set_error(void)
     case EFBIG:     set_error( STATUS_SECTION_TOO_BIG ); break;
     case ENODEV:    set_error( STATUS_NO_SUCH_DEVICE ); break;
     case ENXIO:     set_error( STATUS_NO_SUCH_DEVICE ); break;
+    case EXDEV:     set_error( STATUS_NOT_SAME_DEVICE ); break;
 #ifdef EOVERFLOW
     case EOVERFLOW: set_error( STATUS_INVALID_PARAMETER ); break;
 #endif
-- 
2.5.0

