From 35f58b2470a342db3fe665549ac1983b4877157a Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Tue, 25 Aug 2020 17:37:28 -0500
Subject: [PATCH 06/52] winegstreamer: Insert audioconvert into decoded audio
 streams.

Signed-off-by: Derek Lesho <dlesho@codeweavers.com>
---
 dlls/winegstreamer/media_source.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/dlls/winegstreamer/media_source.c b/dlls/winegstreamer/media_source.c
index 209753608b6..66cfdbdc0fd 100644
--- a/dlls/winegstreamer/media_source.c
+++ b/dlls/winegstreamer/media_source.c
@@ -476,9 +476,18 @@ static HRESULT media_stream_align_with_mf(struct media_stream *stream)
     }
     else if(!strcmp(gst_structure_get_name(gst_caps_get_structure(source_caps, 0)), "audio/x-raw"))
     {
-        stream->my_sink = gst_element_get_static_pad(stream->appsink, "sink");
-        g_object_set(stream->appsink, "caps", source_caps, NULL);
+        GstElement *audioconvert = gst_element_factory_make("audioconvert", NULL);
+
+        gst_bin_add(GST_BIN(stream->parent_source->container), audioconvert);
+
+        stream->my_sink = gst_element_get_static_pad(audioconvert, "sink");
+
         assert(gst_pad_link(stream->their_src, stream->my_sink) == GST_PAD_LINK_OK);
+        assert(gst_element_link(audioconvert, stream->appsink));
+
+        gst_element_sync_state_with_parent(audioconvert);
+
+        g_object_set(stream->appsink, "caps", source_caps, NULL);
     }
     else
     {
-- 
2.28.0

