From f72479c9e576a766f38bcd9c8c377ee53adbdfe2 Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Wed, 13 May 2020 12:32:42 -0500
Subject: [PATCH 43/52] Support stereo down folding.

---
 dlls/winegstreamer/mf_decode.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/dlls/winegstreamer/mf_decode.c b/dlls/winegstreamer/mf_decode.c
index 7ce72a9a6a3..00f69aa76cc 100644
--- a/dlls/winegstreamer/mf_decode.c
+++ b/dlls/winegstreamer/mf_decode.c
@@ -327,9 +327,11 @@ static HRESULT WINAPI mf_decoder_GetOutputAvailableType(IMFTransform *iface, DWO
 
     copy_attr(output_type, decoder->input_type, &MF_MT_FRAME_SIZE);
     copy_attr(output_type, decoder->input_type, &MF_MT_FRAME_RATE);
-    copy_attr(output_type, decoder->input_type, &MF_MT_AUDIO_NUM_CHANNELS);
     copy_attr(output_type, decoder->input_type, &MF_MT_AUDIO_SAMPLES_PER_SECOND);
 
+    /* TODO: support both stereo folding and matching channels */
+    IMFMediaType_SetUINT32(output_type, &MF_MT_AUDIO_NUM_CHANNELS, 2);
+
     if (FAILED(hr = IMFMediaType_SetGUID(output_type, &MF_MT_MAJOR_TYPE, decoder_descs[decoder->type].major_type)))
     {
         IMFMediaType_Release(output_type);
-- 
2.28.0

