From 4ffe94b77e7f50263d8e3ae5f3466a757f9d9d4e Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 11 Jul 2015 15:28:27 +0200
Subject: ntdll: Fix incorrect assignment in assert statement (Coverity).

---
 dlls/ntdll/threadpool.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ntdll/threadpool.c b/dlls/ntdll/threadpool.c
index a4e0619..23091f3 100644
--- a/dlls/ntdll/threadpool.c
+++ b/dlls/ntdll/threadpool.c
@@ -1598,7 +1598,7 @@ static NTSTATUS tp_waitqueue_lock( struct threadpool_object *wait )
     struct waitqueue_bucket *bucket;
     NTSTATUS status;
     HANDLE thread;
-    assert( wait->type = TP_OBJECT_TYPE_WAIT );
+    assert( wait->type == TP_OBJECT_TYPE_WAIT );
 
     wait->u.wait.signaled       = 0;
     wait->u.wait.bucket         = NULL;
-- 
2.4.5

