From f734e1d94b068a78cc10600c190bb2d527ab4866 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Fri, 14 Apr 2017 16:40:43 +0800
Subject: advapi32: If the query is not specified the default query is
 "Global".

---
 dlls/advapi32/registry.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/advapi32/registry.c b/dlls/advapi32/registry.c
index 37462380c40..0aa38c5b9b3 100644
--- a/dlls/advapi32/registry.c
+++ b/dlls/advapi32/registry.c
@@ -1581,9 +1581,13 @@ error:
 
 static DWORD collect_data(struct perf_provider *provider, const WCHAR *query, void **data, DWORD *size, DWORD *obj_count)
 {
+    static const WCHAR globalW[] = { 'G','l','o','b','a','l',0 };
     WCHAR *linkage = provider->linkage[0] ? provider->linkage : NULL;
     DWORD err;
 
+    if (!query || !query[0])
+        query = globalW;
+
     err = provider->pOpen(linkage);
     if (err != ERROR_SUCCESS)
     {
-- 
2.13.1

