From 067bfef2398869f7f2e0e93e43458350f73d733c Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Sat, 27 Feb 2016 18:05:25 +0100
Subject: ntdll: Reserve some more stack space in call_thread_func_wrapper.

---
 dlls/ntdll/signal_i386.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ntdll/signal_i386.c b/dlls/ntdll/signal_i386.c
index 59dca6c..cf3611d 100644
--- a/dlls/ntdll/signal_i386.c
+++ b/dlls/ntdll/signal_i386.c
@@ -2740,7 +2740,7 @@ __ASM_GLOBAL_FUNC(call_thread_func_wrapper,
                   __ASM_CFI(".cfi_rel_offset %ebp,0\n\t")
                   "movl %esp,%ebp\n\t"
                   __ASM_CFI(".cfi_def_cfa_register %ebp\n\t")
-                  "subl $4,%esp\n\t"
+                  "subl $20,%esp\n\t"
                   "pushl 12(%ebp)\n\t"
                   "call *8(%ebp)\n\t"
                   "leal -4(%ebp),%esp\n\t"
-- 
2.7.1

