From efdcaba02d66b65177ce73686f4d4b5fb8f6e4f8 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 13 Jan 2017 01:59:13 +0100
Subject: kernel32: Always start debugger on WinSta0.

---
 dlls/kernel32/except.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/kernel32/except.c b/dlls/kernel32/except.c
index 1a1521a8038..cbc880f9a44 100644
--- a/dlls/kernel32/except.c
+++ b/dlls/kernel32/except.c
@@ -314,6 +314,7 @@ static BOOL	start_debugger(PEXCEPTION_POINTERS epointers, HANDLE hEvent)
     TRACE("Starting debugger %s\n", debugstr_a(cmdline));
     memset(&startup, 0, sizeof(startup));
     startup.cb = sizeof(startup);
+    startup.lpDesktop = (char *)"WinSta0";
     startup.dwFlags = STARTF_USESHOWWINDOW;
     startup.wShowWindow = SW_SHOWNORMAL;
     ret = CreateProcessA(NULL, cmdline, NULL, NULL, TRUE, 0, env, NULL, &startup, &info);
-- 
2.11.0

