From 9e35c8cf89c38c122b877924cb153f4c1748a7d3 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 4 Sep 2014 09:26:33 +0200
Subject: configure: Fix detection of ncurses library on Archlinux.

---
 configure.ac | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index db8b76f..f3fd12d 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1370,7 +1370,7 @@ dnl **** Check which curses lib to use ***
 CURSES_LIBS=""
 if test "$ac_cv_header_ncurses_h" = "yes"
 then
-    WINE_CHECK_SONAME(ncurses,waddch,[CURSES_LIBS="-lncurses"])
+    WINE_CHECK_SONAME(ncurses,waddch,[CURSES_LIBS="-lncurses"],,,libncursesw\\?)
 elif test "$ac_cv_header_curses_h" = "yes"
 then
     WINE_CHECK_SONAME(curses,waddch,[CURSES_LIBS="-lcurses"])
-- 
2.1.0

