From 57462a28a49782e3ef8b0efefab99d336b7d67a2 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Thu, 31 Mar 2016 10:09:03 +0800
Subject: user32: Avoid dereferencing NULL pointer in a trace.

---
 dlls/user32/win.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/user32/win.c b/dlls/user32/win.c
index 33aedbd..43ab5ed 100644
--- a/dlls/user32/win.c
+++ b/dlls/user32/win.c
@@ -3472,7 +3472,7 @@ BOOL WINAPI FlashWindowEx( PFLASHWINFO pfinfo )
 {
     WND *wndPtr;
 
-    TRACE( "%p\n", pfinfo->hwnd );
+    TRACE( "%p\n", pfinfo );
 
     if (!pfinfo)
     {
@@ -3485,7 +3485,7 @@ BOOL WINAPI FlashWindowEx( PFLASHWINFO pfinfo )
         SetLastError( ERROR_INVALID_PARAMETER );
         return FALSE;
     }
-    FIXME( "%p - semi-stub\n", pfinfo );
+    FIXME( "%p - semi-stub\n", pfinfo->hwnd );
 
     if (IsIconic( pfinfo->hwnd ))
     {
-- 
2.7.1

