From d08d8a1233fd4d1fcb269b6e364c2122183fdd3f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sun, 5 Feb 2017 16:34:41 +0100
Subject: wined3d: Replace surface_load_location with texture_load_location.

---
 dlls/wined3d/device.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/wined3d/device.c b/dlls/wined3d/device.c
index 74a4b09ab1a..4d0f6f2c4be 100644
--- a/dlls/wined3d/device.c
+++ b/dlls/wined3d/device.c
@@ -286,7 +286,8 @@ void device_clear_render_targets(struct wined3d_device *device, UINT rt_count, c
     {
         DWORD location = render_offscreen ? dsv->resource->draw_binding : WINED3D_LOCATION_DRAWABLE;
 
-        surface_load_location(depth_stencil, context, location);
+        wined3d_texture_load_location(depth_stencil->container, dsv->sub_resource_idx,
+                context, location);
     }
 
     if (!context_apply_clear_state(context, state, rt_count, fb))
-- 
2.11.0

