From 8dc59ef56211678a681315041c0fd23bd490ad25 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 29 Jul 2015 06:35:07 +0200
Subject: wined3d: Properly initialize format_flags for surfaces and volumes.

---
 dlls/wined3d/surface.c | 2 ++
 dlls/wined3d/volume.c  | 1 +
 2 files changed, 3 insertions(+)

diff --git a/dlls/wined3d/surface.c b/dlls/wined3d/surface.c
index 3b033db..38a233d 100644
--- a/dlls/wined3d/surface.c
+++ b/dlls/wined3d/surface.c
@@ -5603,6 +5603,8 @@ static HRESULT surface_init(struct wined3d_surface *surface, struct wined3d_text
     }
 
     surface->container = container;
+    surface->resource.format_flags = container->resource.format_flags;
+
     surface_validate_location(surface, WINED3D_LOCATION_SYSMEM);
     list_init(&surface->renderbuffers);
     list_init(&surface->overlays);
diff --git a/dlls/wined3d/volume.c b/dlls/wined3d/volume.c
index 424938a..4e1c415 100644
--- a/dlls/wined3d/volume.c
+++ b/dlls/wined3d/volume.c
@@ -779,6 +779,7 @@ static HRESULT volume_init(struct wined3d_volume *volume, struct wined3d_texture
     }
 
     volume->container = container;
+    volume->resource.format_flags = container->resource.format_flags;
 
     return WINED3D_OK;
 }
-- 
2.4.5

