From 61838c8fd1fd47dbbe71c74946b652aa441b7e91 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 19 Feb 2017 00:57:12 +0100
Subject: [PATCH] wined3d: Add additional synchronization CS ops.

---
 dlls/wined3d/view.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/dlls/wined3d/view.c b/dlls/wined3d/view.c
index 59a30ead..f4327373 100644
--- a/dlls/wined3d/view.c
+++ b/dlls/wined3d/view.c
@@ -775,6 +775,8 @@ static void wined3d_shader_resource_view_cs_init(void *object)
                     debug_d3dformat(resource->format->id), debug_d3dformat(view_format->id));
         }
     }
+
+    wined3d_resource_release(resource);
 }
 
 static HRESULT wined3d_shader_resource_view_init(struct wined3d_shader_resource_view *view,
@@ -791,6 +793,7 @@ static HRESULT wined3d_shader_resource_view_init(struct wined3d_shader_resource_
 
     wined3d_resource_incref(view->resource = resource);
 
+    wined3d_resource_acquire(resource);
     wined3d_cs_init_object(resource->device->cs, wined3d_shader_resource_view_cs_init, view);
 
     return WINED3D_OK;
@@ -1145,6 +1148,8 @@ static void wined3d_unordered_access_view_cs_init(void *object)
                     desc, texture_gl, view->format);
         }
     }
+
+    wined3d_resource_release(resource);
 }
 
 static HRESULT wined3d_unordered_access_view_init(struct wined3d_unordered_access_view *view,
@@ -1161,6 +1166,7 @@ static HRESULT wined3d_unordered_access_view_init(struct wined3d_unordered_acces
 
     wined3d_resource_incref(view->resource = resource);
 
+    wined3d_resource_acquire(resource);
     wined3d_cs_init_object(resource->device->cs, wined3d_unordered_access_view_cs_init, view);
 
     return WINED3D_OK;
-- 
2.19.1

