From 25903b870407e8368508fd002f15c509fdd9a66b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Thu, 19 Sep 2013 13:09:58 +0200
Subject: wined3d: Discard the backbuffer in discard presents.

---
 dlls/wined3d/swapchain.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/dlls/wined3d/swapchain.c b/dlls/wined3d/swapchain.c
index a02fdf0..9a36e38 100644
--- a/dlls/wined3d/swapchain.c
+++ b/dlls/wined3d/swapchain.c
@@ -608,6 +608,17 @@ static void swapchain_gl_present(struct wined3d_swapchain *swapchain,
      * up to date and hope WGL flipped front and back buffers and read this data into
      * the FBO. Don't bother about this for now. */
 
+    switch (swapchain->desc.swap_effect)
+    {
+        case WINED3D_SWAP_EFFECT_DISCARD:
+            wined3d_texture_validate_location(swapchain->back_buffers[swapchain->desc.backbuffer_count - 1],
+                    0, WINED3D_LOCATION_DISCARDED);
+            break;
+
+        default:
+            break;
+    }
+
     if (fb->depth_stencil)
     {
         struct wined3d_surface *ds = wined3d_rendertarget_view_get_surface(fb->depth_stencil);
-- 
2.8.0

