From 890fa1752284ee9a459b29eeac7d875cd1a84015 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Tue, 1 Oct 2013 15:30:26 +0200
Subject: wined3d: Give the cs its own state

---
 dlls/wined3d/cs.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index 73352d93b97..d3be8d1ef21 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -603,7 +603,7 @@ static void release_unordered_access_resources(const struct wined3d_shader *shad
 
 static void wined3d_cs_exec_dispatch(struct wined3d_cs *cs, const void *data)
 {
-    struct wined3d_state *state = &cs->device->state;
+    struct wined3d_state *state = &cs->state;
     const struct wined3d_cs_dispatch *op = data;
 
     dispatch_compute(cs->device, state,
@@ -635,7 +635,7 @@ void wined3d_cs_emit_dispatch(struct wined3d_cs *cs,
 
 static void wined3d_cs_exec_draw(struct wined3d_cs *cs, const void *data)
 {
-    struct wined3d_state *state = &cs->device->state;
+    struct wined3d_state *state = &cs->state;
     const struct wined3d_cs_draw *op = data;
     unsigned int i;
 
-- 
2.11.0

