From 2fb94b8d1c47c38586073f1fb8a7355c827e4082 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Wed, 2 Oct 2013 23:28:56 +0200
Subject: wined3d: Unload resources through the CS in uninit_3d.

---
 dlls/wined3d/device.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/wined3d/device.c b/dlls/wined3d/device.c
index 6395aaea6be..082489c295e 100644
--- a/dlls/wined3d/device.c
+++ b/dlls/wined3d/device.c
@@ -1153,6 +1153,9 @@ HRESULT CDECL wined3d_device_uninit_3d(struct wined3d_device *device)
         wined3d_cs_emit_unload_resource(device->cs, resource);
     }
 
+    if (wined3d_settings.cs_multithreaded)
+        device->cs->ops->submit_and_wait(device->cs);
+
     wine_rb_clear(&device->samplers, device_free_sampler, NULL);
 
     /* Destroy the depth blt resources, they will be invalid after the reset. Also free shader
-- 
2.11.0

