From 8599235cc72aa10486d71a15e3059ef0c5478ddc Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Wed, 2 Oct 2013 23:40:20 +0200
Subject: wined3d: Unload resources through the CS in device_reset.

---
 dlls/wined3d/device.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/wined3d/device.c b/dlls/wined3d/device.c
index 082489c295e..7ad88637f08 100644
--- a/dlls/wined3d/device.c
+++ b/dlls/wined3d/device.c
@@ -4674,6 +4674,9 @@ static void delete_opengl_contexts(struct wined3d_device *device, struct wined3d
         wined3d_cs_emit_unload_resource(device->cs, resource);
     }
 
+    if (wined3d_settings.cs_multithreaded)
+        device->cs->ops->submit_and_wait(device->cs);
+
     LIST_FOR_EACH_ENTRY(shader, &device->shaders, struct wined3d_shader, shader_list_entry)
     {
         device->shader_backend->shader_destroy(shader);
-- 
2.11.0

