From f302ec8c029b6abe1a925ce42707bdf1662a2102 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Tue, 17 May 2016 12:49:00 +0200
Subject: wined3d: Do not query available GPU memory on main thread when CSMT
 is enabled.

---
 dlls/wined3d/device.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/wined3d/device.c b/dlls/wined3d/device.c
index 19cdc8d1824..50ae4a928fc 100644
--- a/dlls/wined3d/device.c
+++ b/dlls/wined3d/device.c
@@ -1257,7 +1257,7 @@ UINT CDECL wined3d_device_get_available_texture_mem(const struct wined3d_device
 
     /* We can not acquire the context unless there is a swapchain. */
     if (device->swapchains && gl_info->supported[NVX_GPU_MEMORY_INFO] &&
-            !wined3d_settings.emulated_textureram)
+            !wined3d_settings.emulated_textureram && !wined3d_settings.cs_multithreaded)
     {
         GLint vram_free_kb;
         UINT64 vram_free;
-- 
2.11.0

