From d0f1f1278a8225075bd03c9586112bf29318563c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Mon, 16 Sep 2013 22:22:39 +0200
Subject: wined3d: Allocate sysmem for client storage if it doesn't exist
 already.

---
 dlls/wined3d/texture.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/dlls/wined3d/texture.c b/dlls/wined3d/texture.c
index 41cf61d..8298e53 100644
--- a/dlls/wined3d/texture.c
+++ b/dlls/wined3d/texture.c
@@ -904,8 +904,7 @@ static void texture2d_prepare_texture(struct wined3d_texture *texture, struct wi
         if (gl_info->supported[APPLE_CLIENT_STORAGE])
         {
             if (surface->flags & (SFLAG_NONPOW2 | SFLAG_DIBSECTION)
-                    || texture->flags & WINED3D_TEXTURE_CONVERTED
-                    || !surface->resource.heap_memory)
+                    || texture->flags & WINED3D_TEXTURE_CONVERTED)
             {
                 /* In some cases we want to disable client storage.
                  * SFLAG_NONPOW2 has a bigger opengl texture than the client memory, and different pitches
@@ -917,6 +916,9 @@ static void texture2d_prepare_texture(struct wined3d_texture *texture, struct wi
             }
             else
             {
+                if (!surface->resource.heap_memory)
+                    wined3d_resource_allocate_sysmem(&surface->resource);
+
                 surface->flags |= SFLAG_CLIENT;
                 mem = surface->resource.heap_memory;
 
-- 
2.1.3

