From 13fe67e10d483f6fe5047f693949bdf6560758ed Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 19 Feb 2017 07:01:45 +0100
Subject: wined3d: Run the cs asynchronously.

---
 dlls/wined3d/cs.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index c8394533a86..a656c93684a 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -2637,7 +2637,7 @@ static void *wined3d_cs_mt_require_space(struct wined3d_cs *cs, size_t size)
     return data;
 }
 
-/* static */ void wined3d_cs_mt_submit(struct wined3d_cs *cs)
+static void wined3d_cs_mt_submit(struct wined3d_cs *cs)
 {
     if (cs->thread_id == GetCurrentThreadId())
         return wined3d_cs_st_submit(cs);
@@ -2667,7 +2667,7 @@ static void wined3d_cs_mt_submit_and_wait(struct wined3d_cs *cs)
 static const struct wined3d_cs_ops wined3d_cs_mt_ops =
 {
     wined3d_cs_mt_require_space,
-    wined3d_cs_mt_submit_and_wait,  /* FIXME */
+    wined3d_cs_mt_submit,
     wined3d_cs_mt_submit_and_wait,
 };
 
-- 
2.11.0

