From 900dc644b944426cf580aa03f236043322a1ac20 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Tue, 21 Jan 2014 16:41:33 +0100
Subject: wined3d: Don't delete the buffer in surface_cleanup.

---
 dlls/wined3d/surface.c | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/dlls/wined3d/surface.c b/dlls/wined3d/surface.c
index 4c2a759..7d2d1ea 100644
--- a/dlls/wined3d/surface.c
+++ b/dlls/wined3d/surface.c
@@ -56,12 +56,6 @@ static void surface_cleanup(struct wined3d_surface *surface)
         context = context_acquire(surface->resource.device, NULL);
         gl_info = context->gl_info;
 
-        if (surface->resource.buffer_object)
-        {
-            TRACE("Deleting PBO %u.\n", surface->resource.buffer_object);
-            GL_EXTCALL(glDeleteBuffersARB(1, &surface->resource.buffer_object));
-        }
-
         if (surface->rb_multisample)
         {
             TRACE("Deleting multisample renderbuffer %u.\n", surface->rb_multisample);
-- 
2.1.3

