From b7a9ab5cf76fcde0dcfadfa402587a7282b93718 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Thu, 4 Jul 2013 23:40:56 +0200
Subject: wined3d: Don't preload buffers on unmap

---
 dlls/wined3d/buffer.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/dlls/wined3d/buffer.c b/dlls/wined3d/buffer.c
index a50bba4..2c66a4b 100644
--- a/dlls/wined3d/buffer.c
+++ b/dlls/wined3d/buffer.c
@@ -1133,10 +1133,6 @@ void CDECL wined3d_buffer_unmap(struct wined3d_buffer *buffer)
         buffer_clear_dirty_areas(buffer);
         buffer->map_ptr = NULL;
     }
-    else if (buffer->flags & WINED3D_BUFFER_HASDESC)
-    {
-        wined3d_buffer_preload(buffer);
-    }
 }
 
 HRESULT wined3d_buffer_copy(struct wined3d_buffer *dst_buffer, unsigned int dst_offset,
-- 
2.8.0

