From 25fcdc6e671df8cf7cb68a50e6dc59921a8c2738 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Thu, 4 Jul 2013 23:40:56 +0200
Subject: wined3d: Don't preload buffers on unmap

---
 dlls/wined3d/buffer.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/dlls/wined3d/buffer.c b/dlls/wined3d/buffer.c
index 85886bf..3ecd46b 100644
--- a/dlls/wined3d/buffer.c
+++ b/dlls/wined3d/buffer.c
@@ -1108,10 +1108,6 @@ void CDECL wined3d_buffer_unmap(struct wined3d_buffer *buffer)
         buffer_clear_dirty_areas(buffer);
         buffer->map_ptr = NULL;
     }
-    else if (buffer->flags & WINED3D_BUFFER_HASDESC)
-    {
-        wined3d_resource_preload(&buffer->resource);
-    }
 }
 
 HRESULT wined3d_buffer_copy(struct wined3d_buffer *dst_buffer, unsigned int dst_offset,
-- 
2.9.0

