From 25629ab322f251afa9ae5495212c84a00f5af00f Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 19 Feb 2017 08:29:23 +0100
Subject: wined3d: Reset context before destruction.

---
 dlls/wined3d/device.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/wined3d/device.c b/dlls/wined3d/device.c
index aa24d7f4c37..1bf7db30b3d 100644
--- a/dlls/wined3d/device.c
+++ b/dlls/wined3d/device.c
@@ -1200,6 +1200,7 @@ HRESULT CDECL wined3d_device_uninit_3d(struct wined3d_device *device)
 
     wine_rb_clear(&device->samplers, device_free_sampler, NULL);
 
+    context_set_current(NULL);
     wined3d_device_delete_opengl_contexts(device);
 
     if (device->fb.depth_stencil)
-- 
2.11.0

