From 95126fce6331a34c71f5ff736192d25e78e67dde Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Tue, 2 Apr 2013 16:20:13 +0200
Subject: wined3d: Store the color in clear ops instead of a pointer

---
 dlls/wined3d/cs.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index bac3fdb..fa1ee3a 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -96,7 +96,7 @@ struct wined3d_cs_clear
     enum wined3d_cs_op opcode;
     DWORD rect_count;
     DWORD flags;
-    const struct wined3d_color *color;
+    struct wined3d_color color;
     float depth;
     DWORD stencil;
     RECT rects[1];
@@ -454,7 +454,7 @@ static UINT wined3d_cs_exec_clear(struct wined3d_cs *cs, const void *data)
     wined3d_get_draw_rect(&device->state, &draw_rect);
     device_clear_render_targets(device, device->adapter->gl_info.limits.buffers,
             &cs->state.fb, op->rect_count, op->rect_count ? op->rects : NULL, &draw_rect, op->flags,
-            op->color, op->depth, op->stencil);
+            &op->color, op->depth, op->stencil);
 
     return sizeof(*op) + sizeof(*op->rects) * extra_rects;
 }
@@ -471,7 +471,7 @@ void wined3d_cs_emit_clear(struct wined3d_cs *cs, DWORD rect_count, const RECT *
     if (rect_count)
         memcpy(op->rects, rects, rect_count * sizeof(*rects));
     op->flags = flags;
-    op->color = color;
+    op->color = *color;
     op->depth = depth;
     op->stencil = stencil;
 
-- 
2.1.3

