From a6a1f67d53cb1fe1b84490548ab001516189451d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Wed, 10 Apr 2013 20:12:27 +0200
Subject: wined3d: Run the cs asynchronously

---
 dlls/wined3d/cs.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index c7bb109..03bf85e 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -1777,7 +1777,7 @@ static const struct wined3d_cs_ops wined3d_cs_st_ops =
 static const struct wined3d_cs_ops wined3d_cs_mt_ops =
 {
     wined3d_cs_mt_require_space,
-    wined3d_cs_flush_and_wait,
+    wined3d_cs_flush,
     wined3d_cs_flush_and_wait,
 };
 
-- 
2.8.0

