From 817d317d346d6f91b7affb58ed533ba3c22a397d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Sat, 6 Jul 2013 21:54:45 +0200
Subject: wined3d: Destroy queries through the CS

---
 dlls/wined3d/query.c | 14 +++-----------
 1 file changed, 3 insertions(+), 11 deletions(-)

diff --git a/dlls/wined3d/query.c b/dlls/wined3d/query.c
index 66368cf..c1ba517 100644
--- a/dlls/wined3d/query.c
+++ b/dlls/wined3d/query.c
@@ -237,6 +237,9 @@ static void wined3d_query_destroy_object(void *object)
 {
     struct wined3d_query *query = object;
 
+    if (!list_empty(&query->poll_list_entry))
+        list_remove(&query->poll_list_entry);
+
     /* Queries are specific to the GL context that created them. Not
      * deleting the query will obviously leak it, but that's still better
      * than potentially deleting a different query with the same id in this
@@ -272,18 +275,7 @@ ULONG CDECL wined3d_query_decref(struct wined3d_query *query)
     TRACE("%p decreasing refcount to %u.\n", query, refcount);
 
     if (!refcount)
-    {
-        if (wined3d_settings.cs_multithreaded)
-        {
-            struct wined3d_device *device = query->device;
-
-            FIXME("waiting for cs\n");
-            wined3d_cs_emit_glfinish(device->cs);
-            device->cs->ops->finish(device->cs);
-        }
-
         wined3d_cs_emit_destroy_object(query->device->cs, wined3d_query_destroy_object, query);
-    }
 
     return refcount;
 }
-- 
2.8.0

