From b71ef3360bd48011fa26e8b9e33774d8d2fffb36 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 8 Feb 2017 17:35:18 +0100
Subject: wined3d: Clean up cs lists on shutdown.

---
 dlls/wined3d/cs.c | 21 ++++++++++++++++-----
 1 file changed, 16 insertions(+), 5 deletions(-)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index 5c408ab4f75..218484e5fdd 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -612,6 +612,17 @@ static void wined3d_cs_list_init(struct wined3d_cs_list *list)
     list_init(&list->blocks);
 }
 
+static void wined3d_cs_list_cleanup(struct wined3d_cs_list *list)
+{
+    struct wined3d_cs_block *block, *next;
+
+    LIST_FOR_EACH_ENTRY_SAFE(block, next, &list->blocks, struct wined3d_cs_block, entry)
+    {
+        list_remove(&block->entry);
+        HeapFree(GetProcessHeap(), 0, block);
+    }
+}
+
 static struct wined3d_cs_block *wined3d_cs_get_block(struct wined3d_cs *cs, struct wined3d_cs_list *list)
 {
     struct wined3d_cs_block *block;
@@ -2851,6 +2862,8 @@ static void wined3d_cs_emit_stop(struct wined3d_cs *cs)
 static inline BOOL wined3d_cs_process_block(struct wined3d_cs *cs, struct wined3d_cs_block *block)
 {
     UINT pos = 0;
+    BOOL ret = TRUE;
+
     while (pos < block->pos)
     {
         enum wined3d_cs_op opcode = *(const enum wined3d_cs_op *)&block->data[pos];
@@ -2859,7 +2872,8 @@ static inline BOOL wined3d_cs_process_block(struct wined3d_cs *cs, struct wined3
         {
             if (opcode > WINED3D_CS_OP_STOP)
                 ERR("Invalid opcode %#x.\n", opcode);
-            return FALSE;
+            ret = FALSE;
+            break;
         }
 
         pos += wined3d_cs_op_handlers[opcode](cs, &block->data[pos]);
@@ -2869,7 +2883,7 @@ static inline BOOL wined3d_cs_process_block(struct wined3d_cs *cs, struct wined3
         InterlockedExchange(block->fence, TRUE);
 
     wined3d_cs_list_enqueue(&cs->free_list, block);
-    return TRUE;
+    return ret;
 }
 
 static DWORD WINAPI wined3d_cs_run(void *thread_param)
@@ -2958,12 +2972,9 @@ void wined3d_cs_destroy(struct wined3d_cs *cs)
         if (ret != WAIT_OBJECT_0)
             ERR("Wait failed (%#x).\n", ret);
 
-        /* FIXME: Cleanup the block lists on thread exit. */
-#if 0
         wined3d_cs_list_cleanup(&cs->exec_prio_list);
         wined3d_cs_list_cleanup(&cs->exec_list);
         wined3d_cs_list_cleanup(&cs->free_list);
-#endif
     }
 
     HeapFree(GetProcessHeap(), 0, cs->block_main);
-- 
2.11.0

