From 67a6d3fc2dff7be59c88fdbb9bc4199f20b691b3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Fri, 27 Sep 2013 16:57:01 +0200
Subject: wined3d: Fence blit operations.

---
 dlls/wined3d/cs.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index 329cdce..68c104e 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -1824,6 +1824,10 @@ static UINT wined3d_cs_exec_blt(struct wined3d_cs *cs, const void *data)
             op->src_surface, &op->src_rect,
             op->flags, &op->fx, op->filter);
 
+    wined3d_resource_release(&op->dst_surface->container->resource);
+    if (op->src_surface && op->src_surface != op->dst_surface)
+        wined3d_resource_release(&op->src_surface->container->resource);
+
     return sizeof(*op);
 }
 
@@ -1845,6 +1849,10 @@ void wined3d_cs_emit_blt(struct wined3d_cs *cs, struct wined3d_surface *dst_surf
     if (fx)
         op->fx = *fx;
 
+    wined3d_resource_acquire(&dst_surface->container->resource);
+    if (src_surface && src_surface != dst_surface)
+        wined3d_resource_acquire(&src_surface->container->resource);
+
     cs->ops->submit(cs, sizeof(*op));
 }
 
-- 
2.8.0

