From e7897ceec2a6eb091a1877df85471037bae120b7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Fri, 27 Sep 2013 16:58:13 +0200
Subject: wined3d: Fence color_fill operations.

---
 dlls/wined3d/cs.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index 68c104e..af10384 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -1866,6 +1866,8 @@ static UINT wined3d_cs_exec_clear_rtv(struct wined3d_cs *cs, const void *data)
     else
         op->blitter->depth_fill(device, op->view, &op->rect, op->flags, op->depth, op->stencil);
 
+    wined3d_resource_release(op->view->resource);
+
     return sizeof(*op);
 }
 
@@ -1886,6 +1888,8 @@ void wined3d_cs_emit_clear_rtv(struct wined3d_cs *cs, struct wined3d_rendertarge
     op->stencil = stencil;
     op->blitter = blitter;
 
+    wined3d_resource_acquire(view->resource);
+
     cs->ops->submit(cs, sizeof(*op));
 }
 
-- 
2.8.0

