From 3f09055fac1731ff05bc260ee2f6215ba40bb2cf Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Fri, 30 Aug 2013 10:41:42 +0200
Subject: wined3d: Fence update_texture calls

---
 dlls/wined3d/cs.c      | 6 ++++++
 dlls/wined3d/texture.c | 2 ++
 2 files changed, 8 insertions(+)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index f196f72..ff559de 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -1971,6 +1971,9 @@ static UINT wined3d_cs_exec_update_texture(struct wined3d_cs *cs, const void *da
     device_exec_update_texture(context, op->src, op->dst);
     context_release(context);
 
+    wined3d_resource_release(&op->src->resource);
+    wined3d_resource_release(&op->dst->resource);
+
     return sizeof(*op);
 }
 
@@ -1984,6 +1987,9 @@ void wined3d_cs_emit_update_texture(struct wined3d_cs *cs, struct wined3d_textur
     op->src = src;
     op->dst = dst;
 
+    wined3d_resource_acquire(&op->src->resource);
+    wined3d_resource_acquire(&op->dst->resource);
+
     cs->ops->submit(cs, sizeof(*op));
 }
 
diff --git a/dlls/wined3d/texture.c b/dlls/wined3d/texture.c
index 98d1177..13e8a26 100644
--- a/dlls/wined3d/texture.c
+++ b/dlls/wined3d/texture.c
@@ -1765,6 +1765,8 @@ static HRESULT texture_resource_sub_resource_map(struct wined3d_resource *resour
 
     flags = wined3d_resource_sanitize_map_flags(resource, flags);
 
+    wined3d_resource_wait_idle(&texture->resource);
+
     base_memory = wined3d_cs_emit_texture_map(device->cs, texture, sub_resource_idx, flags);
 
     if (fmt_flags & WINED3DFMT_FLAG_BROKEN_PITCH)
-- 
2.8.0

