From 26566a78da3e11952e265a4999d55279c595162a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Fri, 27 Sep 2013 16:57:01 +0200
Subject: wined3d: Fence blit operations.

---
 dlls/wined3d/cs.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index 98072e2..f34a642 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -1792,6 +1792,10 @@ static UINT wined3d_cs_exec_blt(struct wined3d_cs *cs, const void *data)
             op->src_surface, &op->src_rect,
             op->flags, &op->fx, op->filter);
 
+    wined3d_resource_dec_fence(&op->dst_surface->container->resource);
+    if (op->src_surface && op->src_surface != op->dst_surface)
+        wined3d_resource_dec_fence(&op->src_surface->container->resource);
+
     return sizeof(*op);
 }
 
@@ -1813,6 +1817,10 @@ void wined3d_cs_emit_blt(struct wined3d_cs *cs, struct wined3d_surface *dst_surf
     if (fx)
         op->fx = *fx;
 
+    wined3d_resource_inc_fence(&dst_surface->container->resource);
+    if (src_surface && src_surface != dst_surface)
+        wined3d_resource_inc_fence(&src_surface->container->resource);
+
     cs->ops->submit(cs, sizeof(*op));
 }
 
-- 
2.8.0

