From cccdf39d6f6cc952455a999d466fe868b4640d91 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Fri, 27 Sep 2013 17:04:47 +0200
Subject: wined3d: Fence clear calls.

---
 dlls/wined3d/cs.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index af10384..60ae4a3 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -522,6 +522,7 @@ static UINT wined3d_cs_exec_clear(struct wined3d_cs *cs, const void *data)
     struct wined3d_device *device;
     RECT draw_rect;
     size_t size = FIELD_OFFSET(struct wined3d_cs_clear, rects[op->rect_count]);
+    unsigned int i;
 
     device = cs->device;
     wined3d_get_draw_rect(&cs->state, &draw_rect);
@@ -529,6 +530,17 @@ static UINT wined3d_cs_exec_clear(struct wined3d_cs *cs, const void *data)
             &cs->state.fb, op->rect_count, op->rects, &draw_rect, op->flags,
             &op->color, op->depth, op->stencil);
 
+    if (op->flags & WINED3DCLEAR_TARGET)
+    {
+        for (i = 0; i < device->adapter->gl_info.limits.buffers; i++)
+        {
+            if (cs->state.fb.render_targets[i])
+                wined3d_resource_release(cs->state.fb.render_targets[i]->resource);
+        }
+    }
+    if (op->flags & (WINED3DCLEAR_ZBUFFER | WINED3DCLEAR_STENCIL))
+        wined3d_resource_release(cs->state.fb.depth_stencil->resource);
+
     return size;
 }
 
@@ -537,6 +549,9 @@ void wined3d_cs_emit_clear(struct wined3d_cs *cs, DWORD rect_count, const RECT *
 {
     struct wined3d_cs_clear *op;
     size_t size = FIELD_OFFSET(struct wined3d_cs_clear, rects[rect_count]);
+    const struct wined3d_state *state = &cs->device->state;
+    unsigned int i;
+
     op = cs->ops->require_space(cs, size);
     op->opcode = WINED3D_CS_OP_CLEAR;
     op->flags = flags;
@@ -546,6 +561,17 @@ void wined3d_cs_emit_clear(struct wined3d_cs *cs, DWORD rect_count, const RECT *
     op->rect_count = rect_count;
     memcpy(op->rects, rects, sizeof(*rects) * rect_count);
 
+    if (flags & WINED3DCLEAR_TARGET)
+    {
+        for (i = 0; i < cs->device->adapter->gl_info.limits.buffers; i++)
+        {
+            if (state->fb.render_targets[i])
+                wined3d_resource_acquire(state->fb.render_targets[i]->resource);
+        }
+    }
+    if (flags & (WINED3DCLEAR_ZBUFFER | WINED3DCLEAR_STENCIL))
+        wined3d_resource_acquire(state->fb.depth_stencil->resource);
+
     cs->ops->submit(cs, size);
 }
 
-- 
2.8.0

