From 9aae8336cd11533f9433b2fc305b379d257e5a1a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Fri, 27 Sep 2013 16:55:58 +0200
Subject: wined3d: Fence render targets and depth stencils

---
 dlls/wined3d/cs.c | 32 ++++++++++++++++++++++++++++----
 1 file changed, 28 insertions(+), 4 deletions(-)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index f65797f..070f538 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -590,6 +590,30 @@ void wined3d_cs_emit_clear(struct wined3d_cs *cs, DWORD rect_count, const RECT *
     cs->ops->submit(cs, size);
 }
 
+static inline BOOL wined3d_cs_colorwrite_enabled(const struct wined3d_state *state, unsigned int i)
+{
+    switch (i)
+    {
+        case 0:
+            return !!state->render_states[WINED3D_RS_COLORWRITEENABLE];
+        case 1:
+            return !!state->render_states[WINED3D_RS_COLORWRITEENABLE1];
+        case 2:
+            return !!state->render_states[WINED3D_RS_COLORWRITEENABLE2];
+        case 3:
+            return !!state->render_states[WINED3D_RS_COLORWRITEENABLE3];
+        default:
+            ERR("Unexpected color target %u.\n", i);
+            return TRUE;
+    }
+}
+
+static inline BOOL wined3d_cs_depth_stencil_enabled(const struct wined3d_state *state)
+{
+    return state->render_states[WINED3D_RS_ZENABLE]
+            || state->render_states[WINED3D_RS_STENCILENABLE];
+}
+
 static UINT wined3d_cs_exec_draw(struct wined3d_cs *cs, const void *data)
 {
     struct wined3d_state *state = &cs->state;
@@ -638,10 +662,10 @@ static UINT wined3d_cs_exec_draw(struct wined3d_cs *cs, const void *data)
     }
     for (i = 0; i < cs->device->adapter->gl_info.limits.buffers; ++i)
     {
-        if (state->fb.render_targets[i])
+        if (state->fb.render_targets[i] && wined3d_cs_colorwrite_enabled(state, i))
             wined3d_resource_release(state->fb.render_targets[i]->resource);
     }
-    if (state->fb.depth_stencil)
+    if (state->fb.depth_stencil && wined3d_cs_depth_stencil_enabled(state))
         wined3d_resource_release(state->fb.depth_stencil->resource);
     for (i = 0; i < WINED3D_SHADER_TYPE_COUNT; ++i)
     {
@@ -701,10 +725,10 @@ void wined3d_cs_emit_draw(struct wined3d_cs *cs, int base_vertex_idx, unsigned i
     }
     for (i = 0; i < cs->device->adapter->gl_info.limits.buffers; ++i)
     {
-        if (state->fb.render_targets[i])
+        if (state->fb.render_targets[i] && wined3d_cs_colorwrite_enabled(state, i))
             wined3d_resource_acquire(state->fb.render_targets[i]->resource);
     }
-    if (state->fb.depth_stencil)
+    if (state->fb.depth_stencil && wined3d_cs_depth_stencil_enabled(state))
         wined3d_resource_acquire(state->fb.depth_stencil->resource);
     for (i = 0; i < WINED3D_SHADER_TYPE_COUNT; ++i)
     {
-- 
2.8.0

