From 50911185e5ac185d72315264b0361f441cb2777b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Wed, 2 Oct 2013 12:36:02 +0200
Subject: wined3d: Don't request the frontbuffer to create dummy textures.

---
 dlls/wined3d/device.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/wined3d/device.c b/dlls/wined3d/device.c
index 47bd8c9..3ea4508 100644
--- a/dlls/wined3d/device.c
+++ b/dlls/wined3d/device.c
@@ -1032,7 +1032,7 @@ HRESULT CDECL wined3d_device_init_3d(struct wined3d_device *device,
     device->swapchains[0] = swapchain;
     device_init_swapchain_state(device, swapchain);
 
-    context = context_acquire(device, swapchain->front_buffer->sub_resources[0].u.surface);
+    context = context_acquire(device, NULL);
 
     create_dummy_textures(device, context);
     create_default_sampler(device);
-- 
2.8.0

