From 0603f71135885ebf0666d406380385936ed5ff7b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Thu, 4 Jul 2013 23:40:56 +0200
Subject: wined3d: Don't preload buffers on unmap

---
 dlls/wined3d/buffer.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/dlls/wined3d/buffer.c b/dlls/wined3d/buffer.c
index 5544640..b76a07d 100644
--- a/dlls/wined3d/buffer.c
+++ b/dlls/wined3d/buffer.c
@@ -1129,10 +1129,6 @@ void CDECL wined3d_buffer_unmap(struct wined3d_buffer *buffer)
         buffer_clear_dirty_areas(buffer);
         buffer->map_ptr = NULL;
     }
-    else if (buffer->flags & WINED3D_BUFFER_HASDESC)
-    {
-        wined3d_buffer_preload(buffer);
-    }
 }
 
 static ULONG buffer_resource_incref(struct wined3d_resource *resource)
-- 
2.1.3

