From 8c808d86f613798096da3f113415f2358c63d6bc Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Wed, 31 Jul 2013 23:06:27 +0200
Subject: wined3d: Destroy vertex declarations through the CS

---
 dlls/wined3d/vertexdeclaration.c | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/dlls/wined3d/vertexdeclaration.c b/dlls/wined3d/vertexdeclaration.c
index fcd637d..c8704e9 100644
--- a/dlls/wined3d/vertexdeclaration.c
+++ b/dlls/wined3d/vertexdeclaration.c
@@ -66,11 +66,6 @@ ULONG CDECL wined3d_vertex_declaration_decref(struct wined3d_vertex_declaration
 
     if (!refcount)
     {
-        if (wined3d_settings.cs_multithreaded)
-        {
-            FIXME("Waiting for cs.\n");
-            declaration->device->cs->ops->finish(declaration->device->cs);
-        }
         declaration->parent_ops->wined3d_object_destroyed(declaration->parent);
         wined3d_cs_emit_destroy_object(declaration->device->cs,
                 wined3d_vertex_declaration_destroy_object, declaration);
-- 
2.8.0

