From c41a1ed5fc68a6083416dd263e23fbd72544dfc5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefandoesinger@gmx.at>
Date: Sat, 7 May 2016 13:58:17 +0100
Subject: wined3d: Use surface_blt_ugly in swapchain_gl_present.

---
 dlls/wined3d/swapchain.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/dlls/wined3d/swapchain.c b/dlls/wined3d/swapchain.c
index b885fc7..16dd4c5 100644
--- a/dlls/wined3d/swapchain.c
+++ b/dlls/wined3d/swapchain.c
@@ -514,7 +514,8 @@ static void swapchain_gl_present(struct wined3d_swapchain *swapchain,
         RECT rect = {0, 0, logo_texture->resource.width, logo_texture->resource.height};
 
         /* Blit the logo into the upper left corner of the drawable. */
-        wined3d_texture_blt(back_buffer, 0, &rect, logo_texture, 0, &rect,
+        surface_blt_ugly(back_buffer->sub_resources[0].u.surface, &rect,
+                logo_texture->sub_resources[0].u.surface, &rect,
                 WINED3D_BLT_ALPHA_TEST, NULL, WINED3D_TEXF_POINT);
     }
 
@@ -541,8 +542,8 @@ static void swapchain_gl_present(struct wined3d_swapchain *swapchain,
         if (swapchain->desc.windowed)
             MapWindowPoints(NULL, swapchain->win_handle, (POINT *)&dst_rect, 2);
         if (wined3d_clip_blit(&clip_rect, &dst_rect, &src_rect))
-            wined3d_texture_blt(back_buffer, 0, &dst_rect,
-                    swapchain->device->cursor_texture, 0, &src_rect,
+            surface_blt_ugly(back_buffer->sub_resources[0].u.surface, &dst_rect,
+                    swapchain->device->cursor_texture->sub_resources[0].u.surface, &src_rect,
                     WINED3D_BLT_ALPHA_TEST, NULL, WINED3D_TEXF_POINT);
     }
 
-- 
2.9.0

