From 6e213eabd0b472260b57e0c4584e2804378c761a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Mon, 8 Sep 2014 14:57:37 +0200
Subject: wined3d: Don't discard new buffers.

---
 dlls/wined3d/buffer.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/wined3d/buffer.c b/dlls/wined3d/buffer.c
index 6b01a0a..365ff4a 100644
--- a/dlls/wined3d/buffer.c
+++ b/dlls/wined3d/buffer.c
@@ -1334,6 +1334,7 @@ static HRESULT buffer_init(struct wined3d_buffer *buffer, struct wined3d_device
         return hr;
     }
     buffer->buffer_type_hint = bind_hint;
+    buffer->ignore_discard = TRUE;
 
     TRACE("size %#x, usage %#x, format %s, memory @ %p, iface @ %p.\n", buffer->resource.size, buffer->resource.usage,
             debug_d3dformat(buffer->resource.format->id), buffer->resource.heap_memory, buffer);
-- 
2.8.0

