From 4dc96ab42f9d44017ed6510922c1a324a3e1f2bd Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Thu, 27 Aug 2015 23:43:08 +0200
Subject: wined3d: Destroy samplers through the command stream.

---
 dlls/wined3d/sampler.c | 22 ++++++++++++++--------
 1 file changed, 14 insertions(+), 8 deletions(-)

diff --git a/dlls/wined3d/sampler.c b/dlls/wined3d/sampler.c
index 865198b..8c73466 100644
--- a/dlls/wined3d/sampler.c
+++ b/dlls/wined3d/sampler.c
@@ -33,22 +33,28 @@ ULONG CDECL wined3d_sampler_incref(struct wined3d_sampler *sampler)
     return refcount;
 }
 
+static void wined3d_sampler_destroy_object(void *object)
+{
+    struct wined3d_sampler *sampler = object;
+    struct wined3d_context *context = context_acquire(sampler->device, NULL);
+    const struct wined3d_gl_info *gl_info = context->gl_info;
+
+    GL_EXTCALL(glDeleteSamplers(1, &sampler->name));
+    context_release(context);
+
+    HeapFree(GetProcessHeap(), 0, sampler);
+}
+
 ULONG CDECL wined3d_sampler_decref(struct wined3d_sampler *sampler)
 {
     ULONG refcount = InterlockedDecrement(&sampler->refcount);
-    const struct wined3d_gl_info *gl_info;
-    struct wined3d_context *context;
 
     TRACE("%p decreasing refcount to %u.\n", sampler, refcount);
 
     if (!refcount)
     {
-        context = context_acquire(sampler->device, NULL);
-        gl_info = context->gl_info;
-        GL_EXTCALL(glDeleteSamplers(1, &sampler->name));
-        context_release(context);
-
-        HeapFree(GetProcessHeap(), 0, sampler);
+        struct wined3d_device *device = sampler->device;
+        wined3d_cs_emit_destroy_object(device->cs, wined3d_sampler_destroy_object, sampler);
     }
 
     return refcount;
-- 
2.8.0

