From 44d9ed38c0388ba76f6bcfa79f003754564c5ed4 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Thu, 5 Nov 2015 16:39:01 +0100
Subject: wined3d: Destroy shader resource views through the CS.

---
 dlls/wined3d/view.c | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/dlls/wined3d/view.c b/dlls/wined3d/view.c
index dc68687..6dd7ae2 100644
--- a/dlls/wined3d/view.c
+++ b/dlls/wined3d/view.c
@@ -236,12 +236,6 @@ ULONG CDECL wined3d_shader_resource_view_decref(struct wined3d_shader_resource_v
     {
         struct wined3d_device *device = view->resource->device;
 
-        if (wined3d_settings.cs_multithreaded)
-        {
-            FIXME("Waiting for cs.\n");
-            device->cs->ops->finish(device->cs);
-        }
-
         /* Call wined3d_object_destroyed() before releasing the resource,
          * since releasing the resource may end up destroying the parent. */
         view->parent_ops->wined3d_object_destroyed(view->parent);
-- 
2.8.0

