From 51cdd61c99614b34131f59dfac9ac684f6001e61 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefandoesinger@gmx.at>
Date: Sat, 7 May 2016 22:16:43 +0100
Subject: wined3d: Remove the device_reset CS sync fixme.

---
 dlls/wined3d/device.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/dlls/wined3d/device.c b/dlls/wined3d/device.c
index 95c34b2..24c66a8 100644
--- a/dlls/wined3d/device.c
+++ b/dlls/wined3d/device.c
@@ -4543,7 +4543,6 @@ HRESULT CDECL wined3d_device_reset(struct wined3d_device *device,
 
     if (wined3d_settings.cs_multithreaded)
     {
-        FIXME("Waiting for cs.\n");
         wined3d_cs_emit_glfinish(device->cs);
         device->cs->ops->finish(device->cs);
     }
-- 
2.8.0

