From 83363d459920153f0f603199211a4af01dc1c930 Mon Sep 17 00:00:00 2001
From: Nils Kuhnhenn <nils@volafile.io>
Date: Fri, 5 Aug 2016 20:28:20 +0200
Subject: wined3d: Fix incorrect return value in wined3d_cs_exec_set_light.

---
 dlls/wined3d/cs.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index 3c8a439..7052c19 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -1974,7 +1974,10 @@ static UINT wined3d_cs_exec_set_light(struct wined3d_cs *cs, const void *data)
         TRACE("Adding new light\n");
         object = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(*object));
         if (!object)
-            return E_OUTOFMEMORY;
+        {
+            ERR("Out of memory!\n");
+            return sizeof(*op);
+        }
 
         list_add_head(&cs->state.light_map[hash_idx], &object->entry);
         object->glIndex = -1;
-- 
2.9.0

