From 7ca9013cbb089978d3a6cfa90d73ae37c8bc1df1 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 15 Dec 2016 05:21:00 +0100
Subject: wined3d: Fix occlusion query tests.

---
 dlls/wined3d/cs.c | 7 -------
 1 file changed, 7 deletions(-)

diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
index b480705..c55afad 100644
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -2286,13 +2286,6 @@ static UINT wined3d_cs_exec_query_issue(struct wined3d_cs *cs, const void *data)
         {
             list_add_tail(&cs->query_poll_list, &query->poll_list_entry);
         }
-        else if (!poll && !list_empty(&query->poll_list_entry))
-        {
-            /* Can happen if occlusion queries are restarted. This discards the old
-             * result, polling it could result in a GL error */
-            list_remove(&query->poll_list_entry);
-            list_init(&query->poll_list_entry);
-        }
         else if (op->flags & WINED3DISSUE_END)
         {
             /* Can happen when an occlusion query is ended without being started,
-- 
2.9.0

