From e366ff98587ac212a25fd28f9f26f18ca5f30382 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Fri, 28 Aug 2015 00:01:39 +0200
Subject: wined3d: Hack to reject unsupported color fills.

---
 dlls/wined3d/surface.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dlls/wined3d/surface.c b/dlls/wined3d/surface.c
index e318bf106..f815b65 100644
--- a/dlls/wined3d/surface.c
+++ b/dlls/wined3d/surface.c
@@ -5302,6 +5302,14 @@ HRESULT CDECL wined3d_surface_blt(struct wined3d_surface *dst_surface, const REC
         memset(&src_rect, 0, sizeof(src_rect));
     }
 
+    /* FIXME: We should select the blitter in the main thread, that way we can return an error if the blit
+     * is unsupported without duplicating all the checks... */
+    if (flags & WINEDDBLT_COLORFILL && (dst_surface->container->resource.format_flags & WINED3DFMT_FLAG_BLOCKS))
+    {
+        WARN("Block color fill, returning WINED3DERR_INVALIDCALL\n");
+        return WINED3DERR_INVALIDCALL;
+    }
+
     if (!fx || !(fx->dwDDFX))
         flags &= ~WINEDDBLT_DDFX;
 
-- 
2.5.1

