From 3c2959e90d46d734c2a4f97d3790deb8a8979153 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 4 Sep 2015 13:59:56 +0200
Subject: wined3d: Avoid unnecessary call to wined3d_resource_get_patch.

---
 dlls/wined3d/surface.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/dlls/wined3d/surface.c b/dlls/wined3d/surface.c
index c1413cc..b8be30d 100644
--- a/dlls/wined3d/surface.c
+++ b/dlls/wined3d/surface.c
@@ -2588,9 +2588,6 @@ static void read_from_framebuffer(struct wined3d_surface *surface,
     {
         /* glReadPixels returns the image upside down, and there is no way to prevent this.
          * Flip the lines in software. */
-        UINT pitch, slice_pitch;
-
-        wined3d_resource_get_pitch(&surface->resource, &pitch, &slice_pitch);
 
         if (!(row = HeapAlloc(GetProcessHeap(), 0, pitch)))
             goto error;
-- 
2.5.1

