From e8731fe8f894a10182df1f2f699a0f9f107062bb Mon Sep 17 00:00:00 2001
From: Michael Stefaniuc <mstefani@winehq.org>
Date: Fri, 20 Jul 2018 20:18:06 +0200
Subject: [PATCH] dpnet: Use the ARRAY_SIZE() macro

Signed-off-by: Henri Verbeet <hverbeet@codeweavers.com>
Signed-off-by: Michael Stefaniuc <mstefani@winehq.org>
---
 dlls/dpnet/address.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/dpnet/address.c b/dlls/dpnet/address.c
index 4d55caa47e..0bfa2bca03 100644
--- a/dlls/dpnet/address.c
+++ b/dlls/dpnet/address.c
@@ -150,7 +150,7 @@ static const char *debugstr_SP(const GUID *id) {
 
   if (!id) return "(null)";
 
-  for (i = 0; i < sizeof(guids)/sizeof(guids[0]); i++) {
+  for (i = 0; i < ARRAY_SIZE(guids); i++) {
     if (IsEqualGUID(id, guids[i].guid))
       return guids[i].name;
   }
-- 
2.14.4

