From ab5ec464cf9178ec1466bf6af36ba6584ceda982 Mon Sep 17 00:00:00 2001
From: Michael Stefaniuc <mstefani@winehq.org>
Date: Fri, 20 Jul 2018 20:31:59 +0200
Subject: [PATCH] d3d10: Use the ARRAY_SIZE() macro

Signed-off-by: Henri Verbeet <hverbeet@codeweavers.com>
Signed-off-by: Michael Stefaniuc <mstefani@winehq.org>
---
 dlls/d3d10/effect.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/d3d10/effect.c b/dlls/d3d10/effect.c
index 10d8404559..cd6f4a3e1d 100644
--- a/dlls/d3d10/effect.c
+++ b/dlls/d3d10/effect.c
@@ -1139,7 +1139,7 @@ static const struct d3d10_effect_state_property_info *get_property_info(UINT id)
 {
     unsigned int i;
 
-    for (i = 0; i < sizeof(property_info) / sizeof(*property_info); ++i)
+    for (i = 0; i < ARRAY_SIZE(property_info); ++i)
     {
         if (property_info[i].id == id)
             return &property_info[i];
@@ -1152,7 +1152,7 @@ static const struct d3d10_effect_state_storage_info *get_storage_info(D3D_SHADER
 {
     unsigned int i;
 
-    for (i = 0; i < sizeof(d3d10_effect_state_storage_info) / sizeof(*d3d10_effect_state_storage_info); ++i)
+    for (i = 0; i < ARRAY_SIZE(d3d10_effect_state_storage_info); ++i)
     {
         if (d3d10_effect_state_storage_info[i].id == id)
             return &d3d10_effect_state_storage_info[i];
-- 
2.14.4

