From a12472814a32462fd5e1b23efe0bbae33b76ee1c Mon Sep 17 00:00:00 2001
From: Michael Stefaniuc <mstefani@winehq.org>
Date: Mon, 23 Jul 2018 08:58:41 +0200
Subject: [PATCH] browseui: Use the ARRAY_SIZE() macro

Signed-off-by: Michael Stefaniuc <mstefani@winehq.org>
---
 dlls/browseui/progressdlg.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/browseui/progressdlg.c b/dlls/browseui/progressdlg.c
index 7cba7d6464..2355b0cfa0 100644
--- a/dlls/browseui/progressdlg.c
+++ b/dlls/browseui/progressdlg.c
@@ -114,7 +114,7 @@ static LPWSTR load_string(HINSTANCE hInstance, UINT uiResourceId)
     WCHAR string[256];
     LPWSTR ret;
 
-    LoadStringW(hInstance, uiResourceId, string, sizeof(string)/sizeof(string[0]));
+    LoadStringW(hInstance, uiResourceId, string, ARRAY_SIZE(string));
     ret = HeapAlloc(GetProcessHeap(), 0, (strlenW(string) + 1) * sizeof(WCHAR));
     strcpyW(ret, string);
     return ret;
-- 
2.14.4

