From 26098afa5665e6e3eb70d4560e2059ce97ae3a96 Mon Sep 17 00:00:00 2001
From: Michael Stefaniuc <mstefani@winehq.org>
Date: Mon, 23 Jul 2018 09:18:52 +0200
Subject: [PATCH] itss: Remove a superfluous variable initialization

Signed-off-by: Michael Stefaniuc <mstefani@winehq.org>
---
 dlls/itss/moniker.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/itss/moniker.c b/dlls/itss/moniker.c
index 0509c9e1a5..3315b914d3 100644
--- a/dlls/itss/moniker.c
+++ b/dlls/itss/moniker.c
@@ -287,7 +287,7 @@ static HRESULT WINAPI ITS_IMonikerImpl_GetDisplayName(
     ITS_IMonikerImpl *This = impl_from_IMoniker(iface);
     static const WCHAR szFormat[] = {
         'm','s','-','i','t','s',':','%','s',':',':','%','s',0 };
-    DWORD len = sizeof szFormat / sizeof(WCHAR);
+    DWORD len;
     LPWSTR str;
 
     TRACE("%p %p %p %p\n", iface, pbc, pmkToLeft, ppszDisplayName);
-- 
2.14.4

