From a11a1464b49a115d73c8b787ad2e55327f668c5e Mon Sep 17 00:00:00 2001
From: Michael Stefaniuc <mstefani@winehq.org>
Date: Mon, 23 Jul 2018 16:40:31 +0200
Subject: [PATCH] icinfo: Use the ARRAY_SIZE() macro

Signed-off-by: Michael Stefaniuc <mstefani@winehq.org>
---
 programs/icinfo/icinfo.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/programs/icinfo/icinfo.c b/programs/icinfo/icinfo.c
index 601ec8c8b3..1059183361 100644
--- a/programs/icinfo/icinfo.c
+++ b/programs/icinfo/icinfo.c
@@ -33,7 +33,7 @@ static int mywprintf(const WCHAR *format, ...)
     HANDLE              hout = GetStdHandle(STD_OUTPUT_HANDLE);
 
     va_start(parms, format);
-    vsnprintfW(output_bufW, sizeof(output_bufW)/sizeof(output_bufW[0]), format, parms);
+    vsnprintfW(output_bufW, ARRAY_SIZE(output_bufW), format, parms);
     va_end(parms);
 
     /* Try to write as unicode whenever we think it's a console */
-- 
2.14.4

