From 3c68cac0e75c47c9bc0c3e0265d37d42668af0c7 Mon Sep 17 00:00:00 2001
From: Michael Stefaniuc <mstefani@winehq.org>
Date: Mon, 23 Jul 2018 19:19:29 +0200
Subject: [PATCH] msiexec: Use the ARRAY_SIZE() macro

Signed-off-by: Michael Stefaniuc <mstefani@winehq.org>
---
 programs/msiexec/msiexec.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/programs/msiexec/msiexec.c b/programs/msiexec/msiexec.c
index 5a471a2184..47cef7e0fa 100644
--- a/programs/msiexec/msiexec.c
+++ b/programs/msiexec/msiexec.c
@@ -71,11 +71,11 @@ static void ShowUsage(int ExitCode)
 
     /* MsiGetFileVersion need the full path */
     *filename = 0;
-    res = GetModuleFileNameW(hmsi, filename, sizeof(filename) / sizeof(filename[0]));
+    res = GetModuleFileNameW(hmsi, filename, ARRAY_SIZE(filename));
     if (!res)
         WINE_ERR("GetModuleFileName failed: %d\n", GetLastError());
 
-    len = sizeof(msiexec_version) / sizeof(msiexec_version[0]);
+    len = ARRAY_SIZE(msiexec_version);
     *msiexec_version = 0;
     res = MsiGetFileVersionW(filename, msiexec_version, &len, NULL, NULL);
     if (res)
-- 
2.14.4

