From 1c97d42e6271b8f81a3561d2480a4877014d5fd3 Mon Sep 17 00:00:00 2001
From: Michael Stefaniuc <mstefani@winehq.org>
Date: Tue, 24 Jul 2018 15:35:50 +0200
Subject: [PATCH] uninstaller: Use the ARRAY_SIZE() macro

Signed-off-by: Michael Stefaniuc <mstefani@winehq.org>
---
 programs/uninstaller/main.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/programs/uninstaller/main.c b/programs/uninstaller/main.c
index 736593e27d..142d3d16f7 100644
--- a/programs/uninstaller/main.c
+++ b/programs/uninstaller/main.c
@@ -102,7 +102,7 @@ static void WINAPIV output_message(unsigned int id, ...)
     WCHAR fmt[1024];
     __ms_va_list va_args;
 
-    if (!LoadStringW(GetModuleHandleW(NULL), id, fmt, sizeof(fmt)/sizeof(fmt[0])))
+    if (!LoadStringW(GetModuleHandleW(NULL), id, fmt, ARRAY_SIZE(fmt)))
     {
         WINE_FIXME("LoadString failed with %d\n", GetLastError());
         return;
@@ -342,8 +342,8 @@ static void UninstallProgram(void)
             WCHAR sUninstallFailed[MAX_STRING_LEN];
             HINSTANCE hInst = GetModuleHandleW(0);
 
-            LoadStringW(hInst, IDS_APPNAME, sAppName, sizeof(sAppName)/sizeof(WCHAR));
-            LoadStringW(hInst, IDS_UNINSTALLFAILED, sUninstallFailed, sizeof(sUninstallFailed)/sizeof(WCHAR));
+            LoadStringW(hInst, IDS_APPNAME, sAppName, ARRAY_SIZE(sAppName));
+            LoadStringW(hInst, IDS_UNINSTALLFAILED, sUninstallFailed, ARRAY_SIZE(sUninstallFailed));
             wsprintfW(errormsg, sUninstallFailed, entries[i].command);
             if(MessageBoxW(0, errormsg, sAppName, MB_YESNO | MB_ICONQUESTION)==IDYES)
             {
-- 
2.14.4

