From 34a20bf87deab524a00c1be9d60a6f78306359b4 Mon Sep 17 00:00:00 2001
From: Michael Stefaniuc <mstefani@winehq.org>
Date: Tue, 24 Jul 2018 18:56:16 +0200
Subject: [PATCH] mountmgr.sys: Use the ARRAY_SIZE() macro

Signed-off-by: Michael Stefaniuc <mstefani@winehq.org>
---
 dlls/mountmgr.sys/device.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/mountmgr.sys/device.c b/dlls/mountmgr.sys/device.c
index 1437bfc853..f96b05e4da 100644
--- a/dlls/mountmgr.sys/device.c
+++ b/dlls/mountmgr.sys/device.c
@@ -658,7 +658,7 @@ static void create_drive_devices(void)
             if (!RegQueryValueExW( drives_key, driveW, NULL, &type, (BYTE *)buffer, &size ) &&
                 type == REG_SZ)
             {
-                for (j = 0; j < sizeof(drive_types)/sizeof(drive_types[0]); j++)
+                for (j = 0; j < ARRAY_SIZE(drive_types); j++)
                     if (drive_types[j][0] && !strcmpiW( buffer, drive_types[j] ))
                     {
                         drive_type = j;
@@ -1123,7 +1123,7 @@ static void create_port_devices( DRIVER_OBJECT *driver )
         p[1] = 'o';
         p[2] = 'm';
         search_paths = serial_search_paths;
-        num_search_paths = sizeof(serial_search_paths)/sizeof(serial_search_paths[0]);
+        num_search_paths = ARRAY_SIZE(serial_search_paths);
         windows_ports_key_name = serialcomm_keyW;
     }
     else
@@ -1132,7 +1132,7 @@ static void create_port_devices( DRIVER_OBJECT *driver )
         p[1] = 'p';
         p[2] = 't';
         search_paths = parallel_search_paths;
-        num_search_paths = sizeof(parallel_search_paths)/sizeof(parallel_search_paths[0]);
+        num_search_paths = ARRAY_SIZE(parallel_search_paths);
         windows_ports_key_name = parallel_ports_keyW;
     }
     p += 3;
-- 
2.14.4

