From 5675688cff4eb773a4b531423f72267823caec3b Mon Sep 17 00:00:00 2001
From: Michael Stefaniuc <mstefani@winehq.org>
Date: Wed, 25 Jul 2018 19:17:44 +0200
Subject: [PATCH] scrobj: Use the ARRAY_SIZE() macro

Signed-off-by: Michael Stefaniuc <mstefani@winehq.org>
---
 dlls/scrobj/scrobj.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/scrobj/scrobj.c b/dlls/scrobj/scrobj.c
index 3e6535a607..989a648370 100644
--- a/dlls/scrobj/scrobj.c
+++ b/dlls/scrobj/scrobj.c
@@ -102,7 +102,7 @@ static void release_typelib(void)
     if (!typelib)
         return;
 
-    for (i = 0; i < sizeof(typeinfos)/sizeof(*typeinfos); i++)
+    for (i = 0; i < ARRAY_SIZE(typeinfos); i++)
         if (typeinfos[i])
             ITypeInfo_Release(typeinfos[i]);
 
@@ -369,7 +369,7 @@ static HRESULT WINAPI scriptlet_typelib_get_GUID(IGenScriptletTLib *iface, BSTR
         if (FAILED(hr))
             return hr;
 
-        hr = StringFromGUID2(&guid, guidW, sizeof(guidW)/sizeof(guidW[0]));
+        hr = StringFromGUID2(&guid, guidW, ARRAY_SIZE(guidW));
         if (FAILED(hr))
             return hr;
 
-- 
2.14.4

