From ed22fdf12f19983647bfd60230dc1325998f95fd Mon Sep 17 00:00:00 2001
From: Michael Stefaniuc <mstefani@winehq.org>
Date: Mon, 30 Jul 2018 21:18:08 +0200
Subject: [PATCH] inetmib1: Use the ARRAY_SIZE() macro

Signed-off-by: Michael Stefaniuc <mstefani@winehq.org>
---
 dlls/inetmib1/main.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/inetmib1/main.c b/dlls/inetmib1/main.c
index c0719c1..2f7b9ca 100644
--- a/dlls/inetmib1/main.c
+++ b/dlls/inetmib1/main.c
@@ -1321,7 +1321,7 @@ BOOL WINAPI SnmpExtensionInit(DWORD dwUptimeReference,
         pFirstSupportedRegion);
 
     minSupportedIDLength = UINT_MAX;
-    for (i = 0; i < sizeof(supportedIDs) / sizeof(supportedIDs[0]); i++)
+    for (i = 0; i < ARRAY_SIZE(supportedIDs); i++)
     {
         if (supportedIDs[i].init)
             supportedIDs[i].init();
@@ -1337,7 +1337,7 @@ static void cleanup(void)
 {
     UINT i;
 
-    for (i = 0; i < sizeof(supportedIDs) / sizeof(supportedIDs[0]); i++)
+    for (i = 0; i < ARRAY_SIZE(supportedIDs); i++)
         if (supportedIDs[i].cleanup)
             supportedIDs[i].cleanup();
 }
-- 
1.9.1

