From 3508dde97fce11ef0b8ad570b1305aacec34db2e Mon Sep 17 00:00:00 2001
From: Michael Stefaniuc <mstefani@winehq.org>
Date: Mon, 30 Jul 2018 21:18:10 +0200
Subject: [PATCH] mciwave: Use the ARRAY_SIZE() macro

Signed-off-by: Michael Stefaniuc <mstefani@winehq.org>
---
 dlls/mciwave/mciwave.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/mciwave/mciwave.c b/dlls/mciwave/mciwave.c
index fd4d6f6..0fa9b20 100644
--- a/dlls/mciwave/mciwave.c
+++ b/dlls/mciwave/mciwave.c
@@ -427,7 +427,7 @@ static DWORD create_tmp_file(HMMIO* hFile, LPWSTR* pszTmpFileName)
     szPrefix[2] = 'I';
     szPrefix[3] = '\0';
 
-    if (!GetTempPathW(sizeof(szTmpPath)/sizeof(szTmpPath[0]), szTmpPath)) {
+    if (!GetTempPathW(ARRAY_SIZE(szTmpPath), szTmpPath)) {
         WARN("can't retrieve temp path!\n");
         *pszTmpFileName = NULL;
         return MCIERR_FILE_NOT_FOUND;
-- 
1.9.1

