From e8135dad94110944af7a599dc1310bcbd96eb548 Mon Sep 17 00:00:00 2001
From: Michael Stefaniuc <mstefani@winehq.org>
Date: Mon, 30 Jul 2018 21:18:12 +0200
Subject: [PATCH] mscms: Use the ARRAY_SIZE() macro

Signed-off-by: Michael Stefaniuc <mstefani@winehq.org>
---
 dlls/mscms/profile.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/mscms/profile.c b/dlls/mscms/profile.c
index f0c4087..cb4908c 100644
--- a/dlls/mscms/profile.c
+++ b/dlls/mscms/profile.c
@@ -295,7 +295,7 @@ BOOL WINAPI GetColorDirectoryW( PCWSTR machine, PWSTR buffer, PDWORD size )
 
     if (machine || !size) return FALSE;
 
-    GetSystemDirectoryW( colordir, sizeof(colordir) / sizeof(WCHAR) );
+    GetSystemDirectoryW( colordir, ARRAY_SIZE( colordir ));
     lstrcatW( colordir, colorsubdir );
 
     len = lstrlenW( colordir ) * sizeof(WCHAR);
-- 
1.9.1

