From d6b9f1cef89956e4cc7ddd3e01d843cf9da2f770 Mon Sep 17 00:00:00 2001
From: Michael Stefaniuc <mstefani@winehq.org>
Date: Mon, 30 Jul 2018 21:18:13 +0200
Subject: [PATCH] msctf: Use the ARRAY_SIZE() macro

Signed-off-by: Michael Stefaniuc <mstefani@winehq.org>
---
 dlls/msctf/documentmgr.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/msctf/documentmgr.c b/dlls/msctf/documentmgr.c
index cf4c601..9fa6b0f 100644
--- a/dlls/msctf/documentmgr.c
+++ b/dlls/msctf/documentmgr.c
@@ -188,7 +188,7 @@ static HRESULT WINAPI DocumentMgr_Pop(ITfDocumentMgr *iface, DWORD dwFlags)
     {
         int i;
 
-        for (i = 0; i < sizeof(This->contextStack)/sizeof(This->contextStack[0]); i++)
+        for (i = 0; i < ARRAY_SIZE(This->contextStack); i++)
             if (This->contextStack[i])
             {
                 ITfThreadMgrEventSink_OnPopContext(This->ThreadMgrSink, This->contextStack[i]);
-- 
1.9.1

