From 7c0c2b354c261c898ac9acfea575cc022a1bbb34 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 27 Mar 2015 08:15:14 +0100
Subject: ntoskrnl: Avoid repeated FIXME messages in
 PsLookupProcessByProcessId.

---
 dlls/ntoskrnl.exe/ntoskrnl.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/ntoskrnl.exe/ntoskrnl.c b/dlls/ntoskrnl.exe/ntoskrnl.c
index 95dcdc9..bb26424 100644
--- a/dlls/ntoskrnl.exe/ntoskrnl.c
+++ b/dlls/ntoskrnl.exe/ntoskrnl.c
@@ -1885,7 +1885,8 @@ NTSTATUS WINAPI PsSetLoadImageNotifyRoutine(PLOAD_IMAGE_NOTIFY_ROUTINE routine)
  */
 NTSTATUS WINAPI PsLookupProcessByProcessId(HANDLE processid, PEPROCESS *process)
 {
-    FIXME("(%p %p) stub\n", processid, process);
+    static int once;
+    if (!once++) FIXME("(%p %p) stub\n", processid, process);
     return STATUS_NOT_IMPLEMENTED;
 }
 
-- 
2.3.3

