From a849696bf25a87a84a565d8818fa1d4f5b1ebde3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Fri, 15 May 2015 19:55:13 +0200
Subject: wininet: Fix memory leak by not calling get_cookie_header twice.

---
 dlls/wininet/http.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/dlls/wininet/http.c b/dlls/wininet/http.c
index cd2cdcf..45183cf 100644
--- a/dlls/wininet/http.c
+++ b/dlls/wininet/http.c
@@ -4244,7 +4244,6 @@ static void HTTP_InsertCookies(http_request_t *request)
     if(res != ERROR_SUCCESS || !cookies)
         return;
 
-    get_cookie_header(request->server->name, request->path, &cookies);
     HTTP_HttpAddRequestHeadersW(request, cookies, strlenW(cookies), HTTP_ADDREQ_FLAG_REPLACE);
     heap_free(cookies);
 }
-- 
2.4.0

