From 616ad4e053abdc48810064993caef067a3452587 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <zfigura@codeweavers.com>
Date: Mon, 8 Jul 2019 11:24:39 -0500
Subject: [PATCH] strmbase: Remove unused wine/unicode.h import.

Signed-off-by: Zebediah Figura <zfigura@codeweavers.com>
---
 dlls/strmbase/strmbase_private.h | 1 -
 1 file changed, 1 deletion(-)

diff --git a/dlls/strmbase/strmbase_private.h b/dlls/strmbase/strmbase_private.h
index 439ab9aefa..06fbe62a42 100644
--- a/dlls/strmbase/strmbase_private.h
+++ b/dlls/strmbase/strmbase_private.h
@@ -31,7 +31,6 @@
 #include "wine/heap.h"
 #include "wine/list.h"
 #include "wine/strmbase.h"
-#include "wine/unicode.h"
 
 /* Quality Control */
 typedef struct QualityControlImpl {
-- 
2.17.1

