From b263d3d0958c18b2b8f105b87c27f9cbee176503 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <zfigura@codeweavers.com>
Date: Mon, 8 Jul 2019 11:24:43 -0500
Subject: [PATCH] qedit: Build with msvcrt.

Signed-off-by: Zebediah Figura <zfigura@codeweavers.com>
---
 dlls/qedit/Makefile.in | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/qedit/Makefile.in b/dlls/qedit/Makefile.in
index b47ee260ab..4891328c7a 100644
--- a/dlls/qedit/Makefile.in
+++ b/dlls/qedit/Makefile.in
@@ -1,6 +1,8 @@
 MODULE    = qedit.dll
 IMPORTS   = strmiids strmbase uuid oleaut32 ole32 advapi32
 
+EXTRADLLFLAGS = -mno-cygwin
+
 C_SRCS = \
 	main.c \
 	mediadet.c \
-- 
2.17.1

