From 4018f0ee2e502ded20bc907e3baa47ca348a61b8 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <zfigura@codeweavers.com>
Date: Mon, 8 Jul 2019 11:24:42 -0500
Subject: [PATCH] qedit: Share source with strmbase.

Signed-off-by: Zebediah Figura <zfigura@codeweavers.com>
---
 dlls/qedit/Makefile.in | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/dlls/qedit/Makefile.in b/dlls/qedit/Makefile.in
index 4891328c7a..c5bf8f9cac 100644
--- a/dlls/qedit/Makefile.in
+++ b/dlls/qedit/Makefile.in
@@ -1,12 +1,21 @@
 MODULE    = qedit.dll
-IMPORTS   = strmiids strmbase uuid oleaut32 ole32 advapi32
+IMPORTS   = strmiids uuid oleaut32 ole32 advapi32
+PARENTSRC = ../strmbase
 
 EXTRADLLFLAGS = -mno-cygwin
 
 C_SRCS = \
+	dispatch.c \
+	enumpins.c \
+	filter.c \
 	main.c \
 	mediadet.c \
+	mediatype.c \
 	nullrenderer.c \
+	pin.c \
+	pospass.c \
+	qualitycontrol.c \
+	renderer.c \
 	samplegrabber.c \
 	timeline.c
 
-- 
2.17.1

