From 29294a8ce4ae5114396464e8aa177e267881e734 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <zfigura@codeweavers.com>
Date: Mon, 8 Jul 2019 11:24:40 -0500
Subject: [PATCH] evr: Share source with strmbase.

Signed-off-by: Zebediah Figura <zfigura@codeweavers.com>
---
 dlls/evr/Makefile.in | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/dlls/evr/Makefile.in b/dlls/evr/Makefile.in
index 053d80e61d..e605ce227c 100644
--- a/dlls/evr/Makefile.in
+++ b/dlls/evr/Makefile.in
@@ -1,10 +1,19 @@
 MODULE    = evr.dll
-IMPORTS   = mfuuid strmiids strmbase uuid ole32 oleaut32
+IMPORTS   = mfuuid strmiids uuid ole32 oleaut32
+PARENTSRC = ../strmbase
 
 EXTRADLLFLAGS = -mno-cygwin
 
 C_SRCS = \
+	dispatch.c \
+	enumpins.c \
 	evr.c \
-	main.c
+	filter.c \
+	main.c \
+	mediatype.c \
+	pin.c \
+	pospass.c \
+	qualitycontrol.c \
+	renderer.c
 
 IDL_SRCS = evr_classes.idl
-- 
2.17.1

