From 79c6b2c90c4935a648db1320fb0c354ca3adc138 Mon Sep 17 00:00:00 2001
From: Jeff Smith <whydoubt@gmail.com>
Date: Mon, 8 Jul 2019 14:40:23 -0500
Subject: [PATCH] d3drm: Correct paramater count of several interface defines.

Signed-off-by: Henri Verbeet <hverbeet@codeweavers.com>
---
 include/d3drm.h    | 16 ++++++++--------
 include/d3drmobj.h |  6 +++---
 2 files changed, 11 insertions(+), 11 deletions(-)

diff --git a/include/d3drm.h b/include/d3drm.h
index 126695cea1..c10b9ed18c 100644
--- a/include/d3drm.h
+++ b/include/d3drm.h
@@ -322,7 +322,7 @@ DECLARE_INTERFACE_(IDirect3DRM2,IUnknown)
 #define IDirect3DRM2_EnumerateObjects(p,a,b)                       (p)->EnumerateObjects(a,b)
 #define IDirect3DRM2_Load(p,a,b,c,d,e,f,g,h,i,j)                   (p)->Load(a,b,c,d,e,f,g,h,i,j)
 #define IDirect3DRM2_Tick(p,a)                                     (p)->Tick(a)
-#define IDirect3DRM2_CreateProgressiveMesh(p,a)                    (p)->CreateProgressiveMesh(p,a)
+#define IDirect3DRM2_CreateProgressiveMesh(p,a)                    (p)->CreateProgressiveMesh(a)
 #endif
 
 /*****************************************************************************
@@ -434,7 +434,7 @@ DECLARE_INTERFACE_(IDirect3DRM3,IUnknown)
 #define IDirect3DRM3_CreateProgressiveMesh(p,a)                    (p)->lpVtbl->CreateProgressiveMesh(p,a)
 #define IDirect3DRM3_RegisterClient(p,a,b)                         (p)->lpVtbl->RegisterClient(p,a,b)
 #define IDirect3DRM3_UnregisterClient(p,a)                         (p)->lpVtbl->UnregisterClient(p,a)
-#define IDirect3DRM3_CreateClippedVisual(p,ab)                     (p)->lpVtbl->CreateClippedVisual(p,a,b)
+#define IDirect3DRM3_CreateClippedVisual(p,a,b)                    (p)->lpVtbl->CreateClippedVisual(p,a,b)
 #define IDirect3DRM3_SetOptions(p,a)                               (p)->lpVtbl->SetOptions(p,a)
 #define IDirect3DRM3_GetOptions(p,a)                               (p)->lpVtbl->GetOptions(p,a)
 #else
@@ -475,12 +475,12 @@ DECLARE_INTERFACE_(IDirect3DRM3,IUnknown)
 #define IDirect3DRM3_EnumerateObjects(p,a,b)                       (p)->EnumerateObjects(a,b)
 #define IDirect3DRM3_Load(p,a,b,c,d,e,f,g,h,i,j)                   (p)->Load(a,b,c,d,e,f,g,h,i,j)
 #define IDirect3DRM3_Tick(p,a)                                     (p)->Tick(a)
-#define IDirect3DRM3_CreateProgressiveMesh(p,a)                    (p)->CreateProgressiveMesh(p,a)
-#define IDirect3DRM3_RegisterClient(p,a,b)                         (p)->RegisterClient(p,a,b)
-#define IDirect3DRM3_UnregisterClient(p,a)                         (p)->UnregisterClient(p,a)
-#define IDirect3DRM3_CreateClippedVisual(p,ab)                     (p)->CreateClippedVisual(p,a,b)
-#define IDirect3DRM3_SetOptions(p,a)                               (p)->SetOptions(p,a)
-#define IDirect3DRM3_GetOptions(p,a)                               (p)->GetOptions(p,a)
+#define IDirect3DRM3_CreateProgressiveMesh(p,a)                    (p)->CreateProgressiveMesh(a)
+#define IDirect3DRM3_RegisterClient(p,a,b)                         (p)->RegisterClient(a,b)
+#define IDirect3DRM3_UnregisterClient(p,a)                         (p)->UnregisterClient(a)
+#define IDirect3DRM3_CreateClippedVisual(p,a,b)                    (p)->CreateClippedVisual(a,b)
+#define IDirect3DRM3_SetOptions(p,a)                               (p)->SetOptions(a)
+#define IDirect3DRM3_GetOptions(p,a)                               (p)->GetOptions(a)
 #endif
 
 #define D3DRM_OK                        DD_OK
diff --git a/include/d3drmobj.h b/include/d3drmobj.h
index 5ff33fea01..7e48d9891d 100644
--- a/include/d3drmobj.h
+++ b/include/d3drmobj.h
@@ -3616,9 +3616,9 @@ DECLARE_INTERFACE_(IDirect3DRMWrap, IDirect3DRMObject)
 #define IDirect3DRMWrap_GetName(p,a,b)                          (p)->GetName(a,b)
 #define IDirect3DRMWrap_GetClassName(p,a,b)                     (p)->GetClassName(a,b)
 /*** IDirect3DRMWrap methods ***/
-#define IDirect3DRMWrap_Init(p,a,b,c,d,e,f,g,h,i,j,k,l,m,n,o)   (p)->Init(p,a,b,c,d,e,f,g,h,i,j,k,l,m,n,o)
-#define IDirect3DRMWrap_Apply(p,a)                              (p)->Apply(p,a)
-#define IDirect3DRMWrap_ApplyRelative(p,a,b)                    (p)->ApplyRelative(p,a,b)
+#define IDirect3DRMWrap_Init(p,a,b,c,d,e,f,g,h,i,j,k,l,m,n,o)   (p)->Init(a,b,c,d,e,f,g,h,i,j,k,l,m,n,o)
+#define IDirect3DRMWrap_Apply(p,a)                              (p)->Apply(a)
+#define IDirect3DRMWrap_ApplyRelative(p,a,b)                    (p)->ApplyRelative(a,b)
 #endif
 
 /*****************************************************************************
-- 
2.17.1

