From 008f23dc9713a42a47b68cad54ce3574ef398afb Mon Sep 17 00:00:00 2001
From: Zebediah Figura <z.figura12@gmail.com>
Date: Thu, 18 Jul 2019 18:29:38 -0500
Subject: [PATCH 2/2] services: Fix the arguments to is_root_pnp_service().

Signed-off-by: Zebediah Figura <z.figura12@gmail.com>
---
 programs/services/services.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/programs/services/services.c b/programs/services/services.c
index 67cd4fa855..ec2a8eb05c 100644
--- a/programs/services/services.c
+++ b/programs/services/services.c
@@ -424,7 +424,7 @@ static BOOL schedule_delayed_autostart(struct service_entry **services, unsigned
     return TRUE;
 }
 
-static BOOL is_root_pnp_service(const struct service_entry *service, HDEVINFO set)
+static BOOL is_root_pnp_service(HDEVINFO set, const struct service_entry *service)
 {
     SP_DEVINFO_DATA device = {sizeof(device)};
     WCHAR name[MAX_SERVICE_NAME];
-- 
2.17.1

