From b8177d6adbfeae337189d14a680025ac6277735b Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 26 Oct 2013 19:14:00 +0200
Subject: kernel32/tests: Fix tests compilation

---
 dlls/kernel32/tests/directory.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/kernel32/tests/directory.c b/dlls/kernel32/tests/directory.c
index a3a9580..a7c30ee 100644
--- a/dlls/kernel32/tests/directory.c
+++ b/dlls/kernel32/tests/directory.c
@@ -645,7 +645,7 @@ done:
 
 void init(void)
 {
-    HMODULE hmod = GetModuleHandle("advapi32.dll");
+    HMODULE hmod = GetModuleHandleA("advapi32.dll");
 
     pGetNamedSecurityInfoA = (void *)GetProcAddress(hmod, "GetNamedSecurityInfoA");
     pAddAccessAllowedAceEx = (void *)GetProcAddress(hmod, "AddAccessAllowedAceEx");
-- 
1.7.9.5

