From d65c501e946415d47bb6fc7f084233ac8de43dec Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 26 Sep 2015 17:21:22 +0200
Subject: d3d8: Handle D3DSWAPEFFECT_COPY_VSYNC swapchain effect like
 D3DSWAPEFFECT_COPY.

---
 dlls/d3d8/device.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/d3d8/device.c b/dlls/d3d8/device.c
index e899b34..d006c12 100644
--- a/dlls/d3d8/device.c
+++ b/dlls/d3d8/device.c
@@ -191,13 +191,13 @@ static void present_parameters_from_wined3d_swapchain_desc(D3DPRESENT_PARAMETERS
 static BOOL wined3d_swapchain_desc_from_present_parameters(struct wined3d_swapchain_desc *swapchain_desc,
         const D3DPRESENT_PARAMETERS *present_parameters)
 {
-    if (!present_parameters->SwapEffect || present_parameters->SwapEffect > D3DSWAPEFFECT_COPY)
+    if (!present_parameters->SwapEffect || present_parameters->SwapEffect > D3DSWAPEFFECT_COPY_VSYNC)
     {
         WARN("Invalid swap effect %u passed.\n", present_parameters->SwapEffect);
         return FALSE;
     }
     if (present_parameters->BackBufferCount > 3
-            || (present_parameters->SwapEffect == D3DSWAPEFFECT_COPY
+            || (present_parameters->SwapEffect >= D3DSWAPEFFECT_COPY
             && present_parameters->BackBufferCount > 1))
     {
         WARN("Invalid backbuffer count %u.\n", present_parameters->BackBufferCount);
-- 
2.5.1

