From a4e0737b05a26a85215c06c5c9fe615353540e50 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 13 Sep 2014 01:35:08 +0200
Subject: kernel32: Ignore FIND_FIRST_EX_LARGE_FETCH flag in FindFirstFileExW.

---
 dlls/kernel32/file.c | 3 +++
 include/winbase.h    | 3 +++
 2 files changed, 6 insertions(+)

diff --git a/dlls/kernel32/file.c b/dlls/kernel32/file.c
index 412659b..b277fb9 100644
--- a/dlls/kernel32/file.c
+++ b/dlls/kernel32/file.c
@@ -1859,6 +1859,9 @@ HANDLE WINAPI FindFirstFileExW( LPCWSTR filename, FINDEX_INFO_LEVELS level,
 
     TRACE("%s %d %p %d %p %x\n", debugstr_w(filename), level, data, search_op, filter, flags);
 
+    /* ignore FIND_FIRST_EX_LARGE_FETCH, only a hint for performance optimization */
+    flags &= ~FIND_FIRST_EX_LARGE_FETCH;
+
     if ((search_op != FindExSearchNameMatch && search_op != FindExSearchLimitToDirectories)
 	|| flags != 0)
     {
diff --git a/include/winbase.h b/include/winbase.h
index dd5e816..edd6ad6 100644
--- a/include/winbase.h
+++ b/include/winbase.h
@@ -289,6 +289,9 @@ typedef enum _FINDEX_INFO_LEVELS
 	FindExInfoMaxInfoLevel
 } FINDEX_INFO_LEVELS;
 
+#define FIND_FIRST_EX_CASE_SENSITIVE 1
+#define FIND_FIRST_EX_LARGE_FETCH    2
+
 typedef enum _FINDEX_SEARCH_OPS
 {
 	FindExSearchNameMatch,
-- 
2.1.0

