From d72d7f4a70a9715458b4cbd60c5a951e74770448 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 27 Aug 2015 01:55:42 +0200
Subject: wbemprox: Set bios name in Win32_BIOS.

---
 dlls/wbemprox/builtin.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/dlls/wbemprox/builtin.c b/dlls/wbemprox/builtin.c
index 1c960eb..8f51453 100644
--- a/dlls/wbemprox/builtin.c
+++ b/dlls/wbemprox/builtin.c
@@ -377,7 +377,8 @@ static const struct column col_bios[] =
     { prop_releasedateW,        CIM_DATETIME },
     { prop_serialnumberW,       CIM_STRING },
     { prop_smbiosbiosversionW,  CIM_STRING },
-    { prop_versionW,            CIM_STRING|COL_FLAG_KEY }
+    { prop_versionW,            CIM_STRING|COL_FLAG_KEY },
+    { prop_nameW,               CIM_STRING }
 };
 static const struct column col_cdromdrive[] =
 {
@@ -666,6 +667,8 @@ static const WCHAR bios_smbiosbiosversionW[] =
     {'W','i','n','e',0};
 static const WCHAR bios_versionW[] =
     {'W','I','N','E',' ',' ',' ','-',' ','1',0};
+static const WCHAR bios_nameW[] =
+    {'W','I','N','E',' ','B','I','O','S',0};
 static const WCHAR cdromdrive_mediatypeW[] =
     {'C','D','-','R','O','M',0};
 static const WCHAR cdromdrive_nameW[] =
@@ -765,6 +768,7 @@ struct record_bios
     const WCHAR *serialnumber;
     const WCHAR *smbiosbiosversion;
     const WCHAR *version;
+    const WCHAR *name;
 };
 struct record_cdromdrive
 {
@@ -1041,7 +1045,7 @@ static const struct record_baseboard data_baseboard[] =
 static const struct record_bios data_bios[] =
 {
     { bios_descriptionW, bios_descriptionW, bios_manufacturerW, bios_releasedateW, bios_serialnumberW,
-      bios_smbiosbiosversionW, bios_versionW }
+      bios_smbiosbiosversionW, bios_versionW, bios_nameW }
 };
 static const struct record_computersystemproduct data_compsysproduct[] =
 {
-- 
2.5.0

