From 15e83e517e6b227a55ecee9541cca567b94701b8 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 24 Feb 2016 17:51:48 +0100
Subject: user32/tests: Add additional tests for window station name.

---
 dlls/user32/tests/winstation.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dlls/user32/tests/winstation.c b/dlls/user32/tests/winstation.c
index e49420b..4a553ca 100644
--- a/dlls/user32/tests/winstation.c
+++ b/dlls/user32/tests/winstation.c
@@ -114,6 +114,8 @@ static void test_handles(void)
     DWORD id, flags, le;
     ATOM atom;
     char buffer[20];
+    DWORD size;
+    BOOL ret;
 
     /* win stations */
 
@@ -215,6 +217,12 @@ static void test_handles(void)
     w2 = CreateWindowStationA( "", 0, WINSTA_ALL_ACCESS, NULL );
     ok( w2 != 0, "create station failed err %u\n", GetLastError() );
 
+    memset(buffer, 0, sizeof(buffer));
+    ret = GetUserObjectInformationA( w2, UOI_NAME, buffer, sizeof(buffer), &size );
+    ok( ret, "GetUserObjectInformationA returned %x\n", ret );
+    todo_wine ok( !memcmp(buffer, "Service-0x0-", 12), "unexpected WindowStation name '%s'\n", buffer );
+    todo_wine ok( buffer[strlen(buffer) - 1] == '$', "unexpected WindowStation name '%s'\n", buffer );
+
     SetLastError( 0xdeadbeef );
     w3 = OpenWindowStationA( "", TRUE, WINSTA_ALL_ACCESS );
     todo_wine
-- 
2.7.1

