From 3a96350814cd7fd9d620352e252b02aede20d1cf Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Sat, 30 Mar 2019 12:01:50 -0600
Subject: kernel32: Set error code when attempting to delete file symlinks as
 directories.

Signed-off-by: Erich E. Hoover <erich.e.hoover@gmail.com>
---
 dlls/ntdll/tests/file.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/ntdll/tests/file.c b/dlls/ntdll/tests/file.c
index 58fec76bec4..7f69a9116f2 100644
--- a/dlls/ntdll/tests/file.c
+++ b/dlls/ntdll/tests/file.c
@@ -5208,9 +5208,9 @@ static void test_reparse_points(void)
     bret = RemoveDirectoryW(reparse_path);
     todo_wine ok(!bret, "Succeeded in deleting file symlink as a directory!\n");
     err = GetLastError();
-    todo_wine ok(err == ERROR_DIRECTORY,
-                 "Expected last error 0x%x for RemoveDirectory on file symlink (actually 0x%x)!\n",
-                 ERROR_DIRECTORY, err);
+    ok(err == ERROR_DIRECTORY,
+       "Expected last error 0x%x for RemoveDirectory on file symlink (actually 0x%x)!\n",
+       ERROR_DIRECTORY, err);
     dwret = GetFileAttributesW(reparse_path);
     todo_wine ok(dwret != (DWORD)~0, "Symlink doesn't exist (attributes: 0x%x)!\n", dwret);
     ok(dwret & FILE_ATTRIBUTE_REPARSE_POINT, "File is not a symlink! (attributes: 0x%x)\n", dwret);
-- 
2.17.1

