From f85a2cc8b1911d19140e90dc317a565ef10e3b0f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 26 Feb 2015 05:27:14 +0100
Subject: ntdll: Define a couple more information classes.

---
 dlls/ntdll/file.c  | 20 +++++++++++---------
 include/winternl.h | 28 +++++++++++++++++-----------
 2 files changed, 28 insertions(+), 20 deletions(-)

diff --git a/dlls/ntdll/file.c b/dlls/ntdll/file.c
index 5232027..d05289e 100644
--- a/dlls/ntdll/file.c
+++ b/dlls/ntdll/file.c
@@ -2042,20 +2042,22 @@ NTSTATUS WINAPI NtQueryInformationFile( HANDLE hFile, PIO_STATUS_BLOCK io,
         0,                                             /* FileIdFullDirectoryInformation */
         0,                                             /* FileValidDataLengthInformation */
         0,                                             /* FileShortNameInformation */
-        0,
-        0,
-        0,
+        0,                                             /* FileIoCompletionNotificationInformation, */
+        0,                                             /* FileIoStatusBlockRangeInformation */
+        0,                                             /* FileIoPriorityHintInformation */
         0,                                             /* FileSfioReserveInformation */
         0,                                             /* FileSfioVolumeInformation */
         0,                                             /* FileHardLinkInformation */
-        0,
+        0,                                             /* FileProcessIdsUsingFileInformation */
         0,                                             /* FileNormalizedNameInformation */
-        0,
+        0,                                             /* FileNetworkPhysicalNameInformation */
         0,                                             /* FileIdGlobalTxDirectoryInformation */
-        0,
-        0,
-        0,
-        0                                              /* FileStandardLinkInformation */
+        0,                                             /* FileIsRemoteDeviceInformation */
+        0,                                             /* FileAttributeCacheInformation */
+        0,                                             /* FileNumaNodeInformation */
+        0,                                             /* FileStandardLinkInformation */
+        0,                                             /* FileRemoteProtocolInformation */
+        0,                                             /* FileReplaceCompletionInformation */
     };
 
     struct stat st;
diff --git a/include/winternl.h b/include/winternl.h
index 1a694da..d769ef0 100644
--- a/include/winternl.h
+++ b/include/winternl.h
@@ -411,17 +411,23 @@ typedef enum _FILE_INFORMATION_CLASS {
     FileIdBothDirectoryInformation,
     FileIdFullDirectoryInformation,
     FileValidDataLengthInformation,
-    FileShortNameInformation = 40,
-    /* 41, 42, 43 undocumented */
-    FileSfioReserveInformation = 44,
-    FileSfioVolumeInformation = 45,
-    FileHardLinkInformation = 46,
-    /* 47 undocumented */
-    FileNormalizedNameInformation = 48,
-    /* 49 undocumented */
-    FileIdGlobalTxDirectoryInformation = 50,
-    /* 51, 52, 53 undocumented */
-    FileStandardLinkInformation = 54,
+    FileShortNameInformation,
+    FileIoCompletionNotificationInformation,
+    FileIoStatusBlockRangeInformation,
+    FileIoPriorityHintInformation,
+    FileSfioReserveInformation,
+    FileSfioVolumeInformation,
+    FileHardLinkInformation,
+    FileProcessIdsUsingFileInformation,
+    FileNormalizedNameInformation,
+    FileNetworkPhysicalNameInformation,
+    FileIdGlobalTxDirectoryInformation,
+    FileIsRemoteDeviceInformation,
+    FileAttributeCacheInformation,
+    FileNumaNodeInformation,
+    FileStandardLinkInformation,
+    FileRemoteProtocolInformation,
+    FileReplaceCompletionInformation,
     FileMaximumInformation
 } FILE_INFORMATION_CLASS, *PFILE_INFORMATION_CLASS;
 
-- 
2.3.0

