From 15cc0b19149e0a1c811313ee1a97b82032cf41da Mon Sep 17 00:00:00 2001
From: Bruno Jesus <00cpxxx@gmail.com>
Date: Fri, 12 Dec 2014 05:16:14 +0100
Subject: urlmon: Ignore unsupported flags for CoInternetSetFeatureEnabled

---
 dlls/urlmon/internet.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/dlls/urlmon/internet.c b/dlls/urlmon/internet.c
index e5d41b6..d7deccc 100644
--- a/dlls/urlmon/internet.c
+++ b/dlls/urlmon/internet.c
@@ -536,10 +536,8 @@ static HRESULT set_internet_feature(INTERNETFEATURELIST feature, DWORD flags, BO
     if(feature >= FEATURE_ENTRY_COUNT)
         return E_FAIL;
 
-    if(flags & ~supported_flags) {
+    if(flags & ~supported_flags)
         FIXME("Unsupported flags: %08x\n", flags & ~supported_flags);
-        return E_NOTIMPL;
-    }
 
     if(flags & SET_FEATURE_ON_PROCESS)
         set_feature_on_process(feature, enable);
-- 
2.1.3

