From 3e131a30d2f1b69de3c378b54f82f03f45bba6bb Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sat, 6 Feb 2016 02:24:05 +0100
Subject: msvideo.dll16: Translate 16 bit address in MCIWNDM_SETTIMEFORMATA
 command.

---
 dlls/msvideo.dll16/msvideo16.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/msvideo.dll16/msvideo16.c b/dlls/msvideo.dll16/msvideo16.c
index 1b715e4..f39b52d 100644
--- a/dlls/msvideo.dll16/msvideo16.c
+++ b/dlls/msvideo.dll16/msvideo16.c
@@ -991,6 +991,10 @@ static LRESULT WINAPI MCIWndProc16(HWND hwnd, UINT msg, WPARAM wparam, LPARAM lp
         lparam = (ULONG_PTR)MapSL(lparam);
         break;
 
+    case MCIWNDM_SETTIMEFORMATA:
+        lparam = (ULONG_PTR)MapSL(lparam);
+        break;
+
     default:
         break;
     }
-- 
2.7.0

