From 607d3969a3b2633ce4dd4f99f96d1666c15cd98d Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 15 Jan 2017 22:52:09 +0100
Subject: include: Avoid duplicate definition of REFERENCE_TIME.

---
 include/dmdls.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/dmdls.h b/include/dmdls.h
index b40ef88bf8d..ea7dcf00946 100644
--- a/include/dmdls.h
+++ b/include/dmdls.h
@@ -30,7 +30,10 @@ typedef LONG PCENT;
 typedef LONG PERCENT;
 typedef LONG TCENT;
 
+#ifndef REFERENCE_TIME_DEFINED
+#define REFERENCE_TIME_DEFINED
 typedef LONGLONG REFERENCE_TIME, *LPREFERENCE_TIME;
+#endif
 
 /*****************************************************************************
  * FOURCC definition
-- 
2.11.0

