From 63d289e33bb3bf81aef7d57750048c58cbbedd60 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Tue, 15 Nov 2016 06:59:13 +0000
Subject: dpnet: Avoid redefinition of SOCKADDR.

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 include/dpaddr.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/dpaddr.h b/include/dpaddr.h
index 576caaf..d47dd26 100644
--- a/include/dpaddr.h
+++ b/include/dpaddr.h
@@ -27,7 +27,9 @@ extern "C" {
 #endif /* defined(__cplusplus) */
 
 typedef REFIID	        DPNAREFIID;
+#ifndef _WINSOCK2API_
 typedef struct sockaddr SOCKADDR;
+#endif
 
 /*****************************************************************************
  * DirectPlay8Addr defines
-- 
2.9.0

