From 0b5b7f54630a4e7e2f3ed0fa82fbad727fbc0a77 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <z.figura12@gmail.com>
Date: Thu, 27 Jun 2019 22:30:15 -0500
Subject: [PATCH] wine.inf: Remove registration for the winebus service.

Signed-off-by: Zebediah Figura <z.figura12@gmail.com>
---
 loader/wine.inf.in | 12 ------------
 1 file changed, 12 deletions(-)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 02e46c5a542..eb6cfbbd705 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -135,7 +135,6 @@ AddService=FontCache3.0.0.0,0,WPFFontCacheService
 AddService=LanmanServer,0,LanmanServerService
 AddService=FontCache,0,FontCacheService
 AddService=Schedule,0,TaskSchedulerService
-AddService=WineBus,0,WineBusService
 AddService=Winmgmt,0,WinmgmtService
 AddService=wuauserv,0,wuauService
 
@@ -153,7 +152,6 @@ AddService=FontCache3.0.0.0,0,WPFFontCacheService
 AddService=LanmanServer,0,LanmanServerService
 AddService=FontCache,0,FontCacheService
 AddService=Schedule,0,TaskSchedulerService
-AddService=WineBus,0,WineBusService
 AddService=Winmgmt,0,WinmgmtService
 AddService=wuauserv,0,wuauService
 
@@ -171,7 +169,6 @@ AddService=FontCache3.0.0.0,0,WPFFontCacheService
 AddService=LanmanServer,0,LanmanServerService
 AddService=FontCache,0,FontCacheService
 AddService=Schedule,0,TaskSchedulerService
-AddService=WineBus,0,WineBusService
 AddService=Winmgmt,0,WinmgmtService
 AddService=wuauserv,0,wuauService
 
@@ -3681,15 +3678,6 @@ ServiceType=32
 StartType=3
 ErrorControl=1
 
-[WineBusService]
-Description="Wine Platform Bus Kernel"
-DisplayName="Platform Bus Kernel"
-ServiceBinary="%12%\winebus.sys"
-LoadOrderGroup="WinePlugPlay"
-ServiceType=1
-StartType=2
-ErrorControl=1
-
 [SpoolerService]
 AddReg=SpoolerServiceKeys
 Description="Loads files to memory for later printing"
-- 
2.17.1

