From 6356a5fdb13f638ed0f2072fd12dfc6ca0e66935 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 6 Mar 2016 06:17:31 +0100
Subject: makefiles: Do not generate Makefile for disabled rules.

---
 aclocal.m4 | 1 -
 1 file changed, 1 deletion(-)

diff --git a/aclocal.m4 b/aclocal.m4
index 00eeb98..36c123b 100644
--- a/aclocal.m4
+++ b/aclocal.m4
@@ -284,7 +284,6 @@ wine_fn_disabled_rules ()
 {
     ac_clean=$[@]
 
-    wine_fn_append_file SUBDIRS $ac_dir
     wine_fn_append_file DISABLED_SUBDIRS $ac_dir
     wine_fn_append_rule \
 "__clean__: $ac_dir/clean
-- 
2.7.1

