From 5533c5dcaaecdca695a44018611d37d06ffe6522 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 18 Apr 2015 23:31:02 +0200
Subject: msxml3: Add a missing break (Coverity).

---
 dlls/msxml3/node.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/msxml3/node.c b/dlls/msxml3/node.c
index 7119a43..1ddc9bf 100644
--- a/dlls/msxml3/node.c
+++ b/dlls/msxml3/node.c
@@ -2055,6 +2055,7 @@ IXMLDOMNode *create_node( xmlNodePtr node )
         new_node->ref = 1;
         init_xmlnode(&new_node->node, node, &new_node->IXMLDOMNode_iface, NULL);
         pUnk = (IUnknown*)&new_node->IXMLDOMNode_iface;
+        break;
     }
     default:
         ERR("Called for unsupported node type %d\n", node->type);
-- 
2.3.5

