From 6db27add885e521e7b42829ade2f7de74abdd619 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sun, 7 Feb 2016 01:32:04 +0100
Subject: quartz: Recognize mpeg2 program streams.

---
 dlls/quartz/regsvr.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/quartz/regsvr.c b/dlls/quartz/regsvr.c
index b0f1e9a..196e064 100644
--- a/dlls/quartz/regsvr.c
+++ b/dlls/quartz/regsvr.c
@@ -818,6 +818,11 @@ static struct regsvr_mediatype_parsing const mediatype_parsing_list[] = {
             "0, 10, FFFFFF00000080808080, 494433000000000000",
 	    NULL }
     },
+    {   &MEDIATYPE_Stream,
+        &MEDIASUBTYPE_MPEG2_PROGRAM,
+        {   "0, 5, FFFFFFFFC0, 000001BA40",
+            NULL }
+    },
     {	&MEDIATYPE_Stream,
 	&MEDIASUBTYPE_QTMovie,
 	{   "4, 4, , 6d646174",
-- 
2.7.0

