From 227e86c26da931222508fab0b175db09ce1b4ce2 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 6 May 2017 01:32:51 +0200
Subject: d3dx9_36/tests: Fix some test failures.

---
 dlls/d3dx9_36/tests/math.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/d3dx9_36/tests/math.c b/dlls/d3dx9_36/tests/math.c
index 0e9853abe38..8bc462c48a4 100644
--- a/dlls/d3dx9_36/tests/math.c
+++ b/dlls/d3dx9_36/tests/math.c
@@ -244,7 +244,7 @@ static void D3DXColorTest(void)
 /*_______________D3DXColorLerp________________*/
     expected.r = 0.32f; expected.g = 0.69f; expected.b = 0.356f; expected.a = 0.897f;
     D3DXColorLerp(&got,&color,&color1,scale);
-    expect_color(&expected, &got, 0);
+    expect_color(&expected, &got, 2);
     /* Test the NULL case */
     funcpointer = D3DXColorLerp(&got,NULL,&color1,scale);
     ok(funcpointer == NULL, "Expected: %p, Got: %p\n", NULL, funcpointer);
@@ -2994,7 +2994,7 @@ static void test_D3DXSHEvalDirectionalLight(void)
                         expected = j + test[l].boffset;
                     else
                         expected = test[l].blue_out[startindex + j];
-                    equal = compare_float(expected, blue_out[j], 4);
+                    equal = compare_float(expected, blue_out[j], 8);
                     ok(equal, "Blue: case %u, order %u: expected[%u] = %.8e, received %.8e.\n",
                             l, order, j, expected, blue_out[j]);
                 }
@@ -3339,7 +3339,7 @@ static void test_D3DXSHMultiply3(void)
     D3DXSHMultiply3(c, c, b);
     for (i = 0; i < ARRAY_SIZE(expected_aliased); ++i)
     {
-        equal = compare_float(c[i], expected_aliased[i], 32);
+        equal = compare_float(c[i], expected_aliased[i], 64);
         ok(equal, "Expected[%u] = %.8e, received = %.8e.\n", i, expected_aliased[i], c[i]);
     }
 }
-- 
2.14.2

