From e8f45ee1fb9530fd798d68ae5eeb34190d1de1ff Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 29 Oct 2015 21:22:16 +0100
Subject: makefiles: Use correct source path for ttf install rules.

Signed-off-by: Sebastian Lackner <sebastian@fds-team.de>
---
 tools/makedep.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/makedep.c b/tools/makedep.c
index 30bbc3e..c15f0ac 100644
--- a/tools/makedep.c
+++ b/tools/makedep.c
@@ -2057,7 +2057,7 @@ static struct strarray output_sources( struct makefile *make, struct strarray *t
             }
             if (source->file->flags & FLAG_INSTALL)
             {
-                strarray_add( &make->install_lib_rules, ttf_file );
+                strarray_add( &make->install_lib_rules, strmake( "%s.ttf", obj ));
                 strarray_add( &make->install_lib_rules, strmake( "D$(fontdir)/%s.ttf", obj ));
             }
             if (source->file->flags & FLAG_SFD_FONTS)
-- 
2.6.1

