From d0b101227ad133d8b011a3c5b13e8967d7656abe Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 29 Oct 2015 21:22:24 +0100
Subject: makefiles: Symlinks have no dependencies on source files.

Signed-off-by: Sebastian Lackner <sebastian@fds-team.de>
---
 tools/makedep.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/tools/makedep.c b/tools/makedep.c
index c15f0ac..cbf0d48 100644
--- a/tools/makedep.c
+++ b/tools/makedep.c
@@ -1796,8 +1796,11 @@ static void output_install_rules( struct makefile *make, struct strarray files,
     if (!files.count) return;
 
     for (i = 0; i < files.count; i += 2)
-        if (files.str[i + 1][0] >= 'a' && files.str[i + 1][0] <= 'z')  /* only for files in object dir */
-            strarray_add_uniq( &targets, files.str[i] );
+    {
+        if (files.str[i + 1][0] < 'a' || files.str[i + 1][0] > 'z') continue;  /* only for files in object dir */
+        if (files.str[i + 1][0] == 'y') continue;  /* symlinks have no dependencies on source files */
+        strarray_add_uniq( &targets, files.str[i] );
+    }
 
     output( "install %s::", target );
     output_filenames_obj_dir( make, targets );
-- 
2.6.1

