From a491ae0671f0bf8e482ede7ad5b0538bd10ed87d Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 22 Aug 2015 07:22:56 +0200
Subject: msvcrt: Use constants instead of hardcoded values.

---
 dlls/msvcrt/file.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/msvcrt/file.c b/dlls/msvcrt/file.c
index 49cd4f0..a62f73d 100644
--- a/dlls/msvcrt/file.c
+++ b/dlls/msvcrt/file.c
@@ -393,13 +393,13 @@ static void msvcrt_free_fd(int fd)
   {
     switch (fd)
     {
-    case 0:
+    case MSVCRT_STDIN_FILENO:
         SetStdHandle(STD_INPUT_HANDLE, 0);
         break;
-    case 1:
+    case MSVCRT_STDOUT_FILENO:
         SetStdHandle(STD_OUTPUT_HANDLE, 0);
         break;
-    case 2:
+    case MSVCRT_STDERR_FILENO:
         SetStdHandle(STD_ERROR_HANDLE, 0);
         break;
     }
-- 
2.5.0

