From 47f4a0e0cb0e90695cb863306475e7ac11ef3e4a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 16 Jun 2017 04:53:19 +0200
Subject: advapi32/tests: Add back a dropped test.

---
 dlls/advapi32/tests/security.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/advapi32/tests/security.c b/dlls/advapi32/tests/security.c
index de511af32f..eca83765af 100644
--- a/dlls/advapi32/tests/security.c
+++ b/dlls/advapi32/tests/security.c
@@ -6426,10 +6426,12 @@ static void test_AddMandatoryAce(void)
 
     sacl = (void *)0xdeadbeef;
     present = TRUE;
+    defaulted = TRUE;
     ret = GetSecurityDescriptorSacl(sd2, &present, &sacl, &defaulted);
     ok(ret, "GetSecurityDescriptorSacl failed with error %u\n", GetLastError());
     ok(!present, "SACL is present\n");
     ok(sacl == (void *)0xdeadbeef, "SACL is set\n");
+    todo_wine ok(!defaulted, "SACL defaulted\n");
 
     HeapFree(GetProcessHeap(), 0, sd2);
     CloseHandle(handle);
-- 
2.13.1

