From de7c25e765a81c348ee0df60f9cb311cf5feeb18 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Fri, 13 Jan 2017 22:49:13 +0100
Subject: ntdll: Return success for TokenSessionId in NtSetInformationToken.

---
 dlls/ntdll/nt.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/dlls/ntdll/nt.c b/dlls/ntdll/nt.c
index 7660ea8ce74..9347170a593 100644
--- a/dlls/ntdll/nt.c
+++ b/dlls/ntdll/nt.c
@@ -602,6 +602,20 @@ NTSTATUS WINAPI NtSetInformationToken(
         }
         SERVER_END_REQ;
         break;
+    case TokenSessionId:
+        if (TokenInformationLength < sizeof(DWORD))
+        {
+            ret = STATUS_INFO_LENGTH_MISMATCH;
+            break;
+        }
+        if (!TokenInformation)
+        {
+            ret = STATUS_ACCESS_VIOLATION;
+            break;
+        }
+        FIXME("handling of TokenSessionId not implemented\n");
+        ret = STATUS_SUCCESS;
+        break;
     default:
         FIXME("unimplemented class %u\n", TokenInformationClass);
         break;
-- 
2.11.0

