From 01043db0fa1893a26c59f7bc2972aba437aeb063 Mon Sep 17 00:00:00 2001
From: Andrew Wesie <awesie@gmail.com>
Date: Wed, 3 May 2017 15:00:16 -0500
Subject: ntdll: TokenIntegrityLevel stub in NtSetInformationToken.

Signed-off-by: Andrew Wesie <awesie@gmail.com>
---
 dlls/ntdll/nt.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/ntdll/nt.c b/dlls/ntdll/nt.c
index b0473a5ffd9..93554e929be 100644
--- a/dlls/ntdll/nt.c
+++ b/dlls/ntdll/nt.c
@@ -660,6 +660,10 @@ NTSTATUS WINAPI NtSetInformationToken(
         FIXME("handling of TokenSessionId not implemented\n");
         ret = STATUS_SUCCESS;
         break;
+    case TokenIntegrityLevel:
+        FIXME("TokenIntegrityLevel stub!\n");
+        ret = STATUS_SUCCESS;
+        break;
     default:
         FIXME("unimplemented class %u\n", TokenInformationClass);
         break;
-- 
2.12.2

