From 02c1bbeb9c610518d282c65aaa6f9563c3d80b73 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Tue, 7 Mar 2017 01:57:42 +0100
Subject: ntdll: Implement ProcessImageFileNameWin32 in
 NtQueryInformationProcess.

---
 dlls/ntdll/process.c |  1 +
 include/winternl.h   | 16 ++++++++++++++++
 2 files changed, 17 insertions(+)

diff --git a/dlls/ntdll/process.c b/dlls/ntdll/process.c
index a14b6707167..39cd8be6019 100644
--- a/dlls/ntdll/process.c
+++ b/dlls/ntdll/process.c
@@ -498,6 +498,7 @@ NTSTATUS WINAPI NtQueryInformationProcess(
     case ProcessImageFileName:
         /* FIXME: this will return a DOS path. Windows returns an NT path. Changing this would require also changing kernel32.QueryFullProcessImageName.
          * The latter may be harder because of the lack of RtlNtPathNameToDosPathName. */
+    case ProcessImageFileNameWin32:
         SERVER_START_REQ(get_dll_info)
         {
             UNICODE_STRING *image_file_name_str = ProcessInformation;
diff --git a/include/winternl.h b/include/winternl.h
index 8acb7fd3e11..4a137f42b49 100644
--- a/include/winternl.h
+++ b/include/winternl.h
@@ -829,6 +829,22 @@ typedef enum _PROCESSINFOCLASS {
     ProcessDebugFlags = 31,
     ProcessHandleTracing = 32,
     ProcessExecuteFlags = 34,
+    ProcessTlsInformation = 35,
+    ProcessCookie = 36,
+    ProcessImageInformation = 37,
+    ProcessCycleTime = 38,
+    ProcessPagePriority = 39,
+    ProcessInstrumentationCallback = 40,
+    ProcessThreadStackAllocation = 41,
+    ProcessWorkingSetWatchEx = 42,
+    ProcessImageFileNameWin32 = 43,
+    ProcessImageFileMapping = 44,
+    ProcessAffinityUpdateMode = 45,
+    ProcessMemoryAllocationMode = 46,
+    ProcessGroupInformation = 47,
+    ProcessTokenVirtualizationEnabled = 48,
+    ProcessConsoleHostProcess = 49,
+    ProcessWindowInformation = 50,
     MaxProcessInfoClass
 } PROCESSINFOCLASS, PROCESS_INFORMATION_CLASS;
 
-- 
2.11.0

