From 0fa73610b1527bd2557d34d8575cdaa89f2b23aa Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20Rosenkr=C3=A4nzer?= <bero@lindev.ch>
Date: Sun, 5 Jul 2020 11:33:43 -0500
Subject: [PATCH] include: Correct a spelling error in the definition of
 IXACT3Engine_Initialize.

Signed-off-by: Zebediah Figura <z.figura12@gmail.com>
---
 include/xact3.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/xact3.h b/include/xact3.h
index d66874a511d..80059c124a5 100644
--- a/include/xact3.h
+++ b/include/xact3.h
@@ -679,7 +679,7 @@ DECLARE_INTERFACE_(IXACT3Engine,IUnknown)
 #define IXACT3Engine_GetRendererCount(p,a)         (p)->lpVtbl->GetRendererCount(p,a)
 #define IXACT3Engine_GetRendererDetails(p,a,b)     (p)->lpVtbl->GetRendererDetails(p,a,b)
 #define IXACT3Engine_GetFinalMixFormat(p,a)        (p)->lpVtbl->GetFinalMixFormat(p,a)
-#define IXACT3Engine_Initialize(p,a)               (p)->lpVtbl->Initialze(p,a)
+#define IXACT3Engine_Initialize(p,a)               (p)->lpVtbl->Initialize(p,a)
 #define IXACT3Engine_Shutdown(p)                   (p)->lpVtbl->Shutdown(p)
 #define IXACT3Engine_DoWork(p)                     (p)->lpVtbl->DoWork(p)
 #define IXACT3Engine_CreateSoundBank(p,a,b,c,d,e)  (p)->lpVtbl->CreateSoundBank(p,a,b,c,d,e)
-- 
2.27.0

