From c7189e9d48079b0f3993ed8de8f774f3e0a7a97e Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Tue, 7 Jan 2020 22:14:49 +1100
Subject: [PATCH] xaudio2_7: Support older XACT3Engine interfaces

---
 dlls/xaudio2_7/xact_dll.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/dlls/xaudio2_7/xact_dll.c b/dlls/xaudio2_7/xact_dll.c
index 07f866606d..91e3ef0afd 100644
--- a/dlls/xaudio2_7/xact_dll.c
+++ b/dlls/xaudio2_7/xact_dll.c
@@ -39,6 +39,10 @@ WINE_DEFAULT_DEBUG_CHANNEL(xact3);
 
 static HINSTANCE instance;
 
+#if XACT3_VER >= 0x0301 && XACT3_VER <= 0x304
+DEFINE_GUID(IID_IXACT3Engine301, 0xe72c1b9a, 0xd717, 0x41c0, 0x81, 0xa6, 0x50, 0xeb, 0x56, 0xe8, 0x06, 0x49);
+#endif
+
 BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD reason, void *pReserved)
 {
     TRACE("(%p, %d, %p)\n", hinstDLL, reason, pReserved);
@@ -852,7 +856,12 @@ static HRESULT WINAPI IXACT3EngineImpl_QueryInterface(IXACT3Engine *iface,
     TRACE("(%p)->(%s, %p)\n", This, debugstr_guid(riid), ppvObject);
 
     if(IsEqualGUID(riid, &IID_IUnknown) ||
-            IsEqualGUID(riid, &IID_IXACT3Engine)){
+#if XACT3_VER >= 0x0301 && XACT3_VER <= 0x304
+            IsEqualGUID(riid, &IID_IXACT3Engine301)
+#else
+            IsEqualGUID(riid, &IID_IXACT3Engine)
+#endif
+        ){
         *ppvObject = &This->IXACT3Engine_iface;
     }
     else
-- 
2.25.1

