From 94df3565a9109176ee936416804f325b5fde9627 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 8 Jan 2020 12:34:53 +1100
Subject: [PATCH] xaudio2_7: Always return S_OK in IXACT34Cue Destroy

---
 dlls/xaudio2_7/xact_dll.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/dlls/xaudio2_7/xact_dll.c b/dlls/xaudio2_7/xact_dll.c
index 9cec216002..2bf6eaf374 100644
--- a/dlls/xaudio2_7/xact_dll.c
+++ b/dlls/xaudio2_7/xact_dll.c
@@ -320,13 +320,12 @@ static HRESULT WINAPI IXACT34CueImpl_GetState(IXACT34Cue *iface, DWORD *pdwState
 static HRESULT WINAPI IXACT34CueImpl_Destroy(IXACT34Cue *iface)
 {
     XACT3CueImpl *This = impl_from_IXACT34Cue(iface);
-    HRESULT hr;
 
     TRACE("(%p)\n", iface);
 
-    hr = FACTCue_Destroy(This->fact_cue);
+    FACTCue_Destroy(This->fact_cue);
     HeapFree(GetProcessHeap(), 0, This);
-    return hr;
+    return S_OK;
 }
 
 static HRESULT WINAPI IXACT34CueImpl_SetMatrixCoefficients(IXACT34Cue *iface,
-- 
2.25.1

