From 4963145cb0fd3713de75ac0f0913953980cd7519 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 8 Jan 2020 12:34:53 +1100
Subject: [PATCH] xaudio2_7: Always return S_OK in IXACT34Cue Destroy

---
 dlls/xaudio2_7/xact_dll.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dlls/xaudio2_7/xact_dll.c b/dlls/xaudio2_7/xact_dll.c
index b2cfcfb4068..4900e51802e 100644
--- a/dlls/xaudio2_7/xact_dll.c
+++ b/dlls/xaudio2_7/xact_dll.c
@@ -170,6 +170,7 @@ static HRESULT WINAPI IXACT3CueImpl_GetState(IXACT3Cue *iface, DWORD *pdwState)
 static HRESULT WINAPI IXACT3CueImpl_Destroy(IXACT3Cue *iface)
 {
     XACT3CueImpl *This = impl_from_IXACT3Cue(iface);
+#if XACT3_VER > 0x0304
     HRESULT hr;
 
     TRACE("(%p)\n", iface);
@@ -177,6 +178,13 @@ static HRESULT WINAPI IXACT3CueImpl_Destroy(IXACT3Cue *iface)
     hr = FACTCue_Destroy(This->fact_cue);
     HeapFree(GetProcessHeap(), 0, This);
     return hr;
+#else
+    TRACE("(%p)\n", iface);
+
+    FACTCue_Destroy(This->fact_cue);
+    HeapFree(GetProcessHeap(), 0, This);
+    return S_OK;
+#endif
 }
 
 static HRESULT WINAPI IXACT3CueImpl_SetMatrixCoefficients(IXACT3Cue *iface,
-- 
2.27.0

