From 0b54c7d231cc8ba9cde2bd5cda542ec61ba86fdf Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Tue, 14 Jan 2020 15:47:43 +1100
Subject: [PATCH] xaudio2_7: Trace FAudio version being used

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 dlls/xaudio2_7/xact_dll.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/xaudio2_7/xact_dll.c b/dlls/xaudio2_7/xact_dll.c
index dfe74cebc2..4c576d87ba 100644
--- a/dlls/xaudio2_7/xact_dll.c
+++ b/dlls/xaudio2_7/xact_dll.c
@@ -52,6 +52,8 @@ BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD reason, void *pReserved)
     case DLL_PROCESS_ATTACH:
         instance = hinstDLL;
         DisableThreadLibraryCalls( hinstDLL );
+
+        TRACE("Using FAudio version %d\n", FAudioLinkedVersion() );
         break;
     }
     return TRUE;
-- 
2.25.1

