From 9b14a6c11c5d720682965f0dc92c3f3a5d92a345 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Tue, 14 Jan 2020 15:47:15 +1100
Subject: [PATCH] xaudio2_7: Trace FAudio version being used

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 dlls/xaudio2_7/xaudio_dll.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/xaudio2_7/xaudio_dll.c b/dlls/xaudio2_7/xaudio_dll.c
index a87e0124d3..ae18a72e2b 100644
--- a/dlls/xaudio2_7/xaudio_dll.c
+++ b/dlls/xaudio2_7/xaudio_dll.c
@@ -90,6 +90,8 @@ BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD reason, void *pReserved)
     case DLL_PROCESS_ATTACH:
         instance = hinstDLL;
         DisableThreadLibraryCalls( hinstDLL );
+
+        TRACE("Using FAudio version %d\n", FAudioLinkedVersion() );
         break;
     }
     return TRUE;
-- 
2.25.1

