From e8e4434d2a8edf0b964ed0e93500b20718ba961d Mon Sep 17 00:00:00 2001
From: Zebediah Figura <z.figura12@gmail.com>
Date: Fri, 12 Jun 2020 17:59:11 -0500
Subject: [PATCH] include: Correct the name of WAVEBANKMINIWAVEFORMAT.

Signed-off-by: Zebediah Figura <z.figura12@gmail.com>
---
 include/xact3wb.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/xact3wb.h b/include/xact3wb.h
index c68d7a30fad..81b1518930d 100644
--- a/include/xact3wb.h
+++ b/include/xact3wb.h
@@ -138,7 +138,7 @@ typedef struct WAVEBANKDATA
     DWORD dwEntryMetaDataElementSize;
     DWORD dwEntryNameElementSize;
     DWORD dwAlignment;
-    WAVEBANKMINIFORMAT CompactFormat;
+    WAVEBANKMINIWAVEFORMAT CompactFormat;
     FILETIME BuildTime;
 } WAVEBANKDATA, *LPWAVEBANKDATA;
 typedef const WAVEBANKDATA *LPCWAVEBANKDATA;
-- 
2.27.0

