From 3e2be51876cea141ae48e43341d37d5e676bfbf3 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Thu, 31 May 2018 11:45:10 +1000
Subject: [PATCH] wintab32: Set lcSysExtX/Y for the first index of WTI_DDCTXS

---
 dlls/wintab32/context.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/wintab32/context.c b/dlls/wintab32/context.c
index 8db6217..fe666a5 100644
--- a/dlls/wintab32/context.c
+++ b/dlls/wintab32/context.c
@@ -410,7 +410,7 @@ static UINT WTInfoT(UINT wCategory, UINT nIndex, LPVOID lpOutput, BOOL bUnicode)
             pWTInfoW(wCategory, nIndex, &buf);
 
             /*  Handle system extents here, as we can use user32.dll code to set them */
-            if(wCategory == WTI_DEFSYSCTX)
+            if(wCategory == WTI_DEFSYSCTX || wCategory == WTI_DDCTXS)
             {
                 buf.lcSysExtX = GetSystemMetrics(SM_CXSCREEN);
                 buf.lcSysExtY = GetSystemMetrics(SM_CYSCREEN);
-- 
1.9.1

