From f7577766e160ac55faacd0fadf20fd9ebfb47e3a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Tue, 11 Nov 2014 08:50:46 +0100
Subject: secur32: Return more context attributes in
 schan_InitializeSecurityContextW.

---
 dlls/secur32/schannel.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dlls/secur32/schannel.c b/dlls/secur32/schannel.c
index 5b86a75..12d4c8d 100644
--- a/dlls/secur32/schannel.c
+++ b/dlls/secur32/schannel.c
@@ -908,8 +908,16 @@ static SECURITY_STATUS SEC_ENTRY schan_InitializeSecurityContextW(
     }
 
     *pfContextAttr = 0;
+    if (ctx->req_ctx_attr & ISC_REQ_REPLAY_DETECT)
+        *pfContextAttr |= ISC_RET_REPLAY_DETECT;
+    if (ctx->req_ctx_attr & ISC_REQ_SEQUENCE_DETECT)
+        *pfContextAttr |= ISC_RET_SEQUENCE_DETECT;
+    if (ctx->req_ctx_attr & ISC_REQ_CONFIDENTIALITY)
+        *pfContextAttr |= ISC_RET_CONFIDENTIALITY;
     if (ctx->req_ctx_attr & ISC_REQ_ALLOCATE_MEMORY)
         *pfContextAttr |= ISC_RET_ALLOCATED_MEMORY;
+    if (ctx->req_ctx_attr & ISC_REQ_STREAM)
+        *pfContextAttr |= ISC_RET_STREAM;
 
     return ret;
 }
-- 
2.1.3

