From e2f13a81be20013376928c02d5405c831452cddb Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Fri, 1 Apr 2016 23:58:28 +0200
Subject: vulkan: Try to load libvulkan.so.1.

---
 dlls/vulkan/vulkan_thunks.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/dlls/vulkan/vulkan_thunks.c b/dlls/vulkan/vulkan_thunks.c
index acced5e..aea8d64 100644
--- a/dlls/vulkan/vulkan_thunks.c
+++ b/dlls/vulkan/vulkan_thunks.c
@@ -6572,13 +6572,24 @@ static void *libvulkan_handle;
 
 BOOL init_vulkan( void )
 {
+    static const char *libname[] =
+    {
+        "libvulkan.so.1",
+        "libvulkan.so",
+    };
     void *ptr;
     int i;
 
     if (!(function_heap = HeapCreate( HEAP_CREATE_ENABLE_EXECUTE, 0, 0 )))
         return FALSE;
 
-    if (!(libvulkan_handle = wine_dlopen( "libvulkan.so", RTLD_NOW, NULL, 0 )))
+    for (i = 0; i < ARRAY_SIZE(libname); i++)
+    {
+        libvulkan_handle = wine_dlopen( libname[i], RTLD_NOW, NULL, 0 );
+        if (libvulkan_handle) break;
+    }
+
+    if (!libvulkan_handle)
     {
         ERR_(winediag)( "failed to load libvulkan.so, no support for vulkan\n" );
         HeapDestroy( function_heap );
-- 
2.7.1

