From 1b56ed8fa19d38ff6d17884da2b23ff38a3bd79c Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 6 Jan 2016 15:29:57 +0100
Subject: ntdll: Return STATUS_SUCCESS from NtQuerySystemInformationEx.

---
 dlls/ntdll/nt.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/ntdll/nt.c b/dlls/ntdll/nt.c
index 9ee1923..1214978 100644
--- a/dlls/ntdll/nt.c
+++ b/dlls/ntdll/nt.c
@@ -2180,7 +2180,8 @@ NTSTATUS WINAPI NtQuerySystemInformationEx(SYSTEM_INFORMATION_CLASS SystemInform
 {
     FIXME("(0x%08x,%p,%u,%p,%u,%p) stub\n", SystemInformationClass, Query, QueryLength, SystemInformation,
         Length, ResultLength);
-    return STATUS_NOT_IMPLEMENTED;
+    if (ResultLength) *ResultLength = 0;
+    return STATUS_SUCCESS;
 }
 
 /******************************************************************************
-- 
2.6.4

