From c0cf857381cd4af33ce454a96d9e48f896ed5027 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Wed, 13 Apr 2016 15:36:56 +0800
Subject: gdi32: Trace full contents of DOCINFO in StartDoc.

---
 dlls/gdi32/printdrv.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/gdi32/printdrv.c b/dlls/gdi32/printdrv.c
index 3cfdc47..d2f4076 100644
--- a/dlls/gdi32/printdrv.c
+++ b/dlls/gdi32/printdrv.c
@@ -72,9 +72,9 @@ INT WINAPI StartDocW(HDC hdc, const DOCINFOW* doc)
     INT ret;
     DC *dc = get_dc_ptr( hdc );
 
-    TRACE("DocName = %s Output = %s Datatype = %s\n",
+    TRACE("DocName %s, Output %s, Datatype %s, fwType %#x\n",
           debugstr_w(doc->lpszDocName), debugstr_w(doc->lpszOutput),
-          debugstr_w(doc->lpszDatatype));
+          debugstr_w(doc->lpszDatatype), doc->fwType);
 
     if(!dc) return SP_ERROR;
 
-- 
2.7.1

