From a857ce50c5f1e48d7c08c034374b742292d18344 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 8 Aug 2014 23:40:58 +0200
Subject: user32: Increase MAX_WINPROCS to 16384.

---
 dlls/user32/winproc.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/user32/winproc.c b/dlls/user32/winproc.c
index 57adfc2..e18dd5e 100644
--- a/dlls/user32/winproc.c
+++ b/dlls/user32/winproc.c
@@ -45,7 +45,7 @@ typedef struct tagWINDOWPROC
     WNDPROC        procW;    /* Unicode window proc */
 } WINDOWPROC;
 
-#define MAX_WINPROCS  4096
+#define MAX_WINPROCS  16384
 #define MAX_WINPROC_RECURSION  64
 #define WINPROC_PROC16  ((WINDOWPROC *)1)  /* placeholder for 16-bit window procs */
 
-- 
1.7.9.5

