From 1b110d47948ea020afbfebda288d8362837b8f43 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 16 Mar 2016 01:18:05 +0100
Subject: ieframe: Fix implementation of impl_from_ITargetFrame2.

---
 dlls/ieframe/navigate.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ieframe/navigate.c b/dlls/ieframe/navigate.c
index 1027313eb..88bc7e6 100644
--- a/dlls/ieframe/navigate.c
+++ b/dlls/ieframe/navigate.c
@@ -1243,7 +1243,7 @@ static const IHlinkFrameVtbl HlinkFrameVtbl = {
 
 static inline HlinkFrame *impl_from_ITargetFrame2(ITargetFrame2 *iface)
 {
-    return CONTAINING_RECORD(iface, HlinkFrame, IHlinkFrame_iface);
+    return CONTAINING_RECORD(iface, HlinkFrame, ITargetFrame2_iface);
 }
 
 static HRESULT WINAPI TargetFrame2_QueryInterface(ITargetFrame2 *iface, REFIID riid, void **ppv)
-- 
2.7.1

