From e003cc0c093e2d382cf19b8b322e30abf105ab0f Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Mon, 17 Jul 2017 22:08:46 +0200
Subject: configure.ac: Improve check for if_nameindex.

---
 configure.ac | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/configure.ac b/configure.ac
index 7270ce78525..ea427113249 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2343,6 +2343,9 @@ AC_CACHE_CHECK([for if_nameindex],wine_cv_have_if_nameindex,
                AC_LINK_IFELSE([AC_LANG_PROGRAM([[#ifdef HAVE_SYS_TYPES_H
 #include <sys/types.h>
 #endif
+#ifdef HAVE_SYS_SOCKET_H
+#include <sys/socket.h>
+#endif
 #ifdef HAVE_NET_IF_H
 # include <net/if.h>
 #endif]],
-- 
2.13.1

