From d426832abe3e622e471dddf472f5b24aa53d6fe8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Fri, 29 May 2015 02:41:09 +0200
Subject: mshtml: Do not crash on null window in get_frame_by_name.

---
 dlls/mshtml/htmlwindow.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/dlls/mshtml/htmlwindow.c b/dlls/mshtml/htmlwindow.c
index 618648e..92ed869 100644
--- a/dlls/mshtml/htmlwindow.c
+++ b/dlls/mshtml/htmlwindow.c
@@ -406,6 +406,12 @@ HRESULT get_frame_by_name(HTMLOuterWindow *This, const WCHAR *name, BOOL deep, H
 
         nsIDOMWindow_Release(nswindow);
 
+        if (!window_iter)
+        {
+            FIXME("nsIDOMWindow %p is invalid!\n", nswindow);
+            continue;
+        }
+
         hres = IHTMLElement_get_id(&window_iter->frame_element->element.IHTMLElement_iface, &id);
         if(FAILED(hres)) {
             FIXME("IHTMLElement_get_id failed: 0x%08x\n", hres);
-- 
2.4.2

