From 94fa0ba88100e99d7c3b90b7b17d225e5e519fea Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Wed, 5 Nov 2014 05:06:06 +0100
Subject: ntoskrnl: Add stub for KeSetSystemAffinityThread.

---
 dlls/ntoskrnl.exe/ntoskrnl.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/dlls/ntoskrnl.exe/ntoskrnl.c b/dlls/ntoskrnl.exe/ntoskrnl.c
index a31b186..8d3e3c6 100644
--- a/dlls/ntoskrnl.exe/ntoskrnl.c
+++ b/dlls/ntoskrnl.exe/ntoskrnl.c
@@ -1404,6 +1404,15 @@ KPRIORITY WINAPI KeSetPriorityThread( PKTHREAD Thread, KPRIORITY Priority )
 
 
 /***********************************************************************
+ *           KeSetSystemAffinityThread    (NTOSKRNL.EXE.@)
+ */
+VOID WINAPI KeSetSystemAffinityThread( KAFFINITY affinity )
+{
+    FIXME("(%lx)\n", affinity);
+}
+
+
+/***********************************************************************
  *           KeWaitForSingleObject   (NTOSKRNL.EXE.@)
  */
 NTSTATUS WINAPI KeWaitForSingleObject(PVOID Object,
-- 
1.9.1

