From 765b7ac424a30543d0ca6da8c217d8eb024b870b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Fri, 29 Apr 2016 16:39:50 +0200
Subject: dmusic: Implement IDirectMusicBuffer::PackStructured.

---
 dlls/dmusic/buffer.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/dlls/dmusic/buffer.c b/dlls/dmusic/buffer.c
index bd0ad1a..5be2e0d 100644
--- a/dlls/dmusic/buffer.c
+++ b/dlls/dmusic/buffer.c
@@ -134,9 +134,26 @@ static HRESULT WINAPI IDirectMusicBufferImpl_PackStructured(LPDIRECTMUSICBUFFER
 static HRESULT WINAPI IDirectMusicBufferImpl_PackUnstructured(LPDIRECTMUSICBUFFER iface, REFERENCE_TIME rt, DWORD dwChannelGroup, DWORD cb, LPBYTE lpb)
 {
     IDirectMusicBufferImpl *This = impl_from_IDirectMusicBuffer(iface);
+    DWORD new_write_pos = This->write_pos + sizeof(DMUS_EVENTHEADER) + cb;
+    DMUS_EVENTHEADER header;
 
     FIXME("(%p, 0x%s, %d, %d, %p): stub\n", This, wine_dbgstr_longlong(rt), dwChannelGroup, cb, lpb);
 
+    if (new_write_pos > This->size)
+        return DMUS_E_BUFFER_FULL;
+
+    if (!This->write_pos)
+        This->start_time = rt;
+
+    header.cbEvent = cb;
+    header.dwChannelGroup = dwChannelGroup;
+    header.rtDelta = rt - This->start_time;
+    header.dwFlags = 0;
+
+    memcpy(This->data + This->write_pos, &header, sizeof(header));
+    memcpy(This->data + This->write_pos + sizeof(header), lpb, cb);
+    This->write_pos = new_write_pos;
+
     return S_OK;
 }
 
-- 
2.8.0

