From c586d47aabf8741b8526579868c691e77ff5f7c8 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 21 Nov 2013 03:37:31 +0100
Subject: quartz: Workaround Silverlight problems when multiple monitors are
 found

---
 dlls/quartz/vmr9.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/quartz/vmr9.c b/dlls/quartz/vmr9.c
index 31f3f8e..20da5b9 100644
--- a/dlls/quartz/vmr9.c
+++ b/dlls/quartz/vmr9.c
@@ -1440,7 +1440,7 @@ static HRESULT WINAPI VMR7MonitorConfig_GetAvailableMonitors(IVMRMonitorConfig *
 
     args.info7      = info;
     args.info9      = NULL;
-    args.arraysize  = arraysize;
+    args.arraysize  = 1; /* only return first entry */
     args.numdev     = 0;
     EnumDisplayMonitors(NULL, NULL, get_available_monitors_proc, (LPARAM)&args);
 
@@ -1540,7 +1540,7 @@ static HRESULT WINAPI VMR9MonitorConfig_GetAvailableMonitors(IVMRMonitorConfig9
 
     args.info7      = NULL;
     args.info9      = info;
-    args.arraysize  = arraysize;
+    args.arraysize  = 1; /* only return first entry */
     args.numdev     = 0;
     EnumDisplayMonitors(NULL, NULL, get_available_monitors_proc, (LPARAM)&args);
 
-- 
1.7.9.5

