From cdec706cd4fcff6adee4a40ddcf0c2072982b56e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sat, 21 Jan 2017 22:13:09 +0100
Subject: wined3d: Handle no render target view case when setting scissors.

---
 dlls/wined3d/state.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/dlls/wined3d/state.c b/dlls/wined3d/state.c
index 3c0bb5eb132..b7050be1b3b 100644
--- a/dlls/wined3d/state.c
+++ b/dlls/wined3d/state.c
@@ -4853,8 +4853,16 @@ static void scissorrect(struct wined3d_context *context, const struct wined3d_st
         UINT height;
         UINT width;
 
-        wined3d_rendertarget_view_get_drawable_size(target, context, &width, &height);
-        gl_info->gl_ops.gl.p_glScissor(r->left, height - r->bottom, r->right - r->left, r->bottom - r->top);
+        if (target)
+        {
+            wined3d_rendertarget_view_get_drawable_size(target, context, &width, &height);
+            gl_info->gl_ops.gl.p_glScissor(r->left, height - r->bottom, r->right - r->left, r->bottom - r->top);
+        }
+        else
+        {
+            /* There is nothing to render to anyway */
+            gl_info->gl_ops.gl.p_glScissor(0, 0, 0, 0);
+        }
     }
     checkGLcall("glScissor");
 }
-- 
2.11.0

