From fb6d3997445d70ea0744205f47ede1d3be8c50c1 Mon Sep 17 00:00:00 2001
From: Andrew Wesie <awesie@gmail.com>
Date: Fri, 12 Apr 2019 20:04:03 -0500
Subject: [PATCH] ntdll: Return ntdll.dll as the first entry for
 SystemModuleInformation.

---
 dlls/ntdll/unix/system.c | 28 +++++++++++++++++++++++++++-
 1 file changed, 27 insertions(+), 1 deletion(-)

diff --git a/dlls/ntdll/unix/system.c b/dlls/ntdll/unix/system.c
index 0aa55a69ba1..60a92f39a7c 100644
--- a/dlls/ntdll/unix/system.c
+++ b/dlls/ntdll/unix/system.c
@@ -1630,6 +1630,32 @@ static void get_performance_info( SYSTEM_PERFORMANCE_INFORMATION *info )
     info->TotalCommitLimit    = (totalram + totalswap) / page_size;
 }
 
+static void get_ntdll_system_module(SYSTEM_MODULE *sm)
+{
+    char *ptr;
+    ANSI_STRING str;
+    PLIST_ENTRY entry;
+    LDR_DATA_TABLE_ENTRY *mod;
+
+    /* The first entry must be ntdll. */
+    entry = NtCurrentTeb()->Peb->LdrData->InLoadOrderModuleList.Flink;
+    mod = CONTAINING_RECORD(entry, LDR_DATA_TABLE_ENTRY, InLoadOrderLinks);
+
+    sm->Section = 0;
+    sm->MappedBaseAddress = 0;
+    sm->ImageBaseAddress = mod->DllBase;
+    sm->ImageSize = mod->SizeOfImage;
+    sm->Flags = mod->Flags;
+    sm->LoadOrderIndex = 0;
+    sm->InitOrderIndex = 0;
+    sm->LoadCount = 0;
+    str.Length = 0;
+    str.MaximumLength = MAXIMUM_FILENAME_LENGTH;
+    str.Buffer = (char*)sm->Name;
+    RtlUnicodeStringToAnsiString(&str, &mod->FullDllName, FALSE);
+    ptr = strrchr(str.Buffer, '\\');
+    sm->NameOffset = (ptr != NULL) ? (ptr - str.Buffer + 1) : 0;
+}
 
 /* calculate the mday of dst change date, so that for instance Sun 5 Oct 2007
  * (last Sunday in October of 2007) becomes Sun Oct 28 2007
@@ -2319,7 +2345,7 @@ NTSTATUS WINAPI NtQuerySystemInformation( SYSTEM_INFORMATION_CLASS class,
 
             FIXME("returning fake driver list\n");
             smi->ModulesCount = 1;
-            memset(&smi->Modules[0], 0, sizeof(smi->Modules[0]));
+            get_ntdll_system_module(&smi->Modules[0]);
             ret = STATUS_SUCCESS;
         }
         break;
-- 
2.27.0

