From 5f9c268b7e4022d3dd9cf92f6ab04f1c825d2bfa Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 15 Nov 2015 00:40:07 +0100
Subject: oleaut32/tests: Add a test for TKIND_COCLASS in proxy/stub
 marshalling.

---
 dlls/oleaut32/tests/tmarshal.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/oleaut32/tests/tmarshal.c b/dlls/oleaut32/tests/tmarshal.c
index 758b85f..c28ef8b 100644
--- a/dlls/oleaut32/tests/tmarshal.c
+++ b/dlls/oleaut32/tests/tmarshal.c
@@ -1519,6 +1519,10 @@ static void test_typelibmarshal(void)
         "EXCEPINFO differs from expected: wCode = 0x%x, scode = 0x%08x\n",
         excepinfo.wCode, excepinfo.scode);
     VariantClear(&varresult);
+
+    /* call CoClass (direct) */
+    hr = IWidget_Coclass(pWidget, (void *)V_DISPATCH(&vararg[0]));
+    ok_ole_success(hr, IWidget_Coclass);
     VariantClear(&vararg[0]);
 
     /* call Value with a VT_VARIANT|VT_BYREF type */
-- 
2.6.2

