From bb80e69015a8e69294ebf10ee0f785480df52f4b Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 14 Apr 2017 15:57:18 +0200
Subject: ole32: Fix compilation with recent versions of gcc.

---
 dlls/ole32/storage32.h | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/dlls/ole32/storage32.h b/dlls/ole32/storage32.h
index 4fcfd9c362..feaa093e82 100644
--- a/dlls/ole32/storage32.h
+++ b/dlls/ole32/storage32.h
@@ -528,15 +528,11 @@ StgStreamImpl* StgStreamImpl_Construct(
  */
 #ifdef WORDS_BIGENDIAN
 
-#define htole32(x) RtlUlongByteSwap(x)
-#define htole16(x) RtlUshortByteSwap(x)
 #define lendian32toh(x) RtlUlongByteSwap(x)
 #define lendian16toh(x) RtlUshortByteSwap(x)
 
 #else
 
-#define htole32(x) (x)
-#define htole16(x) (x)
 #define lendian32toh(x) (x)
 #define lendian16toh(x) (x)
 
-- 
2.12.2

