From 0cd3973a47317aff488c63eeb33ddedb2cba2b50 Mon Sep 17 00:00:00 2001
From: Hans Leidekker <hans@codeweavers.com>
Date: Sat, 15 Oct 2016 17:27:41 +0200
Subject: include: Avoid duplicate definition of PEVENT_FILTER_DESCRIPTOR.

---
 include/evntrace.h | 14 ++++++--------
 1 file changed, 6 insertions(+), 8 deletions(-)

diff --git a/include/evntrace.h b/include/evntrace.h
index b54cd11..bb1ce03 100644
--- a/include/evntrace.h
+++ b/include/evntrace.h
@@ -248,16 +248,14 @@ typedef struct _EVENT_TRACE_PROPERTIES
     ULONG LoggerNameOffset;
 } EVENT_TRACE_PROPERTIES, *PEVENT_TRACE_PROPERTIES;
 
-typedef struct _EVENT_FILTER_DESCRIPTOR *PEVENT_FILTER_DESCRIPTOR;
-
 typedef struct _ENABLE_TRACE_PARAMETERS
 {
-    ULONG                    Version;
-    ULONG                    EnableProperty;
-    ULONG                    ControlFlags;
-    GUID                     SourceId;
-    PEVENT_FILTER_DESCRIPTOR EnableFilterDesc;
-    ULONG                    FilterDescCount;
+    ULONG                            Version;
+    ULONG                            EnableProperty;
+    ULONG                            ControlFlags;
+    GUID                             SourceId;
+    struct _EVENT_FILTER_DESCRIPTOR *EnableFilterDesc;
+    ULONG                            FilterDescCount;
 } ENABLE_TRACE_PARAMETERS, *PENABLE_TRACE_PARAMETERS;
 
 #define INVALID_PROCESSTRACE_HANDLE ((TRACEHANDLE)~(ULONG_PTR)0)
-- 
2.9.0

