From 6bb09c3a238180db2b88afeef8f5d08aade863c0 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Mon, 28 Aug 2017 23:28:00 +0200
Subject: kernel32: Fix compile error in load_library function.

---
 dlls/kernel32/module.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/kernel32/module.c b/dlls/kernel32/module.c
index fa6f40b41e3..bfeae34326d 100644
--- a/dlls/kernel32/module.c
+++ b/dlls/kernel32/module.c
@@ -1043,7 +1043,7 @@ static HMODULE load_library( const UNICODE_STRING *libname, DWORD flags )
     NTSTATUS nts;
     HMODULE hModule;
     WCHAR *load_path;
-    static const DWORD unsupported_flags = load_library_search_flags |
+    const DWORD unsupported_flags = load_library_search_flags |
         LOAD_IGNORE_CODE_AUTHZ_LEVEL |
         LOAD_LIBRARY_AS_IMAGE_RESOURCE |
         LOAD_LIBRARY_REQUIRE_SIGNED_TARGET |
-- 
2.14.1

