From 8093970096f440a6accd5930a59fc44d32269356 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 4 Jun 2017 12:52:21 +0200
Subject: d3d11: Remove duplicate const.

---
 dlls/d3d11/device.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/d3d11/device.c b/dlls/d3d11/device.c
index fcf5d10a3b2..e84c3655e08 100644
--- a/dlls/d3d11/device.c
+++ b/dlls/d3d11/device.c
@@ -3324,7 +3324,7 @@ static HRESULT STDMETHODCALLTYPE d3d11_device_CheckFormatSupport(ID3D11Device *i
     struct wined3d *wined3d;
     unsigned int i;
 
-    static const const struct
+    static const struct
     {
         enum wined3d_resource_type rtype;
         unsigned int usage;
-- 
2.13.0

