From 9773d2d2684f1dbc1efd089139ea42a5aacb9306 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 24 Apr 2015 17:03:19 +0200
Subject: d3d9/tests: Avoid warning with gcc 4.9.

---
 dlls/d3d9/tests/visual.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/d3d9/tests/visual.c b/dlls/d3d9/tests/visual.c
index 08362e1..02353e4 100644
--- a/dlls/d3d9/tests/visual.c
+++ b/dlls/d3d9/tests/visual.c
@@ -539,8 +539,8 @@ done:
 
 static void test_specular_lighting(void)
 {
-    static const unsigned int vertices_side = 5;
-    static const unsigned int indices_count = (vertices_side - 1) * (vertices_side - 1) * 2 * 3;
+    const unsigned int vertices_side = 5;
+    const unsigned int indices_count = (vertices_side - 1) * (vertices_side - 1) * 2 * 3;
     static const DWORD fvf = D3DFVF_XYZ | D3DFVF_NORMAL;
     static const D3DMATRIX mat =
     {{{
-- 
2.3.5

