From 450d33400036663b5a0e21fce38a7c449409b4d3 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 30 Jul 2016 23:55:09 +0200
Subject: gdi32: Remove unnecessary debug channel.

---
 dlls/gdi32/enhmfdrv/dc.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/dlls/gdi32/enhmfdrv/dc.c b/dlls/gdi32/enhmfdrv/dc.c
index 64c6136..91fde55 100644
--- a/dlls/gdi32/enhmfdrv/dc.c
+++ b/dlls/gdi32/enhmfdrv/dc.c
@@ -21,9 +21,6 @@
 
 #include <assert.h>
 #include "enhmfdrv/enhmetafiledrv.h"
-#include "wine/debug.h"
-
-WINE_DEFAULT_DEBUG_CHANNEL(enhmetafile);
 
 /* get the emf physdev from the path physdev */
 static inline PHYSDEV get_emfdev( PHYSDEV path )
-- 
2.9.0

