From 33717bde9e702520e23ae014c398bd7076902d43 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 4 Jun 2017 12:56:47 +0200
Subject: shell32: Fix length parameter for ZeroMemory.

---
 dlls/shell32/shfldr_fs.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/shell32/shfldr_fs.c b/dlls/shell32/shfldr_fs.c
index c7259276d1d..10653f92b52 100644
--- a/dlls/shell32/shfldr_fs.c
+++ b/dlls/shell32/shfldr_fs.c
@@ -1315,7 +1315,7 @@ ISFHelper_fnCopyItems (ISFHelper * iface, IShellFolder * pSFFrom, UINT cidl,
 
         if (SUCCEEDED (IPersistFolder2_GetCurFolder (ppf2, &pidl))) {
             SHGetPathFromIDListW (pidl, wszSrcPathRoot);
-            ZeroMemory(wszDstPath, MAX_PATH+1);
+            ZeroMemory(wszDstPath, sizeof(wszDstPath));
             if (This->sPathTarget)
                 lstrcpynW(wszDstPath, This->sPathTarget, MAX_PATH);
             PathAddBackslashW(wszSrcPathRoot);
-- 
2.13.0

