From 2b5e9f330770221eee2eda2aab251eba8d370a60 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 4 Jun 2017 13:17:54 +0200
Subject: fusion: Fix length parameter for ZeroMemory.

---
 dlls/fusion/tests/asmname.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/fusion/tests/asmname.c b/dlls/fusion/tests/asmname.c
index 5fb14a48291..21cd4874edc 100644
--- a/dlls/fusion/tests/asmname.c
+++ b/dlls/fusion/tests/asmname.c
@@ -358,7 +358,7 @@ static void test_assembly_name_props_line(IAssemblyName *name,
         to_widechar(expect, vals[i].val);
 
         size = MAX_PATH;
-        ZeroMemory(str, MAX_PATH);
+        ZeroMemory(str, sizeof(str));
         hr = IAssemblyName_GetProperty(name, i, str, &size);
         to_multibyte(val, str);
 
-- 
2.13.0

