From e733682785c0f2001a972090693c4b49bb8e1090 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 30 Jul 2016 23:55:40 +0200
Subject: mshtml: Avoid misleading indentation warnings.

---
 dlls/mshtml/script.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/mshtml/script.c b/dlls/mshtml/script.c
index 40deda4..19f3541 100644
--- a/dlls/mshtml/script.c
+++ b/dlls/mshtml/script.c
@@ -991,8 +991,8 @@ static HRESULT ScriptBSC_read_data(BSCallback *bsc, IStream *stream)
         if(This->bsc.readed >= This->size) {
 	  void *new_buf;
 	  new_buf = heap_realloc(This->buf, This->size << 1);
-	  if(!new_buf)
-	    return E_OUTOFMEMORY;
+            if(!new_buf)
+                return E_OUTOFMEMORY;
             This->size <<= 1;
             This->buf = new_buf;
         }
-- 
2.9.0

