From 5a398b6537c252196f0f3df6ea4dfbc3956e322b Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 30 Jul 2016 23:56:41 +0200
Subject: msvcirt/tests: Avoid misleading indentation warnings.

---
 dlls/msvcirt/tests/msvcirt.c | 40 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 40 insertions(+)

diff --git a/dlls/msvcirt/tests/msvcirt.c b/dlls/msvcirt/tests/msvcirt.c
index a921b27..11c8b4f 100644
--- a/dlls/msvcirt/tests/msvcirt.c
+++ b/dlls/msvcirt/tests/msvcirt.c
@@ -2896,7 +2896,9 @@ static void test_ostream(void) {
 
     /* flush */
 if (0) /* crashes on native */
+{
     pos = call_func1(p_ostream_flush, &os1);
+}
     os1.base_ios.sb = &fb2.base;
     call_func1(p_filebuf_ctor, &fb2);
     pos = call_func1(p_ostream_flush, &os1);
@@ -2952,7 +2954,9 @@ if (0) /* crashes on native */
     os1.base_ios.tie = &os2;
     os2.base_ios.sb = NULL;
 if (0) /* crashes on native */
+{
     ret = (int) call_func1(p_ostream_opfx, &os1);
+}
     os2.base_ios.sb = &fb2.base;
     os2.base_ios.state = IOSTATE_badbit;
     ret = (int) call_func3(p_streambuf_xsputn, &fb1.base, "We've known each other", 22);
@@ -2979,7 +2983,9 @@ if (0) /* crashes on native */
     ok(ret == 1, "expected 1 got %d\n", ret);
     os1.base_ios.sb = NULL;
 if (0) /* crashes on native */
+{
     call_func1(p_ostream_osfx, &os1);
+}
     os1.base_ios.sb = &fb1.base;
     os1.base_ios.flags = FLAGS_unitbuf;
     call_func1(p_ostream_osfx, &os1);
@@ -3019,7 +3025,9 @@ if (0) /* crashes on native */
     ok(fb1.base.pptr == fb1.base.base + 1, "wrong put pointer, expected %p got %p\n", fb1.base.base + 1, fb1.base.pptr);
     os1.base_ios.sb = NULL;
 if (0) /* crashes on native */
+{
     pos = call_func2(p_ostream_put_char, &os1, 'c');
+}
     os1.base_ios.sb = &fb1.base;
     os1.base_ios.width = 5;
     call_func1(p_filebuf_sync, &fb1);
@@ -3051,7 +3059,9 @@ if (0) /* crashes on native */
     ok(fb1.base.pptr == fb1.base.base + 7, "wrong put pointer, expected %p got %p\n", fb1.base.base + 7, fb1.base.pptr);
     os1.base_ios.sb = NULL;
 if (0) /* crashes on native */
+{
     pos = call_func3(p_ostream_write_char, &os1, "been", 4);
+}
     os1.base_ios.sb = &fb1.base;
     os1.base_ios.width = 5;
     call_func1(p_filebuf_sync, &fb1);
@@ -3167,7 +3177,9 @@ if (0) /* crashes on native */
     ok(pos == &os1, "wrong return, expected %p got %p\n", &os1, pos);
     ok(!strncmp(fb1.base.pptr - 6, "abzzzz", 6), "expected 'abzzzz' got '%s'\n", fb1.base.pptr - 6);
 if (0) /* crashes on native */
+{
     pos = call_func3(p_ostream_writepad, &os1, NULL, "o");
+}
     pos = call_func3(p_ostream_writepad, &os1, "", "hello");
     ok(pos == &os1, "wrong return, expected %p got %p\n", &os1, pos);
     ok(!strncmp(fb1.base.pptr - 6, "helloz", 6), "expected 'helloz' got '%s'\n", fb1.base.pptr - 6);
@@ -3580,7 +3592,9 @@ static void test_istream(void)
     ok(is2.base_ios.precision == 6, "expected 6 got %d\n", is2.base_ios.precision);
     ok(is2.base_ios.width == 0, "expected 0 got %d\n", is2.base_ios.width);
 if (0) /* crashes on native */
+{
     pis = call_func2(p_istream_assign, &is2, NULL);
+}
     is2.extract_delim = is2.count = 0xabababab;
     is2.base_ios.sb = (streambuf*) 0xabababab;
     is2.base_ios.state = 0xabababab;
@@ -3616,7 +3630,9 @@ if (0) /* crashes on native */
     /* eatwhite */
     is1.extract_delim = is1.count = 0;
 if (0) /* crashes on native */
+{
     is1.base_ios.sb = NULL;
+}
     is1.base_ios.state = IOSTATE_badbit;
     is1.base_ios.flags = 0;
     call_func1(p_istream_eatwhite, &is1);
@@ -3684,7 +3700,9 @@ if (0) /* crashes on native */
     ok(is1.base_ios.state == (IOSTATE_badbit|IOSTATE_failbit), "expected %d got %d\n",
         IOSTATE_badbit|IOSTATE_failbit, is1.base_ios.state);
 if (0) /* crashes on native */
+{
     is1.base_ios.sb = NULL;
+}
     is1.base_ios.state = IOSTATE_goodbit;
     is1.base_ios.tie = &os;
     pos = call_func3(p_ostream_sb_ctor, &os, &fb2.base, TRUE);
@@ -3951,7 +3969,9 @@ if (0) /* crashes on native */
     ok(fb1.base.gptr == NULL, "wrong get pointer, expected %p got %p\n", NULL, fb1.base.gptr);
     ok(buffer[0] == 0, "expected 0 got %d\n", buffer[0]);
 if (0) /* crashes on native */
+{
     pis = call_func4(p_istream_get_str_delim, &is1, (char*) 0x1, 5, 0);
+}
 
     /* get_str */
     is1.extract_delim = is1.count = 0xabababab;
@@ -4089,7 +4109,9 @@ if (0) /* crashes on native */
     ok(fb1.base.gptr == fb1.base.base + 3, "wrong get pointer, expected %p got %p\n", fb1.base.base + 3, fb1.base.gptr);
     ok(c == -50, "expected %d got %d\n", -50, c);
 if (0) /* crashes on native */
+{
     pis = call_func2(p_istream_get_char, &is1, NULL);
+}
     fb1.base.gptr = fb1.base.base + 30;
     pis = call_func2(p_istream_get_char, &is1, &c);
     ok(pis == &is1, "wrong return, expected %p got %p\n", &is1, pis);
@@ -4201,7 +4223,9 @@ if (0) /* crashes on native */
     ok(fb2.base.epptr == NULL, "wrong put end, expected %p got %p\n", NULL, fb2.base.epptr);
     is1.base_ios.state = IOSTATE_goodbit;
 if (0) /* crashes on native */
+{
     pis = call_func3(p_istream_get_sb, &is1, NULL, '?');
+}
     *fb1.base.gptr = -50;
     pis = call_func3(p_istream_get_sb, &is1, &fb2.base, -50);
     ok(pis == &is1, "wrong return, expected %p got %p\n", &is1, pis);
@@ -4395,7 +4419,9 @@ if (0) /* crashes on native */
     ok(is1.base_ios.state == IOSTATE_goodbit, "expected %d got %d\n", IOSTATE_goodbit, is1.base_ios.state);
     ok(fb1.base.gptr == fb1.base.base, "wrong get pointer, expected %p got %p\n", fb1.base.base, fb1.base.gptr);
 if (0) /* crashes on native */
+{
     is1.base_ios.sb = NULL;
+}
     fb1.base.gptr = fb1.base.base + 14;
     ret = (int) call_func1(p_istream_peek, &is1);
     ok(ret == 206, "expected 206 got %d\n", ret);
@@ -4440,7 +4466,9 @@ if (0) /* crashes on native */
     ok(fb1.base.gptr == fb1.base.base + 15, "wrong get pointer, expected %p got %p\n", fb1.base.base + 15, fb1.base.gptr);
     is1.base_ios.state = IOSTATE_goodbit;
 if (0) /* crashes on native */
+{
     is1.base_ios.sb = NULL;
+}
     pis = call_func2(p_istream_putback, &is1, -40);
     ok(pis == &is1, "wrong return, expected %p got %p\n", &is1, pis);
     ok(is1.base_ios.state == IOSTATE_goodbit, "expected %d got %d\n", IOSTATE_goodbit, is1.base_ios.state);
@@ -4498,13 +4526,17 @@ if (0) /* crashes on native */
     ok(!strncmp(buffer, fb1.base.base, 30), "unexpected buffer content, got '%s'\n", buffer);
     ok(buffer[30] == 'A', "expected 'A' got %d\n", buffer[30]);
 if (0) /* crashes on native */
+{
     is1.base_ios.sb = NULL;
+}
     is1.base_ios.state = IOSTATE_goodbit;
     fb1.base.eback = fb1.base.gptr = fb1.base.base;
     fb1.base.egptr = fb1.base.base + 30;
     memset(buffer, 'A', sizeof(buffer));
 if (0) /* crashes on native */
+{
     pis = call_func3(p_istream_read, &is1, buffer, -1);
+}
     pis = call_func3(p_istream_read, &is1, buffer, 0);
     ok(pis == &is1, "wrong return, expected %p got %p\n", &is1, pis);
     ok(is1.count == 0, "expected 0 got %d\n", is1.count);
@@ -4520,7 +4552,9 @@ if (0) /* crashes on native */
     ok(fb1.base.gptr == NULL, "wrong get pointer, expected %p got %p\n", NULL, fb1.base.gptr);
     ok(_tell(fb1.fd) == 0, "expected 0 got %d\n", _tell(fb1.fd));
 if (0) /* crashes on native */
+{
     is1.base_ios.sb = NULL;
+}
     pis = call_func2(p_istream_seekg, &is1, -5);
     ok(pis == &is1, "wrong return, expected %p got %p\n", &is1, pis);
     ok(is1.base_ios.state == IOSTATE_failbit, "expected %d got %d\n", IOSTATE_failbit, is1.base_ios.state);
@@ -4553,7 +4587,9 @@ if (0) /* crashes on native */
     ok(is1.base_ios.state == IOSTATE_failbit, "expected %d got %d\n", IOSTATE_failbit, is1.base_ios.state);
     ok(fb1.base.gptr == fb1.base.base, "wrong get pointer, expected %p got %p\n", fb1.base.base, fb1.base.gptr);
 if (0) /* crashes on native */
+{
     is1.base_ios.sb = NULL;
+}
     fb1.base.gptr = fb1.base.egptr;
     pis = call_func3(p_istream_seekg_offset, &is1, 0, SEEKDIR_end);
     ok(pis == &is1, "wrong return, expected %p got %p\n", &is1, pis);
@@ -4574,7 +4610,9 @@ if (0) /* crashes on native */
     ok(ret == 0, "expected 0 got %d\n", ret);
     ok(is1.base_ios.state == IOSTATE_goodbit, "expected %d got %d\n", IOSTATE_goodbit, is1.base_ios.state);
 if (0) /* crashes on native */
+{
     is1.base_ios.sb = NULL;
+}
     fb1.base.eback = fb1.base.gptr = fb1.base.base;
     fb1.base.egptr = fb1.base.base + 30;
     ret = (int) call_func1(p_istream_sync, &is1);
@@ -4607,7 +4645,9 @@ if (0) /* crashes on native */
     ok(ret == 24, "expected 24 got %d\n", ret);
     ok(is1.base_ios.state == IOSTATE_goodbit, "expected %d got %d\n", IOSTATE_goodbit, is1.base_ios.state);
 if (0) /* crashes on native */
+{
     is1.base_ios.sb = NULL;
+}
     fb1.base.eback = fb1.base.gptr = fb1.base.base;
     fb1.base.egptr = fb1.base.base + 30;
     ret = (int) call_func1(p_istream_tellg, &is1);
-- 
2.9.0

