From 3e59710a3091a4a61b7cce00606ed23b7b66dfda Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 4 Jun 2017 12:55:31 +0200
Subject: fusion/tests: Avoid compiler warnings with GCC 7.

---
 dlls/fusion/tests/asmenum.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/fusion/tests/asmenum.c b/dlls/fusion/tests/asmenum.c
index 1dc34a286e5..8b95f2a6bd5 100644
--- a/dlls/fusion/tests/asmenum.c
+++ b/dlls/fusion/tests/asmenum.c
@@ -223,7 +223,7 @@ typedef struct _tagASMNAME
 static BOOL enum_gac_assemblies(struct list *assemblies, int depth, LPSTR path)
 {
     WIN32_FIND_DATAA ffd;
-    CHAR buf[MAX_PATH];
+    CHAR buf[MAX_PATH + 37];
     CHAR disp[MAX_PATH];
     ASMNAME *name;
     HANDLE hfind;
@@ -248,7 +248,7 @@ static BOOL enum_gac_assemblies(struct list *assemblies, int depth, LPSTR path)
         else if (depth == 1)
         {
             char culture[MAX_PATH];
-            char dll[MAX_PATH], exe[MAX_PATH];
+            char dll[MAX_PATH + 6], exe[MAX_PATH + 6];
 
             /* Directories with no dll or exe will not be enumerated */
             sprintf(dll, "%s\\%s\\%s.dll", path, ffd.cFileName, parent);
-- 
2.13.0

