From d93228af9587d69553082d48523a49e74a536704 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 30 Jul 2016 23:57:11 +0200
Subject: oleaut32/tests: Avoid misleading indentation warnings.

---
 dlls/oleaut32/tests/safearray.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/oleaut32/tests/safearray.c b/dlls/oleaut32/tests/safearray.c
index f57b886..787891c 100644
--- a/dlls/oleaut32/tests/safearray.c
+++ b/dlls/oleaut32/tests/safearray.c
@@ -706,7 +706,9 @@ static void test_safearray(void)
     }
 
     if (!pSafeArrayAllocDescriptorEx)
+    {
         return;
+    }
 
 	for (i=0;i<sizeof(vttypes)/sizeof(vttypes[0]);i++) {
 		a = NULL;
-- 
2.9.0

