From 818e3ecdcf86444a2717b135a3316eb938c726a2 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 4 Jun 2017 12:56:10 +0200
Subject: kernel32/tests: Avoid compiler warnings with GCC 7.

---
 dlls/kernel32/tests/heap.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/kernel32/tests/heap.c b/dlls/kernel32/tests/heap.c
index 0acf109dff0..5fee6e39fc3 100644
--- a/dlls/kernel32/tests/heap.c
+++ b/dlls/kernel32/tests/heap.c
@@ -109,6 +109,8 @@ static void test_heap(void)
     }
 
     /* test some border cases of HeapAlloc and HeapReAlloc */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Walloc-size-larger-than="
     mem = HeapAlloc(GetProcessHeap(), 0, 0);
     ok(mem != NULL, "memory not allocated for size 0\n");
     msecond = HeapReAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, mem, ~(SIZE_T)0 - 7);
@@ -125,6 +127,7 @@ static void test_heap(void)
     ok(size == 0 || broken(size == 1) /* some vista and win7 */,
        "HeapSize should have returned 0 instead of %lu\n", size);
     HeapFree(GetProcessHeap(), 0, msecond);
+#pragma GCC diagnostic pop
 
     /* large blocks must be 16-byte aligned */
     mem = HeapAlloc(GetProcessHeap(), 0, 512 * 1024);
-- 
2.13.0

