From 9bda354ab7636fb357f29f777892de3c39ff06ae Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 30 Jul 2016 23:57:37 +0200
Subject: rpcrt4/tests: Avoid misleading indentation warnings.

---
 dlls/rpcrt4/tests/ndr_marshall.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/rpcrt4/tests/ndr_marshall.c b/dlls/rpcrt4/tests/ndr_marshall.c
index aeecf8c..b71f07b 100644
--- a/dlls/rpcrt4/tests/ndr_marshall.c
+++ b/dlls/rpcrt4/tests/ndr_marshall.c
@@ -2434,7 +2434,9 @@ static void test_MesEncodeFixedBufferHandleCreate(void)
 todo_wine
     ok(status == RPC_S_INVALID_ARG, "got %d\n", status);
 if (status == RPC_S_OK)
+{
     MesHandleFree(handle);
+}
 
     status = MesEncodeFixedBufferHandleCreate(buffer, 32, NULL, &handle);
     ok(status == RPC_S_INVALID_ARG, "got %d\n", status);
-- 
2.9.0

