From b79a3f64ec023a17955620362c748320c17a9ed0 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 30 Jul 2016 23:57:59 +0200
Subject: scrrun/tests: Avoid misleading indentation warnings.

---
 dlls/scrrun/tests/dictionary.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/scrrun/tests/dictionary.c b/dlls/scrrun/tests/dictionary.c
index a1b7d06..1880c20 100644
--- a/dlls/scrrun/tests/dictionary.c
+++ b/dlls/scrrun/tests/dictionary.c
@@ -102,7 +102,9 @@ static void test_comparemode(void)
     ok(hr == S_OK, "got 0x%08x\n", hr);
 
 if (0) /* crashes on native */
+{
     hr = IDictionary_get_CompareMode(dict, NULL);
+}
 
     method = 10;
     hr = IDictionary_get_CompareMode(dict, &method);
@@ -934,7 +936,9 @@ static void test_IEnumVARIANT(void)
     ok(hr == S_OK, "got 0x%08x\n", hr);
 
 if (0) /* crashes on native */
+{
     hr = IDictionary__NewEnum(dict, NULL);
+}
 
     hr = IDictionary__NewEnum(dict, &enum1);
     ok(hr == S_OK, "got 0x%08x\n", hr);
-- 
2.9.0

