From 67ed53ca67214d7dbdea7f342b9f2ff0356a220b Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 4 Jun 2017 12:59:10 +0200
Subject: kernel32: Avoid compiler warnings with GCC 7.

---
 dlls/kernel32/oldconfig.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/dlls/kernel32/oldconfig.c b/dlls/kernel32/oldconfig.c
index 6c80dc614f0..3e628757283 100644
--- a/dlls/kernel32/oldconfig.c
+++ b/dlls/kernel32/oldconfig.c
@@ -290,7 +290,8 @@ static void create_hardware_branch(void)
         {
             if (strncmp(dent->d_name, "hd", 2) == 0)
             {
-                sprintf(cStr, procname_ide_media, dent->d_name);
+                result = snprintf(cStr, sizeof(cStr), procname_ide_media, dent->d_name);
+                if (result < 0 || result >= sizeof(cStr)) continue;
                 procfile = fopen(cStr, "r");
                 if (!procfile)
                 {
@@ -306,7 +307,8 @@ static void create_hardware_branch(void)
                     if (nType == DRIVE_UNKNOWN) continue;
                 }
 
-                sprintf(cStr, procname_ide_model, dent->d_name);
+                result = snprintf(cStr, sizeof(cStr), procname_ide_model, dent->d_name);
+                if (result < 0 || result >= sizeof(cStr)) continue;
                 procfile = fopen(cStr, "r");
                 if (!procfile)
                 {
@@ -322,7 +324,8 @@ static void create_hardware_branch(void)
                     cDevModel[strlen(cDevModel) - 1] = 0;
                 }
 
-                sprintf(cUnixDeviceName, "/dev/%s", dent->d_name);
+                result = snprintf(cUnixDeviceName, sizeof(cUnixDeviceName), "/dev/%s", dent->d_name);
+                if (result < 0 || result >= sizeof(cUnixDeviceName)) continue;
                 scsi_addr.PortNumber = (dent->d_name[2] - 'a') / 2;
                 scsi_addr.PathId = 0;
                 scsi_addr.TargetId = (dent->d_name[2] - 'a') % 2;
-- 
2.13.0

