From 173e27263954c449c3657701e9d7489c603f1306 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 30 Jul 2016 23:58:37 +0200
Subject: shell32/tests: Avoid misleading indentation warnings.

---
 dlls/shell32/tests/shellole.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/shell32/tests/shellole.c b/dlls/shell32/tests/shellole.c
index e8cf744..99f0ccb 100644
--- a/dlls/shell32/tests/shellole.c
+++ b/dlls/shell32/tests/shellole.c
@@ -872,7 +872,9 @@ static void test_SHCreateSessionKey(void)
     }
 
 if (0) /* crashes on native */
+{
     hr = pSHCreateSessionKey(KEY_READ, NULL);
+}
 
     hkey = (HKEY)0xdeadbeef;
     hr = pSHCreateSessionKey(0, &hkey);
-- 
2.9.0

