From f590ced17854c6830cd15b63e757ef5798095cb1 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 30 Jul 2016 23:58:52 +0200
Subject: shlwapi/tests: Avoid misleading indentation warnings.

---
 dlls/shlwapi/tests/ordinal.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/shlwapi/tests/ordinal.c b/dlls/shlwapi/tests/ordinal.c
index 74cad7d..b26fd4f 100644
--- a/dlls/shlwapi/tests/ordinal.c
+++ b/dlls/shlwapi/tests/ordinal.c
@@ -3225,7 +3225,9 @@ static void test_IUnknown_GetClassID(void)
     HRESULT hr;
 
 if (0) /* crashes on native systems */
+{
     hr = pIUnknown_GetClassID(NULL, NULL);
+}
 
     memset(&clsid, 0xcc, sizeof(clsid));
     memset(&clsid3, 0xcc, sizeof(clsid3));
-- 
2.9.0

