From 74a5d7a75aaae754ad64fd24947ba3f17d92a00a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 30 Jul 2016 23:59:27 +0200
Subject: winhttp/tests: Avoid misleading indentation warnings.

---
 dlls/winhttp/tests/winhttp.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/winhttp/tests/winhttp.c b/dlls/winhttp/tests/winhttp.c
index 6be7598..d29457a 100644
--- a/dlls/winhttp/tests/winhttp.c
+++ b/dlls/winhttp/tests/winhttp.c
@@ -3823,7 +3823,9 @@ static void test_IWinHttpRequest_Invoke(void)
 
     VariantInit(&ret);
 if (0) /* crashes */
+{
     hr = IWinHttpRequest_Invoke(request, DISPID_HTTPREQUEST_OPTION, &IID_NULL, 0, DISPATCH_PROPERTYPUT, NULL, &ret, NULL, &err);
+}
 
     params.cArgs = 1;
     hr = IWinHttpRequest_Invoke(request, DISPID_HTTPREQUEST_OPTION, &IID_NULL, 0, DISPATCH_PROPERTYPUT, &params, &ret, NULL, &err);
-- 
2.9.0

