From ec626b0fe26ec342091e9db61e78d77acf16f00a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 30 Jul 2016 23:59:41 +0200
Subject: wshom.ocx/tests: Avoid misleading indentation warnings.

---
 dlls/wshom.ocx/tests/wshom.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/wshom.ocx/tests/wshom.c b/dlls/wshom.ocx/tests/wshom.c
index a3d398e..5b6ffd0 100644
--- a/dlls/wshom.ocx/tests/wshom.c
+++ b/dlls/wshom.ocx/tests/wshom.c
@@ -214,7 +214,9 @@ static void test_wshshell(void)
 
     /* current directory */
 if (0) /* crashes on native */
+{
     hr = IWshShell3_get_CurrentDirectory(sh3, NULL);
+}
 
     str = NULL;
     hr = IWshShell3_get_CurrentDirectory(sh3, &str);
-- 
2.9.0

