From 337b1e60966bf902e79d69981c1bf49bf8be6192 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 31 Jul 2016 00:01:24 +0200
Subject: xaudio2_7: Only declare debug channel when needed.

---
 dlls/xaudio2_7/compat.c   | 2 ++
 dlls/xaudio2_7/x3daudio.c | 2 ++
 2 files changed, 4 insertions(+)

diff --git a/dlls/xaudio2_7/compat.c b/dlls/xaudio2_7/compat.c
index 516d6c2..b9e720f 100644
--- a/dlls/xaudio2_7/compat.c
+++ b/dlls/xaudio2_7/compat.c
@@ -101,7 +101,9 @@
 
 #include "wine/debug.h"
 
+#if XAUDIO2_VER <= 7
 WINE_DEFAULT_DEBUG_CHANNEL(xaudio2);
+#endif
 
 /* BEGIN IXAudio2SourceVoice */
 #if XAUDIO2_VER == 0
diff --git a/dlls/xaudio2_7/x3daudio.c b/dlls/xaudio2_7/x3daudio.c
index a182852..58f7062 100644
--- a/dlls/xaudio2_7/x3daudio.c
+++ b/dlls/xaudio2_7/x3daudio.c
@@ -23,7 +23,9 @@
 
 #include "wine/debug.h"
 
+#if XAUDIO2_VER >= 8 || defined X3DAUDIO1_VER
 WINE_DEFAULT_DEBUG_CHANNEL(xaudio2);
+#endif
 
 #ifdef X3DAUDIO1_VER
 BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD reason, void *pReserved)
-- 
2.9.0

