From ced83fdea5e467009d8a4220fab97fb2e7e4e008 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 31 Jul 2016 00:02:13 +0200
Subject: winetest: Remove duplicate if check.

---
 programs/winetest/gui.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/programs/winetest/gui.c b/programs/winetest/gui.c
index a8bcf05..2948d7e 100644
--- a/programs/winetest/gui.c
+++ b/programs/winetest/gui.c
@@ -547,8 +547,7 @@ report (enum report_type t, ...)
     }
         
     va_start (ap, t);
-    if (t < sizeof text_funcs / sizeof text_funcs[0] &&
-        t < sizeof GUI_funcs / sizeof GUI_funcs[0]) ret = funcs[t](ap);
+    if (t < sizeof text_funcs / sizeof text_funcs[0]) ret = funcs[t](ap);
     else report (R_WARNING, "unimplemented report type: %d", t);
     va_end (ap);
     return ret;
-- 
2.9.0

