From e96dbcea16b0f61eeb3547688613774e2c8b778e Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 14 Jan 2016 19:08:08 +0100
Subject: rpcrt4: Fix prototype of RpcBindingServerFromClient.

---
 dlls/rpcrt4/rpc_binding.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/rpcrt4/rpc_binding.c b/dlls/rpcrt4/rpc_binding.c
index 7212a18..2503d31 100644
--- a/dlls/rpcrt4/rpc_binding.c
+++ b/dlls/rpcrt4/rpc_binding.c
@@ -1636,8 +1636,8 @@ RpcBindingInqAuthClientExW( RPC_BINDING_HANDLE ClientBinding, RPC_AUTHZ_HANDLE *
 /***********************************************************************
  *             RpcBindingServerFromClient (RPCRT4.@)
  */
-
-RPC_STATUS RPC_ENTRY RpcBindingServerFromClient(RPC_BINDING_HANDLE ClientBinding, RPC_BINDING_HANDLE *ServerBinding)
+RPCRTAPI RPC_STATUS RPC_ENTRY
+RpcBindingServerFromClient( RPC_BINDING_HANDLE ClientBinding, RPC_BINDING_HANDLE *ServerBinding )
 {
     FIXME("%p %p: stub\n", ClientBinding, ServerBinding);
     return RPC_S_INVALID_BINDING;
-- 
2.6.4

