From 879f9aff357d593c169f2b59728de63792ec7f00 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 21 Aug 2014 05:57:24 +0200
Subject: winebuild: Set a valid major and minor linker version

---
 tools/winebuild/spec32.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tools/winebuild/spec32.c b/tools/winebuild/spec32.c
index 35fe144..c2a3675 100644
--- a/tools/winebuild/spec32.c
+++ b/tools/winebuild/spec32.c
@@ -529,8 +529,8 @@ void output_module( DLLSPEC *spec )
              spec->characteristics );
     output( "\t.short 0x%04x\n",          /* Magic */
              get_ptr_size() == 8 ? IMAGE_NT_OPTIONAL_HDR64_MAGIC : IMAGE_NT_OPTIONAL_HDR32_MAGIC );
-    output( "\t.byte 0\n" );              /* MajorLinkerVersion */
-    output( "\t.byte 0\n" );              /* MinorLinkerVersion */
+    output( "\t.byte 7\n" );              /* MajorLinkerVersion */
+    output( "\t.byte 10\n" );              /* MinorLinkerVersion */
     output( "\t.long 0\n" );              /* SizeOfCode */
     output( "\t.long 0\n" );              /* SizeOfInitializedData */
     output( "\t.long 0\n" );              /* SizeOfUninitializedData */
-- 
1.9.1

