From aa9949cef21f63e2f5ba93515c4e1e17b66b1abf Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Thu, 16 Oct 2014 22:36:58 +0200
Subject: include: Fix definition of SECTION_BASIC_INFORMATION and
 SECTION_IMAGE_INFORMATION.

---
 include/winternl.h | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/include/winternl.h b/include/winternl.h
index 5a27f94..bb27a21 100644
--- a/include/winternl.h
+++ b/include/winternl.h
@@ -1792,7 +1792,7 @@ typedef enum _SECTION_INFORMATION_CLASS
 } SECTION_INFORMATION_CLASS;
 
 typedef struct _SECTION_BASIC_INFORMATION {
-  ULONG BaseAddress;
+  PVOID BaseAddress;
   ULONG Attributes;
   LARGE_INTEGER Size;
 } SECTION_BASIC_INFORMATION, *PSECTION_BASIC_INFORMATION;
@@ -1800,14 +1800,16 @@ typedef struct _SECTION_BASIC_INFORMATION {
 typedef struct _SECTION_IMAGE_INFORMATION {
   PVOID EntryPoint;
   ULONG StackZeroBits;
-  ULONG StackReserved;
-  ULONG StackCommit;
+  SIZE_T StackReserved;
+  SIZE_T StackCommit;
   ULONG ImageSubsystem;
   WORD SubsystemVersionLow;
   WORD SubsystemVersionHigh;
   ULONG Unknown1;
-  ULONG ImageCharacteristics;
-  ULONG ImageMachineType;
+  USHORT ImageCharacteristics;
+  USHORT DllCharacteristics;
+  USHORT ImageMachineType;
+  BOOLEAN ImageContainsCode;
   ULONG Unknown2[3];
 } SECTION_IMAGE_INFORMATION, *PSECTION_IMAGE_INFORMATION;
 
-- 
2.1.2

