From caa20a89174dcf17cd2bd561f20841d4cfb0b7bb Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Mon, 14 Jan 2019 11:07:43 +1100
Subject: [PATCH 2/6] include: Add VSS writer enum

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 include/vswriter.h | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/include/vswriter.h b/include/vswriter.h
index 043b871..ac23861 100644
--- a/include/vswriter.h
+++ b/include/vswriter.h
@@ -44,4 +44,19 @@ typedef enum
     VSS_AWS_THIS_IS_ALTERNATE_WRITER
 } VSS_ALTERNATE_WRITER_STATE;
 
+typedef enum VSS_COMPONENT_TYPE
+{
+    VSS_CT_UNDEFINED = 0,
+    VSS_CT_DATABASE,
+    VSS_CT_FILEGROUP
+} VSS_COMPONENT_TYPE;
+
+typedef enum VSS_FILE_RESTORE_STATUS
+{
+    VSS_RS_UNDEFINED = 0,
+    VSS_RS_NONE,
+    VSS_RS_ALL,
+    VSS_RS_FAILED
+} VSS_FILE_RESTORE_STATUS;
+
 #endif /* ___WINE_VSWRITER_H */
-- 
1.9.1

