From a3867ccbd2fc783982263556d4b9ca525e1ef389 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 28 Sep 2017 07:02:55 +0200
Subject: include: Always define hton/ntoh macros.

---
 include/winsock.h | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/include/winsock.h b/include/winsock.h
index 8b8db83c15b..86ce4ffcf05 100644
--- a/include/winsock.h
+++ b/include/winsock.h
@@ -510,8 +510,6 @@ int WINAPI __WSAFDIsSet(SOCKET,WS(fd_set)*);
 #endif /* WS_DEFINE_SELECT */
 
 /* we have to define hton/ntoh as macros to avoid conflicts with Unix headers */
-#ifndef USE_WS_PREFIX
-
 #undef htonl
 #undef htons
 #undef ntohl
@@ -549,8 +547,6 @@ static inline ULONG __wine_ulong_swap(ULONG l)
 
 #endif  /* WORDS_BIGENDIAN */
 
-#endif  /* USE_WS_PREFIX */
-
 /*
  * Internet address (old style... should be updated)
  */
-- 
2.14.1

