From 17acd93aa2405f5c5032f982e77174e4133c8a36 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Mon, 16 May 2016 22:34:53 +0200
Subject: wininet: Set lpszUrlPath to the end of the string in
 InternetCrackUrlW when dwUrlPathLength > 0.

---
 dlls/wininet/internet.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/wininet/internet.c b/dlls/wininet/internet.c
index 5359794..8149973 100644
--- a/dlls/wininet/internet.c
+++ b/dlls/wininet/internet.c
@@ -1941,6 +1941,8 @@ BOOL WINAPI InternetCrackUrlW(const WCHAR *lpszUrl, DWORD dwUrlLength, DWORD dwF
     {
         if (lpUC->lpszUrlPath && (lpUC->dwUrlPathLength > 0))
             lpUC->lpszUrlPath[0] = 0;
+        else if (lpUC->dwUrlPathLength > 0)
+            lpUC->lpszUrlPath = (WCHAR*)lpszcp;
         lpUC->dwUrlPathLength = 0;
     }
 
-- 
2.8.0

