From d359c245dd87648c1f2002c2a865271c4c9aff8f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sun, 28 May 2017 05:45:01 +0200
Subject: d3d11: Silence ID3D11Device_GetDeviceRemovedReason.

---
 dlls/d3d11/device.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/d3d11/device.c b/dlls/d3d11/device.c
index bc28b4a5446..c78a4a755d8 100644
--- a/dlls/d3d11/device.c
+++ b/dlls/d3d11/device.c
@@ -5749,7 +5749,7 @@ static UINT STDMETHODCALLTYPE d3d11_device_GetCreationFlags(ID3D11Device *iface)
 
 static HRESULT STDMETHODCALLTYPE d3d11_device_GetDeviceRemovedReason(ID3D11Device *iface)
 {
-    FIXME("iface %p stub!\n", iface);
+    TRACE("iface %p stub!\n", iface);
 
     return S_OK;
 }
-- 
2.12.2

