From 67bf62879edea7ee4d6c6f7218c8fa6b92c66163 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sun, 28 May 2017 05:45:31 +0200
Subject: d3d11: Silence depth bias warning in ID3D11DeviceContext_RSSetState.

---
 dlls/d3d11/device.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/d3d11/device.c b/dlls/d3d11/device.c
index c78a4a755d8..14a0b98a32d 100644
--- a/dlls/d3d11/device.c
+++ b/dlls/d3d11/device.c
@@ -1724,7 +1724,7 @@ static void STDMETHODCALLTYPE d3d11_immediate_context_RSSetState(ID3D11DeviceCon
     wined3d_device_set_render_state(device->wined3d_device, WINED3D_RS_CULLMODE, desc->CullMode);
     /* OpenGL style depth bias. */
     if (desc->DepthBias || desc->SlopeScaledDepthBias)
-        FIXME("Ignoring depth bias.\n");
+        WARN("Ignoring depth bias.\n");
     /* GL_DEPTH_CLAMP */
     if (!desc->DepthClipEnable)
         FIXME("Ignoring DepthClipEnable %#x.\n", desc->DepthClipEnable);
-- 
2.12.2

