From 83df2a25df1f0a9e057fbac19c2c1601373b9c38 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 11 Feb 2016 19:07:36 +0100
Subject: wtsapi32: Improve WTSQueryUserToken stub.

---
 dlls/wtsapi32/wtsapi32.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/wtsapi32/wtsapi32.c b/dlls/wtsapi32/wtsapi32.c
index c668604..e1a9cd2 100644
--- a/dlls/wtsapi32/wtsapi32.c
+++ b/dlls/wtsapi32/wtsapi32.c
@@ -243,7 +243,8 @@ BOOL WINAPI WTSQuerySessionInformationW(
 BOOL WINAPI WTSQueryUserToken(ULONG session_id, PHANDLE token)
 {
     FIXME("%u %p\n", session_id, token);
-    return FALSE;
+    if (token) *token = GetCurrentProcessToken();
+    return token != NULL;
 }
 
 /************************************************************
-- 
2.7.1

