From 21a9d3d446f2c30e7f3c2bd3208c3f077c49b7f9 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Tue, 15 Mar 2016 12:32:38 +0800
Subject: tools/wrc: Fix generation of custom dialog control data.

---
 tools/wrc/genres.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/tools/wrc/genres.c b/tools/wrc/genres.c
index 5b7e24c..b97717a 100644
--- a/tools/wrc/genres.c
+++ b/tools/wrc/genres.c
@@ -726,8 +726,7 @@ static res_t *dialog2res(name_id_t *name, dialog_t *dlg)
 				put_word(res, 0);
 			if(ctrl->extra)
 			{
-				put_word(res, ctrl->extra->size+2);
-				put_pad(res);
+				put_word(res, ctrl->extra->size);
 				put_raw_data(res, ctrl->extra, 0);
 			}
 			else
-- 
2.7.1

