From 268ba3b1b3d37eed28e68ec6a64c568d8de7cde2 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 31 Mar 2016 22:33:14 +0200
Subject: winex11: Add 320x240 as supported resolution in desktop mode.

---
 dlls/winex11.drv/desktop.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/winex11.drv/desktop.c b/dlls/winex11.drv/desktop.c
index 72916c8..034d7f3 100644
--- a/dlls/winex11.drv/desktop.c
+++ b/dlls/winex11.drv/desktop.c
@@ -37,8 +37,8 @@ static unsigned int dd_mode_count;
 static unsigned int max_width;
 static unsigned int max_height;
 
-static const unsigned int widths[]  = {320, 400, 512, 640, 800, 1024, 1152, 1280, 1280, 1400, 1600};
-static const unsigned int heights[] = {200, 300, 384, 480, 600,  768,  864,  960, 1024, 1050, 1200};
+static const unsigned int widths[]  = {320, 320, 400, 512, 640, 800, 1024, 1152, 1280, 1280, 1400, 1600};
+static const unsigned int heights[] = {200, 240, 300, 384, 480, 600,  768,  864,  960, 1024, 1050, 1200};
 #define NUM_DESKTOP_MODES (sizeof(widths) / sizeof(widths[0]))
 
 #define _NET_WM_STATE_REMOVE 0
-- 
2.7.1

