From eef0d4fa08355377548c101283b62998dc10884e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sat, 31 May 2014 02:08:45 +0200
Subject: ntdll: Stub TokenAppContainerSid in NtQueryInformationToken

---
 dlls/ntdll/nt.c | 21 ++++++++++++++++++++-
 include/winnt.h | 19 +++++++++++++++++++
 2 files changed, 39 insertions(+), 1 deletion(-)

diff --git a/dlls/ntdll/nt.c b/dlls/ntdll/nt.c
index ce648de..b9d8a52 100644
--- a/dlls/ntdll/nt.c
+++ b/dlls/ntdll/nt.c
@@ -279,7 +279,19 @@ NTSTATUS WINAPI NtQueryInformationToken(
         sizeof(TOKEN_MANDATORY_LABEL) + sizeof(SID), /* TokenIntegrityLevel [sizeof(SID) includes one SubAuthority] */
         0,    /* TokenUIAccess */
         0,    /* TokenMandatoryPolicy */
-        0     /* TokenLogonSid */
+        0,    /* TokenLogonSid */
+        0,    /* TokenIsAppContainer */
+        0,    /* TokenCapabilities */
+        sizeof(TOKEN_APPCONTAINER_INFORMATION) + sizeof(SID), /* TokenAppContainerSid */
+        0,    /* TokenAppContainerNumber */
+        0,    /* TokenUserClaimAttributes*/
+        0,    /* TokenDeviceClaimAttributes */
+        0,    /* TokenRestrictedUserClaimAttributes */
+        0,    /* TokenRestrictedDeviceClaimAttributes */
+        0,    /* TokenDeviceGroups */
+        0,    /* TokenRestrictedDeviceGroups */
+        0,    /* TokenSecurityAttributes */
+        0     /* TokenIsRestricted */
     };
 
     ULONG len = 0;
@@ -521,6 +533,13 @@ NTSTATUS WINAPI NtQueryInformationToken(
             memcpy(psid, &high_level, sizeof(SID));
         }
         break;
+    case TokenAppContainerSid:
+        {
+            TOKEN_APPCONTAINER_INFORMATION *container = tokeninfo;
+            FIXME("QueryInformationToken( ..., TokenAppContainerSid, ...) semi-stub\n");
+            container->TokenAppContainer = NULL;
+        }
+        break;
     default:
         {
             ERR("Unhandled Token Information class %d!\n", tokeninfoclass);
diff --git a/include/winnt.h b/include/winnt.h
index f902ba2..7934022 100644
--- a/include/winnt.h
+++ b/include/winnt.h
@@ -3685,6 +3685,18 @@ typedef enum _TOKEN_INFORMATION_CLASS {
   TokenUIAccess,
   TokenMandatoryPolicy,
   TokenLogonSid,
+  TokenIsAppContainer,
+  TokenCapabilities,
+  TokenAppContainerSid,
+  TokenAppContainerNumber,
+  TokenUserClaimAttributes,
+  TokenDeviceClaimAttributes,
+  TokenRestrictedUserClaimAttributes,
+  TokenRestrictedDeviceClaimAttributes,
+  TokenDeviceGroups,
+  TokenRestrictedDeviceGroups,
+  TokenSecurityAttributes,
+  TokenIsRestricted,
   MaxTokenInfoClass
 } TOKEN_INFORMATION_CLASS;
 
@@ -4411,6 +4423,13 @@ typedef struct _TOKEN_MANDATORY_LABEL {
 } TOKEN_MANDATORY_LABEL, * PTOKEN_MANDATORY_LABEL;
 
 /*
+ * TOKEN_APPCONTAINER_INFORMATION
+ */
+typedef struct _TOKEN_APPCONTAINER_INFORMATION {
+  PSID TokenAppContainer;
+} TOKEN_APPCONTAINER_INFORMATION, * PTOKEN_APPCONTAINER_INFORMATION;
+
+/*
  *	ACLs of NT
  */
 
-- 
1.8.3.2

