From f5fe63051ec417248b5bd263d05b99aa506a61c5 Mon Sep 17 00:00:00 2001
From: Bruno Jesus <00cpxxx@gmail.com>
Date: Fri, 29 May 2015 19:05:12 +0200
Subject: dmstyle: Return the correct variable.

---
 dlls/dmstyle/dmutils.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/dmstyle/dmutils.c b/dlls/dmstyle/dmutils.c
index 5008134..e551410 100644
--- a/dlls/dmstyle/dmutils.c
+++ b/dlls/dmstyle/dmutils.c
@@ -213,7 +213,7 @@ HRESULT IDirectMusicUtils_IPersistStream_ParseReference (LPPERSISTSTREAM iface,
   }
   TRACE("** DM Reference End of Load ***\n");
 
-  return S_OK;
+  return hr;
 }
 
 /* FOURCC to string conversion for debug messages */
-- 
2.4.2

