From 64e1f4ac26867d15c0957c0b0212c9b67a87cb5d Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Tue, 25 Nov 2014 21:26:54 +0100
Subject: wintrust: Reset data->pWintrustData->u.pFile->hFile after closing
 handle.

---
 dlls/wintrust/softpub.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/wintrust/softpub.c b/dlls/wintrust/softpub.c
index 358ba01..4e8582e 100644
--- a/dlls/wintrust/softpub.c
+++ b/dlls/wintrust/softpub.c
@@ -1209,7 +1209,11 @@ HRESULT WINAPI SoftpubCleanup(CRYPT_PROVIDER_DATA *data)
     if (data->fOpenedFile &&
      data->pWintrustData->dwUnionChoice == WTD_CHOICE_FILE &&
      data->pWintrustData->u.pFile)
+    {
         CloseHandle(data->pWintrustData->u.pFile->hFile);
+        data->pWintrustData->u.pFile->hFile = INVALID_HANDLE_VALUE;
+        data->fOpenedFile = FALSE;
+    }
 
     return S_OK;
 }
-- 
2.1.3

