From 4bd593d4d30e913ebda4fa586561dbbc32b4b74a Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Mon, 7 Nov 2016 19:51:40 +0800
Subject: windowscodecs: Fix a typo in 2bppIndexed to 32bppBGRA conversion
 routine.

---
 dlls/windowscodecs/converter.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/windowscodecs/converter.c b/dlls/windowscodecs/converter.c
index 4ee980e..5561d42 100644
--- a/dlls/windowscodecs/converter.c
+++ b/dlls/windowscodecs/converter.c
@@ -257,7 +257,7 @@ static HRESULT copypixels_to_32bppBGRA(struct FormatConverter *This, const WICRe
                         *dstpixel++ = colors[srcval>>6];
                         if (x+1 < prc->Width) *dstpixel++ = colors[srcval>>4&0x3];
                         if (x+2 < prc->Width) *dstpixel++ = colors[srcval>>2&0x3];
-                        if (x+1 < prc->Width) *dstpixel++ = colors[srcval&0x3];
+                        if (x+3 < prc->Width) *dstpixel++ = colors[srcval&0x3];
                     }
                     srcrow += srcstride;
                     dstrow += cbStride;
-- 
2.9.0

