From 3e9426e9eddb491050a780e854ad7d7f889e91bd Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 25 Aug 2016 19:26:07 +0200
Subject: wined3d: Allow creation of 1d shader views.

---
 dlls/wined3d/view.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/wined3d/view.c b/dlls/wined3d/view.c
index 82fac25603..a809344126 100644
--- a/dlls/wined3d/view.c
+++ b/dlls/wined3d/view.c
@@ -46,6 +46,10 @@ static GLenum get_texture_view_target(const struct wined3d_gl_info *gl_info,
     {
         {GL_TEXTURE_CUBE_MAP,  0,                          GL_TEXTURE_CUBE_MAP},
         {GL_TEXTURE_RECTANGLE, 0,                          GL_TEXTURE_RECTANGLE},
+        {GL_TEXTURE_1D,        0,                          GL_TEXTURE_1D},
+        {GL_TEXTURE_1D,        WINED3D_VIEW_TEXTURE_ARRAY, GL_TEXTURE_1D_ARRAY},
+        {GL_TEXTURE_1D_ARRAY,  0,                          GL_TEXTURE_1D},
+        {GL_TEXTURE_1D_ARRAY,  WINED3D_VIEW_TEXTURE_ARRAY, GL_TEXTURE_1D_ARRAY},
         {GL_TEXTURE_2D,        0,                          GL_TEXTURE_2D},
         {GL_TEXTURE_2D,        WINED3D_VIEW_TEXTURE_ARRAY, GL_TEXTURE_2D_ARRAY},
         {GL_TEXTURE_2D_ARRAY,  0,                          GL_TEXTURE_2D},
-- 
2.11.0

