From eba72a3ed5d6f61c92ce2926d8305d19dc66425b Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 30 May 2015 17:08:18 +0200
Subject: ws2_32: Fix uninitialized memory access in do_poll (Coverity).

---
 dlls/ws2_32/socket.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ws2_32/socket.c b/dlls/ws2_32/socket.c
index fdf68a3..60c6ec5 100644
--- a/dlls/ws2_32/socket.c
+++ b/dlls/ws2_32/socket.c
@@ -4794,7 +4794,7 @@ static int do_poll(struct pollfd *pollfds, int count, int timeout)
     struct timeval tv1, tv2;
     int ret, torig = timeout;
 
-    if (timeout > 0) gettimeofday( &tv1, 0 );
+    if (timeout >= 0) gettimeofday( &tv1, 0 );
 
     while ((ret = poll( pollfds, count, timeout )) < 0)
     {
-- 
2.4.2

