From 6addc4f5304995f32525dce5422bc6544d7b313d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sat, 30 May 2015 15:36:43 +0200
Subject: kernel32: Correctly check for an empty short filename in
 GetShortPathNameW (Coverity).

---
 dlls/kernel32/path.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/kernel32/path.c b/dlls/kernel32/path.c
index f74b952..09b8b7f 100644
--- a/dlls/kernel32/path.c
+++ b/dlls/kernel32/path.c
@@ -528,7 +528,7 @@ DWORD WINAPI GetShortPathNameW( LPCWSTR longpath, LPWSTR shortpath, DWORD shortl
 
         /* In rare cases (like "a.abcd") short path may be longer than original path.
          * Make sure we have enough space in temp buffer. */
-        if (wfd.cAlternateFileName && tmplen < strlenW(wfd.cAlternateFileName))
+        if (wfd.cAlternateFileName[0] && tmplen < strlenW(wfd.cAlternateFileName))
         {
             WCHAR *new_buf;
             buf_len += strlenW(wfd.cAlternateFileName) - tmplen;
-- 
2.4.2

