From 1bece0c817599701648b4cd177cf13babe1342e8 Mon Sep 17 00:00:00 2001
From: Jared Smudde <computerwhiz02@hotmail.com>
Date: Fri, 5 Jun 2015 01:43:04 +0200
Subject: shell32: Add caption to Run dialog.

---
 dlls/shell32/shell32.rc | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/shell32/shell32.rc b/dlls/shell32/shell32.rc
index 847d2ce..3a16da3 100644
--- a/dlls/shell32/shell32.rc
+++ b/dlls/shell32/shell32.rc
@@ -330,6 +330,7 @@ FONT 10, "MS Shell Dlg"
 
 SHELL_RUN_DLG DIALOG 0, 0, 235, 83
 STYLE DS_MODALFRAME | WS_POPUP | WS_VISIBLE | WS_CAPTION | WS_SYSMENU
+CAPTION "Run"
 FONT 8, "MS Shell Dlg"
 {
  ICON "", IDC_RUNDLG_ICON, 7, 11, 18, 20, WS_VISIBLE
-- 
2.4.2

