From a88740c95d8250e95bf215169896c0eb54a7eb6c Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Sun, 16 Feb 2014 14:14:13 -0700
Subject: ddraw/tests: Remove broken tests of D3DTRANSFORM_UNCLIPPED.

---
 dlls/ddraw/tests/d3d.c |    6 ------
 1 file changed, 6 deletions(-)

diff --git a/dlls/ddraw/tests/d3d.c b/dlls/ddraw/tests/d3d.c
index 209555a..35987cb 100644
--- a/dlls/ddraw/tests/d3d.c
+++ b/dlls/ddraw/tests/d3d.c
@@ -1256,12 +1256,6 @@ static void Direct3D1Test(void)
            out[i].x, out[i].y, out[i].z, out[i].rhw,
            cmp[i].x, cmp[i].y, cmp[i].z, cmp[i].rhw);
     }
-    for(i = 0; i < sizeof(outH); i++) {
-        if(((unsigned char *) outH)[i] != 0xcc) {
-            ok(FALSE, "Homogeneous output was generated despite UNCLIPPED flag\n");
-            break;
-        }
-    }
 
     SET_VP_DATA(vp_data);
     hr = IDirect3DViewport_SetViewport(Viewport, &vp_data);
-- 
1.7.9.5

