From f62d9448825cd452107a595b5022b58e5ab63efc Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Sun, 16 Feb 2014 14:14:13 -0700
Subject: ddraw/tests: Ignore broken behavior of D3DTRANSFORM_UNCLIPPED.

---
 dlls/ddraw/tests/d3d.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/ddraw/tests/d3d.c b/dlls/ddraw/tests/d3d.c
index 209555a..b3bd3c8 100644
--- a/dlls/ddraw/tests/d3d.c
+++ b/dlls/ddraw/tests/d3d.c
@@ -1257,7 +1257,8 @@ static void Direct3D1Test(void)
            cmp[i].x, cmp[i].y, cmp[i].z, cmp[i].rhw);
     }
     for(i = 0; i < sizeof(outH); i++) {
-        if(((unsigned char *) outH)[i] != 0xcc) {
+        unsigned char byte = ((unsigned char *) outH)[i];
+        if(byte != 0xcc && !broken(byte == 0x0)) {
             ok(FALSE, "Homogeneous output was generated despite UNCLIPPED flag\n");
             break;
         }
-- 
1.7.9.5

