From 8b42b214de22c117d61f2e59f2f3708046e9c9d6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stefan=20D=C3=B6singer?= <stefan@codeweavers.com>
Date: Fri, 20 Mar 2015 12:47:10 +0100
Subject: ddraw: Update the palette before presents to the NULL window.

This fixes bug 38248.
---
 dlls/ddraw/surface.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/ddraw/surface.c b/dlls/ddraw/surface.c
index b8c9872..0d2d731 100644
--- a/dlls/ddraw/surface.c
+++ b/dlls/ddraw/surface.c
@@ -81,6 +81,8 @@ HRESULT ddraw_surface_update_frontbuffer(struct ddraw_surface *surface, const RE
         ERR("Failed to get surface DC, hr %#x.\n", hr);
         return hr;
     }
+    if (surface->palette)
+        wined3d_palette_apply_to_dc(surface->palette->wineD3DPalette, surface_dc);
 
     if (!(screen_dc = GetDC(NULL)))
     {
-- 
2.3.2

