From de11056b08110b83c3ce9e2f90c16280d6c74193 Mon Sep 17 00:00:00 2001
From: Jarkko Korpi <jarkko_korpi@hotmail.com>
Date: Mon, 6 Jul 2015 23:47:07 +0300
Subject: wbemprox: Silence repeated "timeout not supported" fixme.

---
 dlls/wbemprox/class.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/dlls/wbemprox/class.c b/dlls/wbemprox/class.c
index afb9f8c..5ce0381 100644
--- a/dlls/wbemprox/class.c
+++ b/dlls/wbemprox/class.c
@@ -120,7 +120,11 @@ static HRESULT WINAPI enum_class_object_Next(
 
     if (!uCount) return WBEM_S_FALSE;
     if (!apObjects || !puReturned) return WBEM_E_INVALID_PARAMETER;
-    if (lTimeout != WBEM_INFINITE) FIXME("timeout not supported\n");
+    if (lTimeout != WBEM_INFINITE)
+    {
+        static int once;
+        if (!once++) FIXME("timeout not supported\n");
+    }
 
     *puReturned = 0;
     if (ec->index >= view->count) return WBEM_S_FALSE;
-- 
2.4.5

