From ead7f173a46031c20d7771c34e7c231f98332700 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sun, 4 Oct 2015 23:48:16 +0200
Subject: msiexec: Add support for /passive command line option.
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Michael Müller <michael@fds-team.de>
---
 programs/msiexec/msiexec.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/programs/msiexec/msiexec.c b/programs/msiexec/msiexec.c
index cb0984d..22554de 100644
--- a/programs/msiexec/msiexec.c
+++ b/programs/msiexec/msiexec.c
@@ -931,6 +931,10 @@ int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine
 					 wine_dbgstr_w(argvW[i]+2));
 			}
 		}
+		else if(msi_option_equal(argvW[i], "passive"))
+		{
+			InstallUILevel = INSTALLUILEVEL_BASIC|INSTALLUILEVEL_PROGRESSONLY|INSTALLUILEVEL_HIDECANCEL;
+		}
 		else if(msi_option_equal(argvW[i], "y"))
 		{
 			FunctionDllRegisterServer = TRUE;
-- 
2.6.0

