From 8474e07396ee431e79dc57de7a4ba5b337933e8e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Mon, 12 Jan 2015 23:32:45 +0100
Subject: nvapi: Explicity return NULL for 0x33c7358c and 0x593e8644.

---
 dlls/nvapi/nvapi.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/nvapi/nvapi.c b/dlls/nvapi/nvapi.c
index bb9f80f31d7..1e71db854c5 100644
--- a/dlls/nvapi/nvapi.c
+++ b/dlls/nvapi/nvapi.c
@@ -522,6 +522,8 @@ void* CDECL nvapi_QueryInterface(unsigned int offset)
         {0xaea3fa32, NvAPI_GetPhysicalGPUsFromLogicalGPU},
         {0xe5ac921f, NvAPI_EnumPhysicalGPUs},
         {0xceee8e9f, NvAPI_GPU_GetFullName},
+        {0x33c7358c, NULL}, /* This functions seems to be optional */
+        {0x593e8644, NULL}, /* This functions seems to be optional */
     };
     unsigned int i;
     TRACE("(%x)\n", offset);
-- 
2.13.1

