From 9337f83015b33f301ac15b11af5d8c947773e916 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Tue, 13 Jan 2015 00:48:47 +0100
Subject: nvapi: Add stub for NvAPI_Unload.

---
 dlls/nvapi/nvapi.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/dlls/nvapi/nvapi.c b/dlls/nvapi/nvapi.c
index 062c99a6585..44d9d4a5245 100644
--- a/dlls/nvapi/nvapi.c
+++ b/dlls/nvapi/nvapi.c
@@ -531,6 +531,12 @@ static NvAPI_Status CDECL NvAPI_SYS_GetDriverAndBranchVersion(NvU32* pDriverVers
     return NVAPI_OK;
 }
 
+static NvAPI_Status CDECL NvAPI_Unload(void)
+{
+    TRACE("()\n");
+    return NVAPI_OK;
+}
+
 void* CDECL nvapi_QueryInterface(unsigned int offset)
 {
     static const struct
@@ -568,6 +574,7 @@ void* CDECL nvapi_QueryInterface(unsigned int offset)
         {0x1e9d8a31, NvAPI_DISP_GetGDIPrimaryDisplayId},
         {0x9abdd40d, NvAPI_EnumNvidiaDisplayHandle},
         {0x2926aaad, NvAPI_SYS_GetDriverAndBranchVersion},
+        {0xd22bdd7e, NvAPI_Unload},
     };
     unsigned int i;
     TRACE("(%x)\n", offset);
-- 
2.13.1

