From 3927bd79cfd73dde183947e688eb2563ad001705 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Tue, 13 Jan 2015 02:21:07 +0100
Subject: nvapi: Print fixme message for NvAPI_D3D9_StretchRectEx.

---
 dlls/nvapi/nvapi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/nvapi/nvapi.c b/dlls/nvapi/nvapi.c
index 287e681..6852e24 100644
--- a/dlls/nvapi/nvapi.c
+++ b/dlls/nvapi/nvapi.c
@@ -402,7 +402,7 @@ static NvAPI_Status CDECL NvAPI_D3D9_StretchRectEx(IDirect3DDevice9 *pDevice, ID
                                                    const RECT *pSourceRect, IDirect3DResource9 *pDestResource,
                                                    const RECT *pDestRect, D3DTEXTUREFILTERTYPE Filter)
 {
-    TRACE("(%p, %p, %p, %p, %p, %d)\n", pDevice, pSourceResource, pSourceRect, pDestResource, pDestRect, Filter);
+    FIXME("(%p, %p, %p, %p, %p, %d): stub\n", pDevice, pSourceResource, pSourceRect, pDestResource, pDestRect, Filter);
     return NVAPI_UNREGISTERED_RESOURCE;
 }
 
-- 
1.9.1

