From 1710af67c64a6f8ff9f5230404349ec234221308 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Wed, 14 Jan 2015 22:13:57 +0100
Subject: nvapi: Add stub for NvAPI_D3D_GetObjectHandleForResource.

---
 dlls/nvapi/nvapi.c | 7 +++++++
 include/nvapi.h    | 1 +
 2 files changed, 8 insertions(+)

diff --git a/dlls/nvapi/nvapi.c b/dlls/nvapi/nvapi.c
index 52556f18fdc..6bf960f79a8 100644
--- a/dlls/nvapi/nvapi.c
+++ b/dlls/nvapi/nvapi.c
@@ -564,6 +564,12 @@ static NvAPI_Status CDECL NvAPI_GetLogicalGPUFromDisplay(NvDisplayHandle hNvDisp
     return NVAPI_OK;
 }
 
+static NvAPI_Status CDECL NvAPI_D3D_GetObjectHandleForResource(IUnknown *pDevice, IUnknown *pResource, NVDX_ObjectHandle *pHandle)
+{
+    FIXME("(%p, %p, %p): stub\n", pDevice, pResource, pHandle);
+    return NVAPI_ERROR;
+}
+
 void* CDECL nvapi_QueryInterface(unsigned int offset)
 {
     static const struct
@@ -604,6 +610,7 @@ void* CDECL nvapi_QueryInterface(unsigned int offset)
         {0xd22bdd7e, NvAPI_Unload},
         {0x4b708b54, NvAPI_D3D_GetCurrentSLIState},
         {0xee1370cf, NvAPI_GetLogicalGPUFromDisplay},
+        {0xfceac864, NvAPI_D3D_GetObjectHandleForResource},
     };
     unsigned int i;
     TRACE("(%x)\n", offset);
diff --git a/include/nvapi.h b/include/nvapi.h
index e52e9b1d774..ca95a1ac5ee 100644
--- a/include/nvapi.h
+++ b/include/nvapi.h
@@ -55,6 +55,7 @@ typedef void *NvPhysicalGpuHandle;
 typedef void *NvLogicalGpuHandle;
 typedef void *NvDisplayHandle;
 typedef void *StereoHandle;
+typedef void *NVDX_ObjectHandle;
 
 typedef struct
 {
-- 
2.13.1

