From 1da7ea0e73111f1dfbc9f8711e68c6a43eb9a23f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Wed, 14 Jan 2015 22:50:35 +0100
Subject: nvapi: Add stub for NvAPI_D3D9_RegisterResource.

---
 dlls/nvapi/nvapi.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/dlls/nvapi/nvapi.c b/dlls/nvapi/nvapi.c
index 28483be..545bb5a 100644
--- a/dlls/nvapi/nvapi.c
+++ b/dlls/nvapi/nvapi.c
@@ -570,6 +570,12 @@ static NvAPI_Status CDECL NvAPI_D3D_GetObjectHandleForResource(IUnknown *pDevice
     return NVAPI_ERROR;
 }
 
+static NvAPI_Status CDECL NvAPI_D3D9_RegisterResource(IDirect3DResource9* pResource)
+{
+    FIXME("(%p): stub\n", pResource);
+    return NVAPI_ERROR;
+}
+
 void* CDECL nvapi_QueryInterface(unsigned int offset)
 {
     static const struct
@@ -610,7 +616,8 @@ void* CDECL nvapi_QueryInterface(unsigned int offset)
         {0xd22bdd7e, NvAPI_Unload},
         {0x4b708b54, NvAPI_D3D_GetCurrentSLIState},
         {0xee1370cf, NvAPI_GetLogicalGPUFromDisplay},
-        {0xfceac864, NvAPI_D3D_GetObjectHandleForResource}
+        {0xfceac864, NvAPI_D3D_GetObjectHandleForResource},
+        {0xa064bdfc, NvAPI_D3D9_RegisterResource}
     };
     unsigned int i;
     TRACE("(%x)\n", offset);
-- 
1.9.1

