From 79843a592a6b6d7b68e933d19bc59490f276f510 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 20 Mar 2015 22:34:49 +0100
Subject: gdiplus: Add missing returns in initialize_decoder_wic (Coverity).

---
 dlls/gdiplus/image.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/gdiplus/image.c b/dlls/gdiplus/image.c
index 8eb1419..77b9305 100644
--- a/dlls/gdiplus/image.c
+++ b/dlls/gdiplus/image.c
@@ -3425,13 +3425,13 @@ static GpStatus initialize_decoder_wic(IStream *stream, REFGUID container, IWICB
     TRACE("%p,%s\n", stream, wine_dbgstr_guid(container));
 
     hr = WICCreateImagingFactory_Proxy(WINCODEC_SDK_VERSION, &factory);
-    if (FAILED(hr)) hresult_to_status(hr);
+    if (FAILED(hr)) return hresult_to_status(hr);
     hr = IWICImagingFactory_CreateDecoder(factory, container, NULL, decoder);
     IWICImagingFactory_Release(factory);
-    if (FAILED(hr)) hresult_to_status(hr);
+    if (FAILED(hr)) return hresult_to_status(hr);
 
     hr = IWICBitmapDecoder_Initialize(*decoder, stream, WICDecodeMetadataCacheOnLoad);
-    if (FAILED(hr)) hresult_to_status(hr);
+    if (FAILED(hr)) return hresult_to_status(hr);
     return Ok;
 }
 
-- 
2.3.2

