From ccd2fe7c16c909728d157283d38419af6decfdaf Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Fri, 25 Apr 2014 12:13:56 +0900
Subject: windowscodecs: Don't fail to decode GIF if an image has been already
 loaded.

---
 dlls/windowscodecs/ungif.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/windowscodecs/ungif.c b/dlls/windowscodecs/ungif.c
index 427b32f..e08cb4b 100644
--- a/dlls/windowscodecs/ungif.c
+++ b/dlls/windowscodecs/ungif.c
@@ -491,7 +491,7 @@ DGifGetLine(GifFileType * GifFile,
              * image until empty block (size 0) detected. We use GetCodeNext. */
             do
                 if (DGifGetCodeNext(GifFile, &Dummy) == GIF_ERROR)
-                    return GIF_ERROR;
+                    break;
             while (Dummy != NULL) ;
         }
         return GIF_OK;
-- 
2.3.5

