From 68f13c9c9e48dd5907727d1e54680a6ee8684976 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 20 Jul 2017 16:39:29 +0200
Subject: wined3d: Recognize SM4 nop opcode.

---
 dlls/wined3d/shader_sm4.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/wined3d/shader_sm4.c b/dlls/wined3d/shader_sm4.c
index 142c0f89811..123ecbc78a4 100644
--- a/dlls/wined3d/shader_sm4.c
+++ b/dlls/wined3d/shader_sm4.c
@@ -171,6 +171,7 @@ enum wined3d_sm4_opcode
     WINED3D_SM4_OP_MOVC                             = 0x37,
     WINED3D_SM4_OP_MUL                              = 0x38,
     WINED3D_SM4_OP_NE                               = 0x39,
+    WINED3D_SM4_OP_NOP                              = 0x3a,
     WINED3D_SM4_OP_NOT                              = 0x3b,
     WINED3D_SM4_OP_OR                               = 0x3c,
     WINED3D_SM4_OP_RESINFO                          = 0x3d,
@@ -937,6 +938,7 @@ static const struct wined3d_sm4_opcode_info opcode_table[] =
     {WINED3D_SM4_OP_MOVC,                             WINED3DSIH_MOVC,                             "f",    "uff"},
     {WINED3D_SM4_OP_MUL,                              WINED3DSIH_MUL,                              "f",    "ff"},
     {WINED3D_SM4_OP_NE,                               WINED3DSIH_NE,                               "u",    "ff"},
+    {WINED3D_SM4_OP_NOP,                              WINED3DSIH_NOP,                              "",     ""},
     {WINED3D_SM4_OP_NOT,                              WINED3DSIH_NOT,                              "u",    "u"},
     {WINED3D_SM4_OP_OR,                               WINED3DSIH_OR,                               "u",    "uu"},
     {WINED3D_SM4_OP_RESINFO,                          WINED3DSIH_RESINFO,                          "f",    "iR"},
-- 
2.13.1

