From 33f637536046687bc533370397ed7fc704eaa0e2 Mon Sep 17 00:00:00 2001
From: Kimmo Myllyvirta <kimmo.myllyvirta@gmail.com>
Date: Fri, 28 Apr 2017 15:28:55 +0200
Subject: user32: ShowWindow should not send message when window is already
 visible.

---
 dlls/user32/winpos.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/user32/winpos.c b/dlls/user32/winpos.c
index c2b35ec1f5..64fdfe01b0 100644
--- a/dlls/user32/winpos.c
+++ b/dlls/user32/winpos.c
@@ -1239,6 +1239,9 @@ BOOL WINAPI ShowWindow( HWND hwnd, INT cmd )
     if ((cmd == SW_HIDE) && !(GetWindowLongW( hwnd, GWL_STYLE ) & WS_VISIBLE))
         return FALSE;
 
+    if ((cmd == SW_SHOW) && (GetWindowLongW( hwnd, GWL_STYLE ) & WS_VISIBLE))
+        return TRUE;
+
     return SendMessageW( hwnd, WM_WINE_SHOWWINDOW, cmd, 0 );
 }
 
-- 
2.12.2

