From 2ae1a26958e521ade74525d8f001d065138fb5f3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 25 Feb 2016 15:21:56 +0100
Subject: wine.inf: Add WMP 12 as installed windows feature.

---
 loader/wine.inf.in | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index d22b29b..717da09 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -474,6 +474,8 @@ HKLM,%CurrentVersion%\PreviewHandlers,,16
 HKLM,%CurrentVersion%\Run,,16
 HKLM,%CurrentVersion%\Setup,"BootDir",,"%30%"
 HKLM,%CurrentVersion%\Setup,"SharedDir",,"%25%"
+HKLM,%CurrentVersion%\Setup\WindowsFeatures\WindowsMediaVersion,,,"12.0.7601.18840"
+HKLM,%CurrentVersion%\Setup\WindowsFeatures\WindowsMediaVersion,"Version",1,00,00,0c,00,98,49,b1,1d
 HKLM,%CurrentVersion%\Shell Extensions\Approved,,16
 HKLM,%CurrentVersion%\Uninstall,,16
 HKLM,%CurrentVersion%\Winlogon,,16
-- 
2.7.1

