From 78ab3400ea9c3c9ae9d85f3ee2952c20c7d68f73 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Mon, 15 Sep 2014 22:55:49 +0200
Subject: dbghelp: Don't fill KdHelp structure for usermode applications.

---
 dlls/dbghelp/stack.c | 9 ---------
 1 file changed, 9 deletions(-)

diff --git a/dlls/dbghelp/stack.c b/dlls/dbghelp/stack.c
index 3e1ff72..21412f6 100644
--- a/dlls/dbghelp/stack.c
+++ b/dlls/dbghelp/stack.c
@@ -240,15 +240,6 @@ BOOL WINAPI StackWalk64(DWORD MachineType, HANDLE hProcess, HANDLE hThread,
     if (!cpu->stack_walk(&csw, frame, ctx)) return FALSE;
 
     /* we don't handle KdHelp */
-    frame->KdHelp.Thread = 0xC000FADE;
-    frame->KdHelp.ThCallbackStack = 0x10;
-    frame->KdHelp.ThCallbackBStore = 0;
-    frame->KdHelp.NextCallback = 0;
-    frame->KdHelp.FramePointer = 0;
-    frame->KdHelp.KiCallUserMode = 0xD000DAFE;
-    frame->KdHelp.KeUserCallbackDispatcher = 0xE000F000;
-    frame->KdHelp.SystemRangeStart = 0xC0000000;
-    frame->KdHelp.Reserved[0] /* KiUserExceptionDispatcher */ = 0xE0005000;
 
     return TRUE;
 }
-- 
2.1.0

