From 94973c7be88a18b4b4f79cea79f82f450cb659f0 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 14 Jan 2016 07:18:08 +0100
Subject: vcomp/tests: Reenable architecture dependent tests.

---
 dlls/vcomp/tests/vcomp.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/dlls/vcomp/tests/vcomp.c b/dlls/vcomp/tests/vcomp.c
index 4f030d3..5c177df 100644
--- a/dlls/vcomp/tests/vcomp.c
+++ b/dlls/vcomp/tests/vcomp.c
@@ -1406,13 +1406,15 @@ static void test_atomic_integer32(void)
         { p_vcomp_atomic_mul_i4,  0x11223344, -0x77665544,   0xecccdf0 },
         { p_vcomp_atomic_or_i4,   0x11223344,  0x77665544,  0x77667744 },
         { p_vcomp_atomic_shl_i4,  0x11223344,           3, -0x76ee65e0 },
-     /* { p_vcomp_atomic_shl_i4,  0x11223344,          35, -0x76ee65e0 }, */ /* depends on Architecture */
         { p_vcomp_atomic_shl_i4, -0x11223344,           3,  0x76ee65e0 },
         { p_vcomp_atomic_shr_i4,  0x11223344,           3,   0x2244668 },
-     /* { p_vcomp_atomic_shr_i4,  0x11223344,          35,   0x2244668 }, */ /* depends on Architecture */
         { p_vcomp_atomic_shr_i4, -0x11223344,           3,  -0x2244669 },
         { p_vcomp_atomic_sub_i4,  0x11223344,  0x77665544, -0x66442200 },
         { p_vcomp_atomic_xor_i4,  0x11223344,  0x77665544,  0x66446600 },
+    #if defined(__i386__) || defined(__x86_64__)
+        { p_vcomp_atomic_shl_i4,  0x11223344,          35, -0x76ee65e0 },
+        { p_vcomp_atomic_shr_i4,  0x11223344,          35,   0x2244668 },
+    #endif
     };
     struct
     {
@@ -1424,8 +1426,10 @@ static void test_atomic_integer32(void)
         { p_vcomp_atomic_div_ui4, 0x77665544, 0x11223344,          6 },
         { p_vcomp_atomic_div_ui4, 0x77665544, 0xeeddccbc,          0 },
         { p_vcomp_atomic_shr_ui4, 0x11223344,          3,  0x2244668 },
-     /* { p_vcomp_atomic_shr_ui4, 0x11223344,         35,  0x2244668 }, */ /* depends on Architecture */
         { p_vcomp_atomic_shr_ui4, 0xeeddccbc,          3, 0x1ddbb997 },
+    #if defined(__i386__) || defined(__x86_64__)
+        { p_vcomp_atomic_shr_ui4, 0x11223344,         35,  0x2244668 },
+    #endif
     };
     int i;
 
-- 
2.6.4

