From 725ae5202ae6206606a330d3799fb620a09dc8f9 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 4 Jun 2015 16:12:58 +0200
Subject: ntdll/tests: Don't call lstrlenW on non-null-terminated buffer.

---
 dlls/ntdll/tests/directory.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ntdll/tests/directory.c b/dlls/ntdll/tests/directory.c
index 21aa31f..730b0f1 100644
--- a/dlls/ntdll/tests/directory.c
+++ b/dlls/ntdll/tests/directory.c
@@ -350,7 +350,7 @@ static void test_NtQueryDirectoryFile_case(void)
     }
 
     mask.Buffer = testmask;
-    mask.Length = mask.MaximumLength = lstrlenW(testmask) * sizeof(WCHAR);
+    mask.Length = mask.MaximumLength = sizeof(testmask);
     pNtQueryDirectoryFile(dirh, NULL, NULL, NULL, &io, data, data_size,
                           FileBothDirectoryInformation, TRUE, &mask, FALSE);
     ok(U(io).Status == STATUS_SUCCESS, "failed to query directory; status %x\n", U(io).Status);
-- 
2.4.2

