From 15121c7f2af09c906bcfeddd830b61ac1efa3c25 Mon Sep 17 00:00:00 2001
From: Alexandre Julliard <julliard@winehq.org>
Date: Mon, 24 Jul 2017 13:12:50 +0200
Subject: user32: Reduce the cursor height if it also includes the mask.

Signed-off-by: Alexandre Julliard <julliard@winehq.org>
---
 dlls/user32/cursoricon.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/user32/cursoricon.c b/dlls/user32/cursoricon.c
index 5ee5cdfa966..252ea1f2942 100644
--- a/dlls/user32/cursoricon.c
+++ b/dlls/user32/cursoricon.c
@@ -947,6 +947,7 @@ static BOOL CURSORICON_GetResCursorEntry( LPCVOID dir, DWORD size, int n,
     *width = cursor->wWidth;
     *height = cursor->wHeight;
     *bits = resdir->idEntries[n].wBitCount;
+    if (*height == *width * 2) *height /= 2;
     return TRUE;
 }
 
-- 
2.13.1

