From 7fda9638073cbf136c87bfcac507a067c2159b3e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sat, 21 Feb 2015 14:49:30 +0100
Subject: include: Fix an invalid UUID in dxva2api.idl.

---
 include/dxva2api.idl | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/dxva2api.idl b/include/dxva2api.idl
index ebd508b..7a29157 100644
--- a/include/dxva2api.idl
+++ b/include/dxva2api.idl
@@ -434,7 +434,7 @@ interface IDirectXVideoDecoder : IUnknown
  */
 [
     object,
-    uuid(fc51a551-d5e7-11d9-af55-00054e43ff02),
+    uuid(fc51a552-d5e7-11d9-af55-00054e43ff02),
     local
 ]
 interface IDirectXVideoProcessorService : IDirectXVideoAccelerationService
-- 
2.1.0

