From 0dc9223ea631649d09fb6e94c4c01032150c155c Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Tue, 16 Dec 2014 04:10:34 +0100
Subject: kernel32: Return proper char count in get_registry_locale_info.

---
 dlls/kernel32/locale.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/dlls/kernel32/locale.c b/dlls/kernel32/locale.c
index 5fee723..1460f7a 100644
--- a/dlls/kernel32/locale.c
+++ b/dlls/kernel32/locale.c
@@ -1206,7 +1206,9 @@ static INT get_registry_locale_info( struct registry_value *registry_value, LPWS
         {
             if (status == STATUS_BUFFER_OVERFLOW && !buffer)
             {
-                ret = (size - info_size) / sizeof(WCHAR) + 1;
+                ret = (size - info_size) / sizeof(WCHAR);
+                if (!ret || ((WCHAR *)&info->Data)[ret-1])
+                    ret++;
             }
             else if (status == STATUS_OBJECT_NAME_NOT_FOUND)
             {
-- 
2.1.3

