From cb92c3e3611dd8d8d97992b79e1c9fc449ff2569 Mon Sep 17 00:00:00 2001
From: Daniel Jelinski <djelinski1@gmail.com>
Date: Sun, 8 Feb 2015 06:32:34 +0100
Subject: wine.inf: Add registry keys for Windows Performance Library.

---
 loader/wine.inf.in | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 315ef29..2de7e61 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -492,6 +492,10 @@ HKLM,%CurrentVersionNT%\Language Pack,,16
 HKLM,%CurrentVersionNT%\NetworkCards,,16
 HKLM,%CurrentVersionNT%\OpenGLDrivers,,16
 HKLM,%CurrentVersionNT%\Perflib,,16
+HKLM,%CurrentVersionNT%\Perflib,Last Counter,0x10003,1846
+HKLM,%CurrentVersionNT%\Perflib,Last Help,0x10003,1847
+HKLM,%CurrentVersionNT%\Perflib\009,Counter,0x10002,1,1847,1846,End Marker
+HKLM,%CurrentVersionNT%\Perflib\009,Help,0x10002,1847,End Marker
 HKLM,%CurrentVersionNT%\Ports,,16
 HKLM,%CurrentVersionNT%\Print,,16
 HKLM,%CurrentVersionNT%\ProfileList,,16
-- 
2.4.5

