From c3996fd61ba65678d9cfefed715a137b697a72a1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sat, 28 Mar 2015 20:47:52 +0100
Subject: ws2_32: Don't return an error in WS_select when EINTR happens during
 timeout.

---
 dlls/ws2_32/socket.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/dlls/ws2_32/socket.c b/dlls/ws2_32/socket.c
index e7e3de5..d99e566 100644
--- a/dlls/ws2_32/socket.c
+++ b/dlls/ws2_32/socket.c
@@ -4601,7 +4601,11 @@ int WINAPI WS_select(int nfds, WS_fd_set *ws_readfds,
             }
 
             timeout = torig - (tv2.tv_sec * 1000) - (tv2.tv_usec + 999) / 1000;
-            if (timeout <= 0) break;
+            if (timeout <= 0)
+            {
+                ret = 0;
+                break;
+            }
         } else break;
     }
     release_poll_fds( ws_readfds, ws_writefds, ws_exceptfds, pollfds );
-- 
2.1.0

