From e79956abb099aa58f3b4a2859f8595434e3132e1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Fri, 7 Aug 2015 02:44:31 +0200
Subject: wininet: Disabling proxy should return success when no proxy was
 enabled.

---
 dlls/wininet/internet.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/wininet/internet.c b/dlls/wininet/internet.c
index ee04f40..b004c68 100644
--- a/dlls/wininet/internet.c
+++ b/dlls/wininet/internet.c
@@ -359,7 +359,7 @@ static LONG INTERNET_SaveProxySettings( proxyinfo_t *lpwpi )
     }
     else
     {
-        if ((ret = RegDeleteValueW( key, szProxyServer )))
+        if ((ret = RegDeleteValueW( key, szProxyServer )) && ret != ERROR_FILE_NOT_FOUND)
         {
             RegCloseKey( key );
             return ret;
-- 
2.5.0

