From 77ec3e3c1f2346c5b70647ed2e99102a41bad353 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Fri, 18 Aug 2017 12:03:50 +0800
Subject: [PATCH] windowscodecs: Fix stride calculation in JPEG decoder.

---
 dlls/windowscodecs/jpegformat.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/windowscodecs/jpegformat.c b/dlls/windowscodecs/jpegformat.c
index 08f1ee7..27cd880 100644
--- a/dlls/windowscodecs/jpegformat.c
+++ b/dlls/windowscodecs/jpegformat.c
@@ -624,7 +624,7 @@ static HRESULT WINAPI JpegDecoder_Frame_CopyPixels(IWICBitmapFrameDecode *iface,
     else if (This->cinfo.out_color_space == JCS_CMYK) bpp = 32;
     else bpp = 24;
 
-    stride = bpp * This->cinfo.output_width;
+    stride = (bpp * This->cinfo.output_width + 7) / 8;
     data_size = stride * This->cinfo.output_height;
 
     max_row_needed = prc->Y + prc->Height;
-- 
1.9.1

