From 271b0a51b2f1d70fcd7091e4829c77aacfdc0ba1 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Fri, 18 Aug 2017 12:03:50 +0800
Subject: windowscodecs: Fix stride calculation in JPEG decoder.

---
 dlls/windowscodecs/jpegformat.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/windowscodecs/jpegformat.c b/dlls/windowscodecs/jpegformat.c
index abab36ab40a..f31239b8d79 100644
--- a/dlls/windowscodecs/jpegformat.c
+++ b/dlls/windowscodecs/jpegformat.c
@@ -628,7 +628,7 @@ static HRESULT WINAPI JpegDecoder_Frame_CopyPixels(IWICBitmapFrameDecode *iface,
     else if (This->cinfo.out_color_space == JCS_CMYK) bpp = 32;
     else bpp = 24;
 
-    stride = bpp * This->cinfo.output_width;
+    stride = (bpp * This->cinfo.output_width + 7) / 8;
     data_size = stride * This->cinfo.output_height;
 
     max_row_needed = prc->Y + prc->Height;
-- 
2.14.1

