From 2679e91c82514d2f2d90374a06588601498c1baa Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 3 Jun 2015 02:46:57 +0200
Subject: server: Set the closed_fd->unix_name shortly before adding to the
 closed fd list.

---
 server/fd.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/server/fd.c b/server/fd.c
index cc6425c..ad974bb 100644
--- a/server/fd.c
+++ b/server/fd.c
@@ -1473,6 +1473,7 @@ static void fd_destroy( struct object *obj )
     if (fd->poll_index != -1) remove_poll_user( fd, fd->poll_index );
     if (fd->inode)
     {
+        fd->closed->unix_name = fd->unix_name;
         inode_add_closed_fd( fd->inode, fd->closed );
         release_object( fd->inode );
     }
@@ -1667,7 +1668,7 @@ struct fd *dup_fd_object( struct fd *orig, unsigned int access, unsigned int sha
         }
         closed->unix_fd = fd->unix_fd;
         closed->unlink = 0;
-        closed->unix_name = fd->unix_name;
+        closed->unix_name = NULL;
         fd->closed = closed;
         fd->inode = (struct inode *)grab_object( orig->inode );
         list_add_head( &fd->inode->open, &fd->inode_entry );
@@ -1838,7 +1839,7 @@ struct fd *open_fd( struct fd *root, const char *name, int flags, mode_t *mode,
 
     closed_fd->unix_fd = fd->unix_fd;
     closed_fd->unlink = 0;
-    closed_fd->unix_name = fd->unix_name;
+    closed_fd->unix_name = NULL;
 
     if (do_chmod) fchmod( fd->unix_fd, *mode );
     fstat( fd->unix_fd, &st );
-- 
2.4.2

