From 5a4985d6c6748fef78e69367848569e69d46c13f Mon Sep 17 00:00:00 2001
From: Zhaonan Liang <bt2517@126.com>
Date: Tue, 28 Apr 2015 19:06:31 +0800
Subject: include: Add declaration for FILE_LINK_INFORMATION.

---
 include/winternl.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/include/winternl.h b/include/winternl.h
index a84c6d4..d208537 100644
--- a/include/winternl.h
+++ b/include/winternl.h
@@ -567,6 +567,13 @@ typedef struct _FILE_RENAME_INFORMATION {
     WCHAR FileName[1];
 } FILE_RENAME_INFORMATION, *PFILE_RENAME_INFORMATION;
 
+typedef struct _FILE_LINK_INFORMATION {
+    BOOLEAN ReplaceIfExists;
+    HANDLE  RootDirectory;
+    ULONG   FileNameLength;
+    WCHAR   FileName[1];
+} FILE_LINK_INFORMATION, *PFILE_LINK_INFORMATION;
+
 typedef struct _FILE_NAMES_INFORMATION {
     ULONG NextEntryOffset;
     ULONG FileIndex;
-- 
2.4.2

