From 8ff87be88de81b45327a9fc0eeacbc357544be20 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Mon, 17 Aug 2015 01:10:10 +0200
Subject: server: Link named pipes to their device.

---
 server/named_pipe.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/server/named_pipe.c b/server/named_pipe.c
index 53bec02..0ce3e59 100644
--- a/server/named_pipe.c
+++ b/server/named_pipe.c
@@ -743,7 +743,7 @@ static struct named_pipe *create_named_pipe( struct directory *root, const struc
     else
     {
         struct named_pipe_device *dev = (struct named_pipe_device *)obj;
-        if ((pipe = create_object( dev->pipes, &named_pipe_ops, &new_name, NULL )))
+        if ((pipe = create_object( dev->pipes, &named_pipe_ops, &new_name, &dev->obj )))
             clear_error();
     }
 
-- 
2.5.0

