From 3d1d0332cc94bdac774643e1adfe987c534e088b Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 5 Feb 2016 19:07:42 +0100
Subject: server: Fix newly introduced memory leak of object name structure.

---
 server/object.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/server/object.c b/server/object.c
index c8098ca..b8be122 100644
--- a/server/object.c
+++ b/server/object.c
@@ -222,6 +222,7 @@ void *create_object( struct object *parent, const struct object_ops *ops, const
         if (!obj->ops->link_name( obj, name_ptr, parent ))
         {
             free_object( obj );
+            free( name_ptr );
             return NULL;
         }
         name_ptr->obj = obj;
-- 
2.7.0

