From 3df51527f28634f01e0f4f3af284544885d98e99 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 15 Nov 2014 01:04:29 +0100
Subject: server: Close fd if there is no space in thread inflight fd list.

---
 server/thread.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/server/thread.c b/server/thread.c
index 50d6940..97860ee 100644
--- a/server/thread.c
+++ b/server/thread.c
@@ -1047,6 +1047,8 @@ int thread_add_inflight_fd( struct thread *thread, int client, int server )
             thread->inflight[i].server = server;
             return i;
         }
+
+    close( server );
     return -1;
 }
 
-- 
2.1.3

