From 68a94ad2c728c00cc9f81b961b4d47dbe3ccb2f3 Mon Sep 17 00:00:00 2001
From: Christian Costa <titan.costa@gmail.com>
Date: Tue, 6 Sep 2016 00:10:09 +0200
Subject: msvcrt: Display noisy FIXME only once in pf_printf.

---
 dlls/msvcrt/printf.h | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/dlls/msvcrt/printf.h b/dlls/msvcrt/printf.h
index 6c9eeeb..4e59f31 100644
--- a/dlls/msvcrt/printf.h
+++ b/dlls/msvcrt/printf.h
@@ -495,8 +495,10 @@ int FUNC_NAME(pf_printf)(FUNC_NAME(puts_clbk) pf_puts, void *puts_ctx, const API
         } else if(flags.Format == 'c' || flags.Format == 'C') {
             int ch = pf_args(args_ctx, pos, VT_INT, valist).get_int;
 
-            if((ch&0xff) != ch)
-                FIXME("multibyte characters printing not supported\n");
+            if((ch&0xff) != ch) {
+                static int once;
+                if (!once++) FIXME("multibyte characters printing not supported\n");
+            }
 
             i = FUNC_NAME(pf_handle_string)(pf_puts, puts_ctx, &ch, 1, &flags, locinfo, legacy_wide);
         } else if(flags.Format == 'p') {
-- 
2.9.0

