From 9b5bb1a278bbdf33a1402603bae315fabfa1d411 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Mon, 9 Oct 2017 21:54:22 +0200
Subject: ntdll: Fix code to make DOS area accessible.

---
 dlls/ntdll/virtual.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ntdll/virtual.c b/dlls/ntdll/virtual.c
index d28b138e575..3ae72af8cd9 100644
--- a/dlls/ntdll/virtual.c
+++ b/dlls/ntdll/virtual.c
@@ -1622,7 +1622,7 @@ void virtual_init(void)
 
     /* make the DOS area accessible (except the low 64K) to hide bugs in broken apps like Excel 2003 */
     size = (char *)address_space_start - (char *)0x10000;
-    if (size && wine_mmap_is_in_reserved_area( (void*)0x10000, size ) == 1)
+    if (address_space_start > (void *)0x10000 && wine_mmap_is_in_reserved_area( (void *)0x10000, size ) == 1)
         wine_anon_mmap( (void *)0x10000, size, PROT_READ | PROT_WRITE, MAP_FIXED );
 }
 
-- 
2.14.1

