From 29ac40e7943195bda7acbc0ffa5cc966d8c1a40a Mon Sep 17 00:00:00 2001
From: Christian Costa <titan.costa@gmail.com>
Date: Mon, 31 Aug 2015 22:54:59 -0300
Subject: crypt32: Print CryptUnprotectMemory FIXME only once.

---
 dlls/crypt32/main.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/crypt32/main.c b/dlls/crypt32/main.c
index 241a1d9..696195a 100644
--- a/dlls/crypt32/main.c
+++ b/dlls/crypt32/main.c
@@ -259,6 +259,7 @@ BOOL WINAPI CryptProtectMemory(void *data, DWORD len, DWORD flags)
 
 BOOL WINAPI CryptUnprotectMemory(void *data, DWORD len, DWORD flags)
 {
-    FIXME("(%p %u %08x): stub\n", data, len, flags);
+    static int fixme_once;
+    if (!fixme_once++) FIXME("(%p %u %08x): stub\n", data, len, flags);
     return TRUE;
 }
-- 
2.5.0

