From ffc4081c4705362d89632dc82ed251c440e1c00f Mon Sep 17 00:00:00 2001
From: Lucian Poston <lucianposton@pm.me>
Date: Mon, 21 May 2018 19:40:03 -0700
Subject: [PATCH 3/6] dwrite: Skip failing font metrics test for Goha

Signed-off-by: Lucian Poston <lucianposton@pm.me>
---
 dlls/dwrite/tests/layout.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/dlls/dwrite/tests/layout.c b/dlls/dwrite/tests/layout.c
index 7542cad8d7..c8f3f5a00f 100644
--- a/dlls/dwrite/tests/layout.c
+++ b/dlls/dwrite/tests/layout.c
@@ -620,6 +620,7 @@ static HRESULT WINAPI testrenderer_DrawUnderline(IDWriteTextRenderer *iface,
 {
     struct renderer_context *ctxt = (struct renderer_context*)context;
     struct drawcall_entry entry = { 0 };
+    static const WCHAR gohaW[] = {'G','o','h','a','-','T','i','b','e','b',' ','Z','e','m','e','n',0};
 
     if (ctxt)
         TEST_MEASURING_MODE(ctxt, underline->measuringMode);
@@ -635,9 +636,13 @@ static HRESULT WINAPI testrenderer_DrawUnderline(IDWriteTextRenderer *iface,
         IDWriteFontFace_GetMetrics(fontface, &metrics);
 
         ok(emsize == metrics.designUnitsPerEm, "Unexpected font size %f\n", emsize);
-        /* Expected height is in design units, allow some absolute difference from it. Seems to only happen on Vista */
-        ok(fabs(metrics.capHeight - underline->runHeight) < 2.0f, "Expected runHeight %u, got %f, family %s\n",
-                metrics.capHeight, underline->runHeight, wine_dbgstr_w(ctxt->familyW));
+        if (lstrcmpW(gohaW, ctxt->familyW)) {
+            /* Expected height is in design units, allow some absolute difference from it. Seems to only happen on Vista */
+            ok(fabs(metrics.capHeight - underline->runHeight) < 2.0f, "Expected runHeight %u, got %f, family %s\n",
+                    metrics.capHeight, underline->runHeight, wine_dbgstr_w(ctxt->familyW));
+        } else {
+            skip("%s is an invalid font, rejected by windows.\n", wine_dbgstr_w(gohaW));
+        }
 
         IDWriteFontFace_Release(fontface);
     }
-- 
2.18.0

