From 3b4f9505365d6803c0205c6f16f19a15bf0cc24d Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 31 Aug 2018 15:32:13 +1000
Subject: [PATCH 2/5] hnetcfg: Register NATUPnP interface

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 dlls/hnetcfg/hnetcfg_tlb.idl | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/hnetcfg/hnetcfg_tlb.idl b/dlls/hnetcfg/hnetcfg_tlb.idl
index 33e8cc8..e368fe6 100644
--- a/dlls/hnetcfg/hnetcfg_tlb.idl
+++ b/dlls/hnetcfg/hnetcfg_tlb.idl
@@ -21,3 +21,4 @@
 #pragma makedep regtypelib
 
 #include "netfw.idl"
+#include "natupnp.idl"
-- 
1.9.1

