From 4b74342ce0dd261a93f5a2d30140078696b5a46d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Tue, 7 Mar 2017 15:26:50 +0100
Subject: include: Update LDR_MODULE to more recent windows versions.

---
 include/winternl.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/include/winternl.h b/include/winternl.h
index 5cd1664f1f..ef9a5bad1c 100644
--- a/include/winternl.h
+++ b/include/winternl.h
@@ -2103,6 +2103,13 @@ typedef struct _LDR_MODULE
     ULONG               CheckSum;
     ULONG               TimeDateStamp;
     HANDLE              ActivationContext;
+    PVOID               PatchInformation;
+    LIST_ENTRY          ForwarderLinks;
+    LIST_ENTRY          ServiceTagLinks;
+    LIST_ENTRY          StaticLinks;
+    PVOID               ContextInformation;
+    ULONG_PTR           OriginalBase;
+    LARGE_INTEGER       LoadTime;
 } LDR_MODULE, *PLDR_MODULE;
 
 /* those defines are (some of the) regular LDR_MODULE.Flags values */
-- 
2.11.0

