From 85dae537010f21064aea63c24888609169cf249a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Tue, 7 Mar 2017 15:26:50 +0100
Subject: [PATCH] include: Update LDR_MODULE to more recent windows versions.

---
 include/winternl.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/include/winternl.h b/include/winternl.h
index 7612f8a..3659db1 100644
--- a/include/winternl.h
+++ b/include/winternl.h
@@ -2168,6 +2168,13 @@ typedef struct _LDR_MODULE
     ULONG               CheckSum;
     ULONG               TimeDateStamp;
     HANDLE              ActivationContext;
+    PVOID               PatchInformation;
+    LIST_ENTRY          ForwarderLinks;
+    LIST_ENTRY          ServiceTagLinks;
+    LIST_ENTRY          StaticLinks;
+    PVOID               ContextInformation;
+    ULONG_PTR           OriginalBase;
+    LARGE_INTEGER       LoadTime;
 } LDR_MODULE, *PLDR_MODULE;
 
 typedef struct _LDR_DLL_LOADED_NOTIFICATION_DATA
-- 
1.9.1

