From af02399ac80682e6ca5c093150460dca20c7323b Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 4 Mar 2015 06:57:53 +0100
Subject: ntdll: Add support for threadpool group cancel callback.

---
 dlls/ntdll/threadpool2.c | 17 ++++++++++++++---
 1 file changed, 14 insertions(+), 3 deletions(-)

diff --git a/dlls/ntdll/threadpool2.c b/dlls/ntdll/threadpool2.c
index 31b5e2e..5233271 100644
--- a/dlls/ntdll/threadpool2.c
+++ b/dlls/ntdll/threadpool2.c
@@ -81,6 +81,7 @@ struct threadpool_object
     struct threadpool       *pool;
     struct threadpool_group *group;
     PVOID                   userdata;
+    PTP_CLEANUP_GROUP_CANCEL_CALLBACK group_cancel_callback;
     /* information about the group, locked via .group->cs */
     struct list             group_entry;
     BOOL                    is_group_member;
@@ -317,6 +318,7 @@ static void tp_object_initialize( struct threadpool_object *object, struct threa
     object->pool                    = pool;
     object->group                   = NULL;
     object->userdata                = userdata;
+    object->group_cancel_callback   = NULL;
 
     memset( &object->group_entry, 0, sizeof(object->group_entry) );
     object->is_group_member         = FALSE;
@@ -332,6 +334,7 @@ static void tp_object_initialize( struct threadpool_object *object, struct threa
             FIXME("unsupported environment version %u\n", environment->Version);
 
         object->group = impl_from_TP_CLEANUP_GROUP( environment->CleanupGroup );
+        object->group_cancel_callback   = environment->CleanupGroupCancelCallback;
 
         WARN("environment not fully implemented yet\n");
     }
@@ -460,7 +463,7 @@ out:
     RtlLeaveCriticalSection( &pool->cs );
 }
 
-static void tp_object_cancel( struct threadpool_object *object )
+static void tp_object_cancel( struct threadpool_object *object, BOOL group_cancel, PVOID userdata )
 {
     struct threadpool *pool = object->pool;
     LONG pending_callbacks = 0;
@@ -477,6 +480,14 @@ static void tp_object_cancel( struct threadpool_object *object )
 
     RtlLeaveCriticalSection( &pool->cs );
 
+    /* Execute group cancellation callback if defined, and if this was actually a group cancel. */
+    if (pending_callbacks && group_cancel && object->group_cancel_callback)
+    {
+        TRACE( "executing group cancel callback %p(%p, %p)\n", object->group_cancel_callback, object, userdata );
+        object->group_cancel_callback( object, userdata );
+        TRACE( "callback %p returned\n", object->group_cancel_callback );
+    }
+
     /* Release references */
     while (pending_callbacks--)
         tp_object_release( object );
@@ -627,7 +638,7 @@ static void tp_group_release_members( struct threadpool_group *group, BOOL cance
     {
         LIST_FOR_EACH_ENTRY( object, &members, struct threadpool_object, group_entry )
         {
-            tp_object_cancel( object );
+            tp_object_cancel( object, TRUE, userdata );
         }
     }
 
@@ -808,7 +819,7 @@ VOID WINAPI TpWaitForWork( TP_WORK *work, BOOL cancel_pending )
     if (this)
     {
         if (cancel_pending)
-            tp_object_cancel( this );
+            tp_object_cancel( this, FALSE, NULL );
         tp_object_wait( this );
     }
 }
-- 
2.3.0

