From 0aa4ad9970796cbbbe4fc5f81209e6d5d505a385 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 6 Feb 2015 20:09:41 +0100
Subject: ntdll/tests: Add tests for threadpool wait objects.

---
 dlls/ntdll/tests/threadpool.c | 287 ++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 287 insertions(+)

diff --git a/dlls/ntdll/tests/threadpool.c b/dlls/ntdll/tests/threadpool.c
index 4be3a8b..fd587ba 100644
--- a/dlls/ntdll/tests/threadpool.c
+++ b/dlls/ntdll/tests/threadpool.c
@@ -869,6 +869,291 @@ static void test_tp_window_length(void)
     pTpReleasePool(pool);
 }
 
+static void CALLBACK wait_cb(TP_CALLBACK_INSTANCE *instance, void *userdata, TP_WAIT *wait, TP_WAIT_RESULT result)
+{
+    trace("Running wait callback\n");
+
+    if (result == WAIT_OBJECT_0)
+        InterlockedIncrement((LONG *)userdata);
+    else if (result == WAIT_TIMEOUT)
+        InterlockedExchangeAdd((LONG *)userdata, 0x10000);
+    else
+        ok(0, "unexpected result %u\n", result);
+}
+
+static void test_tp_wait(void)
+{
+    TP_CALLBACK_ENVIRON environment;
+    HANDLE semaphore;
+    TP_WAIT *wait, *wait2;
+    TP_POOL *pool;
+    NTSTATUS status;
+    LONG userdata;
+    LARGE_INTEGER when;
+    DWORD ret;
+
+    /* Allocate new threadpool */
+    pool = NULL;
+    status = pTpAllocPool(&pool, NULL);
+    ok(!status, "TpAllocPool failed with status %x\n", status);
+    ok(pool != NULL, "expected pool != NULL\n");
+
+    /* Allocate new wait items */
+    wait = NULL;
+    memset(&environment, 0, sizeof(environment));
+    environment.Version = 1;
+    environment.Pool = pool;
+    status = pTpAllocWait(&wait, wait_cb, &userdata, &environment);
+    ok(!status, "TpAllocWait failed with status %x\n", status);
+    ok(wait != NULL, "expected wait != NULL\n");
+
+    wait2 = NULL;
+    status = pTpAllocWait(&wait2, wait_cb, &userdata, &environment);
+    ok(!status, "TpAllocWait failed with status %x\n", status);
+    ok(wait != NULL, "expected wait != NULL\n");
+
+    semaphore = CreateSemaphoreW(NULL, 0, 1, NULL);
+    ok(semaphore != NULL, "failed to create semaphore\n");
+
+    /* Infinite timeout, signal the semaphore immediately */
+    userdata = 0;
+    pTpSetWait(wait, semaphore, NULL);
+    ReleaseSemaphore(semaphore, 1, NULL);
+    Sleep(50);
+    ok(userdata == 1, "expected userdata = 1, got %u\n", userdata);
+
+    /* Relative timeout, no event */
+    userdata = 0;
+    when.QuadPart = (ULONGLONG)50 * -10000;
+    pTpSetWait(wait, semaphore, &when);
+    Sleep(100);
+    pTpWaitForWait(wait, FALSE);
+    ok(userdata == 0x10000, "expected userdata = 0x10000, got %u\n", userdata);
+    ret = WaitForSingleObject(semaphore, 50);
+    ok(ret == WAIT_TIMEOUT, "expected ret = WAIT_TIMEOUT, got %u\n", ret);
+
+    /* Relative timeout, with event */
+    userdata = 0;
+    when.QuadPart = (ULONGLONG)500 * -10000;
+    pTpSetWait(wait, semaphore, &when);
+    pTpWaitForWait(wait, TRUE);
+    Sleep(250);
+    ReleaseSemaphore(semaphore, 1, NULL);
+    Sleep(50);
+    pTpWaitForWait(wait, FALSE);
+    ok(userdata == 1, "expected userdata = 1, got %u\n", userdata);
+    ret = WaitForSingleObject(semaphore, 50);
+    ok(ret == WAIT_TIMEOUT, "expected ret = WAIT_TIMEOUT, got %u\n", ret);
+
+    /* Absolute timeout, no event */
+    userdata = 0;
+    NtQuerySystemTime( &when );
+    when.QuadPart += (ULONGLONG)50 * 10000;
+    pTpSetWait(wait, semaphore, &when);
+    Sleep(100);
+    pTpWaitForWait(wait, FALSE);
+    ok(userdata == 0x10000, "expected userdata = 0x10000, got %u\n", userdata);
+    ret = WaitForSingleObject(semaphore, 50);
+    ok(ret == WAIT_TIMEOUT, "expected ret = WAIT_TIMEOUT, got %u\n", ret);
+
+    /* Absolute timeout, with event */
+    userdata = 0;
+    NtQuerySystemTime( &when );
+    when.QuadPart += (ULONGLONG)500 * 10000;
+    pTpSetWait(wait, semaphore, &when);
+    pTpWaitForWait(wait, TRUE);
+    Sleep(250);
+    ReleaseSemaphore(semaphore, 1, NULL);
+    Sleep(50);
+    pTpWaitForWait(wait, FALSE);
+    ok(userdata == 1, "expected userdata = 1, got %u\n", userdata);
+    ret = WaitForSingleObject(semaphore, 50);
+    ok(ret == WAIT_TIMEOUT, "expected ret = WAIT_TIMEOUT, got %u\n", ret);
+
+    /* Trigger event immediately */
+    userdata = 0;
+    when.QuadPart = 0;
+    pTpSetWait(wait, semaphore, &when);
+    pTpWaitForWait(wait, FALSE);
+    ok(userdata == 0x10000, "expected userdata = 0x10000, got %u\n", userdata);
+    ret = WaitForSingleObject(semaphore, 50);
+    ok(ret == WAIT_TIMEOUT, "expected ret = WAIT_TIMEOUT, got %u\n", ret);
+
+    /* Cancel a pending wait */
+    userdata = 0;
+    when.QuadPart = (ULONGLONG)500 * -10000;
+    pTpSetWait(wait, semaphore, &when);
+    pTpWaitForWait(wait, TRUE);
+    Sleep(250);
+    pTpSetWait(wait, NULL, (void *)0xdeadbeef);
+    Sleep(50);
+    ReleaseSemaphore(semaphore, 1, NULL);
+    Sleep(50);
+    ok(userdata == 0, "expected userdata = 0, got %u\n", userdata);
+    ret = WaitForSingleObject(semaphore, 1000);
+    ok(ret == WAIT_OBJECT_0, "expected ret = WAIT_OBJECT_0, got %u\n", ret);
+
+    /* Test with INVALID_HANDLE_VALUE */
+    userdata = 0;
+    when.QuadPart = 0;
+    pTpSetWait(wait, INVALID_HANDLE_VALUE, &when);
+    Sleep(50);
+    ok(userdata == 0x10000, "expected userdata = 0x10000, got %u\n", userdata);
+
+    /* Cancel a pending wait with INVALID_HANDLE_VALUE */
+    userdata = 0;
+    when.QuadPart = (ULONGLONG)500 * -10000;
+    pTpSetWait(wait, semaphore, &when);
+    pTpWaitForWait(wait, TRUE);
+    Sleep(250);
+    when.QuadPart = (ULONGLONG)100 * -10000;
+    pTpSetWait(wait, INVALID_HANDLE_VALUE, &when);
+    Sleep(250);
+    ok(userdata == 0x10000, "expected userdata = 0x10000, got %u\n", userdata);
+
+    /* Add two objects with the same semaphore */
+    userdata = 0;
+    pTpSetWait(wait, semaphore, NULL);
+    pTpSetWait(wait2, semaphore, NULL);
+    ok(userdata == 0, "expected userdata = 0, got %u\n", userdata);
+    ReleaseSemaphore(semaphore, 1, NULL);
+    Sleep(10);
+    pTpWaitForWait(wait, FALSE);
+    pTpWaitForWait(wait2, FALSE);
+    ok(userdata == 1, "expected userdata = 1, got %u\n", userdata);
+    ret = WaitForSingleObject(semaphore, 50);
+    ok(ret == WAIT_TIMEOUT, "expected ret = WAIT_TIMEOUT, got %u\n", ret);
+
+    CloseHandle(semaphore);
+    semaphore = CreateSemaphoreW(NULL, 0, 2, NULL);
+    ok(semaphore != NULL, "failed to create semaphore\n");
+
+    /* Repeat test above, but with a semaphore of count 2 */
+    userdata = 0;
+    pTpSetWait(wait, semaphore, NULL);
+    pTpSetWait(wait2, semaphore, NULL);
+    ok(userdata == 0, "expected userdata = 0, got %u\n", userdata);
+    ReleaseSemaphore(semaphore, 2, NULL);
+    Sleep(10);
+    pTpWaitForWait(wait, FALSE);
+    pTpWaitForWait(wait2, FALSE);
+    ok(userdata == 2, "expected userdata = 2, got %u\n", userdata);
+    ret = WaitForSingleObject(semaphore, 50);
+    ok(ret == WAIT_TIMEOUT, "expected ret = WAIT_TIMEOUT, got %u\n", ret);
+
+    CloseHandle(semaphore);
+
+    /* Cleanup */
+    pTpReleaseWait(wait2);
+    pTpReleaseWait(wait);
+    pTpReleasePool(pool);
+}
+
+static LONG multi_wait_callbacks;
+static DWORD multi_wait_result;
+
+static void CALLBACK multi_wait_cb(TP_CALLBACK_INSTANCE *instance, void *userdata, TP_WAIT *wait, TP_WAIT_RESULT result)
+{
+    DWORD index = (DWORD)(DWORD_PTR)userdata;
+    InterlockedIncrement(&multi_wait_callbacks);
+
+    if (result == WAIT_OBJECT_0)
+        multi_wait_result = index;
+    else if (result == WAIT_TIMEOUT)
+        multi_wait_result = 0x10000 | index;
+    else
+        ok(0, "unexpected result %u\n", result);
+}
+
+static void test_tp_multi_wait(void)
+{
+    TP_CALLBACK_ENVIRON environment;
+    HANDLE semaphores[512];
+    TP_WAIT *waits[512];
+    TP_POOL *pool;
+    NTSTATUS status;
+    LARGE_INTEGER when;
+    int i;
+
+    /* Allocate new threadpool */
+    pool = NULL;
+    status = pTpAllocPool(&pool, NULL);
+    ok(!status, "TpAllocPool failed with status %x\n", status);
+    ok(pool != NULL, "expected pool != NULL\n");
+
+    memset(&environment, 0, sizeof(environment));
+    environment.Version = 1;
+    environment.Pool = pool;
+
+    /* Create semaphores, wait objects and enable them */
+    for (i = 0; i < sizeof(semaphores)/sizeof(semaphores[0]); i++)
+    {
+        semaphores[i] = CreateSemaphoreW(NULL, 0, 1, NULL);
+        ok(semaphores[i] != NULL, "failed to create semaphores[%d]\n", i);
+
+        waits[i] = NULL;
+        status = pTpAllocWait(&waits[i], multi_wait_cb, (void *)(DWORD_PTR)i, &environment);
+        ok(!status, "TpAllocWait failed with status %x\n", status);
+        ok(waits[i] != NULL, "expected waits[%d] != NULL\n", i);
+
+        pTpSetWait(waits[i], semaphores[i], NULL);
+    }
+
+    /* Now test releasing the semaphores */
+    for (i = 0; i < sizeof(semaphores)/sizeof(semaphores[0]); i++)
+    {
+        multi_wait_callbacks = 0;
+        multi_wait_result = 0;
+
+        ReleaseSemaphore(semaphores[i], 1, NULL);
+        while (multi_wait_callbacks == 0) Sleep(10);
+        ok(multi_wait_callbacks == 1, "expected multi_wait_callbacks = 1, got %u\n", multi_wait_callbacks);
+        ok(multi_wait_result == i, "expected multi_wait_result = %u, got %u\n", multi_wait_result, i);
+
+        pTpSetWait(waits[i], semaphores[i], NULL);
+    }
+
+    /* Now again in the reversed order */
+    for (i = sizeof(semaphores)/sizeof(semaphores[0]) - 1; i >= 0; i--)
+    {
+        multi_wait_callbacks = 0;
+        multi_wait_result = 0;
+
+        ReleaseSemaphore(semaphores[i], 1, NULL);
+        while (multi_wait_callbacks == 0) Sleep(10);
+        ok(multi_wait_callbacks == 1, "expected multi_wait_callbacks = 1, got %u\n", multi_wait_callbacks);
+        ok(multi_wait_result == i, "expected multi_wait_result = %u, got %u\n", multi_wait_result, i);
+
+        pTpSetWait(waits[i], semaphores[i], NULL);
+    }
+
+    /* Now test with a timeout */
+    multi_wait_callbacks = 0;
+    multi_wait_result = 0;
+    for (i = 0; i < sizeof(semaphores)/sizeof(semaphores[0]); i++)
+    {
+        when.QuadPart = 0;
+        pTpSetWait(waits[i], semaphores[i], &when);
+    }
+    Sleep(50);
+    ok(multi_wait_callbacks == sizeof(semaphores)/sizeof(semaphores[0]),
+       "got wrong multi_wait_callbacks %u\n", multi_wait_callbacks);
+    ok(multi_wait_result >> 16, "expected multi_wait_result >> 16 != 0\n");
+
+    /* Add them all again, we want that the wait is pending while destroying it */
+    for (i = 0; i < sizeof(semaphores)/sizeof(semaphores[0]); i++)
+        pTpSetWait(waits[i], semaphores[i], NULL);
+
+    /* Destroy the objects and semaphores */
+    for (i = 0; i < sizeof(semaphores)/sizeof(semaphores[0]); i++)
+    {
+        pTpReleaseWait(waits[i]);
+        NtClose(semaphores[i]);
+    }
+
+    pTpReleasePool(pool);
+}
+
 START_TEST(threadpool)
 {
     if (!init_threadpool())
@@ -882,6 +1167,8 @@ START_TEST(threadpool)
     test_tp_disassociate();
     test_tp_timer();
     test_tp_window_length();
+    test_tp_wait();
+    test_tp_multi_wait();
 
     /* FIXME: Make sure worker threads have terminated before. */
     Sleep(100);
-- 
2.4.4

