From f7356c55fbeb2cc4d4ffa0b684e371ffc8183731 Mon Sep 17 00:00:00 2001
From: Amine Khaldi <amine.khaldi@reactos.org>
Date: Mon, 23 Feb 2015 21:40:25 +0100
Subject: dinput/tests: Declare pDirectInputCreateEx in a MSVC compatible way.

---
 dlls/dinput/tests/dinput.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/dinput/tests/dinput.c b/dlls/dinput/tests/dinput.c
index 0490642..5ac55cf 100644
--- a/dlls/dinput/tests/dinput.c
+++ b/dlls/dinput/tests/dinput.c
@@ -50,7 +50,7 @@ static const DWORD directinput_version_list[] =
     DIRECTINPUT_VERSION_700,
 };
 
-static HRESULT WINAPI (*pDirectInputCreateEx)(HINSTANCE, DWORD, REFIID, LPVOID *, LPUNKNOWN);
+static HRESULT (WINAPI *pDirectInputCreateEx)(HINSTANCE, DWORD, REFIID, LPVOID *, LPUNKNOWN);
 
 static BOOL CALLBACK dummy_callback(const DIDEVICEINSTANCEA *instance, void *context)
 {
-- 
2.3.0

