From 348d1eafe2d80017509ebfb79832a7465fd1693b Mon Sep 17 00:00:00 2001
From: Amine Khaldi <amine.khaldi@reactos.org>
Date: Sun, 22 Feb 2015 22:39:25 +0100
Subject: d3drm: Mark DllCanUnloadNow and DllGetClassObject as private.

---
 dlls/d3drm/d3drm.spec | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/d3drm/d3drm.spec b/dlls/d3drm/d3drm.spec
index 4108a72..e83fb0b 100644
--- a/dlls/d3drm/d3drm.spec
+++ b/dlls/d3drm/d3drm.spec
@@ -19,5 +19,5 @@
 @ stdcall D3DRMVectorScale(ptr ptr float)
 @ stdcall D3DRMVectorSubtract(ptr ptr ptr)
 @ stdcall Direct3DRMCreate(ptr)
-@ stub DllCanUnloadNow
-@ stub DllGetClassObject
+@ stub -private DllCanUnloadNow
+@ stub -private DllGetClassObject
-- 
2.3.0

