From cbb471762f116b18b77288009cd1c7002ed9c2cb Mon Sep 17 00:00:00 2001
From: Kimmo Myllyvirta <kimmo.myllyvirta@gmail.com>
Date: Sun, 23 Jul 2017 22:22:41 +0300
Subject: dxgi: Add stub implementation for DXGI_PRESENT_TEST.

---
 dlls/dxgi/swapchain.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/dlls/dxgi/swapchain.c b/dlls/dxgi/swapchain.c
index 2b61327f0d3..90c8ab55094 100644
--- a/dlls/dxgi/swapchain.c
+++ b/dlls/dxgi/swapchain.c
@@ -171,6 +171,13 @@ static HRESULT STDMETHODCALLTYPE dxgi_swapchain_Present(IDXGISwapChain *iface, U
     TRACE("iface %p, sync_interval %u, flags %#x\n", iface, sync_interval, flags);
 
     if (sync_interval) FIXME("Unimplemented sync interval %u\n", sync_interval);
+
+    if (flags == DXGI_PRESENT_TEST)
+    {
+        WARN("DXGI_PRESENT_TEST stub!\n");
+        return S_OK;
+    }
+
     if (flags) FIXME("Unimplemented flags %#x\n", flags);
 
     wined3d_mutex_lock();
-- 
2.13.1

