From 4b55f147ef88c192985081fc5a023de2b198cfb1 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Mon, 7 Mar 2016 02:40:56 +0100
Subject: configure.ac: Fix cross-compiling of 32-bit Wine for OSX.

---
 configure.ac | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/configure.ac b/configure.ac
index b57ff6a..fd65810 100644
--- a/configure.ac
+++ b/configure.ac
@@ -175,6 +175,13 @@ case $host in
     AC_SUBST(TARGETFLAGS,"-marm")
     ;;
   i[[3456789]]86*)
+    if test "$cross_compiling" = "yes"; then
+      CC="$CC -m32"
+      CXX="$CXX -m32"
+      host_cpu="i386"
+      notice_platform="32-bit "
+      AC_SUBST(TARGETFLAGS,"-m32")
+    fi
     enable_win16=${enable_win16:-yes}
     ;;
 esac
-- 
2.7.1

