From 9743c07fc7a759f1e7c893414be8bd95a095921f Mon Sep 17 00:00:00 2001
From: Zebediah Figura <z.figura12@gmail.com>
Date: Wed, 24 Apr 2019 23:21:25 -0500
Subject: [PATCH] server: Use default_fd_get_esync_fd() for directory change
 notification objects.

---
 server/change.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/server/change.c b/server/change.c
index 2be6a8360..9f07be705 100644
--- a/server/change.c
+++ b/server/change.c
@@ -115,7 +115,7 @@ static const struct object_ops dir_ops =
     add_queue,                /* add_queue */
     remove_queue,             /* remove_queue */
     default_fd_signaled,      /* signaled */
-    NULL,                     /* get_esync_fd */
+    default_fd_get_esync_fd,  /* get_esync_fd */
     no_satisfied,             /* satisfied */
     no_signal,                /* signal */
     dir_get_fd,               /* get_fd */
-- 
2.21.0

