From 86412d7f8bd47ab0228e3a2375e353003f27c199 Mon Sep 17 00:00:00 2001
From: Jacek Caban <jacek@codeweavers.com>
Date: Sat, 18 Apr 2015 21:48:00 +0200
Subject: mshtml: Avoid crash when trying to bind event scripts to window.

---
 dlls/mshtml/script.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/dlls/mshtml/script.c b/dlls/mshtml/script.c
index 4590854..e38e5bf 100644
--- a/dlls/mshtml/script.c
+++ b/dlls/mshtml/script.c
@@ -1506,8 +1506,10 @@ void bind_event_scripts(HTMLDocumentNode *doc)
 
                 if(SUCCEEDED(hres))
                     bind_activex_event(doc, plugin_container, event, event_disp);
-                else
+                else if(target_node)
                     bind_node_event(doc, event_target, target_node, event, event_disp);
+                else
+                    FIXME("binding to window not supported.\n");
 
                 if(target_node) {
                     IHTMLDOMNode_Release(&target_node->IHTMLDOMNode_iface);
-- 
2.3.5

