From 0c70b7b664bc3a1b997cd8ff72f08deae099a879 Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Thu, 3 Apr 2014 21:36:33 -0600
Subject: wined3d: Silence repeated resource_check_usage FIXME.

---
 dlls/wined3d/resource.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/dlls/wined3d/resource.c b/dlls/wined3d/resource.c
index 999dc5c..5a06346 100644
--- a/dlls/wined3d/resource.c
+++ b/dlls/wined3d/resource.c
@@ -56,9 +56,14 @@ static void resource_check_usage(DWORD usage)
             | WINED3DUSAGE_STATICDECL
             | WINED3DUSAGE_OVERLAY
             | WINED3DUSAGE_TEXTURE;
+    static DWORD notified = 0;
+    DWORD unhandled = usage & ~handled;
 
-    if (usage & ~handled)
-        FIXME("Unhandled usage flags %#x.\n", usage & ~handled);
+    if (unhandled && (unhandled & ~notified))
+    {
+        FIXME("Unhandled usage flags %#x.\n", unhandled);
+        notified |= unhandled;
+    }
 }
 
 HRESULT resource_init(struct wined3d_resource *resource, struct wined3d_device *device,
-- 
1.7.9.5

