From abc7d034d2a7ba515b01386166c1ef51d7c53eaf Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 3 Apr 2014 05:25:38 +0200
Subject: kernel32: Silence SetWaitableTimerEx fixme message

---
 dlls/kernel32/sync.c |    8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/dlls/kernel32/sync.c b/dlls/kernel32/sync.c
index 6c452a4..d363e1f 100644
--- a/dlls/kernel32/sync.c
+++ b/dlls/kernel32/sync.c
@@ -1183,8 +1183,12 @@ BOOL WINAPI SetWaitableTimer( HANDLE handle, const LARGE_INTEGER *when, LONG per
 BOOL WINAPI SetWaitableTimerEx( HANDLE handle, const LARGE_INTEGER *when, LONG period,
                               PTIMERAPCROUTINE callback, LPVOID arg, REASON_CONTEXT *context, ULONG tolerabledelay )
 {
-    FIXME("(%p, %p, %d, %p, %p, %p, %d) semi-stub\n",
-          handle, when, period, callback, arg, context, tolerabledelay);
+    static int once;
+    if (!once++)
+    {
+        FIXME("(%p, %p, %d, %p, %p, %p, %d) semi-stub\n",
+              handle, when, period, callback, arg, context, tolerabledelay);
+    }
     return SetWaitableTimer(handle, when, period, callback, arg, FALSE);
 }
 
-- 
1.7.9.5

