From 950ff9091599bb19c702b3d679d67d3f679b7814 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 7 Mar 2014 18:31:25 +0100
Subject: shell32: Register user administrative tools shell folder

---
 dlls/shell32/shellpath.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/shell32/shellpath.c b/dlls/shell32/shellpath.c
index 947ef29..e9b6fa8 100644
--- a/dlls/shell32/shellpath.c
+++ b/dlls/shell32/shellpath.c
@@ -2526,7 +2526,8 @@ static HRESULT _SHRegisterUserShellFolders(BOOL bDefault)
      CSIDL_COOKIES,
      CSIDL_HISTORY,
      CSIDL_MYPICTURES,
-     CSIDL_FONTS
+     CSIDL_FONTS,
+     CSIDL_ADMINTOOLS
     };
     WCHAR userShellFolderPath[MAX_PATH], shellFolderPath[MAX_PATH];
     LPCWSTR pUserShellFolderPath, pShellFolderPath;
-- 
1.7.9.5

