From 13d5795515836c4b8674bf96643333d68cd3d3a0 Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@gmail.com>
Date: Thu, 3 Apr 2014 21:36:33 -0600
Subject: wined3d: Silence repeated resource_check_usage FIXME.

---
 dlls/wined3d/resource.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/dlls/wined3d/resource.c b/dlls/wined3d/resource.c
index 999dc5c..3558440 100644
--- a/dlls/wined3d/resource.c
+++ b/dlls/wined3d/resource.c
@@ -49,7 +49,7 @@ static DWORD resource_access_from_pool(enum wined3d_pool pool)
 
 static void resource_check_usage(DWORD usage)
 {
-    static const DWORD handled = WINED3DUSAGE_RENDERTARGET
+    static DWORD handled = WINED3DUSAGE_RENDERTARGET
             | WINED3DUSAGE_DEPTHSTENCIL
             | WINED3DUSAGE_DYNAMIC
             | WINED3DUSAGE_AUTOGENMIPMAP
@@ -58,7 +58,10 @@ static void resource_check_usage(DWORD usage)
             | WINED3DUSAGE_TEXTURE;
 
     if (usage & ~handled)
+    {
         FIXME("Unhandled usage flags %#x.\n", usage & ~handled);
+        handled |= usage;
+    }
 }
 
 HRESULT resource_init(struct wined3d_resource *resource, struct wined3d_device *device,
-- 
1.7.9.5

