From a8f0b6092f5f1a52ca87bf6fc2822fe0b4edcc84 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 13 Sep 2015 23:03:43 +0200
Subject: ntdll: Ignore higher bits in selector for ThreadDescriptorTableEntry
 info query.

---
 dlls/ntdll/thread.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/ntdll/thread.c b/dlls/ntdll/thread.c
index 0be5585..5799f27 100644
--- a/dlls/ntdll/thread.c
+++ b/dlls/ntdll/thread.c
@@ -1062,7 +1062,7 @@ NTSTATUS WINAPI NtQueryInformationThread( HANDLE handle, THREADINFOCLASS class,
                 status = STATUS_INFO_LENGTH_MISMATCH;
             else if (!(tdi->Selector & 4))  /* GDT selector */
             {
-                unsigned sel = tdi->Selector & ~3;  /* ignore RPL */
+                unsigned sel = LOWORD(tdi->Selector) & ~3;  /* ignore RPL */
                 status = STATUS_SUCCESS;
                 if (!sel)  /* null selector */
                     memset( &tdi->Entry, 0, sizeof(tdi->Entry) );
@@ -1093,7 +1093,7 @@ NTSTATUS WINAPI NtQueryInformationThread( HANDLE handle, THREADINFOCLASS class,
                 SERVER_START_REQ( get_selector_entry )
                 {
                     req->handle = wine_server_obj_handle( handle );
-                    req->entry = tdi->Selector >> 3;
+                    req->entry = LOWORD(tdi->Selector) >> 3;
                     status = wine_server_call( req );
                     if (!status)
                     {
-- 
2.5.1

