From d4dc0e7e0c45cce29e4b6e968184eb6c946b825e Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 25 May 2017 01:50:05 +0200
Subject: winex11.drv: Fix compilation without XFixes.

---
 dlls/winex11.drv/clipboard.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/winex11.drv/clipboard.c b/dlls/winex11.drv/clipboard.c
index 3da40811794..8e6e05b8288 100644
--- a/dlls/winex11.drv/clipboard.c
+++ b/dlls/winex11.drv/clipboard.c
@@ -1954,21 +1954,21 @@ static BOOL wait_clipboard_mutex(void)
 
 
 /**************************************************************************
- *              handle_selection_notify_event
+ *              selection_notify_event
  *
  * Called when x11 clipboard content changes
  */
+#ifdef SONAME_LIBXFIXES
 static BOOL selection_notify_event( HWND hwnd, XEvent *event )
 {
-#ifdef SONAME_LIBXFIXES
     XFixesSelectionNotifyEvent *req = (XFixesSelectionNotifyEvent*)event;
 
     if (!is_clipboard_owner) return FALSE;
     if (req->owner == selection_window) return FALSE;
     request_selection_contents( req->display, TRUE );
     return FALSE;
-#endif
 }
+#endif
 
 /**************************************************************************
  *		xfixes_init
-- 
2.12.2

