From 479993311ae3467ba359b76057744dec56fa2fea Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 26 Jul 2015 20:54:23 +0200
Subject: kernel32: Fake success in SetFileCompletionNotificationModes.

---
 dlls/kernel32/file.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/dlls/kernel32/file.c b/dlls/kernel32/file.c
index e43829e..569ed22 100644
--- a/dlls/kernel32/file.c
+++ b/dlls/kernel32/file.c
@@ -1046,8 +1046,7 @@ BOOL WINAPI SetEndOfFile( HANDLE hFile )
 BOOL WINAPI SetFileCompletionNotificationModes( HANDLE handle, UCHAR flags )
 {
     FIXME("%p %x - stub\n", handle, flags);
-    SetLastError(ERROR_CALL_NOT_IMPLEMENTED);
-    return FALSE;
+    return TRUE;
 }
 
 
-- 
2.4.5

