From a713a9bc504a7c87cf003caff225fd1dd035e6aa Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Tue, 28 Mar 2017 13:31:34 +0800
Subject: windowscodecs: Implement IWICMetadataQueryReader::GetContainerFormat.

---
 dlls/windowscodecs/metadataquery.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/dlls/windowscodecs/metadataquery.c b/dlls/windowscodecs/metadataquery.c
index bf2668ea4b..f84d595851 100644
--- a/dlls/windowscodecs/metadataquery.c
+++ b/dlls/windowscodecs/metadataquery.c
@@ -90,12 +90,13 @@ static ULONG WINAPI mqr_Release(IWICMetadataQueryReader *iface)
     return ref;
 }
 
-static HRESULT WINAPI mqr_GetContainerFormat(IWICMetadataQueryReader *iface,
-        GUID *pguidContainerFormat)
+static HRESULT WINAPI mqr_GetContainerFormat(IWICMetadataQueryReader *iface, GUID *format)
 {
     QueryReader *This = impl_from_IWICMetadataQueryReader(iface);
-    FIXME("(%p,%p)\n", This, pguidContainerFormat);
-    return E_NOTIMPL;
+
+    TRACE("(%p,%p)\n", This, format);
+
+    return IWICMetadataBlockReader_GetContainerFormat(This->block, format);
 }
 
 static HRESULT WINAPI mqr_GetLocation(IWICMetadataQueryReader *iface,
-- 
2.12.2

