From 835b3f4dbfff6eb52e605933387200b488ded25a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sun, 22 Jan 2017 15:00:38 +0100
Subject: wined3d: Add more blend modes.

---
 dlls/wined3d/state.c   | 8 ++++++++
 include/wine/wined3d.h | 4 ++++
 2 files changed, 12 insertions(+)

diff --git a/dlls/wined3d/state.c b/dlls/wined3d/state.c
index 5952c6cc610..1a8c0431511 100644
--- a/dlls/wined3d/state.c
+++ b/dlls/wined3d/state.c
@@ -412,6 +412,14 @@ static GLenum gl_blend_factor(enum wined3d_blend factor, const struct wined3d_fo
             return GL_CONSTANT_COLOR_EXT;
         case WINED3D_BLEND_INVBLENDFACTOR:
             return GL_ONE_MINUS_CONSTANT_COLOR_EXT;
+        case WINED3D_BLEND_SRC1COLOR:
+            return GL_SRC1_COLOR;
+        case WINED3D_BLEND_INVSRC1COLOR:
+            return GL_ONE_MINUS_SRC1_COLOR;
+        case WINED3D_BLEND_SRC1ALPHA:
+            return GL_SRC1_ALPHA;
+        case WINED3D_BLEND_INVSRC1ALPHA:
+            return GL_ONE_MINUS_SRC1_ALPHA;
         default:
         {
             static int once;
diff --git a/include/wine/wined3d.h b/include/wine/wined3d.h
index dc333448772..ffad17fa6e8 100644
--- a/include/wine/wined3d.h
+++ b/include/wine/wined3d.h
@@ -400,6 +400,10 @@ enum wined3d_blend
     WINED3D_BLEND_BOTHINVSRCALPHA           = 13,
     WINED3D_BLEND_BLENDFACTOR               = 14,
     WINED3D_BLEND_INVBLENDFACTOR            = 15,
+    WINED3D_BLEND_SRC1COLOR                 = 16,
+    WINED3D_BLEND_INVSRC1COLOR              = 17,
+    WINED3D_BLEND_SRC1ALPHA                 = 18,
+    WINED3D_BLEND_INVSRC1ALPHA              = 19,
 };
 
 enum wined3d_blend_op
-- 
2.11.0

