From ad76cf41e0a79a75c0acb3a8aae6caa6dab9549c Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Fri, 31 Mar 2017 13:39:41 +0800
Subject: propsys: Add support for VT_I1 and VT_UI1 to PropVariantChangeType.

---
 dlls/propsys/propvar.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/dlls/propsys/propvar.c b/dlls/propsys/propvar.c
index 3fee67ad276..007fd94b840 100644
--- a/dlls/propsys/propvar.c
+++ b/dlls/propsys/propvar.c
@@ -264,6 +264,32 @@ HRESULT WINAPI PropVariantChangeType(PROPVARIANT *ppropvarDest, REFPROPVARIANT p
 
     switch (vt)
     {
+    case VT_I1:
+    {
+        LONGLONG res;
+
+        hr = PROPVAR_ConvertNumber(propvarSrc, 8, TRUE, &res);
+        if (SUCCEEDED(hr))
+        {
+            ppropvarDest->vt = VT_I1;
+            ppropvarDest->u.cVal = (char)res;
+        }
+        return hr;
+    }
+
+    case VT_UI1:
+    {
+        LONGLONG res;
+
+        hr = PROPVAR_ConvertNumber(propvarSrc, 8, FALSE, &res);
+        if (SUCCEEDED(hr))
+        {
+            ppropvarDest->vt = VT_UI1;
+            ppropvarDest->u.bVal = (UCHAR)res;
+        }
+        return hr;
+    }
+
     case VT_I2:
     {
         SHORT res;
-- 
2.12.2

