From a862cb25cefe885d4de212c21fc8a62dc8b8c682 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Fri, 31 Mar 2017 13:45:23 +0800
Subject: propsys: Add support for VT_BSTR to PropVariantChangeType.

---
 dlls/propsys/propvar.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/propsys/propvar.c b/dlls/propsys/propvar.c
index a359bc724fc..92399f56576 100644
--- a/dlls/propsys/propvar.c
+++ b/dlls/propsys/propvar.c
@@ -369,7 +369,9 @@ HRESULT WINAPI PropVariantChangeType(PROPVARIANT *ppropvarDest, REFPROPVARIANT p
         }
         return hr;
     }
+
     case VT_LPWSTR:
+    case VT_BSTR:
     {
         WCHAR *res;
         hr = PropVariantToStringAlloc(propvarSrc, &res);
-- 
2.12.2

