From de2e5f34189173bb594b68fc392268c8820589e6 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 9 Mar 2018 16:27:51 +1100
Subject: [PATCH] include: Add more types to ntifs.h

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 include/ddk/ntifs.h | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/include/ddk/ntifs.h b/include/ddk/ntifs.h
index 32c9e3084b..5e7dd18c7e 100644
--- a/include/ddk/ntifs.h
+++ b/include/ddk/ntifs.h
@@ -19,8 +19,24 @@
 #ifndef __NTIFS_H__
 #define __NTIFS_H__
 
+#include "ntddk.h"
+
 typedef struct _EX_PUSH_LOCK EX_PUSH_LOCK, *PEX_PUSH_LOCK;
 
+typedef enum _FS_FILTER_SECTION_SYNC_TYPE
+{
+    SyncTypeOther         = 0,
+    SyncTypeCreateSection
+} FS_FILTER_SECTION_SYNC_TYPE, *PFS_FILTER_SECTION_SYNC_TYPE;
+
+typedef struct _FS_FILTER_SECTION_SYNC_OUTPUT
+{
+    ULONG StructureSize;
+    ULONG SizeReturned;
+    ULONG Flags;
+    ULONG DesiredReadAlignment;
+} FS_FILTER_SECTION_SYNC_OUTPUT, *PFS_FILTER_SECTION_SYNC_OUTPUT;
+
 typedef struct _KQUEUE
 {
   DISPATCHER_HEADER Header;
@@ -30,6 +46,7 @@ typedef struct _KQUEUE
   LIST_ENTRY ThreadListHead;
 } KQUEUE, *PKQUEUE, *RESTRICTED_POINTER PRKQUEUE;
 
+
 NTSTATUS WINAPI ObQueryNameString(PVOID,POBJECT_NAME_INFORMATION,ULONG,PULONG);
 
 #endif
-- 
2.16.1

