From 179b74aeee8d4bf7a35b3a57d9395e4040175900 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Sun, 10 Jan 2016 16:01:55 +0800
Subject: widl: Write correct syskind by SLTG typelib generator.

---
 tools/widl/write_sltg.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/widl/write_sltg.c b/tools/widl/write_sltg.c
index 4b86bf9..61632ee 100644
--- a/tools/widl/write_sltg.c
+++ b/tools/widl/write_sltg.c
@@ -232,7 +232,7 @@ static void init_library(struct sltg_typelib *sltg)
     sltg->library.name = add_name(sltg, sltg->typelib->name);
     sltg->library.helpstring = NULL;
     sltg->library.helpcontext = 0;
-    sltg->library.syskind = SYS_WIN32;
+    sltg->library.syskind = typelib_kind;
     sltg->library.lcid = 0x0409;
     sltg->library.libflags = 0;
     sltg->library.version = 0;
-- 
2.6.4

