From 5e54bd9690e6b969ea5cccfd58b8f6ffb8800fef Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Sun, 10 Jan 2016 16:03:33 +0800
Subject: widl: Add support for VT_VOID and VT_VARIANT to SLTG typelib
 generator.

---
 tools/widl/write_sltg.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/tools/widl/write_sltg.c b/tools/widl/write_sltg.c
index 61632ee..08bb25e 100644
--- a/tools/widl/write_sltg.c
+++ b/tools/widl/write_sltg.c
@@ -549,6 +549,12 @@ static int get_element_size(type_t *type)
     case VT_LPWSTR:
         return pointer_size;
 
+    case VT_VOID:
+        return 0;
+
+    case VT_VARIANT:
+        return pointer_size == 8 ? 24 : 16;
+
     default:
         error("get_element_size: unrecognized vt %d\n", vt);
         break;
-- 
2.6.4

