From 3f9d49c152b1116b2349ad9cc9acf98da017f39a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 21 Jan 2016 02:51:56 +0100
Subject: widl: Add --oldtlb switch in usage message.

---
 tools/widl/widl.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/tools/widl/widl.c b/tools/widl/widl.c
index afeea39..c90b806 100644
--- a/tools/widl/widl.c
+++ b/tools/widl/widl.c
@@ -65,6 +65,7 @@ static const char usage[] =
 "   -m32, -m64         Set the kind of typelib to build (Win32 or Win64)\n"
 "   -N                 Do not preprocess input\n"
 "   --oldnames         Use old naming conventions\n"
+"   --oldtlb           Use old typelib (SLTG) format\n"
 "   -o, --output=NAME  Set the output file name\n"
 "   -Otype             Type of stubs to generate (-Os, -Oi, -Oif)\n"
 "   -p                 Generate proxy\n"
-- 
2.6.4

