From bd3806f58fda1d13e1359cc06854b81e22ce15e2 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 21 Jan 2016 02:51:56 +0100
Subject: widl: Add --oldtlb switch in usage message.

---
 tools/widl/widl.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/tools/widl/widl.c b/tools/widl/widl.c
index 9761fa4f..a8a76db6 100644
--- a/tools/widl/widl.c
+++ b/tools/widl/widl.c
@@ -61,6 +61,7 @@ static const char usage[] =
 "   -m32, -m64         Set the target architecture (Win32 or Win64)\n"
 "   -N                 Do not preprocess input\n"
 "   --oldnames         Use old naming conventions\n"
+"   --oldtlb           Use old typelib (SLTG) format\n"
 "   -o, --output=NAME  Set the output file name\n"
 "   -Otype             Type of stubs to generate (-Os, -Oi, -Oif)\n"
 "   -p                 Generate proxy\n"
-- 
2.19.1

