From a004652281da81bace202bed4fe451f1d60f83f0 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 21 Jan 2016 02:53:22 +0100
Subject: stdole32.tlb: Compile typelib with --oldtlb.

---
 dlls/stdole32.tlb/Makefile.in | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/stdole32.tlb/Makefile.in b/dlls/stdole32.tlb/Makefile.in
index 6422325..226235a 100644
--- a/dlls/stdole32.tlb/Makefile.in
+++ b/dlls/stdole32.tlb/Makefile.in
@@ -1,4 +1,5 @@
 MODULE    = stdole32.tlb
+EXTRAIDLFLAGS = --oldtlb
 
 RC_SRCS = rsrc.rc
 IDL_SRCS = std_ole_v1.idl
-- 
2.6.4

