From e6b772e0000dac0c50a8070b46413352e9c86174 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 6 Feb 2016 03:26:10 +0100
Subject: ole32: Set DebugInfo->Spare[0] for handle_wrapper lock.

---
 dlls/ole32/hglobalstream.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/ole32/hglobalstream.c b/dlls/ole32/hglobalstream.c
index 5eb6320..8b7e960 100644
--- a/dlls/ole32/hglobalstream.c
+++ b/dlls/ole32/hglobalstream.c
@@ -71,6 +71,7 @@ static void handle_release(struct handle_wrapper *handle)
             handle->hglobal = NULL;
         }
 
+        handle->lock.DebugInfo->Spare[0] = 0;
         DeleteCriticalSection(&handle->lock);
         HeapFree(GetProcessHeap(), 0, handle);
     }
@@ -130,6 +131,7 @@ static struct handle_wrapper *handle_create(HGLOBAL hglobal, BOOL delete_on_rele
         handle->size = GlobalSize(hglobal);
         handle->delete_on_release = delete_on_release;
         InitializeCriticalSection(&handle->lock);
+        handle->lock.DebugInfo->Spare[0] = (DWORD_PTR)(__FILE__ ": handle_wrapper.lock");
     }
     return handle;
 }
-- 
2.7.0

