From 67313bff113c1ea8df68d6117fa17bfa10a260d2 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Mon, 30 May 2016 17:14:13 +0800
Subject: user32: Add a check if the menu text is a valid string.

Fixes #40704.
---
 dlls/user32/menu.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/user32/menu.c b/dlls/user32/menu.c
index e42d8f7..e1b78ca 100644
--- a/dlls/user32/menu.c
+++ b/dlls/user32/menu.c
@@ -1658,7 +1658,7 @@ static void MENU_DrawMenuItem( HWND hwnd, HMENU hmenu, HWND hwndOwner, HDC hdc,
         SetViewportOrgEx( hdc, origorg.x, origorg.y, NULL);
     }
     /* process text if present */
-    if (lpitem->text)
+    if (!IS_INTRESOURCE(lpitem->text))
     {
 	int i;
 	HFONT hfontOld = 0;
-- 
2.8.0

