From 0051325fd29e84924ad64047a2476329f0e1abbf Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 16 Aug 2015 07:17:48 +0200
Subject: server: Initialize irp->thread immediately after creation of irp_call
 object.

---
 server/device.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/server/device.c b/server/device.c
index e6dc15e..812bce3 100644
--- a/server/device.c
+++ b/server/device.c
@@ -259,6 +259,7 @@ static struct irp_call *create_irp( struct device_file *file, const irp_params_t
     if ((irp = alloc_object( &irp_call_ops )))
     {
         irp->file     = (struct device_file *)grab_object( file );
+        irp->thread   = NULL;
         irp->async    = NULL;
         irp->params   = *params;
         irp->status   = STATUS_PENDING;
-- 
2.5.0

