From dfa61702f635d6ae7773e9e846800a77f179337c Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Thu, 2 Aug 2018 14:01:53 +0800
Subject: [PATCH 2/3] server: Add support for FILE_WRITE_THROUGH to
 create_file() request.

Signed-off-by: Dmitry Timoshkov <dmitry@baikal.ru>
---
 server/file.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/server/file.c b/server/file.c
index 446621a..9299530 100644
--- a/server/file.c
+++ b/server/file.c
@@ -224,6 +224,11 @@ static struct object *create_file( struct fd *root, const char *nameptr, data_si
     default:                set_error( STATUS_INVALID_PARAMETER ); goto done;
     }
 
+#ifdef O_DIRECT
+    if (options & FILE_WRITE_THROUGH)
+        flags |= O_DIRECT;
+#endif
+
     if (sd)
     {
         const SID *owner = sd_get_owner( sd );
-- 
1.9.1

