From 83896a0c6bf2756273793f5d998259e7545fe15e Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Thu, 2 Aug 2018 14:04:13 +0800
Subject: [PATCH 3/3] kernel32: Add support for FILE_FLAG_WRITE_THROUGH to
 CreateFile.

Signed-off-by: Dmitry Timoshkov <dmitry@baikal.ru>
---
 dlls/kernel32/file.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/kernel32/file.c b/dlls/kernel32/file.c
index 1e5b9fe..1b33cfb 100644
--- a/dlls/kernel32/file.c
+++ b/dlls/kernel32/file.c
@@ -1560,6 +1560,8 @@ HANDLE WINAPI CreateFileW( LPCWSTR filename, DWORD access, DWORD sharing,
         options |= FILE_SYNCHRONOUS_IO_NONALERT;
     if (attributes & FILE_FLAG_RANDOM_ACCESS)
         options |= FILE_RANDOM_ACCESS;
+    if (attributes & FILE_FLAG_WRITE_THROUGH)
+        options |= FILE_WRITE_THROUGH;
     attributes &= FILE_ATTRIBUTE_VALID_FLAGS;
 
     attr.Length = sizeof(attr);
-- 
1.9.1

