From 09c8ebbd33c38c40a769790e6ef028b035ad428a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Mon, 16 Oct 2017 16:39:28 +0200
Subject: server: Fix double-free in build_shared_mapping.

---
 server/mapping.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/server/mapping.c b/server/mapping.c
index bfd34f2b983..b868b3cf15a 100644
--- a/server/mapping.c
+++ b/server/mapping.c
@@ -585,6 +585,7 @@ static int build_shared_mapping( struct mapping *mapping, int fd,
         if (pwrite( shared_fd, buffer, file_size, write_pos ) != file_size) goto error;
     }
     free( buffer );
+    buffer = NULL;
 
     if (!(shared = alloc_object( &shared_map_ops ))) goto error;
     shared->fd = (struct fd *)grab_object( mapping->fd );
-- 
2.14.1

