From 3d297d901c57a96aa031b55983f548e77e7bc807 Mon Sep 17 00:00:00 2001
From: Mark Harmstone <mark@harmstone.com>
Date: Sat, 11 Apr 2015 23:18:16 +0100
Subject: [PATCH 13/19] uxtheme: Parse v4 size properties.

---
 dlls/uxtheme/msstyles.c | 4 ++++
 dlls/uxtheme/msstyles.h | 1 +
 2 files changed, 5 insertions(+)

diff --git a/dlls/uxtheme/msstyles.c b/dlls/uxtheme/msstyles.c
index b672adc..2903889 100644
--- a/dlls/uxtheme/msstyles.c
+++ b/dlls/uxtheme/msstyles.c
@@ -1054,6 +1054,10 @@ void MSSTYLES_ParseV4Properties(PTHEME_FILE tf, const WCHAR* variant_name)
                     memcpy(&prop->u.margins, &buf[i], min(sizeof(MARGINS), ph->datalen));
                 break;
 
+                case TMT_SIZE:
+                    memcpy(&prop->u.size, &buf[i], min(sizeof(SIZE), ph->datalen));
+                break;
+
                 default:
                     FIXME("Unprocessed non-resource property type %u\n", ph->type);
                 break;
diff --git a/dlls/uxtheme/msstyles.h b/dlls/uxtheme/msstyles.h
index 61a4e78..8a9ad79 100644
--- a/dlls/uxtheme/msstyles.h
+++ b/dlls/uxtheme/msstyles.h
@@ -40,6 +40,7 @@ typedef struct _THEME_PROPERTY {
         signed long integer;
         COLORREF colour;
         MARGINS margins;
+        SIZE size;
     } u;
 
     struct _THEME_PROPERTY *next;
-- 
2.0.5

