From fedb6517eef718f91ed4171c5a41901638bf6a46 Mon Sep 17 00:00:00 2001
From: Mark Harmstone <mark@harmstone.com>
Date: Sun, 12 Apr 2015 00:00:22 +0100
Subject: [PATCH 19/19] uxtheme: Implement GetThemeMetric for v4 themes.

---
 dlls/uxtheme/property.c | 33 +++++++++++++++++++++++++++++++--
 1 file changed, 31 insertions(+), 2 deletions(-)

diff --git a/dlls/uxtheme/property.c b/dlls/uxtheme/property.c
index aace3d5..3ab0b1b 100644
--- a/dlls/uxtheme/property.c
+++ b/dlls/uxtheme/property.c
@@ -263,8 +263,37 @@ HRESULT WINAPI GetThemeMetric(HTHEME hTheme, HDC hdc, int iPartId,
         return E_PROP_ID_UNSUPPORTED;
 
     if (tp->version == 4) {
-        FIXME("Not yet supported for v4 themes\n");
-        return E_NOTIMPL;
+        switch (tp->iPrimitiveType) {
+            case TMT_INT:
+            case TMT_BOOL:
+            case TMT_ENUM:
+                *piVal = tp->u.integer;
+                return S_OK;
+
+            case TMT_COLOR:
+                *piVal = tp->u.colour;
+                return S_OK;
+
+            case TMT_SIZE:
+                *piVal = tp->u.size.cx;
+                return S_OK;
+
+            case TMT_POSITION:
+                *piVal = tp->u.point.x;
+                return S_OK;
+
+            case TMT_MARGINS:
+                *piVal = tp->u.margins.cxLeftWidth;
+                return S_OK;
+
+            case TMT_INTLIST:
+                *piVal = tp->u.intlist->iValues[0];
+                return S_OK;
+
+            default:
+                FIXME("Unsupported primitive type %u\n", tp->iPrimitiveType);
+                break;
+        }
     } else {
         switch(tp->iPrimitiveType) {
             case TMT_POSITION: /* Only the X coord is retrieved */
-- 
2.0.5

