From 538e7170f2c8927428c7a4fb7a383ab85af7e318 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 27 Jan 2016 08:16:10 +0100
Subject: [PATCH 5/5] uxtheme: Fix some incorrect error codes.

---
 dlls/uxtheme/gtk.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/uxtheme/gtk.c b/dlls/uxtheme/gtk.c
index a5465356..74d55ada 100644
--- a/dlls/uxtheme/gtk.c
+++ b/dlls/uxtheme/gtk.c
@@ -457,7 +457,7 @@ HRESULT uxtheme_gtk_SetWindowTheme(HWND hwnd, LPCWSTR sub_app_name, LPCWSTR sub_
     FIXME("(%p, %s, %s)\n", hwnd, debugstr_w(sub_app_name),
           debugstr_w(sub_id_list));
 
-    return E_NOTIMPL;
+    return S_OK;
 }
 
 HRESULT uxtheme_gtk_GetThemeBool(HTHEME htheme, int part_id, int state_id,
@@ -924,7 +924,7 @@ BOOL uxtheme_gtk_IsThemePartDefined(HTHEME htheme, int part_id, int state_id)
 
     if (theme->vtable == NULL)
     {
-        SetLastError(ERROR_INVALID_HANDLE);
+        SetLastError(E_HANDLE);
         return FALSE;
     }
 
-- 
2.20.1

