From 1b00043f71c020e59d28bbc79f78a17bc3b4fd86 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 26 Jun 2019 10:38:23 +1000
Subject: [PATCH] uxtheme: Dont build with msvcrt

---
 dlls/uxtheme/Makefile.in | 2 --
 1 file changed, 2 deletions(-)

diff --git a/dlls/uxtheme/Makefile.in b/dlls/uxtheme/Makefile.in
index d0409300230..e09f891ee48 100644
--- a/dlls/uxtheme/Makefile.in
+++ b/dlls/uxtheme/Makefile.in
@@ -4,8 +4,6 @@ IMPORTS   = user32 gdi32 advapi32
 DELAYIMPORTS = msimg32
 EXTRAINCL = $(GTK3_CFLAGS)
 
-EXTRADLLFLAGS = -mno-cygwin
-
 C_SRCS = \
 	buffer.c \
 	draw.c \
-- 
2.20.1

