From c6bb50c511a23de00e0a5975c30e3d01415fc612 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Mon, 19 Aug 2019 09:35:43 +1000
Subject: [PATCH] Added wine/unicode

---
 dlls/uxtheme/msstyles.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/uxtheme/msstyles.c b/dlls/uxtheme/msstyles.c
index 5030e52a597..456256216a4 100644
--- a/dlls/uxtheme/msstyles.c
+++ b/dlls/uxtheme/msstyles.c
@@ -32,6 +32,7 @@
 
 #include "msstyles.h"
 
+#include "wine/unicode.h"
 #include "wine/exception.h"
 #include "wine/debug.h"
 #include "wine/heap.h"
-- 
2.17.1

