From 60dd76b04fa8aad2200f91660d0277a1c403dbab Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 27 Jan 2016 08:16:10 +0100
Subject: uxthemegtk: Fix some incorrect error codes.

---
 dlls/uxtheme-gtk/uxtheme.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/uxtheme-gtk/uxtheme.c b/dlls/uxtheme-gtk/uxtheme.c
index ab369f5..631c4da 100644
--- a/dlls/uxtheme-gtk/uxtheme.c
+++ b/dlls/uxtheme-gtk/uxtheme.c
@@ -519,7 +519,7 @@ HTHEME WINAPI OpenThemeDataEx(HWND hwnd, LPCWSTR classlist, DWORD flags)
     /* comctl32.dll likes to send NULL */
     if (classlist == NULL)
     {
-        SetLastError(ERROR_INVALID_PARAMETER);
+        SetLastError(E_POINTER);
         return NULL;
     }
 
@@ -578,7 +578,7 @@ HRESULT WINAPI SetWindowTheme(HWND hwnd, LPCWSTR sub_app_name,
     FIXME("(%p, %s, %s)\n", hwnd, debugstr_w(sub_app_name),
           debugstr_w(sub_id_list));
 
-    return E_NOTIMPL;
+    return S_OK;
 }
 
 HRESULT WINAPI GetThemeBool(HTHEME htheme, int part_id, int state_id,
@@ -1145,7 +1145,7 @@ BOOL WINAPI IsThemePartDefined(HTHEME htheme, int part_id, int state_id)
 
     if (theme == NULL || theme->vtable == NULL)
     {
-        SetLastError(ERROR_INVALID_HANDLE);
+        SetLastError(E_HANDLE);
         return FALSE;
     }
 
-- 
2.6.4

