From 3d3dd8491369119ee0dea8ec571b917f8613910a Mon Sep 17 00:00:00 2001
From: Martin Storsjo <martin@martin.st>
Date: Mon, 3 Aug 2015 22:26:03 +0300
Subject: vcruntime140: Hook up a function with a new name to the existing
 implementation

---
 dlls/vcruntime140/vcruntime140.spec | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/vcruntime140/vcruntime140.spec b/dlls/vcruntime140/vcruntime140.spec
index 9027955..adfb0f0 100644
--- a/dlls/vcruntime140/vcruntime140.spec
+++ b/dlls/vcruntime140/vcruntime140.spec
@@ -46,7 +46,7 @@
 @ stub __uncaught_exceptions
 @ stub __vcrt_GetModuleFileNameW
 @ stub __vcrt_GetModuleHandleW
-@ stub __vcrt_InitializeCriticalSectionEx
+@ cdecl -arch=i386,win64 __vcrt_InitializeCriticalSectionEx(ptr long long) MSVCR110__crtInitializeCriticalSectionEx
 @ stub __vcrt_LoadLibraryExW
 @ cdecl -arch=i386 -norelay _chkesp()
 @ cdecl -arch=i386 _except_handler2(ptr ptr ptr ptr)
-- 
2.5.0

