From b47092c85b03f1cbea95408f7472587c0ad93253 Mon Sep 17 00:00:00 2001
From: Martin Storsjo <martin@martin.st>
Date: Mon, 3 Aug 2015 22:26:06 +0300
Subject: make_specfiles: Use cdecl for stub redirects to ucrtbase, just like
 to msvcr*

---
 tools/make_specfiles | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/make_specfiles b/tools/make_specfiles
index 00e609c..54495cf 100755
--- a/tools/make_specfiles
+++ b/tools/make_specfiles
@@ -359,7 +359,7 @@ sub update_spec_file($)
         if ($parent{callconv} ne "stub" || $parent{args})
         {
             my $callconv = $parent{callconv} ne "stub" ? $parent{callconv} :
-                           $parent{spec} =~ /msvc/ ? "cdecl" : "stdcall";  # hack
+                           $parent{spec} =~ /(msvc|ucrtbase)/ ? "cdecl" : "stdcall";  # hack
             $_ = sprintf "$descr{ordinal} %s %s%s", $callconv, $flags, $func;
 
             if ($parent{target} =~ /$group_head\./)  # use the same forward as parent if possible
-- 
2.5.0

