From 945050ac121f63a56af7d0d740d84481c454ba31 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 11 Apr 2020 17:15:18 +1000
Subject: [PATCH] secur32: Add schan_imp_get_application_protocol for macos

---
 dlls/secur32/schannel_macosx.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/dlls/secur32/schannel_macosx.c b/dlls/secur32/schannel_macosx.c
index f86a561cf8d..378ab30ad1d 100644
--- a/dlls/secur32/schannel_macosx.c
+++ b/dlls/secur32/schannel_macosx.c
@@ -849,6 +849,24 @@ void schan_imp_set_session_target(schan_imp_session session, const char *target)
     SSLSetPeerDomainName( s->context, target, strlen(target) );
 }
 
+void schan_imp_set_application_protocols(schan_imp_session session, unsigned char *buffer, unsigned int buflen)
+{
+    struct mac_session *s = (struct mac_session*)session;
+
+    TRACE("(%p/%p, %p, %d)\n", s, s->context, buffer, buflen);
+}
+
+SECURITY_STATUS schan_imp_get_application_protocol(schan_imp_session session,
+                                                   SecPkgContext_ApplicationProtocol *protocol)
+{
+    struct mac_session *s = (struct mac_session*)session;
+
+    TRACE("(%p/%p, %p)\n", s, s->context, protocol);
+
+    memset(protocol, 0, sizeof(*protocol));
+    return SEC_E_OK;
+}
+
 SECURITY_STATUS schan_imp_handshake(schan_imp_session session)
 {
     struct mac_session *s = (struct mac_session*)session;
-- 
2.25.1

