From 35edeed55faa516480528c5c5e64ceedd45e29fb Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 6 Aug 2017 02:50:23 +0200
Subject: packager: Prefer native version.

---
 dlls/packager/packager_main.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/packager/packager_main.c b/dlls/packager/packager_main.c
index c1d194e3684..af1ee224ca6 100644
--- a/dlls/packager/packager_main.c
+++ b/dlls/packager/packager_main.c
@@ -675,6 +675,8 @@ BOOL WINAPI DllMain(HINSTANCE instance, DWORD reason, void *reserved)
     TRACE("(%p, %u, %p)\n", instance, reason, reserved);
 
     switch(reason){
+        case DLL_WINE_PREATTACH:
+            return FALSE;  /* prefer native version */
         case DLL_PROCESS_ATTACH:
             g_instance = instance;
             DisableThreadLibraryCalls(instance);
-- 
2.13.1

