From 5f05e5517f75081777b26d6f0e300a12121d094a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 17 Aug 2014 02:56:57 +0200
Subject: kernel32/tests: Add additional tests for PIPE_NOWAIT in overlapped
 mode.

---
 dlls/kernel32/tests/pipe.c | 293 ++++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 292 insertions(+), 1 deletion(-)

diff --git a/dlls/kernel32/tests/pipe.c b/dlls/kernel32/tests/pipe.c
index 3f753f8..c3f0b19 100644
--- a/dlls/kernel32/tests/pipe.c
+++ b/dlls/kernel32/tests/pipe.c
@@ -2748,7 +2748,7 @@ static void test_readfileex_pending(void)
 {
     HANDLE server, client, event;
     BOOL ret;
-    DWORD err, wait, num_bytes;
+    DWORD err, wait, num_bytes, lpmode;
     OVERLAPPED overlapped;
     char read_buf[1024];
     char write_buf[1024];
@@ -2903,6 +2903,159 @@ todo_wine
     CloseHandle(client);
     CloseHandle(server);
 
+    /* On Windows versions > 2000 it is not possible to add PIPE_NOWAIT to a byte-mode
+     * PIPE after creating. Create a new pipe for the following tests. */
+    server = CreateNamedPipeA(PIPENAME, FILE_FLAG_OVERLAPPED | PIPE_ACCESS_DUPLEX,
+        /* dwOpenMode */ PIPE_TYPE_BYTE | PIPE_NOWAIT,
+        /* nMaxInstances */ 1,
+        /* nOutBufSize */ 1024,
+        /* nInBufSize */ 1024,
+        /* nDefaultWait */ NMPWAIT_USE_DEFAULT_WAIT,
+        /* lpSecurityAttrib */ NULL);
+    ok(server != INVALID_HANDLE_VALUE, "cf failed\n");
+
+    client = CreateFileA(PIPENAME, GENERIC_READ|GENERIC_WRITE, 0, NULL,
+        OPEN_EXISTING, 0, NULL);
+    ok(client != INVALID_HANDLE_VALUE, "cf failed\n");
+
+    memset(&overlapped, 0, sizeof(overlapped));
+    overlapped.hEvent = event;
+
+    /* Initial check with empty pipe */
+    memset(read_buf, 0, sizeof(read_buf));
+    completion_called = 0;
+    ResetEvent(event);
+    ret = ReadFileEx(server, read_buf, 4, &overlapped, completion_routine);
+    ok(ret == FALSE, "ReadFileEx succeded\n");
+    ok(completion_called == 0, "completion routine called before ReadFileEx returned\n");
+    ok(GetLastError() == ERROR_NO_DATA, "expected ERROR_NO_DATA, got %d\n", GetLastError());
+    wait = WaitForSingleObjectEx(event, 0, TRUE);
+    ok(wait == WAIT_TIMEOUT, "WaitForSingleObjectEx returned %x\n", wait);
+    ok(completion_called == 0, "completion routine called before writing to file\n");
+
+    /* Call ReadFileEx after writing content to the pipe */
+    num_bytes = 0xdeadbeef;
+    ret = WriteFile(client, long_test_string, strlen(long_test_string), &num_bytes, NULL);
+    ok(ret, "WriteFile failed, err=%i\n", GetLastError());
+    ok(num_bytes == strlen(long_test_string), "bytes %u\n", num_bytes);
+
+    memset(read_buf, 0, sizeof(read_buf));
+    completion_called = 0;
+    ResetEvent(event);
+    ret = ReadFileEx(server, read_buf, 4, &overlapped, completion_routine);
+    ok(ret == TRUE, "ReadFileEx failed, err=%i\n", GetLastError());
+    ok(completion_called == 0, "completion routine called before ReadFileEx returned\n");
+    wait = WaitForSingleObjectEx(event, 0, TRUE);
+    ok(wait == WAIT_IO_COMPLETION || wait == WAIT_OBJECT_0, "WaitForSingleObjectEx returned %x\n", wait);
+    ok(completion_called == 1, "completion not called after writing pipe\n");
+    ok(completion_errorcode == 0, "completion called with error %x\n", completion_errorcode);
+    ok(completion_num_bytes == 4, "ReadFileEx returned only %d bytes\n", completion_num_bytes);
+    ok(completion_lpoverlapped == &overlapped, "completion called with wrong overlapped pointer\n");
+
+    ret = ReadFile(server, read_buf + 4, sizeof(read_buf) - 4, &num_bytes, NULL);
+    ok(ret == TRUE, "ReadFile succeeded\n");
+    ok(num_bytes == strlen(long_test_string)-4, "ReadFile returned only %d bytes\n", num_bytes);
+    ok(!memcmp(long_test_string, read_buf, strlen(long_test_string)), "ReadFile read wrong bytes\n");
+
+    /* Same again, but read as a single part */
+    num_bytes = 0xdeadbeef;
+    ret = WriteFile(client, long_test_string, strlen(long_test_string), &num_bytes, NULL);
+    ok(ret, "WriteFile failed, err=%i\n", GetLastError());
+    ok(num_bytes == strlen(long_test_string), "bytes %u\n", num_bytes);
+
+    memset(read_buf, 0, sizeof(read_buf));
+    completion_called = 0;
+    ResetEvent(event);
+    ret = ReadFileEx(server, read_buf, sizeof(read_buf), &overlapped, completion_routine);
+    ok(ret == TRUE, "ReadFileEx failed, err=%i\n", GetLastError());
+    ok(completion_called == 0, "completion routine called before ReadFileEx returned\n");
+    wait = WaitForSingleObjectEx(event, 0, TRUE);
+    ok(wait == WAIT_IO_COMPLETION || wait == WAIT_OBJECT_0, "WaitForSingleObjectEx returned %x\n", wait);
+    ok(completion_called == 1, "completion not called after writing pipe\n");
+    ok(completion_errorcode == 0, "completion called with error %x\n", completion_errorcode);
+    ok(completion_num_bytes == strlen(long_test_string), "ReadFileEx returned only %d bytes\n", completion_num_bytes);
+    ok(completion_lpoverlapped == &overlapped, "completion called with wrong overlapped pointer\n");
+    ok(!memcmp(long_test_string, read_buf, strlen(long_test_string)), "ReadFile read wrong bytes\n");
+
+    /* Check content of overlapped structure */
+    memset(read_buf, 0, sizeof(read_buf));
+    S(U(overlapped)).Offset = 0;
+    S(U(overlapped)).OffsetHigh = 0;
+    overlapped.Internal = -1;
+    overlapped.InternalHigh = -1;
+    overlapped.hEvent = event;
+    num_bytes = 0xdeadbeef;
+    SetLastError(0xdeadbeef);
+    ret = ReadFile(server, read_buf, 4, &num_bytes, &overlapped);
+    ok(ret == FALSE, "ReadFile succeeded\n");
+    ok(GetLastError() == ERROR_NO_DATA, "expected ERROR_NO_DATA, got %d\n", GetLastError());
+    ok(num_bytes == 0, "bytes %u\n", num_bytes);
+    ok((NTSTATUS)overlapped.Internal == STATUS_PENDING, "expected STATUS_PENDING, got %#lx\n", overlapped.Internal);
+    todo_wine
+    ok(overlapped.InternalHigh == -1, "expected -1, got %lu\n", overlapped.InternalHigh);
+    wait = WaitForSingleObject(event, 100);
+    ok(wait == WAIT_TIMEOUT, "WaitForSingleObjectEx returned %x\n", wait);
+    ok((NTSTATUS)overlapped.Internal == STATUS_PENDING, "expected STATUS_PENDING, got %#lx\n", overlapped.Internal);
+    todo_wine
+    ok(overlapped.InternalHigh == -1, "expected -1, got %lu\n", overlapped.InternalHigh);
+
+    /* Call ReadFile after writing to the pipe */
+    num_bytes = 0xdeadbeef;
+    ret = WriteFile(client, long_test_string, strlen(long_test_string), &num_bytes, NULL);
+    ok(ret, "WriteFile failed, err=%i\n", GetLastError());
+    ok(num_bytes == strlen(long_test_string), "bytes %u\n", num_bytes);
+
+    memset(read_buf, 0, sizeof(read_buf));
+    S(U(overlapped)).Offset = 0;
+    S(U(overlapped)).OffsetHigh = 0;
+    overlapped.Internal = -1;
+    overlapped.InternalHigh = -1;
+    overlapped.hEvent = event;
+    num_bytes = 0xdeadbeef;
+    SetLastError(0xdeadbeef);
+    ret = ReadFile(server, read_buf, 4, &num_bytes, &overlapped);
+    ok(ret == TRUE, "ReadFile failed, err=%i\n", GetLastError());
+    ok(num_bytes == 4, "bytes %u\n", num_bytes);
+    ok((NTSTATUS)overlapped.Internal == STATUS_SUCCESS, "expected STATUS_SUCCESS, got %#lx\n", overlapped.Internal);
+    ok(overlapped.InternalHigh == 4, "expected 4, got %lu\n", overlapped.InternalHigh);
+    wait = WaitForSingleObject(event, 100);
+    ok(wait == WAIT_IO_COMPLETION || wait == WAIT_OBJECT_0, "WaitForSingleObject returned %x\n", wait);
+    ok((NTSTATUS)overlapped.Internal == STATUS_SUCCESS, "expected STATUS_SUCCESS, got %#lx\n", overlapped.Internal);
+    ok(overlapped.InternalHigh == 4, "expected 4, got %lu\n", overlapped.InternalHigh);
+
+    ret = ReadFile(server, read_buf + 4, sizeof(read_buf) - 4, &num_bytes, NULL);
+    ok(ret == TRUE, "ReadFile failed\n");
+    ok(num_bytes == strlen(long_test_string)-4, "ReadFile returned only %d bytes\n", num_bytes);
+    ok(!memcmp(long_test_string, read_buf, strlen(long_test_string)), "ReadFile read wrong bytes\n");
+
+    /* Same again, but read as a single part */
+    num_bytes = 0xdeadbeef;
+    ret = WriteFile(client, long_test_string, strlen(long_test_string), &num_bytes, NULL);
+    ok(ret, "WriteFile failed, err=%i\n", GetLastError());
+    ok(num_bytes == strlen(long_test_string), "bytes %u\n", num_bytes);
+
+    memset(read_buf, 0, sizeof(read_buf));
+    S(U(overlapped)).Offset = 0;
+    S(U(overlapped)).OffsetHigh = 0;
+    overlapped.Internal = -1;
+    overlapped.InternalHigh = -1;
+    overlapped.hEvent = event;
+    num_bytes = 0xdeadbeef;
+    SetLastError(0xdeadbeef);
+    ret = ReadFile(server, read_buf, sizeof(read_buf), &num_bytes, &overlapped);
+    ok(ret == TRUE, "ReadFile failed, err=%i\n", GetLastError());
+    ok(num_bytes == strlen(long_test_string), "bytes %u\n", num_bytes);
+    ok((NTSTATUS)overlapped.Internal == 0, "expected 0, got %#lx\n", overlapped.Internal);
+    ok(overlapped.InternalHigh == strlen(long_test_string), "expected %u, got %lu\n", (DWORD)strlen(long_test_string), overlapped.InternalHigh);
+    wait = WaitForSingleObject(event, 100);
+    ok(wait == WAIT_IO_COMPLETION || wait == WAIT_OBJECT_0, "WaitForSingleObject returned %x\n", wait);
+    ok((NTSTATUS)overlapped.Internal == 0, "expected 0, got %#lx\n", overlapped.Internal);
+    ok(overlapped.InternalHigh == strlen(long_test_string), "expected %u, got %lu\n", (DWORD)strlen(long_test_string), overlapped.InternalHigh);
+    ok(!memcmp(long_test_string, read_buf, strlen(long_test_string)), "ReadFile read wrong bytes\n");
+
+    CloseHandle(client);
+    CloseHandle(server);
+
     server = CreateNamedPipeA(PIPENAME, FILE_FLAG_OVERLAPPED | PIPE_ACCESS_DUPLEX,
         /* dwOpenMode */ PIPE_TYPE_MESSAGE | PIPE_READMODE_MESSAGE | PIPE_WAIT,
         /* nMaxInstances */ 1,
@@ -3041,6 +3194,144 @@ todo_wine
     ok(num_bytes == strlen(long_test_string)-4, "ReadFile returned only %d bytes\n", num_bytes);
     ok(!memcmp(long_test_string, read_buf, strlen(long_test_string)), "ReadFile read wrong bytes\n");
 
+    /* Tests for PIPE_NOWAIT in message mode */
+    lpmode = PIPE_READMODE_MESSAGE | PIPE_NOWAIT;
+    ok(SetNamedPipeHandleState(server, &lpmode, NULL, NULL), "Change mode\n");
+
+    /* Initial check with empty pipe */
+    memset(read_buf, 0, sizeof(read_buf));
+    completion_called = 0;
+    ResetEvent(event);
+    ret = ReadFileEx(server, read_buf, 4, &overlapped, completion_routine);
+    ok(ret == FALSE, "ReadFileEx succeded\n");
+    ok(completion_called == 0, "completion routine called before ReadFileEx returned\n");
+    ok(GetLastError() == ERROR_NO_DATA, "expected ERROR_NO_DATA, got %d\n", GetLastError());
+    wait = WaitForSingleObjectEx(event, 0, TRUE);
+    ok(wait == WAIT_TIMEOUT, "WaitForSingleObjectEx returned %x\n", wait);
+    ok(completion_called == 0, "completion routine called before writing to file\n");
+
+    /* Call ReadFileEx after writing content to the pipe */
+    num_bytes = 0xdeadbeef;
+    ret = WriteFile(client, long_test_string, strlen(long_test_string), &num_bytes, NULL);
+    ok(ret, "WriteFile failed, err=%i\n", GetLastError());
+    ok(num_bytes == strlen(long_test_string), "bytes %u\n", num_bytes);
+
+    memset(read_buf, 0, sizeof(read_buf));
+    completion_called = 0;
+    ResetEvent(event);
+    ret = ReadFileEx(server, read_buf, 4, &overlapped, completion_routine);
+    ok(ret == TRUE, "ReadFileEx failed, err=%i\n", GetLastError());
+    ok(completion_called == 0, "completion routine called before ReadFileEx returned\n");
+    wait = WaitForSingleObjectEx(event, 0, TRUE);
+    ok(wait == WAIT_IO_COMPLETION || wait == WAIT_OBJECT_0, "WaitForSingleObjectEx returned %x\n", wait);
+    ok(completion_called == 1, "completion not called after writing pipe\n");
+    ok(completion_errorcode == 0, "completion called with error %x\n", completion_errorcode);
+    ok(completion_num_bytes == 4, "ReadFileEx returned only %d bytes\n", completion_num_bytes);
+    ok(completion_lpoverlapped == &overlapped, "completion called with wrong overlapped pointer\n");
+
+    ret = ReadFile(server, read_buf + 4, sizeof(read_buf) - 4, &num_bytes, NULL);
+    ok(ret == TRUE, "ReadFile succeeded\n");
+    ok(num_bytes == strlen(long_test_string)-4, "ReadFile returned only %d bytes\n", num_bytes);
+    ok(!memcmp(long_test_string, read_buf, strlen(long_test_string)), "ReadFile read wrong bytes\n");
+
+    /* Same again, but read as a single part */
+    num_bytes = 0xdeadbeef;
+    ret = WriteFile(client, long_test_string, strlen(long_test_string), &num_bytes, NULL);
+    ok(ret, "WriteFile failed, err=%i\n", GetLastError());
+    ok(num_bytes == strlen(long_test_string), "bytes %u\n", num_bytes);
+
+    memset(read_buf, 0, sizeof(read_buf));
+    completion_called = 0;
+    ResetEvent(event);
+    ret = ReadFileEx(server, read_buf, sizeof(read_buf), &overlapped, completion_routine);
+    ok(ret == TRUE, "ReadFileEx failed, err=%i\n", GetLastError());
+    ok(completion_called == 0, "completion routine called before ReadFileEx returned\n");
+    wait = WaitForSingleObjectEx(event, 0, TRUE);
+    ok(wait == WAIT_IO_COMPLETION || wait == WAIT_OBJECT_0, "WaitForSingleObjectEx returned %x\n", wait);
+    ok(completion_called == 1, "completion not called after writing pipe\n");
+    ok(completion_errorcode == 0, "completion called with error %x\n", completion_errorcode);
+    ok(completion_num_bytes == strlen(long_test_string), "ReadFileEx returned only %d bytes\n", completion_num_bytes);
+    ok(completion_lpoverlapped == &overlapped, "completion called with wrong overlapped pointer\n");
+    ok(!memcmp(long_test_string, read_buf, strlen(long_test_string)), "ReadFile read wrong bytes\n");
+
+    /* Check content of overlapped structure */
+    memset(read_buf, 0, sizeof(read_buf));
+    S(U(overlapped)).Offset = 0;
+    S(U(overlapped)).OffsetHigh = 0;
+    overlapped.Internal = -1;
+    overlapped.InternalHigh = -1;
+    overlapped.hEvent = event;
+    num_bytes = 0xdeadbeef;
+    SetLastError(0xdeadbeef);
+    ret = ReadFile(server, read_buf, 4, &num_bytes, &overlapped);
+    ok(ret == FALSE, "ReadFile succeeded\n");
+    ok(GetLastError() == ERROR_NO_DATA, "expected ERROR_NO_DATA, got %d\n", GetLastError());
+    ok(num_bytes == 0, "bytes %u\n", num_bytes);
+    ok((NTSTATUS)overlapped.Internal == STATUS_PENDING, "expected STATUS_PENDING, got %#lx\n", overlapped.Internal);
+    todo_wine
+    ok(overlapped.InternalHigh == -1, "expected -1, got %lu\n", overlapped.InternalHigh);
+    wait = WaitForSingleObject(event, 100);
+    ok(wait == WAIT_TIMEOUT, "WaitForSingleObjectEx returned %x\n", wait);
+    ok((NTSTATUS)overlapped.Internal == STATUS_PENDING, "expected STATUS_PENDING, got %#lx\n", overlapped.Internal);
+    todo_wine
+    ok(overlapped.InternalHigh == -1, "expected -1, got %lu\n", overlapped.InternalHigh);
+
+    /* Call ReadFile after writing to the pipe */
+    num_bytes = 0xdeadbeef;
+    ret = WriteFile(client, long_test_string, strlen(long_test_string), &num_bytes, NULL);
+    ok(ret, "WriteFile failed, err=%i\n", GetLastError());
+    ok(num_bytes == strlen(long_test_string), "bytes %u\n", num_bytes);
+
+    memset(read_buf, 0, sizeof(read_buf));
+    S(U(overlapped)).Offset = 0;
+    S(U(overlapped)).OffsetHigh = 0;
+    overlapped.Internal = -1;
+    overlapped.InternalHigh = -1;
+    overlapped.hEvent = event;
+    num_bytes = 0xdeadbeef;
+    SetLastError(0xdeadbeef);
+    ret = ReadFile(server, read_buf, 4, &num_bytes, &overlapped);
+    ok(ret == FALSE, "ReadFile succeeded\n");
+    ok(GetLastError() == ERROR_MORE_DATA, "expected ERROR_MORE_DATA, got %d\n", GetLastError());
+    todo_wine
+    ok(num_bytes == 0, "bytes %u\n", num_bytes);
+    ok((NTSTATUS)overlapped.Internal == STATUS_BUFFER_OVERFLOW, "expected STATUS_BUFFER_OVERFLOW, got %#lx\n", overlapped.Internal);
+    ok(overlapped.InternalHigh == 4, "expected 4, got %lu\n", overlapped.InternalHigh);
+    wait = WaitForSingleObject(event, 100);
+    ok(wait == WAIT_IO_COMPLETION || wait == WAIT_OBJECT_0, "WaitForSingleObject returned %x\n", wait);
+    ok((NTSTATUS)overlapped.Internal == STATUS_BUFFER_OVERFLOW, "expected STATUS_BUFFER_OVERFLOW, got %#lx\n", overlapped.Internal);
+    ok(overlapped.InternalHigh == 4, "expected 4, got %lu\n", overlapped.InternalHigh);
+
+    ret = ReadFile(server, read_buf + 4, sizeof(read_buf) - 4, &num_bytes, NULL);
+    ok(ret == TRUE, "ReadFile failed\n");
+    ok(num_bytes == strlen(long_test_string)-4, "ReadFile returned only %d bytes\n", num_bytes);
+    ok(!memcmp(long_test_string, read_buf, strlen(long_test_string)), "ReadFile read wrong bytes\n");
+
+    /* Same again, but read as a single part */
+    num_bytes = 0xdeadbeef;
+    ret = WriteFile(client, long_test_string, strlen(long_test_string), &num_bytes, NULL);
+    ok(ret, "WriteFile failed, err=%i\n", GetLastError());
+    ok(num_bytes == strlen(long_test_string), "bytes %u\n", num_bytes);
+
+    memset(read_buf, 0, sizeof(read_buf));
+    S(U(overlapped)).Offset = 0;
+    S(U(overlapped)).OffsetHigh = 0;
+    overlapped.Internal = -1;
+    overlapped.InternalHigh = -1;
+    overlapped.hEvent = event;
+    num_bytes = 0xdeadbeef;
+    SetLastError(0xdeadbeef);
+    ret = ReadFile(server, read_buf, sizeof(read_buf), &num_bytes, &overlapped);
+    ok(ret == TRUE, "ReadFile failed, err=%i\n", GetLastError());
+    ok(num_bytes == strlen(long_test_string), "bytes %u\n", num_bytes);
+    ok((NTSTATUS)overlapped.Internal == 0, "expected 0, got %#lx\n", overlapped.Internal);
+    ok(overlapped.InternalHigh == strlen(long_test_string), "expected %u, got %lu\n", (DWORD)strlen(long_test_string), overlapped.InternalHigh);
+    wait = WaitForSingleObject(event, 100);
+    ok(wait == WAIT_IO_COMPLETION || wait == WAIT_OBJECT_0, "WaitForSingleObject returned %x\n", wait);
+    ok((NTSTATUS)overlapped.Internal == 0, "expected 0, got %#lx\n", overlapped.Internal);
+    ok(overlapped.InternalHigh == strlen(long_test_string), "expected %u, got %lu\n", (DWORD)strlen(long_test_string), overlapped.InternalHigh);
+    ok(!memcmp(long_test_string, read_buf, strlen(long_test_string)), "ReadFile read wrong bytes\n");
+
     CloseHandle(client);
     CloseHandle(server);
     CloseHandle(event);
-- 
2.3.0

