From 03e1e23b99c29bb63cdb5808ff07374d38003efa Mon Sep 17 00:00:00 2001
From: Sergey Isakov <sergey.slice@gmail.com>
Date: Sat, 16 Apr 2016 19:14:43 +0200
Subject: wined3d: Fix typo in TRACE message.

---
 dlls/wined3d/directx.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/wined3d/directx.c b/dlls/wined3d/directx.c
index 8358930..dfa1cdd 100644
--- a/dlls/wined3d/directx.c
+++ b/dlls/wined3d/directx.c
@@ -1444,7 +1444,7 @@ static const struct gpu_description *query_gpu_description(const struct wined3d_
     if (wined3d_settings.pci_device_id != PCI_DEVICE_NONE)
     {
         device = wined3d_settings.pci_device_id;
-        TRACE("Overriding device PCI ID with 0x%04x.\n", vendor);
+        TRACE("Overriding device PCI ID with 0x%04x.\n", device);
     }
 
     if (wined3d_settings.emulated_textureram)
-- 
2.7.1

