From d4fcd702f9d7eda8209511b5f6522be544b794c4 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 30 Nov 2014 18:18:23 +0100
Subject: ntdll: Fix condition handling in RtlVerifyVersionInfo.

---
 dlls/ntdll/version.c | 15 +++++----------
 1 file changed, 5 insertions(+), 10 deletions(-)

diff --git a/dlls/ntdll/version.c b/dlls/ntdll/version.c
index a6340eb..fa40613 100644
--- a/dlls/ntdll/version.c
+++ b/dlls/ntdll/version.c
@@ -732,38 +732,33 @@ NTSTATUS WINAPI RtlVerifyVersionInfo( const RTL_OSVERSIONINFOEXW *info,
 
     if(dwTypeMask & (VER_MAJORVERSION|VER_MINORVERSION|VER_SERVICEPACKMAJOR|VER_SERVICEPACKMINOR))
     {
-        unsigned char condition = 0;
+        unsigned char condition;
         BOOLEAN do_next_check = TRUE;
 
         if(dwTypeMask & VER_MAJORVERSION)
-            condition = dwlConditionMask >> 1*3 & 0x07;
-        else if(dwTypeMask & VER_MINORVERSION)
-            condition = dwlConditionMask >> 0*3 & 0x07;
-        else if(dwTypeMask & VER_SERVICEPACKMAJOR)
-            condition = dwlConditionMask >> 5*3 & 0x07;
-        else if(dwTypeMask & VER_SERVICEPACKMINOR)
-            condition = dwlConditionMask >> 4*3 & 0x07;
-
-        if(dwTypeMask & VER_MAJORVERSION)
         {
+            condition = dwlConditionMask >> 1*3 & 0x07;
             status = version_compare_values(ver.dwMajorVersion, info->dwMajorVersion, condition);
             do_next_check = (ver.dwMajorVersion == info->dwMajorVersion) &&
                 ((condition != VER_EQUAL) || (status == STATUS_SUCCESS));
         }
         if((dwTypeMask & VER_MINORVERSION) && do_next_check)
         {
+            condition = dwlConditionMask >> 0*3 & 0x07;
             status = version_compare_values(ver.dwMinorVersion, info->dwMinorVersion, condition);
             do_next_check = (ver.dwMinorVersion == info->dwMinorVersion) &&
                 ((condition != VER_EQUAL) || (status == STATUS_SUCCESS));
         }
         if((dwTypeMask & VER_SERVICEPACKMAJOR) && do_next_check)
         {
+            condition = dwlConditionMask >> 5*3 & 0x07;
             status = version_compare_values(ver.wServicePackMajor, info->wServicePackMajor, condition);
             do_next_check = (ver.wServicePackMajor == info->wServicePackMajor) &&
                 ((condition != VER_EQUAL) || (status == STATUS_SUCCESS));
         }
         if((dwTypeMask & VER_SERVICEPACKMINOR) && do_next_check)
         {
+            condition = dwlConditionMask >> 4*3 & 0x07;
             status = version_compare_values(ver.wServicePackMinor, info->wServicePackMinor, condition);
         }
 
-- 
2.1.3

