From e86c56743909d0e2f77dc58f2724a6b517671f28 Mon Sep 17 00:00:00 2001
From: Amine Khaldi <amine.khaldi@reactos.org>
Date: Sat, 18 Oct 2014 19:57:48 +0100
Subject: strmbase: Remove unused TransformFilter_Vtbl (Clang)

Regards,
Amine.

From 53611893c65e0470cf8d81b7900a02f9640c6dcc Mon Sep 17 00:00:00 2001
From: Amine Khaldi <amine.khaldi@reactos.org>
Date: Sat, 18 Oct 2014 19:53:56 +0100
Subject: [PATCH] strmbase: Remove unused TransformFilter_Vtbl (Clang).
---
 dlls/strmbase/transform.c | 20 --------------------
 1 file changed, 20 deletions(-)

diff --git a/dlls/strmbase/transform.c b/dlls/strmbase/transform.c
index 0fc7515..4d872be 100644
--- a/dlls/strmbase/transform.c
+++ b/dlls/strmbase/transform.c
@@ -42,7 +42,6 @@ WINE_DEFAULT_DEBUG_CHANNEL(strmbase);
 static const WCHAR wcsInputPinName[] = {'i','n','p','u','t',' ','p','i','n',0};
 static const WCHAR wcsOutputPinName[] = {'o','u','t','p','u','t',' ','p','i','n',0};
 
-static const IBaseFilterVtbl TransformFilter_Vtbl;
 static const IPinVtbl TransformFilter_InputPin_Vtbl;
 static const IPinVtbl TransformFilter_OutputPin_Vtbl;
 static const IQualityControlVtbl TransformFilter_QualityControl_Vtbl;
@@ -439,25 +438,6 @@ HRESULT WINAPI TransformFilterImpl_FindPin(IBaseFilter * iface, LPCWSTR Id, IPin
     return E_NOTIMPL;
 }
 
-static const IBaseFilterVtbl TransformFilter_Vtbl =
-{
-    TransformFilterImpl_QueryInterface,
-    BaseFilterImpl_AddRef,
-    TransformFilterImpl_Release,
-    BaseFilterImpl_GetClassID,
-    TransformFilterImpl_Stop,
-    TransformFilterImpl_Pause,
-    TransformFilterImpl_Run,
-    BaseFilterImpl_GetState,
-    BaseFilterImpl_SetSyncSource,
-    BaseFilterImpl_GetSyncSource,
-    BaseFilterImpl_EnumPins,
-    TransformFilterImpl_FindPin,
-    BaseFilterImpl_QueryFilterInfo,
-    BaseFilterImpl_JoinFilterGraph,
-    BaseFilterImpl_QueryVendorInfo
-};
-
 static HRESULT WINAPI TransformFilter_InputPin_EndOfStream(IPin * iface)
 {
     BaseInputPin* This = impl_BaseInputPin_from_IPin(iface);
-- 
2.1.2

