From d9c7dc0ec91492b7e57682ea3ddb10feaffcc18b Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 21 Mar 2015 04:22:10 +0100
Subject: user32: Fix padding in user_thread_info structure.

---
 dlls/user32/user_private.h | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/dlls/user32/user_private.h b/dlls/user32/user_private.h
index adf3f7d..c389414 100644
--- a/dlls/user32/user_private.h
+++ b/dlls/user32/user_private.h
@@ -169,6 +169,7 @@ struct wm_char_mapping_data
 
 /* this is the structure stored in TEB->Win32ClientInfo */
 /* no attempt is made to keep the layout compatible with the Windows one */
+#include <pshpack4.h>
 struct user_thread_info
 {
     HANDLE                        server_queue;           /* Handle to server-side queue */
@@ -190,8 +191,9 @@ struct user_thread_info
     HWND                          msg_window;             /* HWND_MESSAGE parent window */
     RAWINPUT                     *rawinput;
 
-    ULONG                         pad[6];                 /* Available for more data */
+    ULONG                         pad[5];                 /* Available for more data */
 };
+#include <poppack.h>
 
 struct hook_extra_info
 {
-- 
2.3.2

