From 497be9c7987e7cd08a70cd33864729d06b8d1882 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 12 Jul 2017 05:26:04 +0200
Subject: webservices: Remove dead code in read_attribute_value_bin.

---
 dlls/webservices/reader.c | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/dlls/webservices/reader.c b/dlls/webservices/reader.c
index 5bc862e5bf1..b94c35947c4 100644
--- a/dlls/webservices/reader.c
+++ b/dlls/webservices/reader.c
@@ -1582,11 +1582,6 @@ static HRESULT read_attribute_value_bin( struct reader *reader, WS_XML_ATTRIBUTE
         return WS_E_NOT_SUPPORTED;
     }
 
-    if (type >= RECORD_INT8_TEXT && type <= RECORD_INT64_TEXT)
-    {
-        attr->value = &text_int64->text;
-        return S_OK;
-    }
     if (type >= RECORD_BYTES8_TEXT && type <= RECORD_BYTES32_TEXT)
     {
         attr->value = &text_base64->text;
-- 
2.13.1

