From 93eaea5931b6aa60b1d4edf9c39cfc7274c47e39 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 29 Nov 2014 09:55:39 +0100
Subject: ws2_32/tests: Fix several copy and paste errors.

---
 dlls/ws2_32/tests/sock.c | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/dlls/ws2_32/tests/sock.c b/dlls/ws2_32/tests/sock.c
index 106588d..16f98c0 100644
--- a/dlls/ws2_32/tests/sock.c
+++ b/dlls/ws2_32/tests/sock.c
@@ -3358,14 +3358,14 @@ static void test_select(void)
     ret = select(maxfd+1, &readfds, &readfds, NULL, &select_timeout);
     ok(ret == 2, "select returned %d\n", ret);
     ok(FD_ISSET(fdWrite, &readfds), "fdWrite socket is not in the set\n");
-    ok(FD_ISSET(fdRead, &readfds), "fdWrite socket is not in the set\n");
+    ok(FD_ISSET(fdRead, &readfds), "fdRead socket is not in the set\n");
 
     ok(send(fdWrite, "test", 4, 0) == 4, "failed to send data\n");
     FD_ZERO(&readfds);
     FD_SET(fdRead, &readfds);
     ret = select(fdRead+1, &readfds, NULL, NULL, &select_timeout);
     ok(ret == 1, "select returned %d\n", ret);
-    ok(FD_ISSET(fdRead, &readfds), "fdWrite socket is not in the set\n");
+    ok(FD_ISSET(fdRead, &readfds), "fdRead socket is not in the set\n");
 
     FD_ZERO(&readfds);
     FD_SET(fdWrite, &readfds);
@@ -3373,7 +3373,7 @@ static void test_select(void)
     ret = select(maxfd+1, &readfds, &readfds, NULL, &select_timeout);
     ok(ret == 2, "select returned %d\n", ret);
     ok(FD_ISSET(fdWrite, &readfds), "fdWrite socket is not in the set\n");
-    ok(FD_ISSET(fdRead, &readfds), "fdWrite socket is not in the set\n");
+    ok(FD_ISSET(fdRead, &readfds), "fdRead socket is not in the set\n");
 
     while(1) {
         FD_ZERO(&writefds);
@@ -3388,7 +3388,7 @@ static void test_select(void)
     ret = select(maxfd+1, &readfds, &readfds, NULL, &select_timeout);
     ok(ret == 1, "select returned %d\n", ret);
     ok(!FD_ISSET(fdWrite, &readfds), "fdWrite socket is in the set\n");
-    ok(FD_ISSET(fdRead, &readfds), "fdWrite socket is not in the set\n");
+    ok(FD_ISSET(fdRead, &readfds), "fdRead socket is not in the set\n");
 
     ok(send(fdRead, "test", 4, 0) == 4, "failed to send data\n");
     Sleep(100);
@@ -3398,7 +3398,7 @@ static void test_select(void)
     ret = select(maxfd+1, &readfds, &readfds, NULL, &select_timeout);
     ok(ret == 2, "select returned %d\n", ret);
     ok(FD_ISSET(fdWrite, &readfds), "fdWrite socket is not in the set\n");
-    ok(FD_ISSET(fdRead, &readfds), "fdWrite socket is not in the set\n");
+    ok(FD_ISSET(fdRead, &readfds), "fdRead socket is not in the set\n");
 
     closesocket(fdRead);
     closesocket(fdWrite);
-- 
2.1.3

