From 7847b8fd4f68d229e5fc478e755fb212fdff3431 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 18 Aug 2016 18:37:29 +0200
Subject: setupapi: Add spec file stub entry for SetupDiInstallDeviceInterfaces
 and SetupDiRegisterCoDeviceInstallers.

---
 dlls/setupapi/setupapi.spec | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/setupapi/setupapi.spec b/dlls/setupapi/setupapi.spec
index 0eb8359..4a4d193 100644
--- a/dlls/setupapi/setupapi.spec
+++ b/dlls/setupapi/setupapi.spec
@@ -367,6 +367,7 @@
 @ stub SetupDiInstallClassExW
 @ stdcall SetupDiInstallClassW(long wstr long ptr)
 @ stub SetupDiInstallDevice
+@ stub SetupDiInstallDeviceInterfaces
 @ stub SetupDiInstallDriverFiles
 @ stdcall SetupDiLoadClassIcon(ptr ptr ptr)
 @ stub SetupDiMoveDuplicateDevice
@@ -379,6 +380,7 @@
 @ stdcall SetupDiOpenDeviceInterfaceA(ptr str long ptr)
 @ stub SetupDiOpenDeviceInterfaceRegKey
 @ stdcall SetupDiOpenDeviceInterfaceW(ptr wstr long ptr)
+@ stub SetupDiRegisterCoDeviceInstallers
 @ stdcall SetupDiRegisterDeviceInfo(ptr ptr long ptr ptr ptr)
 @ stdcall SetupDiRemoveDevice(ptr ptr)
 @ stdcall SetupDiRemoveDeviceInterface(ptr ptr)
-- 
2.9.0

