From 987688e2ad1576c9a31311e745c0161274a46f15 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 31 Jul 2015 05:42:05 +0200
Subject: opencl: Silence deprecation warning for various OpenCL functions.

---
 dlls/opencl/Makefile.in | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/opencl/Makefile.in b/dlls/opencl/Makefile.in
index 1f6e76c..3523d0a 100644
--- a/dlls/opencl/Makefile.in
+++ b/dlls/opencl/Makefile.in
@@ -1,5 +1,6 @@
 MODULE    = opencl.dll
 EXTRALIBS = $(OPENCL_LIBS)
+EXTRADEFS = -Wno-deprecated-declarations
 
 C_SRCS = \
 	opencl.c
-- 
2.5.0

