From 905baa4b802d589ea8a3f52c5566187c4e2dc1d9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Mon, 15 May 2017 17:56:48 +0200
Subject: winebuild: Add syscall thunks in fake dlls.

---
 tools/winebuild/spec32.c | 32 ++++++++++++++++++++++++++++++--
 1 file changed, 30 insertions(+), 2 deletions(-)

diff --git a/tools/winebuild/spec32.c b/tools/winebuild/spec32.c
index 2d6117df691..d1bd1683334 100644
--- a/tools/winebuild/spec32.c
+++ b/tools/winebuild/spec32.c
@@ -722,10 +722,14 @@ static void create_stub_exports_x86( DLLSPEC *spec )
     for (i = 0; i < spec->nb_entry_points; i++)
     {
         ORDDEF *odp = &spec->entry_points[i];
-        const char *name = get_stub_name( odp, spec );
+        const char *name;
         size_t rva;
 
+        if (odp->flags & FLAG_SYSCALL)
+            continue;
+
         align_output_rva( 16, 16 );
+        name = get_stub_name( odp, spec );
         put_label( name );
 
         put_byte( 0x8b ); put_byte( 0xff );                           /* mov edi, edi */
@@ -750,6 +754,30 @@ static void create_stub_exports_x86( DLLSPEC *spec )
         put_byte( 0xc3 );                                             /* ret */
     }
 
+    /* output syscalls */
+    for (i = 0; i < spec->nb_syscalls; i++)
+    {
+        ORDDEF *odp = spec->syscalls[i];
+
+        align_output_rva( 16, 16 );
+        put_label( odp->link_name );
+
+        put_byte( 0xb8 ); put_dword( i );                     /* mov eax, SYSCALL */
+        put_byte( 0x33 ); put_byte( 0xc9 );                   /* xor ecx, ecx */
+        put_byte( 0x8d ); put_byte( 0x54 );                   /* lea edx, [esp + 4] */
+        put_byte( 0x24 ); put_byte( 0x04 );
+        put_byte( 0x64 ); put_byte( 0xff );                   /* call dword ptr fs:[0C0h] */
+        put_byte( 0x15 ); put_dword( 0xc0 );
+        put_byte( 0x83 ); put_byte( 0xc4 ); put_byte( 0x04 ); /* add esp, 4 */
+        put_byte( 0xc2 ); put_word( get_args_size(odp) );     /* ret X */
+    }
+
+    if (spec->nb_syscalls)
+    {
+        for (i = 0; i < 0x20; i++)
+            put_byte( 0 );
+    }
+
     /* name to show in stub message */
     align_output_rva( 16, 16 );
     put_label( "dll_fake_name" );
@@ -775,7 +803,7 @@ static void create_stub_exports_x86( DLLSPEC *spec )
         ORDDEF *odp = spec->ordinals[i];
         if (odp)
         {
-            const char *name = get_stub_name( odp, spec );
+            const char *name = (odp->flags & FLAG_SYSCALL) ? odp->link_name : get_stub_name( odp, spec );
             put_dword( label_rva( name ) );
         }
         else
-- 
2.12.2

