From d77c31b1567bfbde1ce4ceb526ad596d0be3e788 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 25 May 2017 03:22:25 +0200
Subject: [PATCH] winebuild: Fix size of relocation information in fake dlls.

---
 tools/winebuild/spec32.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tools/winebuild/spec32.c b/tools/winebuild/spec32.c
index 6968e41260f..f95b45b9a0d 100644
--- a/tools/winebuild/spec32.c
+++ b/tools/winebuild/spec32.c
@@ -1238,8 +1238,8 @@ static void output_fake_module_pass( DLLSPEC *spec )
     /* .reloc contents */
     align_output_rva( file_align, section_align );
     put_label( "reloc_start" );
-    put_dword( 0 );   /* VirtualAddress */
-    put_dword( 0 );   /* SizeOfBlock */
+    put_dword( label_rva("text_start") );   /* VirtualAddress */
+    put_dword( 8 );                         /* SizeOfBlock */
     put_label( "reloc_end" );
 
     /* .rsrc contents */
-- 
2.26.2

