From 72e217f735af13a8a7026688fd46f353c24e93ab Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 25 May 2017 03:22:25 +0200
Subject: winebuild: Fix size of relocation information in fake dlls.

---
 tools/winebuild/spec32.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tools/winebuild/spec32.c b/tools/winebuild/spec32.c
index 5a245b5ae81..4dda31c349b 100644
--- a/tools/winebuild/spec32.c
+++ b/tools/winebuild/spec32.c
@@ -1100,8 +1100,8 @@ static void output_fake_module_pass( DLLSPEC *spec )
     /* .reloc contents */
     align_output_rva( file_align, section_align );
     put_label( "reloc_start" );
-    put_dword( 0 );   /* VirtualAddress */
-    put_dword( 0 );   /* SizeOfBlock */
+    put_dword( label_rva("text_start") );   /* VirtualAddress */
+    put_dword( 8 );                         /* SizeOfBlock */
     put_label( "reloc_end" );
 
     /* .rsrc contents */
-- 
2.12.2

