From 8c4a9cd1290fe9240a8de0d3ce5771f528153b06 Mon Sep 17 00:00:00 2001
From: Mark Harmstone <mark@harmstone.com>
Date: Tue, 18 Nov 2014 18:39:58 +0000
Subject: [PATCH 4/7] winepulse: fix segfault in pulse_rd_loop

---
 dlls/winepulse.drv/mmdevdrv.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/winepulse.drv/mmdevdrv.c b/dlls/winepulse.drv/mmdevdrv.c
index e521794..583690b 100644
--- a/dlls/winepulse.drv/mmdevdrv.c
+++ b/dlls/winepulse.drv/mmdevdrv.c
@@ -818,6 +818,7 @@ static void pulse_rd_loop(ACImpl *This, size_t bytes)
         size_t src_len, copy, rem = This->period_bytes;
         if (!(p = (ACPacket*)list_head(&This->packet_free_head))) {
             p = (ACPacket*)list_head(&This->packet_filled_head);
+            if (!p) return;
             if (!p->discont) {
                 next = (ACPacket*)p->entry.next;
                 next->discont = 1;
-- 
1.9.1

