From 9d4e023a3c2f57e3e85ac406135a8a41064b3311 Mon Sep 17 00:00:00 2001
From: Andrew Eikum <aeikum@codeweavers.com>
Date: Tue, 3 Nov 2015 16:22:53 -0600
Subject: winepulse: Prefer PulseAudio driver

Signed-off-by: Andrew Eikum <aeikum@codeweavers.com>
---
 dlls/winepulse.drv/mmdevdrv.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/winepulse.drv/mmdevdrv.c b/dlls/winepulse.drv/mmdevdrv.c
index 9aa1537..b2f83c8 100644
--- a/dlls/winepulse.drv/mmdevdrv.c
+++ b/dlls/winepulse.drv/mmdevdrv.c
@@ -869,7 +869,7 @@ int WINAPI AUDDRV_GetPriority(void)
     pthread_mutex_lock(&pulse_lock);
     hr = pulse_test_connect();
     pthread_mutex_unlock(&pulse_lock);
-    return SUCCEEDED(hr) ? Priority_Low : Priority_Unavailable;
+    return SUCCEEDED(hr) ? Priority_Preferred : Priority_Unavailable;
 }
 
 HRESULT WINAPI AUDDRV_GetAudioEndpoint(GUID *guid, IMMDevice *dev, IAudioClient **out)
-- 
2.6.2

