From f0fe2f14297ba745299f3dd11871dad288666c63 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 4 Nov 2015 02:20:43 +0100
Subject: winepulse.drv: Use delay import for winealsa.drv.

---
 dlls/winepulse.drv/Makefile.in | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/winepulse.drv/Makefile.in b/dlls/winepulse.drv/Makefile.in
index d660063..5671c0d 100644
--- a/dlls/winepulse.drv/Makefile.in
+++ b/dlls/winepulse.drv/Makefile.in
@@ -1,5 +1,6 @@
 MODULE    = winepulse.drv
 IMPORTS   = dxguid uuid winmm user32 advapi32 ole32
+DELAYIMPORTS = winealsa.drv
 EXTRALIBS = $(PULSE_LIBS) $(PTHREAD_LIBS)
 EXTRAINCL = $(PULSE_CFLAGS)
 
-- 
2.6.2

