From f9b183ba340f5dd0f8cf558c91e5a30bf2d09dfd Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 26 Oct 2013 18:40:07 +0200
Subject: winex11: Update gl_drawable for embedded windows

---
 dlls/winex11.drv/window.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index ebbfd45..6a8d63b 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -2286,7 +2286,7 @@ void CDECL X11DRV_WindowPosChanged( HWND hwnd, HWND insert_after, UINT swp_flags
 
     sync_client_position( data, &old_client_rect, &old_whole_rect );
 
-    if (!data->whole_window)
+    if (data->embedded || !data->whole_window)
     {
         release_win_data( data );
         sync_gl_drawable( hwnd, visible_rect, rectClient );
-- 
1.7.9.5

