From d0c4bb573627e7e8ea1295bd781ee58c25a8c769 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 9 Jan 2016 21:20:28 +0100
Subject: vmm.vxd: Fix protection flags passed to VirtualAlloc.

---
 dlls/vmm.vxd/vmm.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/vmm.vxd/vmm.c b/dlls/vmm.vxd/vmm.c
index 7327f78..833a8d1 100644
--- a/dlls/vmm.vxd/vmm.c
+++ b/dlls/vmm.vxd/vmm.c
@@ -159,7 +159,7 @@ DWORD WINAPI VMM_VxDCall( DWORD service, CONTEXT *context )
         if ( page == PR_PRIVATE || page == PR_SHARED ) page = 0;
         /* FIXME: Handle flags in some way */
         address = (LPVOID )(page * page_size);
-        ret = VirtualAlloc ( address, npages * page_size, MEM_RESERVE, 0 );
+        ret = VirtualAlloc ( address, npages * page_size, MEM_RESERVE, PAGE_EXECUTE_READWRITE );
         TRACE("PageReserve: returning: %p\n", ret );
         if ( ret == NULL )
           return -1;
-- 
2.6.4

