From d9ca35b4d4489e101caa4950d086413a3a08a0e7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sun, 28 May 2017 03:59:24 +0200
Subject: wined3d: Don't leak free_so_statistics_queries on device destruction.

---
 dlls/wined3d/context.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/wined3d/context.c b/dlls/wined3d/context.c
index bdcdf007a42..8d674c1cf88 100644
--- a/dlls/wined3d/context.c
+++ b/dlls/wined3d/context.c
@@ -1343,6 +1343,7 @@ static void context_destroy_gl_resources(struct wined3d_context *context)
         checkGLcall("context cleanup");
     }
 
+    HeapFree(GetProcessHeap(), 0, context->free_so_statistics_queries);
     HeapFree(GetProcessHeap(), 0, context->free_timestamp_queries);
     HeapFree(GetProcessHeap(), 0, context->free_occlusion_queries);
     HeapFree(GetProcessHeap(), 0, context->free_event_queries);
-- 
2.12.2

