From fc95e404034572b4335b8239dfd7206aa5c53ce2 Mon Sep 17 00:00:00 2001
From: Maarten Lankhorst <maarten.lankhorst@canonical.com>
Date: Sat, 4 Jan 2014 07:08:55 +0100
Subject: [PATCH 29/42] winepulse v21: return early if padding didn't update

---
 dlls/winepulse.drv/mmdevdrv.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/dlls/winepulse.drv/mmdevdrv.c b/dlls/winepulse.drv/mmdevdrv.c
index 554a9fc..a4575d5 100644
--- a/dlls/winepulse.drv/mmdevdrv.c
+++ b/dlls/winepulse.drv/mmdevdrv.c
@@ -531,7 +531,10 @@ static void pulse_wr_callback(pa_stream *s, size_t bytes, void *userdata)
     else
         This->pad = 0;
 
-    assert(oldpad >= This->pad);
+    if (oldpad == This->pad)
+        return;
+
+    assert(oldpad > This->pad);
 
     This->clock_written += oldpad - This->pad;
     TRACE("New pad: %zu (-%zu)\n", This->pad / pa_frame_size(&This->ss), (oldpad - This->pad) / pa_frame_size(&This->ss));
-- 
1.8.5.2

