From 0f7b5851573569ddf40e36e48670bf9d09765920 Mon Sep 17 00:00:00 2001
From: Mark Harmstone <mark@harmstone.com>
Date: Tue, 24 Jun 2014 08:48:49 +0200
Subject: [PATCH 43/43] winepulse: handle stream create failing correctly

---
 dlls/winepulse.drv/mmdevdrv.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/dlls/winepulse.drv/mmdevdrv.c b/dlls/winepulse.drv/mmdevdrv.c
index f052a08..e755e8a 100644
--- a/dlls/winepulse.drv/mmdevdrv.c
+++ b/dlls/winepulse.drv/mmdevdrv.c
@@ -701,6 +701,12 @@ static HRESULT pulse_stream_connect(ACImpl *This, UINT32 period_bytes) {
     ret = InterlockedIncrement(&number);
     sprintf(buffer, "audio stream #%i", ret);
     This->stream = pa_stream_new(pulse_ctx, buffer, &This->ss, &This->map);
+
+    if (!This->stream) {
+        WARN("pa_stream_new returned error %i\n", pa_context_errno(pulse_ctx));
+        return AUDCLNT_E_ENDPOINT_CREATE_FAILED;
+    }
+
     pa_stream_set_state_callback(This->stream, pulse_stream_state, This);
     pa_stream_set_buffer_attr_callback(This->stream, pulse_attr_update, This);
     pa_stream_set_moved_callback(This->stream, pulse_attr_update, This);
-- 
2.0.0

