From 5c62e8187c7af1dbf7ad25e521f1a53e84f1c6d3 Mon Sep 17 00:00:00 2001
From: Andrew Wesie <awesie@gmail.com>
Date: Sun, 13 Nov 2016 12:35:04 -0600
Subject: ntdll: Fix privileged instruction exception code.

Signed-off-by: Andrew Wesie <awesie@gmail.com>
---
 dlls/ntdll/signal_x86_64.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/dlls/ntdll/signal_x86_64.c b/dlls/ntdll/signal_x86_64.c
index 1c5ab158a3a..f434775df00 100644
--- a/dlls/ntdll/signal_x86_64.c
+++ b/dlls/ntdll/signal_x86_64.c
@@ -2790,7 +2790,6 @@ static void segv_handler( int signal, siginfo_t *siginfo, void *sigcontext )
             WORD err = ERROR_sig(ucontext);
             if ((err & 7) == 2 && handle_interrupt( err >> 3, rec, win_context )) break;
             rec->ExceptionCode = err ? EXCEPTION_ACCESS_VIOLATION : EXCEPTION_PRIV_INSTRUCTION;
-            rec->ExceptionCode = EXCEPTION_ACCESS_VIOLATION;
         }
         break;
     case TRAP_x86_PAGEFLT:  /* Page fault */
-- 
2.14.1

