From f3bc9d9828f84d9899fd1b9ef42af3ea7027e258 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Thu, 22 Mar 2018 09:35:41 +1100
Subject: [PATCH] winex11.drv: Fix compile error when Vulkan isn't avaiable

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 dlls/winex11.drv/vulkan.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/winex11.drv/vulkan.c b/dlls/winex11.drv/vulkan.c
index 27ff722..f96ef64 100644
--- a/dlls/winex11.drv/vulkan.c
+++ b/dlls/winex11.drv/vulkan.c
@@ -37,10 +37,10 @@
 #include "wine/vulkan.h"
 #include "wine/vulkan_driver.h"
 
-#ifdef SONAME_LIBVULKAN
-
 WINE_DEFAULT_DEBUG_CHANNEL(vulkan);
 
+#ifdef SONAME_LIBVULKAN
+
 typedef VkFlags VkXlibSurfaceCreateFlagsKHR;
 #define VK_STRUCTURE_TYPE_XLIB_SURFACE_CREATE_INFO_KHR 1000004000
 
-- 
1.9.1

