From 90d826bf3bfe4cb92334bdab29fd0b71d5cb947b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 16 Apr 2015 21:23:10 +0200
Subject: libwine: Add process specific debug channels.

---
 libs/wine/debug.c | 27 +++++++++++++++++++++++++--
 1 file changed, 25 insertions(+), 2 deletions(-)

diff --git a/libs/wine/debug.c b/libs/wine/debug.c
index 548ef22..caaf383 100644
--- a/libs/wine/debug.c
+++ b/libs/wine/debug.c
@@ -125,10 +125,22 @@ static void add_option( const char *name, unsigned char set, unsigned char clear
     nb_debug_options++;
 }
 
+/* get name of the current process */
+static const char *get_process_name( void )
+{
+    const char *ret, *tmp;
+    if (__wine_main_argc < 2) return NULL;
+    ret = __wine_main_argv[1];
+    if ((tmp = strrchr(ret, '/'))) ret = ++tmp;
+    if ((tmp = strrchr(ret, '\\'))) ret = ++tmp;
+    return ret;
+}
+
 /* parse a set of debugging option specifications and add them to the option list */
 static void parse_options( const char *str )
 {
-    char *opt, *next, *options;
+    char *opt, *next, *popt, *options;
+    const char *process = get_process_name();
     unsigned int i;
 
     if (!(options = strdup(str))) return;
@@ -139,6 +151,17 @@ static void parse_options( const char *str )
 
         if ((next = strchr( opt, ',' ))) *next++ = 0;
 
+        if ((popt = strchr( opt, ':' )))
+        {
+            unsigned int inv = 0;
+            *popt = 0;
+            if (!process) continue;
+            if (*opt == '-' || *opt == '+')
+                inv = (*opt++ == '-');
+            if (inv == !strcmp( opt, process )) continue;
+            opt = ++popt;
+        }
+
         p = opt + strcspn( opt, "+-" );
         if (!p[0]) p = opt;  /* assume it's a debug channel name */
 
@@ -180,7 +203,7 @@ static void debug_usage(void)
 {
     static const char usage[] =
         "Syntax of the WINEDEBUG variable:\n"
-        "  WINEDEBUG=[class]+xxx,[class]-yyy,...\n\n"
+        "  WINEDEBUG=[+process:][class]+xxx,[-process:][class]-yyy,...\n\n"
         "Example: WINEDEBUG=+all,warn-heap\n"
         "    turns on all messages except warning heap messages\n"
         "Available message classes: err, warn, fixme, trace\n";
-- 
2.3.5

