From 76ff2b8644ec3e3b7c17ff3bc058dbdfc11a2ea0 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 17 May 2019 08:16:47 +1000
Subject: [PATCH] xaudio2_7: Fix build

---
 dlls/xaudio2_7/xaudio_dll.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/xaudio2_7/xaudio_dll.c b/dlls/xaudio2_7/xaudio_dll.c
index 8fc2364..3dc0d6c 100644
--- a/dlls/xaudio2_7/xaudio_dll.c
+++ b/dlls/xaudio2_7/xaudio_dll.c
@@ -30,6 +30,7 @@
 
 #include "xapofx.h"
 
+#include "wine/asm.h"
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(xaudio2);
-- 
1.9.1

