From ed6d82faea2efb04dda99738f8eea9ac8c9de1df Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 9 Jan 2014 02:54:31 +0100
Subject: user32: Decrease minimum SetTimer interval from 15 to 5 ms

---
 dlls/user32/message.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/user32/message.c b/dlls/user32/message.c
index 2cd1472..8a4aa7d 100644
--- a/dlls/user32/message.c
+++ b/dlls/user32/message.c
@@ -4407,7 +4407,7 @@ UINT_PTR WINAPI SetTimer( HWND hwnd, UINT_PTR id, UINT timeout, TIMERPROC proc )
 
     /* MSDN states that the minimum timeout should be USER_TIMER_MINIMUM (10.0 ms), but testing
      * indicates that the true minimum is closer to 15.6 ms. */
-    timeout = min( max( 15, timeout ), USER_TIMER_MAXIMUM );
+    timeout = min( max( 5, timeout ), USER_TIMER_MAXIMUM );
 
     SERVER_START_REQ( set_win_timer )
     {
-- 
1.8.3.2

