From 34551d5eb1b88021766cafcb9046d1e76a2b3bf8 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Mon, 13 Jan 2014 23:37:02 +0100
Subject: wined3d: Allow to set strictDrawOrdering via environment variable

---
 dlls/wined3d/wined3d_main.c |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/wined3d/wined3d_main.c b/dlls/wined3d/wined3d_main.c
index a0ecfa7..fa89700 100644
--- a/dlls/wined3d/wined3d_main.c
+++ b/dlls/wined3d/wined3d_main.c
@@ -140,6 +140,7 @@ static BOOL wined3d_dll_init(HINSTANCE hInstDLL)
     HKEY appkey = 0;
     DWORD len, tmpvalue;
     WNDCLASSA wc;
+    const char *env;
 
     wined3d_context_tls_idx = TlsAlloc();
     if (wined3d_context_tls_idx == TLS_OUT_OF_INDEXES)
@@ -309,6 +310,9 @@ static BOOL wined3d_dll_init(HINSTANCE hInstDLL)
     if (appkey) RegCloseKey( appkey );
     if (hkey) RegCloseKey( hkey );
 
+    if ((env = getenv("WINE_STRICT_DRAW_ORDERING")))
+        wined3d_settings.strict_draw_ordering = atoi(env);
+
     return TRUE;
 }
 
-- 
1.7.9.5

