From 5d8bbfbf4bc409af8de7ad580e6cde74797c170c Mon Sep 17 00:00:00 2001
From: Andrew Church <achurch+wine@achurch.org>
Date: Thu, 4 Dec 2014 14:57:46 +0100
Subject: user32: Set last error when GetRawInputDeviceList fails.

---
 dlls/user32/input.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/user32/input.c b/dlls/user32/input.c
index ec81e60..967b114 100644
--- a/dlls/user32/input.c
+++ b/dlls/user32/input.c
@@ -493,6 +493,7 @@ UINT WINAPI GetRawInputDeviceList(RAWINPUTDEVICELIST *devices, UINT *device_coun
     if (*device_count < 2)
     {
         *device_count = 2;
+        SetLastError(ERROR_INSUFFICIENT_BUFFER);
         return ~0U;
     }
 
-- 
2.1.3

