From 69a209f88bb8a21ec48638e6d6b6f805a258d788 Mon Sep 17 00:00:00 2001
From: Thomas Crider <gloriouseggroll@gmail.com>
Date: Fri, 13 Apr 2018 18:09:21 -0400
Subject: [PATCH] xaudio2: Return S_OK in IXAudio2 in CommitChanges

---
 dlls/xaudio2_7/xaudio_dll.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/xaudio2_7/xaudio_dll.c b/dlls/xaudio2_7/xaudio_dll.c
index 28cf985..647d2fd 100644
--- a/dlls/xaudio2_7/xaudio_dll.c
+++ b/dlls/xaudio2_7/xaudio_dll.c
@@ -1895,7 +1895,7 @@ static HRESULT WINAPI IXAudio2Impl_CommitChanges(IXAudio2 *iface,
 
     TRACE("(%p)->(0x%x): stub!\n", This, operationSet);
 
-    return E_NOTIMPL;
+    return S_OK;
 }
 
 static void WINAPI IXAudio2Impl_GetPerformanceData(IXAudio2 *iface,
-- 
1.9.1

