From ef42bee07f37038aa690ddf8c164c7fff8512a25 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 12 Jul 2015 03:36:20 +0200
Subject: dwrite: Avoid dereferencing NULL pointer for fonts without VDMX.

---
 dlls/dwrite/opentype.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/dwrite/opentype.c b/dlls/dwrite/opentype.c
index 63e1b74..8d90c04 100644
--- a/dlls/dwrite/opentype.c
+++ b/dlls/dwrite/opentype.c
@@ -1347,7 +1347,7 @@ static const struct VDMX_group *find_vdmx_group(const struct VDMX_Header *hdr)
 BOOL opentype_get_vdmx_size(const void *data, INT emsize, UINT16 *ascent, UINT16 *descent)
 {
     const struct VDMX_Header *hdr = (const struct VDMX_Header*)data;
-    const struct VDMX_group *group = find_vdmx_group(hdr);
+    const struct VDMX_group *group;
     const struct VDMX_vTable *tables;
     WORD recs, i;
 
-- 
2.4.5

