From 3d286a43424fc20c3a8ccdfbbbf7ce84a2bb4300 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 26 Sep 2014 20:24:50 +0200
Subject: user32: Call UpdateWindow() during DIALOG_CreateIndirect.

---
 dlls/user32/dialog.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/user32/dialog.c b/dlls/user32/dialog.c
index 497d545..819af7a 100644
--- a/dlls/user32/dialog.c
+++ b/dlls/user32/dialog.c
@@ -714,6 +714,7 @@ static HWND DIALOG_CreateIndirect( HINSTANCE hInst, LPCVOID dlgTemplate,
         if (template.style & WS_VISIBLE && !(GetWindowLongW( hwnd, GWL_STYLE ) & WS_VISIBLE))
         {
            ShowWindow( hwnd, SW_SHOWNORMAL );   /* SW_SHOW doesn't always work */
+            UpdateWindow( hwnd );
         }
         return hwnd;
     }
-- 
2.1.1

