From cef8020c7a1f4f8224513dac2017189c2774d1c8 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 3 Sep 2015 23:30:25 +0200
Subject: opencl: Avoid deprecation warning for OpenCL 1.2 APIs.

---
 dlls/opencl/opencl.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/opencl/opencl.c b/dlls/opencl/opencl.c
index f690733..2d145bf 100644
--- a/dlls/opencl/opencl.c
+++ b/dlls/opencl/opencl.c
@@ -32,6 +32,7 @@ WINE_DEFAULT_DEBUG_CHANNEL(opencl);
 
 #if defined(HAVE_CL_CL_H)
 #define CL_USE_DEPRECATED_OPENCL_1_1_APIS
+#define CL_USE_DEPRECATED_OPENCL_1_2_APIS
 #define CL_USE_DEPRECATED_OPENCL_2_0_APIS
 #include <CL/cl.h>
 #elif defined(HAVE_OPENCL_OPENCL_H)
-- 
2.5.1

