From 43743ce1c92eeb824eb489f0a4b06e81e28a3426 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 14 Aug 2014 01:39:13 +0200
Subject: msi: Fix ITERATE_MoveFiles when no source- and destname is
 specified.

---
 dlls/msi/files.c |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/dlls/msi/files.c b/dlls/msi/files.c
index ec46ae3..62e5f43 100644
--- a/dlls/msi/files.c
+++ b/dlls/msi/files.c
@@ -843,7 +843,13 @@ static UINT ITERATE_MoveFiles( MSIRECORD *rec, LPVOID param )
     {
         if (!wildcards)
         {
-            destname = strdupW(sourcename);
+            WCHAR *p;
+            if (sourcename)
+                destname = strdupW(sourcename);
+            else if ((p = strrchrW(sourcedir, '\\')))
+                destname = strdupW(p + 1);
+            else
+                destname = strdupW(sourcedir);
             if (!destname)
                 goto done;
         }
-- 
1.7.9.5

