From 992dbbf53f5c8d372fcc211e6b3205ec2e096489 Mon Sep 17 00:00:00 2001
From: Yanis Lukes <Pendrokar@gmail.com>
Date: Sun, 28 Sep 2014 17:04:45 +0200
Subject: wine.inf: Add fake ProductId to HKLM\CurrentVersionNT.

---
 loader/wine.inf.in | 1 +
 1 file changed, 1 insertion(+)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 42a750e..64947c8 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -468,6 +468,7 @@ HKLM,%CurrentVersion%\Uninstall,,16
 HKCU,%CurrentVersion%\Run,,16
 HKLM,%CurrentVersion%\Run,,16
 HKLM,%CurrentVersionNT%,"InstallDate",0x10003,1273299354
+HKLM,%CurrentVersionNT%,"ProductId",,"12345-oem-0000001-54321"
 HKLM,%CurrentVersionNT%,"RegisteredOrganization",2,""
 HKLM,%CurrentVersionNT%,"RegisteredOwner",2,""
 HKLM,%CurrentVersionNT%,"SystemRoot",,"%10%"
-- 
1.9.1

