From 1f127d55e0e6d473e6fe5069f2957f9db71fdc23 Mon Sep 17 00:00:00 2001
From: Hugh McMaster <hugh.mcmaster@outlook.com>
Date: Mon, 8 Jun 2015 22:22:21 +1000
Subject: kernel32: Set console InsertMode immediately.

---
 dlls/kernel32/editline.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/kernel32/editline.c b/dlls/kernel32/editline.c
index 7664b55..3c2e8de 100644
--- a/dlls/kernel32/editline.c
+++ b/dlls/kernel32/editline.c
@@ -987,6 +987,10 @@ WCHAR* CONSOLE_Readline(HANDLE hConsoleIn, BOOL can_pos_cursor)
 	    }
 	}
 
+        if (GetConsoleMode(hConsoleIn, &mode))
+            ctx.insert = (mode & (ENABLE_INSERT_MODE|ENABLE_EXTENDED_FLAGS)) ==
+                         (ENABLE_INSERT_MODE|ENABLE_EXTENDED_FLAGS) ? 1 : 0;
+
 	if (func)
 	    (func)(&ctx);
 	else if (!(ir.Event.KeyEvent.dwControlKeyState & LEFT_ALT_PRESSED))
-- 
2.4.2

