From fcb6d21ede52ef770a04c59ea6bd5b9483c31f32 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 17 May 2017 12:03:15 +0200
Subject: appwiz.cpl: Copy addons to cache instead of moving.

---
 dlls/appwiz.cpl/addons.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/appwiz.cpl/addons.c b/dlls/appwiz.cpl/addons.c
index c3847b54948..91cd485b026 100644
--- a/dlls/appwiz.cpl/addons.c
+++ b/dlls/appwiz.cpl/addons.c
@@ -510,7 +510,7 @@ static HRESULT WINAPI InstallCallback_OnStopBinding(IBindStatusCallback *iface,
 
         cache_file_name = get_cache_file_name(TRUE);
         if(cache_file_name) {
-            MoveFileW(msi_file, cache_file_name);
+            CopyFileW(msi_file, cache_file_name, FALSE);
             heap_free(cache_file_name);
         }
     }else {
-- 
2.12.2

