From c908afaed01d4ca53ef66a7fb649402f9ae1a3d8 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 26 Feb 2015 04:23:28 +0100
Subject: server: Fall back to global key state when thread doesn't have a
 queue.

---
 server/queue.c | 25 +++++++++++++------------
 1 file changed, 13 insertions(+), 12 deletions(-)

diff --git a/server/queue.c b/server/queue.c
index 3a321cd..6cbf8e1 100644
--- a/server/queue.c
+++ b/server/queue.c
@@ -2783,27 +2783,28 @@ DECL_HANDLER(get_key_state)
     struct desktop *desktop;
     data_size_t size = min( 256, get_reply_max_size() );
 
-    if (!req->tid)  /* get global async key state */
-    {
-        if (!(desktop = get_thread_desktop( current, 0 ))) return;
-        if (req->key >= 0)
-        {
-            reply->state = desktop->keystate[req->key & 0xff];
-            desktop->keystate[req->key & 0xff] &= ~0x40;
-        }
-        set_reply_data( desktop->keystate, size );
-        release_object( desktop );
-    }
-    else
+    if (req->tid)
     {
         if (!(thread = get_thread_from_id( req->tid ))) return;
         if (thread->queue)
         {
             if (req->key >= 0) reply->state = thread->queue->input->keystate[req->key & 0xff];
             set_reply_data( thread->queue->input->keystate, size );
+            release_object( thread );
+            return;
         }
         release_object( thread );
     }
+
+    /* get global async key state */
+    if (!(desktop = get_thread_desktop( current, 0 ))) return;
+    if (req->key >= 0)
+    {
+        reply->state = desktop->keystate[req->key & 0xff];
+        desktop->keystate[req->key & 0xff] &= ~0x40;
+    }
+    set_reply_data( desktop->keystate, size );
+    release_object( desktop );
 }
 
 
-- 
2.3.0

