From fe236be1d5f8285d4438afde1bd5b7b7d502341d Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 24 Oct 2015 22:52:44 +0200
Subject: combase: Simplify check for NULL pointer in
 WindowsCreateString[Reference].

---
 dlls/combase/string.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/dlls/combase/string.c b/dlls/combase/string.c
index 766e765..4344d36 100644
--- a/dlls/combase/string.c
+++ b/dlls/combase/string.c
@@ -79,13 +79,13 @@ HRESULT WINAPI WindowsCreateString(LPCWSTR ptr, UINT32 len,
 
     if (out == NULL)
         return E_INVALIDARG;
-    if (ptr == NULL && len > 0)
-        return E_POINTER;
     if (len == 0)
     {
         *out = NULL;
         return S_OK;
     }
+    if (ptr == NULL)
+        return E_POINTER;
     if (!alloc_string(len, out))
         return E_OUTOFMEMORY;
     priv = impl_from_HSTRING(*out);
@@ -105,13 +105,13 @@ HRESULT WINAPI WindowsCreateStringReference(LPCWSTR ptr, UINT32 len,
 
     if (out == NULL || header == NULL)
         return E_INVALIDARG;
-    if (ptr == NULL && len > 0)
-        return E_POINTER;
     if (len == 0)
     {
         *out = NULL;
         return S_OK;
     }
+    if (ptr == NULL)
+        return E_POINTER;
     if (ptr[len] != '\0')
         return E_INVALIDARG;
     priv->buffer = (LPWSTR)ptr;
-- 
2.6.1

