From 39a3448740eddb5083e3508e51c3ec6922275c5b Mon Sep 17 00:00:00 2001
From: Jarkko Korpi <jarkko_korpi@hotmail.com>
Date: Sat, 1 Aug 2015 06:45:35 +0300
Subject: wininet: Silence wininet no support on this platform message.

---
 dlls/wininet/internet.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/wininet/internet.c b/dlls/wininet/internet.c
index bd380dc..0680c7d 100644
--- a/dlls/wininet/internet.c
+++ b/dlls/wininet/internet.c
@@ -2311,7 +2311,8 @@ static WCHAR *get_proxy_autoconfig_url(void)
     CFRelease( settings );
     return ret;
 #else
-    FIXME( "no support on this platform\n" );
+    static int once;
+    if (!once++) FIXME( "no support on this platform\n" );
     return NULL;
 #endif
 }
-- 
2.8.0

