From 781e0d3e9a3c1818d59d09125e5f7bf503ea6015 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Thu, 18 Jul 2019 09:46:30 +1000
Subject: [PATCH 3/4] uuid: Add directmanipulation.h

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 dlls/uuid/uuid.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/uuid/uuid.c b/dlls/uuid/uuid.c
index aa3b30c3c7..eacb83301d 100644
--- a/dlls/uuid/uuid.c
+++ b/dlls/uuid/uuid.c
@@ -90,6 +90,7 @@ DEFINE_GUID(GUID_NULL,0,0,0,0,0,0,0,0,0,0,0);
 #include "netcfgx.h"
 #include "shimgdata.h"
 #include "comsvcs.h"
+#include "directmanipulation.h"
 
 /* FIXME: cguids declares GUIDs but does not define their values */
 
-- 
2.17.1

