From 6ee51b2ae31c53d05fef808c0aa1cfd47534e6af Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Fri, 19 Jul 2019 08:46:01 +1000
Subject: [PATCH 5/6] include: Add DCompManipulationCompositor coclass and
 supporting interfaces.

---
 include/directmanipulation.idl | 79 ++++++++++++++++++++++++++++++++++
 1 file changed, 79 insertions(+)

diff --git a/include/directmanipulation.idl b/include/directmanipulation.idl
index 9dc2366cc3..c2fd831c6c 100644
--- a/include/directmanipulation.idl
+++ b/include/directmanipulation.idl
@@ -66,6 +66,76 @@ interface IDirectManipulationManager : IUnknown
                           [in] REFIID riid, [out, iid_is(riid), retval] void **obj);
 }
 
+[
+    local,
+    object,
+    uuid(b89962cb-3d89-442b-bb58-5098fa0f9f16),
+    pointer_default(unique)
+]
+interface IDirectManipulationContent : IUnknown
+{
+    HRESULT GetContentRect([out] RECT *size);
+
+    HRESULT SetContentRect([in] const RECT *size);
+
+    HRESULT GetViewport([in] REFIID riid, [out, iid_is(riid)] void **object);
+
+    HRESULT GetTag([in] REFIID riid, [out, iid_is(riid)] void **object, [out] UINT32 *id);
+
+    HRESULT SetTag([in, unique] IUnknown *object, [in] UINT32 id);
+
+    HRESULT GetOutputTransform([out] float *matrix, [in] DWORD count);
+
+    HRESULT GetContentTransform([out] float *matrix, [in] DWORD count);
+
+    HRESULT SyncContentTransform([in] const float *matrix, [in] DWORD count);
+}
+
+[
+    local,
+    object,
+    uuid(790b6337-64f8-4ff5-a269-b32bc2af27a7),
+    pointer_default(unique)
+]
+interface IDirectManipulationUpdateHandler : IUnknown
+{
+    HRESULT Update();
+}
+
+[
+    local,
+    object,
+    uuid(b0ae62fd-be34-46e7-9caa-d361facbb9cc),
+    pointer_default(unique)
+]
+interface IDirectManipulationUpdateManager : IUnknown
+{
+    HRESULT RegisterWaitHandleCallback([in] HANDLE handle, [in] IDirectManipulationUpdateHandler *handler,
+                                       [out] DWORD *cookie);
+
+    HRESULT UnregisterWaitHandleCallback([in] DWORD cookie);
+
+    HRESULT Update([in] IDirectManipulationFrameInfoProvider *provider);
+}
+
+[
+    local,
+    object,
+    uuid(537a0825-0387-4efa-b62f-71eb1f085a7e),
+    pointer_default(unique)
+]
+interface IDirectManipulationCompositor : IUnknown
+{
+    HRESULT AddContent([in] IDirectManipulationContent *content, [in] IUnknown *device,
+                       [in] IUnknown *parent, [in] IUnknown *child);
+
+    HRESULT RemoveContent([in] IDirectManipulationContent *content);
+
+    HRESULT SetUpdateManager([in] IDirectManipulationUpdateManager *manager);
+
+    HRESULT Flush();
+}
+
 [
     local,
     object,
@@ -100,4 +170,13 @@ library DirectManipulation
         interface IDirectManipulationManager2;
         [default] interface IDirectManipulationManager;
     }
+
+    [
+        uuid(79dea627-a08a-43ac-8ef5-6900b9299126)
+    ]
+    coclass DCompManipulationCompositor
+    {
+        [default] interface IDirectManipulationCompositor;
+        interface IDirectManipulationFrameInfoProvider;
+    }
 }
-- 
2.17.1

