From d71134e27429531e7ebf25f4b5db8e9047ecbaad Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 24 Jul 2019 10:05:35 +1000
Subject: [PATCH] include: Add IDirectManipulationViewport2 interface

---
 include/directmanipulation.idl | 173 +++++++++++++++++++++++++++++++++
 1 file changed, 173 insertions(+)

diff --git a/include/directmanipulation.idl b/include/directmanipulation.idl
index c2fd831c6c..443b31a0ce 100644
--- a/include/directmanipulation.idl
+++ b/include/directmanipulation.idl
@@ -23,6 +23,8 @@ cpp_quote("#if 0")
 typedef void* HWND;
 cpp_quote("#endif")
 
+interface IDirectManipulationViewportEventHandler;
+
 typedef enum DIRECTMANIPULATION_HITTEST_TYPE
 {
     DIRECTMANIPULATION_HITTEST_TYPE_ASYNCHRONOUS     = 0x00000000,
@@ -30,6 +32,68 @@ typedef enum DIRECTMANIPULATION_HITTEST_TYPE
     DIRECTMANIPULATION_HITTEST_TYPE_AUTO_SYNCHRONOUS = 0x00000002
 } DIRECTMANIPULATION_HITTEST_TYPE;
 
+typedef enum DIRECTMANIPULATION_STATUS
+{
+    DIRECTMANIPULATION_BUILDING  = 0,
+    DIRECTMANIPULATION_ENABLED   = 1,
+    DIRECTMANIPULATION_DISABLED  = 2,
+    DIRECTMANIPULATION_RUNNING   = 3,
+    DIRECTMANIPULATION_INERTIA   = 4,
+    DIRECTMANIPULATION_READY     = 5,
+    DIRECTMANIPULATION_SUSPENDED = 6
+} DIRECTMANIPULATION_STATUS;
+
+typedef enum DIRECTMANIPULATION_VIEWPORT_OPTIONS
+{
+    DIRECTMANIPULATION_VIEWPORT_OPTIONS_DEFAULT              = 0x0000,
+    DIRECTMANIPULATION_VIEWPORT_OPTIONS_AUTODISABLE          = 0x0001,
+    DIRECTMANIPULATION_VIEWPORT_OPTIONS_MANUALUPDATE         = 0x0002,
+    DIRECTMANIPULATION_VIEWPORT_OPTIONS_INPUT                = 0x0004,
+    DIRECTMANIPULATION_VIEWPORT_OPTIONS_EXPLICITHITTEST      = 0x0008,
+    DIRECTMANIPULATION_VIEWPORT_OPTIONS_DISABLEPIXELSNAPPING = 0x0010,
+} DIRECTMANIPULATION_VIEWPORT_OPTIONS;
+
+typedef enum DIRECTMANIPULATION_CONFIGURATION
+{
+    DIRECTMANIPULATION_CONFIGURATION_NONE                = 0x00000000,
+    DIRECTMANIPULATION_CONFIGURATION_INTERACTION         = 0x00000001,
+    DIRECTMANIPULATION_CONFIGURATION_TRANSLATION_X       = 0x00000002,
+    DIRECTMANIPULATION_CONFIGURATION_TRANSLATION_Y       = 0x00000004,
+    DIRECTMANIPULATION_CONFIGURATION_SCALING             = 0x00000010,
+    DIRECTMANIPULATION_CONFIGURATION_TRANSLATION_INERTIA = 0x00000020,
+    DIRECTMANIPULATION_CONFIGURATION_SCALING_INERTIA     = 0x00000080,
+    DIRECTMANIPULATION_CONFIGURATION_RAILS_X             = 0x00000100,
+    DIRECTMANIPULATION_CONFIGURATION_RAILS_Y             = 0x00000200,
+} DIRECTMANIPULATION_CONFIGURATION;
+
+typedef enum DIRECTMANIPULATION_GESTURE_CONFIGURATION
+{
+    DIRECTMANIPULATION_GESTURE_NONE                   = 0x00000000,
+    DIRECTMANIPULATION_GESTURE_DEFAULT                = 0x00000000,
+    DIRECTMANIPULATION_GESTURE_CROSS_SLIDE_VERTICAL   = 0x00000008,
+    DIRECTMANIPULATION_GESTURE_CROSS_SLIDE_HORIZONTAL = 0x00000010,
+    DIRECTMANIPULATION_GESTURE_PINCH_ZOOM             = 0x00000020,
+} DIRECTMANIPULATION_GESTURE_CONFIGURATION;
+
+typedef enum DIRECTMANIPULATION_MOTION_TYPES
+{
+    DIRECTMANIPULATION_MOTION_NONE       = 0x0000,
+    DIRECTMANIPULATION_MOTION_TRANSLATEX = 0x0001,
+    DIRECTMANIPULATION_MOTION_TRANSLATEY = 0x0002,
+    DIRECTMANIPULATION_MOTION_ZOOM       = 0x0004,
+    DIRECTMANIPULATION_MOTION_CENTERX    = 0x0010,
+    DIRECTMANIPULATION_MOTION_CENTERY    = 0x0020,
+    DIRECTMANIPULATION_MOTION_ALL        = DIRECTMANIPULATION_MOTION_TRANSLATEX | DIRECTMANIPULATION_MOTION_TRANSLATEY |
+                                           DIRECTMANIPULATION_MOTION_ZOOM | DIRECTMANIPULATION_MOTION_CENTERX |
+                                           DIRECTMANIPULATION_MOTION_CENTERY
+} DIRECTMANIPULATION_MOTION_TYPES;
+
+typedef enum DIRECTMANIPULATION_INPUT_MODE
+{
+    DIRECTMANIPULATION_INPUT_MODE_AUTOMATIC = 0,
+    DIRECTMANIPULATION_INPUT_MODE_MANUAL    = 1
+} DIRECTMANIPULATION_INPUT_MODE;
+
 [
     local,
     object,
@@ -136,6 +200,104 @@ interface IDirectManipulationCompositor : IUnknown
     HRESULT Flush();
 }
 
+[
+    local,
+    object,
+    uuid(28b85a3d-60a0-48bd-9ba1-5ce8d9ea3a6d),
+    pointer_default(unique)
+]
+interface IDirectManipulationViewport : IUnknown
+{
+    HRESULT Enable();
+
+    HRESULT Disable();
+
+    HRESULT SetContact([in] UINT32 id);
+
+    HRESULT ReleaseContact([in] UINT32 id);
+
+    HRESULT ReleaseAllContacts();
+
+    HRESULT GetStatus([out] DIRECTMANIPULATION_STATUS *status);
+
+    HRESULT GetTag([in] REFIID riid, [out, iid_is(riid)] void **object, [out] UINT32 *id);
+
+    HRESULT SetTag([in, unique] IUnknown *object, [in] UINT32 id);
+
+    HRESULT GetViewportRect([out, retval] RECT *viewport);
+
+    HRESULT SetViewportRect([in] const RECT *viewport);
+
+    HRESULT ZoomToRect([in] const float left, [in] const float top, [in] const float right,
+                        [in] const float bottom, [in] BOOL animate);
+
+    HRESULT SetViewportTransform([in] const float *matrix, [in] DWORD count);
+
+    HRESULT SyncDisplayTransform([in] const float *matrix, [in] DWORD count);
+
+    HRESULT GetPrimaryContent([in] REFIID riid, [out, iid_is(riid)] void **object);
+
+    HRESULT AddContent([in, unique] IDirectManipulationContent *content);
+
+    HRESULT RemoveContent([in, unique] IDirectManipulationContent *content);
+
+    HRESULT SetViewportOptions([in] DIRECTMANIPULATION_VIEWPORT_OPTIONS options);
+
+    HRESULT AddConfiguration([in] DIRECTMANIPULATION_CONFIGURATION configuration);
+
+    HRESULT RemoveConfiguration([in] DIRECTMANIPULATION_CONFIGURATION configuration);
+
+    HRESULT ActivateConfiguration([in] DIRECTMANIPULATION_CONFIGURATION configuration);
+
+    HRESULT SetManualGesture([in] DIRECTMANIPULATION_GESTURE_CONFIGURATION configuration);
+
+    HRESULT SetChaining([in] DIRECTMANIPULATION_MOTION_TYPES enabledTypes);
+
+    HRESULT AddEventHandler([in] HWND window, [in] IDirectManipulationViewportEventHandler *eventHandler,
+                        [out, retval] DWORD *cookie);
+
+    HRESULT RemoveEventHandler([in] DWORD cookie);
+
+    HRESULT SetInputMode([in] DIRECTMANIPULATION_INPUT_MODE mode);
+
+    HRESULT SetUpdateMode([in] DIRECTMANIPULATION_INPUT_MODE mode);
+
+    HRESULT Stop();
+
+    HRESULT Abandon();
+}
+
+[
+    local,
+    object,
+    uuid(923ccaac-61e1-4385-b726-017af189882a),
+    pointer_default(unique)
+]
+interface IDirectManipulationViewport2 : IDirectManipulationViewport
+{
+    HRESULT AddBehavior([in] IUnknown *behavior, [out, retval] DWORD *cookie);
+
+    HRESULT RemoveBehavior([in] DWORD cookie);
+
+    HRESULT RemoveAllBehaviors();
+}
+
+[
+    local,
+    object,
+    uuid(952121DA-D69F-45F9-B0F9-F23944321A6D),
+    pointer_default(unique)
+]
+interface IDirectManipulationViewportEventHandler : IUnknown
+{
+    HRESULT OnViewportStatusChanged([in] IDirectManipulationViewport *viewport, [in] DIRECTMANIPULATION_STATUS current,
+                                    [in] DIRECTMANIPULATION_STATUS previous);
+
+    HRESULT OnViewportUpdated([in] IDirectManipulationViewport *viewport);
+
+    HRESULT OnContentUpdated([in] IDirectManipulationViewport *viewport, [in] IDirectManipulationContent *content);
+}
+
 [
     local,
     object,
@@ -179,4 +341,15 @@ library DirectManipulation
         [default] interface IDirectManipulationCompositor;
         interface IDirectManipulationFrameInfoProvider;
     }
+
+    [
+        noncreatable,
+        hidden,
+        uuid(34e211b6-3650-4f75-8334-fa359598e1c5)
+    ]
+    coclass DirectManipulationViewport
+    {
+        interface IDirectManipulationViewport2;
+        [default] interface IDirectManipulationViewport;
+    }
 }
-- 
2.17.1

