From 2f249a6a8c5906813d05470df61928d43eb289c8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Wed, 20 Sep 2017 21:33:30 +0200
Subject: d3d11: Implement d3d10_device_GenerateMips.

---
 dlls/d3d11/device.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/dlls/d3d11/device.c b/dlls/d3d11/device.c
index 22adf361feb..d389c2625a0 100644
--- a/dlls/d3d11/device.c
+++ b/dlls/d3d11/device.c
@@ -7550,12 +7550,16 @@ static void STDMETHODCALLTYPE d3d10_devi
 static void STDMETHODCALLTYPE d3d10_device_GenerateMips(ID3D10Device1 *iface,
         ID3D10ShaderResourceView *view)
 {
-    struct d3d_shader_resource_view *srv = unsafe_impl_from_ID3D10ShaderResourceView(view);
+    struct d3d_device *device = impl_from_ID3D10Device(iface);
+    struct d3d_shader_resource_view *view = unsafe_impl_from_ID3D10ShaderResourceView(shader_resource_view);
 
-    TRACE("iface %p, view %p.\n", iface, view);
+    TRACE("iface %p, shader_resource_view %p.\n", iface, shader_resource_view);
+
+    if (!view)
+        return;
 
     wined3d_mutex_lock();
-    wined3d_shader_resource_view_generate_mipmaps(srv->wined3d_view);
+    wined3d_device_generate_mips_view(device->wined3d_device, view->wined3d_view);
     wined3d_mutex_unlock();
 }
 
-- 
2.14.1
