From cc25929d3a152ee70f0f779091d82da88010f2ae Mon Sep 17 00:00:00 2001
From: Nikolay Sivov <nsivov@codeweavers.com>
Date: Mon, 3 Apr 2017 08:14:08 +0300
Subject: dwrite: Added missing max bitmap size initialization.

Signed-off-by: Nikolay Sivov <nsivov@codeweavers.com>
Signed-off-by: Alexandre Julliard <julliard@winehq.org>
---
 dlls/dwrite/font.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/dwrite/font.c b/dlls/dwrite/font.c
index e89d0a825a..4e35ceab2e 100644
--- a/dlls/dwrite/font.c
+++ b/dlls/dwrite/font.c
@@ -5203,6 +5203,7 @@ HRESULT create_glyphrunanalysis(const struct glyphrunanalysis_desc *desc, IDWrit
 
     analysis->flags = 0;
     analysis->bitmap = NULL;
+    analysis->max_glyph_bitmap_size = 0;
     analysis->ppdip = desc->ppdip;
     analysis->origin.x = desc->origin_x * desc->ppdip;
     analysis->origin.y = desc->origin_y * desc->ppdip;
-- 
2.11.0

