From ab3de1775fa8321218d5712b7562b8575a222a31 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sun, 17 Jan 2016 01:37:09 +0100
Subject: include/roapi.h: Add further typedefs.

---
 include/roapi.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/include/roapi.h b/include/roapi.h
index 0421fe9..f4154f8 100644
--- a/include/roapi.h
+++ b/include/roapi.h
@@ -1,5 +1,6 @@
 /*
  * Copyright (C) 2014 Martin Storsjo
+ * Copyright (C) 2016 Michael Müller
  *
  * This library is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -20,6 +21,7 @@
 #define __WINE_ROAPI_H
 
 #include <windef.h>
+#include <activation.h>
 
 typedef enum
 {
@@ -27,6 +29,11 @@ typedef enum
     RO_INIT_MULTITHREADED  = 1,
 } RO_INIT_TYPE;
 
+DECLARE_HANDLE(APARTMENT_SHUTDOWN_REGISTRATION_COOKIE);
+
+typedef struct {} *RO_REGISTRATION_COOKIE;
+typedef HRESULT (WINAPI *PFNGETACTIVATIONFACTORY)(HSTRING, IActivationFactory **);
+
 #ifdef __cplusplus
 extern "C" {
 #endif
-- 
2.6.4

