From 2ad226c2f25a9bd0130f5ae8393e719b37a36afc Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sun, 17 Jan 2016 02:00:04 +0100
Subject: include/objidl.idl: Add IApartmentShutdown interface.

---
 include/objidl.idl | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/include/objidl.idl b/include/objidl.idl
index c18442f..4388bab 100644
--- a/include/objidl.idl
+++ b/include/objidl.idl
@@ -2451,6 +2451,17 @@ interface IDummyHICONIncluder : IUnknown
     HRESULT Dummy([in] HICON hIcon, [in] HDC hdc);
 }
 
+[
+    object,
+    local,
+    pointer_default(unique),
+    uuid(a2f05a09-27a2-42b5-bc0e-ac163ef49d9b)
+]
+interface IApartmentShutdown : IUnknown
+{
+    void OnUninitialize([in] UINT64 identifier);
+}
+
 cpp_quote("#ifdef USE_COM_CONTEXT_DEF")
 
 typedef DWORD CPFLAGS;
-- 
2.6.4

