From fa3088d5c689f7403df2525394a8d6837681959a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 28 Nov 2015 23:29:49 +0100
Subject: services: Start SERVICE_FILE_SYSTEM_DRIVER services with winedevice.

---
 programs/services/services.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/programs/services/services.c b/programs/services/services.c
index 1335837..53a3846 100644
--- a/programs/services/services.c
+++ b/programs/services/services.c
@@ -631,7 +631,8 @@ static DWORD get_service_binary_path(const struct service_entry *service_entry,
 
     ExpandEnvironmentStringsW(service_entry->config.lpBinaryPathName, *path, size);
 
-    if (service_entry->config.dwServiceType == SERVICE_KERNEL_DRIVER)
+    if (service_entry->config.dwServiceType == SERVICE_KERNEL_DRIVER ||
+        service_entry->config.dwServiceType == SERVICE_FILE_SYSTEM_DRIVER)
     {
         static const WCHAR winedeviceW[] = {'\\','w','i','n','e','d','e','v','i','c','e','.','e','x','e',' ',0};
         WCHAR system_dir[MAX_PATH];
-- 
2.6.2

