From f79b3408935d1f4650c7369e0507c223c5f9ed33 Mon Sep 17 00:00:00 2001
From: David Torok <dt@zeroitlab.com>
Date: Tue, 19 Nov 2019 23:01:46 +0100
Subject: [PATCH] ntdll: Stub NtQueryInformationThread(ThreadHideFromDebugger).

---
 dlls/ntdll/thread.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/ntdll/thread.c b/dlls/ntdll/thread.c
index b913326f796..840b8ffd0b1 100644
--- a/dlls/ntdll/thread.c
+++ b/dlls/ntdll/thread.c
@@ -1460,6 +1460,11 @@ NTSTATUS WINAPI NtQueryInformationThread( HANDLE handle, THREADINFOCLASS class,
                 *ret_len = sizeof(*info) + desc_len;
         }
         return status;
+    case ThreadHideFromDebugger:
+        if (length != sizeof(char)) return STATUS_INFO_LENGTH_MISMATCH;
+        *(BOOLEAN *)data = TRUE;
+        if (ret_len) *ret_len = sizeof(BOOLEAN);
+        return STATUS_SUCCESS;
     case ThreadPriority:
     case ThreadBasePriority:
     case ThreadImpersonationToken:
-- 
2.24.0

