From 66039784d648883d899f33278fd8e31f946edb05 Mon Sep 17 00:00:00 2001
From: Andrew Wesie <awesie@gmail.com>
Date: Thu, 26 Jul 2018 11:34:32 -0500
Subject: [PATCH] ntdll: Stub for ProcessCookie in NtQueryInformationProcess.

---
 dlls/ntdll/process.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/dlls/ntdll/process.c b/dlls/ntdll/process.c
index b299287..4e08c62 100644
--- a/dlls/ntdll/process.c
+++ b/dlls/ntdll/process.c
@@ -572,6 +572,18 @@ NTSTATUS WINAPI NtQueryInformationProcess(
         else
             ret = STATUS_INFO_LENGTH_MISMATCH;
         break;
+    case ProcessCookie:
+        FIXME("(%p,info_class=%d,%p,0x%08x,%p) stub\n",
+              ProcessHandle,ProcessInformationClass,
+              ProcessInformation,ProcessInformationLength,
+              ReturnLength);
+
+        len = sizeof(ULONG);
+        if (ProcessInformationLength == len)
+            *(ULONG *)ProcessInformation = 0;
+        else
+            ret = STATUS_INFO_LENGTH_MISMATCH;
+        break;
     default:
         FIXME("(%p,info_class=%d,%p,0x%08x,%p) Unknown information class\n",
               ProcessHandle,ProcessInformationClass,
-- 
1.9.1

