From ea81f88b5641ff317dbc8c674a70e0c3be373f6f Mon Sep 17 00:00:00 2001
From: Bruno Jesus <00cpxxx@gmail.com>
Date: Sat, 21 Jan 2017 21:19:20 +0100
Subject: winmm: Fix the ordinal number of PlaySoundA.

---
 dlls/winmm/winmm.spec | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/winmm/winmm.spec b/dlls/winmm/winmm.spec
index a1eaf816a28..87908b3d021 100644
--- a/dlls/winmm/winmm.spec
+++ b/dlls/winmm/winmm.spec
@@ -1,5 +1,5 @@
 # ordinal exports
-1 stdcall @(ptr long long) PlaySoundA
+2 stdcall @(ptr long long) PlaySoundA
 3 stub @
 4 stub @
 
-- 
2.11.0

