From 91bd7aea87eea74498a3ea39fbc095cbdbbd1a6d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Tue, 3 Nov 2015 17:36:54 +0100
Subject: sfc_os: Set an error code in SfcGetNextProtectedFile stub.

---
 dlls/sfc_os/sfc_os.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/sfc_os/sfc_os.c b/dlls/sfc_os/sfc_os.c
index 9df38f8..f76b3a2 100644
--- a/dlls/sfc_os/sfc_os.c
+++ b/dlls/sfc_os/sfc_os.c
@@ -54,6 +54,8 @@ BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
 BOOL WINAPI SfcGetNextProtectedFile(HANDLE handle, PROTECTED_FILE_DATA *data)
 {
     FIXME("%p %p\n", handle, data);
+
+    SetLastError(ERROR_NO_MORE_FILES);
     return FALSE;
 }
 
-- 
2.6.2

