From 095360e2e7faec0135c0f7256c683cd28691f05a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 10 May 2015 02:23:46 +0200
Subject: ntdll: Use POSIX implementation to enumerate directory content.

---
 dlls/ntdll/directory.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/dlls/ntdll/directory.c b/dlls/ntdll/directory.c
index cbfc022..82244d5 100644
--- a/dlls/ntdll/directory.c
+++ b/dlls/ntdll/directory.c
@@ -23,6 +23,12 @@
 #include "config.h"
 #include "wine/port.h"
 
+/* The FreeBSD implementation to enumerate directory content is completely
+ * broken, which causes test failures in kernel32/file and ntdll/directory
+ * tests, and also causes bug 35397. Fallback to the POSIX implementation
+ * until this issue is fixed. */
+#undef HAVE_GETDIRENTRIES
+
 #include <assert.h>
 #include <sys/types.h>
 #ifdef HAVE_DIRENT_H
-- 
2.4.0

