From 5df655b72d55c2c4071fd7b84d2705f922bcc1a8 Mon Sep 17 00:00:00 2001
From: Vijay Kiran Kamuju <infyquest@gmail.com>
Date: Fri, 18 Sep 2020 17:31:59 +0200
Subject: [PATCH 11/23] xactengine2_10: Implement IXACTWaveBank::Play function.

Signed-off-by: Vijay Kiran Kamuju <infyquest@gmail.com>
---
 dlls/xactengine2_10/xact2_dll.c | 35 +++++++++++++++++++++++++++++++--
 1 file changed, 33 insertions(+), 2 deletions(-)

diff --git a/dlls/xactengine2_10/xact2_dll.c b/dlls/xactengine2_10/xact2_dll.c
index 7bb8134fd81..2e06341b5e9 100644
--- a/dlls/xactengine2_10/xact2_dll.c
+++ b/dlls/xactengine2_10/xact2_dll.c
@@ -580,9 +580,40 @@ static HRESULT WINAPI IXACTWaveBankImpl_Play(IXACTWaveBank *iface,
         XACTLOOPCOUNT nLoopCount, IXACTWave** ppWave)
 {
     XACTWaveBankImpl *This = impl_from_IXACTWaveBank(iface);
-    FIXME("(%p)->(0x%x, %u, 0x%x, %u, %p): stub!\n", This, nWaveIndex, dwFlags, dwPlayOffset,
+    XACTWaveImpl *wave;
+    FACTWave *fwave;
+    HRESULT hr;
+
+    TRACE("(%p)->(0x%x, %u, 0x%x, %u, %p)!\n", This, nWaveIndex, dwFlags, dwPlayOffset,
             nLoopCount, ppWave);
-    return E_NOTIMPL;
+
+    /* If the application doesn't want a handle, don't generate one at all.
+     * Let the engine handle that memory instead.
+     * -flibit
+     */
+    if (ppWave == NULL){
+        hr = FACTWaveBank_Play(This->fact_wavebank, nWaveIndex, dwFlags,
+                dwPlayOffset, nLoopCount, NULL);
+    }else{
+        hr = FACTWaveBank_Play(This->fact_wavebank, nWaveIndex, dwFlags,
+                dwPlayOffset, nLoopCount, &fwave);
+        if(FAILED(hr))
+            return hr;
+
+        wave = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(*wave));
+        if (!wave)
+        {
+            FACTWave_Destroy(fwave);
+            ERR("Failed to allocate XACT3WaveImpl!");
+            return E_OUTOFMEMORY;
+        }
+
+        wave->IXACTWave_iface.lpVtbl = &XACTWave_Vtbl;
+        wave->fact_wave = fwave;
+        *ppWave = &wave->IXACTWave_iface;
+    }
+
+    return hr;
 }
 
 static HRESULT WINAPI IXACTWaveBankImpl_Stop(IXACTWaveBank *iface,
-- 
2.28.0

