From d4751813371cb7f3b8c15261ff00f2d5afc9feb2 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Tue, 28 Apr 2015 16:35:44 +0200
Subject: wine.inf: Update timezone information.

---
 loader/wine.inf.in | 204 +++++++++++++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 198 insertions(+), 6 deletions(-)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 1e4de94..7d6be2f 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -2707,7 +2707,12 @@ HKLM,%CurrentVersionNT%\Time Zones\E. Europe Standard Time,"TZI",1,4c,ff,ff,ff,0
 HKLM,%CurrentVersionNT%\Time Zones\E. South America Standard Time,"Display",,"America/Sao_Paulo"
 HKLM,%CurrentVersionNT%\Time Zones\E. South America Standard Time,"Dlt",,"E. South America Daylight Time"
 HKLM,%CurrentVersionNT%\Time Zones\E. South America Standard Time,"Std",,"E. South America Standard Time"
-HKLM,%CurrentVersionNT%\Time Zones\E. South America Standard Time,"TZI",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,00,00,02,00,00,00,03,00,00,00,00,00,00,00,00,00,00,00,0a,00,00,00,02,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\E. South America Standard Time,"TZI",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,00,00,02,00,00,00,03,00,00,00,00,00,00,00,00,00,00,00,0a,00,00,00,03,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\E. South America Standard Time\Dynamic DST,"2015",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,df,07,02,00,00,00,16,00,00,00,00,00,00,00,00,00,df,07,0a,00,00,00,12,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\E. South America Standard Time\Dynamic DST,"2023",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,e7,07,02,00,00,00,1a,00,00,00,00,00,00,00,00,00,e7,07,0a,00,00,00,0f,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\E. South America Standard Time\Dynamic DST,"2026",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,ea,07,02,00,00,00,16,00,00,00,00,00,00,00,00,00,ea,07,0a,00,00,00,12,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\E. South America Standard Time\Dynamic DST,"2034",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,f2,07,02,00,00,00,1a,00,00,00,00,00,00,00,00,00,f2,07,0a,00,00,00,0f,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\E. South America Standard Time\Dynamic DST,"2037",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,f5,07,02,00,00,00,16,00,00,00,00,00,00,00,00,00,f5,07,0a,00,00,00,12,00,00,00,00,00,00,00,00,00
 HKLM,%CurrentVersionNT%\Time Zones\Eastern Standard Time,"Display",,"America/New_York"
 HKLM,%CurrentVersionNT%\Time Zones\Eastern Standard Time,"Dlt",,"Eastern Daylight Time"
 HKLM,%CurrentVersionNT%\Time Zones\Eastern Standard Time,"Std",,"Eastern Standard Time"
@@ -2723,7 +2728,87 @@ HKLM,%CurrentVersionNT%\Time Zones\Ekaterinburg Standard Time,"TZI",1,d4,fe,ff,f
 HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time,"Display",,"Pacific/Fiji"
 HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time,"Dlt",,"Fiji Daylight Time"
 HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time,"Std",,"Fiji Standard Time"
-HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time,"TZI",1,30,fd,ff,ff,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time,"TZI",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,00,00,01,00,00,00,03,00,03,00,00,00,00,00,00,00,00,00,0b,00,00,00,01,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2016",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,e0,07,01,00,00,00,18,00,03,00,00,00,00,00,00,00,e0,07,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2017",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,e1,07,01,00,00,00,16,00,03,00,00,00,00,00,00,00,e1,07,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2021",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,e5,07,01,00,00,00,18,00,03,00,00,00,00,00,00,00,e5,07,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2022",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,e6,07,01,00,00,00,17,00,03,00,00,00,00,00,00,00,e6,07,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2023",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,e7,07,01,00,00,00,16,00,03,00,00,00,00,00,00,00,e7,07,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2027",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,eb,07,01,00,00,00,18,00,03,00,00,00,00,00,00,00,eb,07,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2028",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,ec,07,01,00,00,00,17,00,03,00,00,00,00,00,00,00,ec,07,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2033",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,f1,07,01,00,00,00,17,00,03,00,00,00,00,00,00,00,f1,07,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2034",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,f2,07,01,00,00,00,16,00,03,00,00,00,00,00,00,00,f2,07,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2038",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,f6,07,01,00,00,00,18,00,03,00,00,00,00,00,00,00,f6,07,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2039",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,f7,07,01,00,00,00,17,00,03,00,00,00,00,00,00,00,f7,07,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2040",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,f8,07,01,00,00,00,16,00,03,00,00,00,00,00,00,00,f8,07,0b,00,00,00,04,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2044",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,fc,07,01,00,00,00,18,00,03,00,00,00,00,00,00,00,fc,07,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2045",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,fd,07,01,00,00,00,16,00,03,00,00,00,00,00,00,00,fd,07,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2049",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,01,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,01,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2050",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,02,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,02,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2051",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,03,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,03,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2055",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,07,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,07,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2056",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,08,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,08,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2061",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,0d,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,0d,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2062",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,0e,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,0e,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2066",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,12,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,12,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2067",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,13,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,13,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2068",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,14,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,14,08,0b,00,00,00,04,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2072",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,18,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,18,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2073",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,19,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,19,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2077",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,1d,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,1d,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2078",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,1e,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,1e,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2079",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,1f,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,1f,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2083",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,23,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,23,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2084",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,24,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,24,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2089",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,29,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,29,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2090",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,2a,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,2a,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2094",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,2e,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,2e,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2095",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,2f,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,2f,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2096",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,30,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,30,08,0b,00,00,00,04,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2100",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,34,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,34,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2101",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,35,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,35,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2102",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,36,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,36,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2106",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,3a,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,3a,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2107",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,3b,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,3b,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2108",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,3c,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,3c,08,0b,00,00,00,04,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2112",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,40,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,40,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2113",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,41,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,41,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2117",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,45,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,45,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2118",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,46,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,46,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2119",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,47,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,47,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2123",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,4b,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,4b,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2124",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,4c,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,4c,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2129",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,51,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,51,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2130",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,52,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,52,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2134",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,56,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,56,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2135",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,57,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,57,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2136",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,58,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,58,08,0b,00,00,00,04,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2140",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,5c,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,5c,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2141",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,5d,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,5d,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2145",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,61,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,61,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2146",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,62,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,62,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2147",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,63,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,63,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2151",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,67,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,67,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2152",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,68,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,68,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2157",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,6d,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,6d,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2158",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,6e,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,6e,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2162",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,72,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,72,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2163",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,73,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,73,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2164",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,74,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,74,08,0b,00,00,00,04,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2168",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,78,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,78,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2169",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,79,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,79,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2173",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,7d,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,7d,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2174",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,7e,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,7e,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2175",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,7f,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,7f,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2179",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,83,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,83,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2180",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,84,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,84,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2185",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,89,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,89,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2186",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,8a,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,8a,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2190",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,8e,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,8e,08,0b,00,00,00,07,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2191",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,8f,08,01,00,00,00,17,00,03,00,00,00,00,00,00,00,8f,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2192",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,90,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,90,08,0b,00,00,00,04,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2196",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,94,08,01,00,00,00,18,00,03,00,00,00,00,00,00,00,94,08,0b,00,00,00,06,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Fiji Standard Time\Dynamic DST,"2197",1,30,fd,ff,ff,00,00,00,00,c4,ff,ff,ff,95,08,01,00,00,00,16,00,03,00,00,00,00,00,00,00,95,08,0b,00,00,00,05,00,02,00,00,00,00,00,00,00
 HKLM,%CurrentVersionNT%\Time Zones\FLE Standard Time,"Display",,"Europe/Helsinki"
 HKLM,%CurrentVersionNT%\Time Zones\FLE Standard Time,"Dlt",,"FLE Daylight Time"
 HKLM,%CurrentVersionNT%\Time Zones\FLE Standard Time,"Std",,"FLE Standard Time"
@@ -2739,7 +2824,61 @@ HKLM,%CurrentVersionNT%\Time Zones\GMT Standard Time,"TZI",1,00,00,00,00,00,00,0
 HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time,"Display",,"America/Godthab"
 HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time,"Dlt",,"Greenland Daylight Time"
 HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time,"Std",,"Greenland Standard Time"
-HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time,"TZI",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,00,00,0a,00,06,00,04,00,17,00,00,00,00,00,00,00,00,00,03,00,06,00,05,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time,"TZI",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,00,00,0a,00,06,00,05,00,17,00,00,00,00,00,00,00,00,00,03,00,06,00,05,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2015",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,df,07,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,df,07,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2018",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,e2,07,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,e2,07,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2020",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,e4,07,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,e4,07,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2026",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,ea,07,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,ea,07,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2029",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,ed,07,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,ed,07,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2035",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,f3,07,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,f3,07,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2037",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,f5,07,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,f5,07,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2040",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,f8,07,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,f8,07,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2043",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,fb,07,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,fb,07,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2046",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,fe,07,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,fe,07,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2048",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,00,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,00,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2054",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,06,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,06,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2057",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,09,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,09,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2063",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,0f,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,0f,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2065",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,11,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,11,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2068",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,14,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,14,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2071",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,17,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,17,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2074",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,1a,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,1a,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2076",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,1c,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,1c,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2082",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,22,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,22,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2085",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,25,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,25,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2091",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,2b,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,2b,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2093",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,2d,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,2d,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2096",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,30,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,30,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2099",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,33,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,33,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2103",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,37,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,37,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2105",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,39,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,39,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2108",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,3c,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,3c,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2111",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,3f,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,3f,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2114",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,42,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,42,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2116",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,44,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,44,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2122",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,4a,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,4a,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2125",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,4d,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,4d,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2131",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,53,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,53,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2133",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,55,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,55,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2136",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,58,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,58,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2139",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,5b,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,5b,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2142",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,5e,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,5e,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2144",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,60,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,60,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2150",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,66,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,66,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2153",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,69,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,69,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2159",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,6f,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,6f,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2161",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,71,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,71,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2164",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,74,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,74,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2167",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,77,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,77,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2170",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,7a,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,7a,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2172",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,7c,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,7c,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2178",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,82,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,82,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2181",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,85,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,85,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2187",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,8b,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,8b,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2189",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,8d,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,8d,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2192",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,90,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,90,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2195",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,93,08,0a,00,06,00,18,00,17,00,00,00,00,00,00,00,93,08,03,00,06,00,1c,00,16,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Greenland Standard Time\Dynamic DST,"2198",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,96,08,0a,00,06,00,1b,00,17,00,00,00,00,00,00,00,96,08,03,00,06,00,18,00,16,00,00,00,00,00,00,00
 HKLM,%CurrentVersionNT%\Time Zones\Greenwich Standard Time,"Display",,"Greenwich"
 HKLM,%CurrentVersionNT%\Time Zones\Greenwich Standard Time,"Dlt",,"Greenwich Daylight Time"
 HKLM,%CurrentVersionNT%\Time Zones\Greenwich Standard Time,"Std",,"Greenwich Standard Time"
@@ -2763,11 +2902,64 @@ HKLM,%CurrentVersionNT%\Time Zones\Iran Standard Time,"TZI",1,2e,ff,ff,ff,00,00,
 HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time,"Display",,"Asia/Jerusalem"
 HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time,"Dlt",,"Israel Daylight Time"
 HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time,"Std",,"Israel Standard Time"
-HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time,"TZI",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,00,00,0a,00,00,00,05,00,02,00,00,00,00,00,00,00,00,00,03,00,05,00,05,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time,"TZI",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,00,00,0a,00,00,00,05,00,02,00,00,00,00,00,00,00,00,00,03,00,05,00,04,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2019",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,e3,07,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,e3,07,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2024",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,e8,07,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,e8,07,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2030",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,ee,07,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,ee,07,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2041",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,f9,07,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,f9,07,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2047",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,ff,07,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,ff,07,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2052",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,04,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,04,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2058",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,0a,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,0a,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2069",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,15,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,15,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2075",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,1b,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,1b,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2080",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,20,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,20,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2086",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,26,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,26,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2097",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,31,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,31,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2109",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,3d,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,3d,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2115",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,43,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,43,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2120",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,48,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,48,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2126",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,4e,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,4e,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2137",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,59,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,59,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2143",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,5f,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,5f,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2148",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,64,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,64,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2154",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,6a,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,6a,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2165",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,75,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,75,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2171",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,7b,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,7b,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2176",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,80,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,80,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2182",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,86,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,86,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2193",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,91,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,91,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Israel Standard Time\Dynamic DST,"2199",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,97,08,0a,00,00,00,1b,00,02,00,00,00,00,00,00,00,97,08,03,00,05,00,1d,00,02,00,00,00,00,00,00,00
 HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time,"Display",,"Asia/Amman"
 HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time,"Dlt",,"Jordan Daylight Time"
 HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time,"Std",,"Jordan Standard Time"
-HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time,"TZI",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,00,00,0a,00,05,00,05,00,01,00,00,00,00,00,00,00,00,00,03,00,04,00,05,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time,"TZI",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,00,00,0a,00,05,00,05,00,01,00,00,00,00,00,00,00,00,00,03,00,05,00,05,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2016",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,e0,07,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,e0,07,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2022",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,e6,07,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,e6,07,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2033",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,f1,07,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,f1,07,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2039",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,f7,07,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,f7,07,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2044",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,fc,07,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,fc,07,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2050",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,02,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,02,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2061",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,0d,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,0d,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2067",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,13,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,13,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2072",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,18,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,18,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2078",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,1e,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,1e,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2089",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,29,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,29,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2095",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,2f,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,2f,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2101",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,35,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,35,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2107",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,3b,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,3b,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2112",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,40,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,40,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2118",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,46,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,46,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2129",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,51,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,51,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2135",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,57,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,57,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2140",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,5c,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,5c,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2146",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,62,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,62,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2157",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,6d,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,6d,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2163",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,73,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,73,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2168",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,78,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,78,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2174",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,7e,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,7e,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2185",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,89,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,89,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2191",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,8f,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,8f,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\Jordan Standard Time\Dynamic DST,"2196",1,88,ff,ff,ff,00,00,00,00,c4,ff,ff,ff,94,08,0a,00,05,00,1c,00,01,00,00,00,00,00,00,00,94,08,04,00,05,00,01,00,00,00,00,00,00,00,00,00
 HKLM,%CurrentVersionNT%\Time Zones\Korea Standard Time,"Display",,"Asia/Seoul"
 HKLM,%CurrentVersionNT%\Time Zones\Korea Standard Time,"Dlt",,"Korea Daylight Time"
 HKLM,%CurrentVersionNT%\Time Zones\Korea Standard Time,"Std",,"Korea Standard Time"
@@ -2855,7 +3047,7 @@ HKLM,%CurrentVersionNT%\Time Zones\Russian Standard Time,"TZI",1,4c,ff,ff,ff,00,
 HKLM,%CurrentVersionNT%\Time Zones\SA Eastern Standard Time,"Display",,"America/Buenos_Aires"
 HKLM,%CurrentVersionNT%\Time Zones\SA Eastern Standard Time,"Dlt",,"SA Eastern Daylight Time"
 HKLM,%CurrentVersionNT%\Time Zones\SA Eastern Standard Time,"Std",,"SA Eastern Standard Time"
-HKLM,%CurrentVersionNT%\Time Zones\SA Eastern Standard Time,"TZI",1,b4,00,00,00,00,00,00,00,c4,ff,ff,ff,00,00,03,00,00,00,03,00,00,00,00,00,00,00,00,00,00,00,0a,00,00,00,01,00,00,00,00,00,00,00,00,00
+HKLM,%CurrentVersionNT%\Time Zones\SA Eastern Standard Time,"TZI",1,b4,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00
 HKLM,%CurrentVersionNT%\Time Zones\SA Pacific Standard Time,"Display",,"America/Bogota"
 HKLM,%CurrentVersionNT%\Time Zones\SA Pacific Standard Time,"Dlt",,"SA Pacific Daylight Time"
 HKLM,%CurrentVersionNT%\Time Zones\SA Pacific Standard Time,"Std",,"SA Pacific Standard Time"
-- 
2.3.5

