From 2e080477fe74a95265befb02dee6bd1c07d712cc Mon Sep 17 00:00:00 2001
From: Andrew Cook <ariscop@gmail.com>
Date: Sat, 4 Apr 2015 18:02:53 +1100
Subject: ntdll: Return STATUS_SUCCESS from SetInformationJobObject

Fixes #38351
---
 dlls/ntdll/sync.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ntdll/sync.c b/dlls/ntdll/sync.c
index 5d1c5f2..8566799 100644
--- a/dlls/ntdll/sync.c
+++ b/dlls/ntdll/sync.c
@@ -728,7 +728,7 @@ NTSTATUS WINAPI NtSetInformationJobObject( HANDLE handle, JOBOBJECTINFOCLASS cla
 
     default:
         FIXME( "stub: %p %u %p %u\n", handle, class, info, len );
-        return STATUS_NOT_IMPLEMENTED;
+        return STATUS_SUCCESS;
     }
 
     return status;
-- 
2.3.3

