From c3de2aeaf6101fb30a99bc8e7eaea614d97c4049 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 28 Feb 2015 06:34:24 +0100
Subject: kernel32/tests: Add tests for adding a terminated process to a job
 object.

---
 dlls/kernel32/tests/process.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/dlls/kernel32/tests/process.c b/dlls/kernel32/tests/process.c
index 466c231..a464db2 100644
--- a/dlls/kernel32/tests/process.c
+++ b/dlls/kernel32/tests/process.c
@@ -2260,6 +2260,23 @@ static void test_IsProcessInJob(void)
 
     CloseHandle(pi.hProcess);
     CloseHandle(pi.hThread);
+
+    /* Test adding an already terminated process to a job object */
+    create_process("exit", &pi);
+
+    dwret = WaitForSingleObject(pi.hProcess, 500);
+    ok(dwret == WAIT_OBJECT_0, "WaitForSingleObject returned %u\n", dwret);
+
+    SetLastError(0xdeadbeef);
+    ret = pAssignProcessToJobObject(job, pi.hProcess);
+    todo_wine
+    ok(!ret, "AssignProcessToJobObject unexpectedly succeeded\n");
+    todo_wine
+    expect_eq_d(ERROR_ACCESS_DENIED, GetLastError());
+
+    CloseHandle(pi.hProcess);
+    CloseHandle(pi.hThread);
+
     CloseHandle(job);
 }
 
-- 
2.3.0

