From e3ea811519e1c58b73d1208aaefd34f8069bbd90 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 14 Sep 2014 07:07:13 +0200
Subject: glu32: Fix wrapper function for glu[Begin|End]Polygon and gluNextContour.

---
 dlls/glu32/glu.c | 61 ++++++++++++++++++++++++++++++++++----------------------
 1 file changed, 37 insertions(+), 24 deletions(-)

diff --git a/dlls/glu32/glu.c b/dlls/glu32/glu.c
index 46a2336..16eb355 100644
--- a/dlls/glu32/glu.c
+++ b/dlls/glu32/glu.c
@@ -352,30 +352,6 @@ int WINAPI wine_gluNurbsCallback(void *arg0,int arg1,void *arg2) {
 }
 
 /***********************************************************************
- *		gluBeginPolygon (GLU32.@)
- */
-extern int gluBeginPolygon(void *arg0);
-int WINAPI wine_gluBeginPolygon(void *arg0) {
-	return gluBeginPolygon(arg0);
-}
-
-/***********************************************************************
- *		gluEndPolygon (GLU32.@)
- */
-extern int gluEndPolygon(void *arg0);
-int WINAPI wine_gluEndPolygon(void *arg0) {
-	return gluEndPolygon(arg0);
-}
-
-/***********************************************************************
- *		gluNextContour (GLU32.@)
- */
-extern int gluNextContour(void *arg0,int arg1);
-int WINAPI wine_gluNextContour(void *arg0,int arg1) {
-	return gluNextContour(arg0,arg1);
-}
-
-/***********************************************************************
  *		gluGetString (GLU32.@)
  */
 extern int gluGetString(int arg0);
@@ -617,3 +593,40 @@ void WINAPI wine_gluTessNormal(void *tess, double arg1, double arg2, double arg3
     wine_tess_t *wine_tess = tess;
     gluTessNormal(wine_tess->tess, arg1, arg2, arg3);
 }
+
+/***********************************************************************
+ *      gluBeginPolygon (GLU32.@)
+ */
+int WINAPI wine_gluBeginPolygon(void *tess)
+{
+    wine_tess_t *wine_tess = tess;
+    wine_tess->polygon_data = NULL;
+
+    gluTessBeginPolygon(wine_tess->tess, wine_tess);
+    gluTessBeginContour(wine_tess->tess);
+    return 0;
+}
+
+/***********************************************************************
+ *      gluEndPolygon (GLU32.@)
+ */
+int WINAPI wine_gluEndPolygon(void *tess)
+{
+    wine_tess_t *wine_tess = tess;
+
+    gluTessEndContour(wine_tess->tess);
+    gluTessEndPolygon(wine_tess->tess);
+    return 0;
+}
+
+/***********************************************************************
+ *      gluNextContour (GLU32.@)
+ */
+int WINAPI wine_gluNextContour(void *tess, int arg1)
+{
+    wine_tess_t *wine_tess = tess;
+
+    gluTessEndContour(wine_tess->tess);
+    gluTessBeginContour(wine_tess->tess);
+    return 0;
+}
-- 
2.1.0

