From b80174a0a4cb6ebd18b236681127f0c5da4e9073 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Mon, 24 Jul 2017 18:57:29 +0200
Subject: wined3d: Fix context backup window handling.

---
 dlls/wined3d/context.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/wined3d/context.c b/dlls/wined3d/context.c
index be7650003cc..036e8c261f8 100644
--- a/dlls/wined3d/context.c
+++ b/dlls/wined3d/context.c
@@ -1109,7 +1109,7 @@ static BOOL context_set_pixel_format(struct wined3d_context *context, HDC dc, BO
     if (dc == context->hdc && context->hdc_is_private && context->hdc_has_format)
         return TRUE;
 
-    if (dc == context->hdc && !context->hdc_is_private && WindowFromDC(dc) != context->win_handle)
+    if (dc == context->hdc && !private && WindowFromDC(dc) != context->win_handle)
         return FALSE;
 
     current = gl_info->gl_ops.wgl.p_wglGetPixelFormat(dc);
-- 
2.13.1

