From 7f6dd6cba12884b026a3aafce30173341d2f610b Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Mon, 7 Nov 2016 19:50:54 +0800
Subject: oleaut32: Add an arguments check to LoadTypeLibEx.

---
 dlls/oleaut32/typelib.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/oleaut32/typelib.c b/dlls/oleaut32/typelib.c
index e9f1507..b0ae6b6 100644
--- a/dlls/oleaut32/typelib.c
+++ b/dlls/oleaut32/typelib.c
@@ -482,6 +482,9 @@ HRESULT WINAPI LoadTypeLibEx(
 
     TRACE("(%s,%d,%p)\n",debugstr_w(szFile), regkind, pptLib);
 
+    if (!szFile || !pptLib)
+        return E_INVALIDARG;
+
     *pptLib = NULL;
 
     res = TLB_ReadTypeLib(szFile, szPath, MAX_PATH + 1, (ITypeLib2**)pptLib);
-- 
2.9.0

