From 87053e0d99e5e8691b571729e66f9c3ff8272569 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 16 Apr 2015 21:27:22 +0200
Subject: wbemprox: Allow connecting to localhost.

---
 dlls/wbemprox/wbemlocator.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/wbemprox/wbemlocator.c b/dlls/wbemprox/wbemlocator.c
index 463e6d4..b50d7cb 100644
--- a/dlls/wbemprox/wbemlocator.c
+++ b/dlls/wbemprox/wbemlocator.c
@@ -89,10 +89,11 @@ static HRESULT WINAPI wbem_locator_QueryInterface(
 static BOOL is_local_machine( const WCHAR *server )
 {
     static const WCHAR dotW[] = {'.',0};
+    static const WCHAR localhostW[] = {'l','o','c','a','l','h','o','s','t',0};
     WCHAR buffer[MAX_COMPUTERNAME_LENGTH + 1];
     DWORD len = sizeof(buffer) / sizeof(buffer[0]);
 
-    if (!server || !strcmpW( server, dotW )) return TRUE;
+    if (!server || !strcmpW( server, dotW ) || !strcmpW( server, localhostW )) return TRUE;
     if (GetComputerNameW( buffer, &len ) && !strcmpiW( server, buffer )) return TRUE;
     return FALSE;
 }
-- 
2.3.5

