From 0563e84a0e506bcbb6cf989d57bf74264cfb5ed7 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 2 Apr 2016 02:50:59 +0200
Subject: wintrust/tests: Add some additional tests.

---
 dlls/wintrust/tests/softpub.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/dlls/wintrust/tests/softpub.c b/dlls/wintrust/tests/softpub.c
index 2b51c30..a2fa764 100644
--- a/dlls/wintrust/tests/softpub.c
+++ b/dlls/wintrust/tests/softpub.c
@@ -1154,6 +1154,16 @@ static void test_wintrust_digest(void)
             { CERT_E_CHAINING, TRUE }, { S_OK, FALSE }
         },
         {
+            {{ SelfSignedFile32, sizeof(SelfSignedFile32) },
+             { Dummy, sizeof(Dummy) }},
+            { TRUST_E_NOSIGNATURE, TRUE }, { TRUST_E_NOSIGNATURE, TRUE }
+        },
+        {
+            {{ Dummy, sizeof(Dummy) },
+             { SelfSignedFile32 + sizeof(Dummy), sizeof(SelfSignedFile32) - sizeof(Dummy) }},
+            { TRUST_E_SUBJECT_FORM_UNKNOWN, FALSE }, { TRUST_E_NOSIGNATURE, TRUE }
+        },
+        {
             {{ SelfSignedFile32, 19 },
              { Dummy, sizeof(Dummy) },
              { SelfSignedFile32 + 19 + sizeof(Dummy), sizeof(SelfSignedFile32) - 19 - sizeof(Dummy) }},
@@ -1174,6 +1184,16 @@ static void test_wintrust_digest(void)
             { CERT_E_CHAINING, TRUE }, { S_OK, FALSE }
         },
         {
+            {{ SelfSignedFile64, sizeof(SelfSignedFile64) },
+             { Dummy, sizeof(Dummy) }},
+            { TRUST_E_NOSIGNATURE, TRUE }, { TRUST_E_NOSIGNATURE, TRUE }
+        },
+        {
+            {{ Dummy, sizeof(Dummy) },
+             { SelfSignedFile64 + sizeof(Dummy), sizeof(SelfSignedFile64) - sizeof(Dummy) }},
+            { TRUST_E_SUBJECT_FORM_UNKNOWN, FALSE }, { TRUST_E_NOSIGNATURE, TRUE }
+        },
+        {
             {{ SelfSignedFile64, 19 },
              { Dummy, sizeof(Dummy) },
              { SelfSignedFile64 + 19 + sizeof(Dummy), sizeof(SelfSignedFile64) - 19 - sizeof(Dummy) }},
-- 
2.7.1

