From d67655d6733fdf62a10f330d9292ced0aa99f809 Mon Sep 17 00:00:00 2001
From: Kimmo Myllyvirta <kimmo.myllyvirta@gmail.com>
Date: Fri, 7 Jul 2017 11:21:32 +0300
Subject: [PATCH] d3d11: Implement HSSetSamplers for deferred contexts.

---
 dlls/d3d11/device.c | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/dlls/d3d11/device.c b/dlls/d3d11/device.c
index aabbe38..2d03464 100644
--- a/dlls/d3d11/device.c
+++ b/dlls/d3d11/device.c
@@ -57,6 +57,7 @@ enum deferred_cmd
     DEFERRED_CSSETSAMPLERS,             /* samplers_info */
     DEFERRED_DSSETSAMPLERS,             /* samplers_info */
     DEFERRED_GSSETSAMPLERS,             /* samplers_info */
+    DEFERRED_HSSETSAMPLERS,             /* samplers_info */
     DEFERRED_PSSETSAMPLERS,             /* samplers_info */
 
     DEFERRED_CSSETCONSTANTBUFFERS,      /* constant_buffers_info */
@@ -296,8 +297,8 @@ static void add_deferred_set_shader_resources(struct d3d11_deferred_context *con
     }
 }
 
-/* for DEFERRED_CSSETSAMPLERS, DEFERRED_DSSETSAMPLERS, DEFERRED_GSSETSAMPLERS and
- * DEFERRED_PSSETSAMPLERS */
+/* for DEFERRED_CSSETSAMPLERS, DEFERRED_DSSETSAMPLERS, DEFERRED_GSSETSAMPLERS,
+ * DEFERRED_HSSETSAMPLERS and DEFERRED_PSSETSAMPLERS */
 static void add_deferred_set_samplers(struct d3d11_deferred_context *context, enum deferred_cmd cmd,
         UINT start_slot, UINT sampler_count, ID3D11SamplerState *const *samplers)
 {
@@ -460,6 +461,7 @@ static void free_deferred_calls(struct list *commands)
             case DEFERRED_CSSETSAMPLERS:
             case DEFERRED_DSSETSAMPLERS:
             case DEFERRED_GSSETSAMPLERS:
+            case DEFERRED_HSSETSAMPLERS:
             case DEFERRED_PSSETSAMPLERS:
             {
                 for (i = 0; i < call->samplers_info.num_samplers; i++)
@@ -679,6 +681,12 @@ static void exec_deferred_calls(ID3D11DeviceContext1 *iface, struct list *comman
                         call->samplers_info.num_samplers, call->samplers_info.samplers);
                 break;
             }
+            case DEFERRED_HSSETSAMPLERS:
+            {
+                ID3D11DeviceContext1_HSSetSamplers(iface, call->samplers_info.start_slot,
+                        call->samplers_info.num_samplers, call->samplers_info.samplers);
+                break;
+            }
             case DEFERRED_PSSETSAMPLERS:
             {
                 ID3D11DeviceContext1_PSSetSamplers(iface, call->samplers_info.start_slot,
@@ -4599,8 +4607,12 @@ static void STDMETHODCALLTYPE d3d11_deferred_context_HSSetShader(ID3D11DeviceCon
 static void STDMETHODCALLTYPE d3d11_deferred_context_HSSetSamplers(ID3D11DeviceContext *iface,
         UINT start_slot, UINT sampler_count, ID3D11SamplerState *const *samplers)
 {
-    FIXME("iface %p, start_slot %u, sampler_count %u, samplers %p stub!\n",
+    struct d3d11_deferred_context *context = impl_from_deferred_ID3D11DeviceContext(iface);
+
+    TRACE("iface %p, start_slot %u, sampler_count %u, samplers %p.\n",
             iface, start_slot, sampler_count, samplers);
+
+    add_deferred_set_samplers(context, DEFERRED_HSSETSAMPLERS, start_slot, sampler_count, samplers);
 }
 
 static void STDMETHODCALLTYPE d3d11_deferred_context_HSSetConstantBuffers(ID3D11DeviceContext *iface,
-- 
1.9.1

