From f35990fe3eaccfbccf458b79f814018e126a46c5 Mon Sep 17 00:00:00 2001
From: Johannes Specht <jojos_band@gmx.net>
Date: Mon, 4 Sep 2017 04:22:04 +0200
Subject: d3d11: Implement DispatchIndirect for deferred contexts.

---
 dlls/d3d11/device.c | 32 +++++++++++++++++++++++++++++++-
 1 file changed, 31 insertions(+), 1 deletion(-)

diff --git a/dlls/d3d11/device.c b/dlls/d3d11/device.c
index e0a65d5fc60..5a4fd13d611 100644
--- a/dlls/d3d11/device.c
+++ b/dlls/d3d11/device.c
@@ -90,6 +90,7 @@ enum deferred_cmd
 
     DEFERRED_MAP,                       /* map_info */
     DEFERRED_DISPATCH,                  /* dispatch_info */
+    DEFERRED_DISPATCHINDIRECT,          /* dispatch_indirect_info */
 
     DEFERRED_CLEARSTATE,
     DEFERRED_CLEARRENDERTARGETVIEW,     /* clear_rtv_info */
@@ -313,6 +314,11 @@ struct deferred_call
             UINT count_z;
         } dispatch_info;
         struct
+        {
+            ID3D11Buffer *buffer;
+            UINT offset;
+        } dispatch_indirect_info;
+        struct
         {
             ID3D11ShaderResourceView *view;
         } generate_mips_info;
@@ -704,6 +710,12 @@ static void free_deferred_calls(struct list *commands)
             {
                 break; /* nothing to do */
             }
+            case DEFERRED_DISPATCHINDIRECT:
+            {
+                if (call->dispatch_indirect_info.buffer)
+                    ID3D11Buffer_Release(call->dispatch_indirect_info.buffer);
+                break;
+            }
             case DEFERRED_CLEARSTATE:
             {
                 break; /* nothing to do */
@@ -1102,6 +1114,13 @@ static void exec_deferred_calls(ID3D11DeviceContext *iface, struct list *command
                         call->dispatch_info.count_y, call->dispatch_info.count_z);
                 break;
             }
+            case DEFERRED_DISPATCHINDIRECT:
+            {
+                ID3D11DeviceContext_DispatchIndirect(iface,
+                        call->dispatch_indirect_info.buffer,
+                        call->dispatch_indirect_info.offset);
+                break;
+            }
             case DEFERRED_CLEARSTATE:
             {
                 ID3D11DeviceContext_ClearState(iface);
@@ -4700,7 +4719,18 @@ static void STDMETHODCALLTYPE d3d11_deferred_context_Dispatch(ID3D11DeviceContex
 static void STDMETHODCALLTYPE d3d11_deferred_context_DispatchIndirect(ID3D11DeviceContext *iface,
         ID3D11Buffer *buffer, UINT offset)
 {
-    FIXME("iface %p, buffer %p, offset %u stub!\n", iface, buffer, offset);
+    struct d3d11_deferred_context *context = impl_from_deferred_ID3D11DeviceContext(iface);
+    struct deferred_call *call;
+
+    TRACE("iface %p, buffer %p, offset %u.\n", iface, buffer, offset);
+
+    if (!(call = add_deferred_call(context, 0)))
+        return;
+
+    call->cmd = DEFERRED_DISPATCHINDIRECT;
+    if (buffer) ID3D11Buffer_AddRef(buffer);
+    call->dispatch_indirect_info.buffer = buffer;
+    call->dispatch_indirect_info.offset = offset;
 }
 
 static void STDMETHODCALLTYPE d3d11_deferred_context_RSSetState(ID3D11DeviceContext *iface,
-- 
2.14.1

