From 9d18943140ce1ee26d238bf257a8b52cc6332e8e Mon Sep 17 00:00:00 2001
From: Christian Costa <titan.costa@gmail.com>
Date: Fri, 19 Feb 2016 10:01:14 +0100
Subject: quartz: Don't print FIXME for IAMFilterMiscFlags in parsers.

---
 dlls/quartz/parser.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/dlls/quartz/parser.c b/dlls/quartz/parser.c
index fd98787..fcc88a7 100644
--- a/dlls/quartz/parser.c
+++ b/dlls/quartz/parser.c
@@ -155,8 +155,12 @@ HRESULT WINAPI Parser_QueryInterface(IBaseFilter * iface, REFIID riid, LPVOID *
         return S_OK;
     }
 
-    if (!IsEqualIID(riid, &IID_IPin) && !IsEqualIID(riid, &IID_IVideoWindow))
+    if (!IsEqualIID(riid, &IID_IPin) &&
+        !IsEqualIID(riid, &IID_IVideoWindow) &&
+        !IsEqualIID(riid, &IID_IAMFilterMiscFlags))
+    {
         FIXME("No interface for %s!\n", qzdebugstr_guid(riid));
+    }
 
     return E_NOINTERFACE;
 }
-- 
2.7.0

