From fb7b3dde4dc3687643cbf9d66c0e0e3da27d1ad2 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 8 May 2015 18:02:20 +0200
Subject: server: Fix leak of async handle in pipe_server_flush.

---
 server/named_pipe.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/server/named_pipe.c b/server/named_pipe.c
index c698a87..f5e536c 100644
--- a/server/named_pipe.c
+++ b/server/named_pipe.c
@@ -563,6 +563,7 @@ static obj_handle_t pipe_server_flush( struct fd *fd, const async_data_t *async_
         if (!server->flush_poll)
             server->flush_poll = add_timeout_user( -TICKS_PER_SEC / 10, check_flushed, server );
         if (blocking) handle = alloc_handle( current->process, async, SYNCHRONIZE, 0 );
+        release_object( async );
         set_error( STATUS_PENDING );
     }
     return handle;
-- 
2.4.0

