From ddf59a2c4bed6165d79c2cc7655c8151d99e7965 Mon Sep 17 00:00:00 2001
From: Christian Costa <titan.costa@gmail.com>
Date: Sun, 14 Feb 2016 19:32:45 +0100
Subject: wined3d: Display FIXME for cmp function 0 only once.

---
 dlls/wined3d/state.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/dlls/wined3d/state.c b/dlls/wined3d/state.c
index 21e49eb..8f7c9bd 100644
--- a/dlls/wined3d/state.c
+++ b/dlls/wined3d/state.c
@@ -241,8 +241,11 @@ GLenum wined3d_gl_compare_func(enum wined3d_cmp_func f)
         case WINED3D_CMP_ALWAYS:
             return GL_ALWAYS;
         default:
-            FIXME("Unrecognized compare function %#x.\n", f);
+        {
+            static int once;
+            if (f || !once++) FIXME("Unrecognized compare function %#x.\n", f);
             return GL_NONE;
+        }
     }
 }
 
-- 
2.7.1

