From d6feff32b27e0661406b4f24ec7932f919a9628a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 30 Oct 2015 23:43:02 +0100
Subject: winegcc: Pass '-read_only_relocs suppress' to the linker on OSX.

---
 tools/winegcc/winegcc.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/tools/winegcc/winegcc.c b/tools/winegcc/winegcc.c
index ad139f5..220784b 100644
--- a/tools/winegcc/winegcc.c
+++ b/tools/winegcc/winegcc.c
@@ -377,7 +377,12 @@ static const strarray* get_lddllflags( const struct options *opts, const strarra
         strarray_add( flags, "-bundle" );
         strarray_add( flags, "-multiply_defined" );
         strarray_add( flags, "suppress" );
-        if (opts->target_cpu == CPU_POWERPC)
+        if (opts->target_cpu == CPU_x86)
+        {
+            strarray_add( flags, "-read_only_relocs" );
+            strarray_add( flags, "suppress" );
+        }
+        else if (opts->target_cpu == CPU_POWERPC)
         {
             strarray_add( flags, "-read_only_relocs" );
             strarray_add( flags, "warning" );
-- 
2.6.1

