From a3e9e92ee5c436b3704b505ae4ea4cc076526b0e Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 31 Oct 2015 02:41:47 +0100
Subject: ntdll: Use close_handle instead of NtClose for internal memory
 management functions.

---
 dlls/ntdll/virtual.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/ntdll/virtual.c b/dlls/ntdll/virtual.c
index f30d94a..54c4cf4 100644
--- a/dlls/ntdll/virtual.c
+++ b/dlls/ntdll/virtual.c
@@ -1279,7 +1279,7 @@ static NTSTATUS map_image( HANDLE hmapping, int fd, char *base, SIZE_T total_siz
  error:
     if (view) delete_view( view );
     server_leave_uninterrupted_section( &csVirtual, &sigset );
-    if (dup_mapping) NtClose( dup_mapping );
+    if (dup_mapping) close_handle( dup_mapping );
     return status;
 }
 
@@ -2663,7 +2663,7 @@ NTSTATUS WINAPI SYSCALL(NtMapViewOfSection)( HANDLE handle, HANDLE process, PVOI
             res = map_image( handle, unix_handle, base, size, mask, header_size,
                              shared_fd, dup_mapping, map_vprot, addr_ptr );
             if (shared_needs_close) close( shared_fd );
-            NtClose( shared_file );
+            close_handle( shared_file );
         }
         else
         {
@@ -2731,7 +2731,7 @@ NTSTATUS WINAPI SYSCALL(NtMapViewOfSection)( HANDLE handle, HANDLE process, PVOI
     server_leave_uninterrupted_section( &csVirtual, &sigset );
 
 done:
-    if (dup_mapping) NtClose( dup_mapping );
+    if (dup_mapping) close_handle( dup_mapping );
     if (needs_close) close( unix_handle );
     return res;
 }
-- 
2.6.1

