From cfe794189edc600ca60308be5ba9e4a965c5112a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 16 Oct 2016 02:01:16 +0200
Subject: server: Map MAXIMUM_ALLOWED even if skipping access check.

---
 server/handle.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/server/handle.c b/server/handle.c
index 6b68ff4..b4114ef 100644
--- a/server/handle.c
+++ b/server/handle.c
@@ -278,6 +278,7 @@ obj_handle_t alloc_handle_no_access_check( struct process *process, void *ptr, u
 {
     struct object *obj = ptr;
     access = obj->ops->map_access( obj, access ) & ~RESERVED_ALL;
+    if ((access & MAXIMUM_ALLOWED) && !check_object_access( obj, &access )) return 0;
     return alloc_handle_entry( process, ptr, access, attr );
 }
 
-- 
2.9.0

