From 168d4c1da4c5b7ce71eb2d07bb3369f2c0ad5df7 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 5 Nov 2015 17:56:24 +0100
Subject: include: Remove check for __WINE_WNE_PORT_H in winsock2.h.

---
 include/winsock2.h | 2 --
 1 file changed, 2 deletions(-)

diff --git a/include/winsock2.h b/include/winsock2.h
index 461b90c..212352c 100644
--- a/include/winsock2.h
+++ b/include/winsock2.h
@@ -113,7 +113,6 @@ extern "C" {
 #define SD_BOTH                    0x02
 
 /* Constants for WSAPoll() */
-#ifndef __WINE_WINE_PORT_H
 #ifndef USE_WS_PREFIX
 #define POLLERR                    0x0001
 #define POLLHUP                    0x0002
@@ -137,7 +136,6 @@ extern "C" {
 #define WS_POLLIN                  (WS_POLLRDNORM|WS_POLLRDBAND)
 #define WS_POLLOUT                 (WS_POLLWRNORM)
 #endif
-#endif
 
 /* Constants for WSAIoctl() */
 #ifdef USE_WS_PREFIX
-- 
2.6.2

