From e921966b9073171d4a76ef6a96ba93413023e029 Mon Sep 17 00:00:00 2001
From: Matteo Bruni <mbruni@codeweavers.com>
Date: Thu, 7 May 2015 16:07:53 +0200
Subject: wined3d: Dirtify vertex shader on transformed <-> untransformed
 transition.

---
 dlls/wined3d/glsl_shader.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/wined3d/glsl_shader.c b/dlls/wined3d/glsl_shader.c
index 0a130e1..8e618e5 100644
--- a/dlls/wined3d/glsl_shader.c
+++ b/dlls/wined3d/glsl_shader.c
@@ -7682,6 +7682,10 @@ static void glsl_vertex_pipe_vdecl(struct wined3d_context *context,
 
     context->last_was_rhw = transformed;
 
+    /* Transformed position forces FFP draws. */
+    if (transformed != wasrhw)
+        context->shader_update_mask |= 1 << WINED3D_SHADER_TYPE_VERTEX;
+
     if (!use_vs(state))
     {
         if (context->last_was_vshader)
-- 
2.4.0

