From 4bcc72aa747a76c9300fbbd59a3c629206a7a554 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 1 Mar 2015 05:05:04 +0100
Subject: msctf: Always use interlocked functions when accessing
 MSCTF_refCount.

---
 dlls/msctf/msctf.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/msctf/msctf.c b/dlls/msctf/msctf.c
index 78992f7..231246a 100644
--- a/dlls/msctf/msctf.c
+++ b/dlls/msctf/msctf.c
@@ -110,7 +110,7 @@ static void ClassFactory_Destructor(ClassFactory *This)
 {
     TRACE("Destroying class factory %p\n", This);
     HeapFree(GetProcessHeap(),0,This);
-    MSCTF_refCount--;
+    InterlockedDecrement(&MSCTF_refCount);
 }
 
 static HRESULT WINAPI ClassFactory_QueryInterface(IClassFactory *iface, REFIID riid, LPVOID *ppvOut)
@@ -190,7 +190,7 @@ static HRESULT ClassFactory_Constructor(LPFNCONSTRUCTOR ctor, LPVOID *ppvOut)
     This->ctor = ctor;
     *ppvOut = This;
     TRACE("Created class factory %p\n", This);
-    MSCTF_refCount++;
+    InterlockedIncrement(&MSCTF_refCount);
     return S_OK;
 }
 
-- 
2.3.0

