From ec7c2cf03581d8b1c43bc87cba8f190165a3cc02 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Alex=20Villac=C3=ADs=20Lasso?= <a_villacis@palosanto.com>
Date: Sun, 31 May 2009 22:53:00 -0500
Subject: user.exe16: Fix handling of DDL_DRIVES flag in DlgDirList.

---
 dlls/user.exe16/dialog.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/user.exe16/dialog.c b/dlls/user.exe16/dialog.c
index fead42d..ce17ed8 100644
--- a/dlls/user.exe16/dialog.c
+++ b/dlls/user.exe16/dialog.c
@@ -636,7 +636,7 @@ INT16 WINAPI DlgDirList16( HWND16 hDlg, LPSTR spec, INT16 idLBox,
      * be set automatically (this is different in Win32, and
      * DIALOG_DlgDirList sends Win32 messages to the control,
      * so do it here) */
-    if (attrib & DDL_DRIVES) attrib |= DDL_EXCLUSIVE;
+    if (attrib == DDL_DRIVES) attrib |= DDL_EXCLUSIVE;
     return DlgDirListA( WIN_Handle32(hDlg), spec, idLBox, idStatic, attrib );
 }
 
-- 
2.7.1

