From 56d7b4030df9de1de6fb35ec4f31dad2661f46f9 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Tue, 20 Sep 2016 17:17:42 +0800
Subject: windowscodecs: Use V_UI1() instead of V_UNION() to assign a VT_UI1
 variant member.

---
 dlls/windowscodecs/tiffformat.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/windowscodecs/tiffformat.c b/dlls/windowscodecs/tiffformat.c
index d94be78..1f757d8 100644
--- a/dlls/windowscodecs/tiffformat.c
+++ b/dlls/windowscodecs/tiffformat.c
@@ -1971,7 +1971,7 @@ static HRESULT WINAPI TiffEncoder_CreateNewFrame(IWICBitmapEncoder *iface,
             VARIANT v;
             VariantInit(&v);
             V_VT(&v) = VT_UI1;
-            V_UNION(&v, bVal) = WICTiffCompressionDontCare;
+            V_UI1(&v) = WICTiffCompressionDontCare;
             hr = IPropertyBag2_Write(*ppIEncoderOptions, 1, opts, &v);
             VariantClear(&v);
             if (FAILED(hr))
-- 
2.9.0

