From be07df750862699f2515c0ac0ceb7a6c25e9458a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Author=20Name?= <author@email.com>
Subject: [PATCH v3] component: Replace arg1 with arg2.

Signed-off-by: =?UTF-8?q?Author=20Name?= <author@email.com>
Signed-off-by: Other Developer <other@email.com>
---
 test.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/test.txt b/test.txt
index d54375d..0078e66 100644
--- a/test.txt
+++ b/test.txt
@@ -1,7 +1,7 @@
 line1();
 line2();
 line3();
-function(arg1);
+function(arg2);
 line5();
 line6();
 line7();
-- 
2.7.1
