From d96ffdba8bf208acda52a30d969e496414b5fb30 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 20 Jan 2016 21:42:35 +0100
Subject: server: Fix invalid memory accesss caused by destroying winstation
 before desktop.

---
 server/winstation.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/server/winstation.c b/server/winstation.c
index 7a6b9fc..ad28476f 100644
--- a/server/winstation.c
+++ b/server/winstation.c
@@ -257,6 +257,7 @@ static void desktop_destroy( struct object *obj )
     if (desktop->global_hooks) release_object( desktop->global_hooks );
     if (desktop->close_timeout) remove_timeout_user( desktop->close_timeout );
     list_remove( &desktop->entry );
+    unlink_named_object( obj );
     release_object( desktop->winstation );
 }
 
-- 
2.6.4

