From 7148e6c97c72c999f56f811d2501b7bfc90fac5a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Tue, 11 Aug 2015 03:42:07 +0200
Subject: imagehlp: Remove unused structure.

---
 dlls/imagehlp/access.c | 17 -----------------
 1 file changed, 17 deletions(-)

diff --git a/dlls/imagehlp/access.c b/dlls/imagehlp/access.c
index 9e1f187..6a33c0c 100644
--- a/dlls/imagehlp/access.c
+++ b/dlls/imagehlp/access.c
@@ -36,23 +36,6 @@ WINE_DEFAULT_DEBUG_CHANNEL(imagehlp);
 
 static PLOADED_IMAGE IMAGEHLP_pFirstLoadedImage=NULL;
 
-static LOADED_IMAGE IMAGEHLP_EmptyLoadedImage = {
-  NULL,       /* ModuleName */
-  0,          /* hFile */
-  NULL,       /* MappedAddress */
-  NULL,       /* FileHeader */
-  NULL,       /* LastRvaSection */
-  0,          /* NumberOfSections */
-  NULL,       /* Sections */
-  1,          /* Characteristics */
-  FALSE,      /* fSystemImage */
-  FALSE,      /* fDOSImage */
-  FALSE,      /* fReadOnly */
-  0,          /* Version */
-  { &IMAGEHLP_EmptyLoadedImage.Links, &IMAGEHLP_EmptyLoadedImage.Links }, /* Links */
-  148,        /* SizeOfImage; */
-};
-
 DECLSPEC_HIDDEN extern HANDLE IMAGEHLP_hHeap;
 
 /***********************************************************************
-- 
2.5.0

