From 5748aa15dd5e619714e9a04a033b30bd1fdae7d7 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 29 Jan 2015 00:18:47 +0100
Subject: vcomp: Add a warning that multithreading is not yet supported.

---
 dlls/vcomp/main.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/vcomp/main.c b/dlls/vcomp/main.c
index e37cf16..1dd57dd 100644
--- a/dlls/vcomp/main.c
+++ b/dlls/vcomp/main.c
@@ -126,6 +126,8 @@ BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
                 ERR("Failed to allocate TLS index\n");
                 return FALSE;
             }
+
+            FIXME("Builtin vcomp doesn't support multithreading, use native library for better performance.\n");
             break;
         case DLL_PROCESS_DETACH:
             TlsFree(vcomp_context_tls);
-- 
2.2.1

