From 507cba6fbcec06846ac90e62a76aaa61117f21fd Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 7 Sep 2014 01:46:52 +0200
Subject: ntdll: Use call_dll_entry_point to execute TLS callbacks.

---
 dlls/ntdll/loader.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index dfe7516..bd160d9 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -1030,7 +1030,7 @@ static void call_tls_callbacks( HMODULE module, UINT reason )
                     GetCurrentThreadId(), *callback, module, reason_names[reason] );
         __TRY
         {
-            (*callback)( module, reason, NULL );
+            call_dll_entry_point( (DLLENTRYPROC)(*callback), module, reason, NULL );
         }
         __EXCEPT_ALL
         {
-- 
2.1.0

