From ae3350ac731b45fd0b47d50f0e64525ec48803f9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Mon, 21 Dec 2015 01:46:34 +0100
Subject: [PATCH 3/7] wusa: Treat empty update list as error.

---
 programs/wusa/main.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/programs/wusa/main.c b/programs/wusa/main.c
index 0d6e5c345c..837729554f 100644
--- a/programs/wusa/main.c
+++ b/programs/wusa/main.c
@@ -997,6 +997,12 @@ static BOOL install_msu(WCHAR *filename, struct installer_state *state)
         }
     }
 
+    if (list_empty(&state->updates))
+    {
+        WINE_ERR("No updates found, probably incompatible MSU file format?\n");
+        goto done;
+    }
+
     /* perform dryrun */
     set_assembly_status(&state->assemblies, ASSEMBLY_STATUS_NONE);
     if (!install_updates(state, TRUE))
-- 
2.23.0.rc1

