From e489fac14e207ba0973c145ede43587457cc0be1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Mon, 21 Dec 2015 01:46:34 +0100
Subject: wusa: Treat empty update list as error.

---
 programs/wusa/main.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/programs/wusa/main.c b/programs/wusa/main.c
index 7c1dfef..9cd8a2d 100644
--- a/programs/wusa/main.c
+++ b/programs/wusa/main.c
@@ -998,6 +998,12 @@ static BOOL install_msu(WCHAR *filename, struct installer_state *state)
         }
     }
 
+    if (list_empty(&state->updates))
+    {
+        WINE_ERR("No updates found, probably incompatible MSU file format?\n");
+        goto done;
+    }
+
     /* perform dryrun */
     set_assembly_status(&state->assemblies, ASSEMBLY_STATUS_NONE);
     if (!install_updates(state, TRUE))
-- 
2.6.4

