From b1dbdbafa126589fdaf8f92eb0706eab0ff85f56 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Mon, 21 Dec 2015 04:09:11 +0100
Subject: wusa: Print warning when encountering msdelta compressed files.

---
 programs/wusa/main.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/programs/wusa/main.c b/programs/wusa/main.c
index 6a893ec..299aff2 100644
--- a/programs/wusa/main.c
+++ b/programs/wusa/main.c
@@ -892,6 +892,7 @@ static void installer_cleanup(struct installer_state *state)
 
 static BOOL load_assemblies_from_cab(const WCHAR *filename, struct installer_state *state)
 {
+    static const WCHAR manifest_cix_xmlW[] = {'_','m','a','n','i','f','e','s','t','_','.','c','i','x','.','x','m','l',0};
     static const WCHAR manifestW[] = {'.','m','a','n','i','f','e','s','t',0};
     static const WCHAR mumW[] = {'.','m','u','m',0};
     static const WCHAR starW[] = {'*',0};
@@ -910,6 +911,14 @@ static BOOL load_assemblies_from_cab(const WCHAR *filename, struct installer_sta
         return FALSE;
     }
 
+    if (!(path = path_combine(temp_path, manifest_cix_xmlW))) return FALSE;
+    if (GetFileAttributesW(path) != INVALID_FILE_ATTRIBUTES)
+    {
+        WINE_FIXME("Cabinet uses proprietary msdelta file compression which is not (yet) supported.\n");
+        WINE_FIXME("Installation of msu file will most likely fail.\n");
+    }
+    heap_free(path);
+
     if (!(path = path_combine(temp_path, starW))) return FALSE;
     search = FindFirstFileW(path, &data);
     heap_free(path);
-- 
2.6.4

