From 9a131a310a10bb8189117d96dc2f53b4ae117442 Mon Sep 17 00:00:00 2001
From: "Erich E. Hoover" <erich.e.hoover@wine-staging.com>
Date: Thu, 3 Sep 2015 16:24:45 -0600
Subject: wineps.drv: Add support for PostScript Format 2 custom glyph names.

---
 dlls/wineps.drv/download.c | 38 +++++++++++++++++++++++++++++++++++++-
 1 file changed, 37 insertions(+), 1 deletion(-)

diff --git a/dlls/wineps.drv/download.c b/dlls/wineps.drv/download.c
index febb9dd..7f39303 100644
--- a/dlls/wineps.drv/download.c
+++ b/dlls/wineps.drv/download.c
@@ -644,6 +644,41 @@ WORD get_post2_name_index(BYTE *post2header, DWORD size, WORD index)
     return GET_BE_WORD(post2header + offset);
 }
 
+void get_post2_custom_glyph_name(BYTE *post2header, DWORD size, WORD index, char *name)
+{
+    USHORT numberOfGlyphs = GET_BE_WORD(post2header);
+    int i, name_offset = (1 + numberOfGlyphs) * sizeof(USHORT);
+    BYTE name_length;
+
+    if(name_offset + sizeof(BYTE) > size)
+    {
+        FIXME("Pascal name offset '%d' exceeds PostScript Format 2 table size (%d)\n",
+              name_offset + sizeof(BYTE), size);
+        return;
+    }
+    for(i = 0; i < index; i++)
+    {
+        name_length = (post2header + name_offset)[0];
+        name_offset += name_length + sizeof(BYTE);
+        if(name_offset + sizeof(BYTE) > size)
+        {
+            FIXME("Pascal name offset '%d' exceeds PostScript Format 2 table size (%d)\n",
+                  name_offset + sizeof(BYTE), size);
+            return;
+        }
+    }
+    name_length = min((post2header + name_offset)[0], MAX_G_NAME);
+    name_offset += sizeof(BYTE);
+    if(name_offset + name_length > size)
+    {
+        FIXME("Pascal name offset '%d' exceeds PostScript Format 2 table size (%d)\n",
+              name_offset + name_length, size);
+        return;
+    }
+    memcpy(name, post2header + name_offset, name_length);
+    name[name_length] = 0;
+}
+
 void get_glyph_name(HDC hdc, WORD index, char *name)
 {
     struct
@@ -694,7 +729,8 @@ void get_glyph_name(HDC hdc, WORD index, char *name)
         if(glyphNameIndex < 258)
             get_standard_glyph_name(glyphNameIndex, name);
         else
-            FIXME("PostScript Format 2 custom glyphs are currently unsupported.\n");
+            get_post2_custom_glyph_name(post2header, size - sizeof(*post_header),
+                                        glyphNameIndex - 258, name);
     }
     else
         FIXME("PostScript Format %d.%d glyph names are currently unsupported.\n",
-- 
2.5.1

