From b9eb0aeed5bc09d8a54c383c9d5149cacf36f1c3 Mon Sep 17 00:00:00 2001
From: Louis Lenders <xerox.xerox2000x@gmail.com>
Date: Wed, 5 Dec 2018 13:18:26 +0100
Subject: [PATCH] dwmapi: Add tests for DwmGetTransportAttributes().

Signed-off-by: Louis Lenders <xerox.xerox2000x@gmail.com>
---
 dlls/dwmapi/tests/dwmapi.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/dlls/dwmapi/tests/dwmapi.c b/dlls/dwmapi/tests/dwmapi.c
index 1904e283bba..d79861f53ed 100644
--- a/dlls/dwmapi/tests/dwmapi.c
+++ b/dlls/dwmapi/tests/dwmapi.c
@@ -33,7 +33,18 @@ static void test_DwmIsCompositionEnabled(void)
     ok(enabled == TRUE || enabled == FALSE, "Got unexpected %#x.\n", enabled);
 }
 
+static void test_dwm_get_transport_attributes(void)
+{
+    BOOL isremoting, isconnected;
+    DWORD generation;
+    HRESULT hr;
+
+    hr = DwmGetTransportAttributes(&isremoting, &isconnected, &generation);
+    ok(hr == S_OK || hr == DWM_E_COMPOSITIONDISABLED, "Got unexpected %#x.\n", hr);
+}
+
 START_TEST(dwmapi)
 {
     test_DwmIsCompositionEnabled();
+    test_dwm_get_transport_attributes();
 }
-- 
2.27.0

