From 0d627c069b85cfcf18429e349b6a4166bd646451 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 9 Jul 2017 15:59:59 +0200
Subject: reg/tests: Don't call RegCloseKey on uninitialized memory.

---
 programs/reg/tests/reg.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/programs/reg/tests/reg.c b/programs/reg/tests/reg.c
index af092908cbc..941d3257c95 100644
--- a/programs/reg/tests/reg.c
+++ b/programs/reg/tests/reg.c
@@ -793,7 +793,7 @@ static void test_import(void)
     DWORD r, dword = 0x123, type, size;
     char test1_reg[MAX_PATH], test2_reg[MAX_PATH], cmdline[MAX_PATH];
     char test_string[] = "Test string", buffer[24];
-    HKEY hkey, subkey;
+    HKEY hkey, subkey = NULL;
     LONG err;
     BYTE hex[8];
 
-- 
2.13.1

