From 1332e777815c624db046e5e9ece93a789ff72870 Mon Sep 17 00:00:00 2001
From: Qian Hong <qhong@codeweavers.com>
Date: Thu, 22 Oct 2015 15:54:30 +0800
Subject: ntdll: Implement FileNamesInformation class support.

---
 dlls/ntdll/directory.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dlls/ntdll/directory.c b/dlls/ntdll/directory.c
index 223b842..2b54e14 100644
--- a/dlls/ntdll/directory.c
+++ b/dlls/ntdll/directory.c
@@ -159,6 +159,7 @@ union file_directory_info
     FILE_FULL_DIRECTORY_INFORMATION    full;
     FILE_ID_BOTH_DIRECTORY_INFORMATION id_both;
     FILE_ID_FULL_DIRECTORY_INFORMATION id_full;
+    FILE_NAMES_INFORMATION names;
 };
 
 struct dir_data_buffer
@@ -277,6 +278,8 @@ static inline unsigned int dir_info_size( FILE_INFORMATION_CLASS class, unsigned
         return offsetof( FILE_ID_BOTH_DIRECTORY_INFORMATION, FileName[len] );
     case FileIdFullDirectoryInformation:
         return offsetof( FILE_ID_FULL_DIRECTORY_INFORMATION, FileName[len] );
+    case FileNamesInformation:
+        return offsetof( FILE_NAMES_INFORMATION, FileName[len] );
     default:
         assert(0);
         return 0;
@@ -1600,6 +1603,10 @@ static NTSTATUS get_dir_data_entry( struct dir_data *dir_data, void *info_ptr, I
         info->id_both.FileNameLength = name_len;
         break;
 
+    case FileNamesInformation:
+        info->names.FileNameLength = name_len;
+        break;
+
     default:
         assert(0);
         return 0;
@@ -1990,6 +1997,7 @@ NTSTATUS WINAPI NtQueryDirectoryFile( HANDLE handle, HANDLE event,
     case FileFullDirectoryInformation:
     case FileIdBothDirectoryInformation:
     case FileIdFullDirectoryInformation:
+    case FileNamesInformation:
         if (length < dir_info_align( dir_info_size( info_class, 1 ))) return STATUS_INFO_LENGTH_MISMATCH;
         if (!buffer) return STATUS_ACCESS_VIOLATION;
         break;
-- 
2.8.0

