From 11fa3c3d26b860036633ca2e1437586d5bcb99d1 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 15 May 2016 17:00:46 +0200
Subject: ws2_32/tests: Hide random test failures in socket tests.

---
 dlls/ws2_32/tests/sock.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/dlls/ws2_32/tests/sock.c b/dlls/ws2_32/tests/sock.c
index 9e0c2f9..0b9e245 100644
--- a/dlls/ws2_32/tests/sock.c
+++ b/dlls/ws2_32/tests/sock.c
@@ -2530,6 +2530,7 @@ static void test_WSASocket(void)
     if (sock == INVALID_SOCKET)
     {
         err = WSAGetLastError();
+        todo_wine_if (err == WSAEPROTONOSUPPORT)
         ok(err == WSAEAFNOSUPPORT || broken(err == WSAEPROTONOSUPPORT), "Expected 10047, received %d\n", err);
         skip("IPX is not supported\n");
     }
@@ -2799,13 +2800,13 @@ static void test_WSAEnumNetworkEvents(void)
                     "Test[%d]: WSAEnumNetworkEvents failed\n", i);
                 if (i >= 1 && j == 0) /* FD_WRITE is SET on first try for UDP and connected TCP */
                 {
-                    todo_wine_if (i == 0) /* Remove when fixed */
+                    todo_wine_if (i == 0 || net_events.lNetworkEvents == 0)
                         ok (net_events.lNetworkEvents == FD_WRITE, "Test[%d]: expected 2, got %d\n",
                             i, net_events.lNetworkEvents);
                 }
                 else
                 {
-                    todo_wine_if (i != 0) /* Remove when fixed */
+                    todo_wine_if (i != 0 && net_events.lNetworkEvents != 0)
                         ok (net_events.lNetworkEvents == 0, "Test[%d]: expected 0, got %d\n",
                             i, net_events.lNetworkEvents);
                 }
-- 
2.8.0

