From 6f484b8e9cb7dcfcc8c3f0103012576b43572422 Mon Sep 17 00:00:00 2001
From: Jarkko Korpi <jarkko_korpi@hotmail.com>
Date: Fri, 10 Jul 2015 18:49:44 +0300
Subject: widl: Increase buffer size in typegen.c.

---
 tools/widl/typegen.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/widl/typegen.c b/tools/widl/typegen.c
index 4bb4cc4..045a5b7 100644
--- a/tools/widl/typegen.c
+++ b/tools/widl/typegen.c
@@ -1142,7 +1142,7 @@ static unsigned char get_func_oi2_flags( const var_t *func )
 static unsigned int write_new_procformatstring_type(FILE *file, int indent, const var_t *var,
                                                     int is_return, unsigned int *stack_offset)
 {
-    char buffer[64];
+    char buffer[128];
     unsigned int stack_size, typestring_offset;
     unsigned short flags;
     unsigned char fc = get_parameter_fc( var, is_return, &flags, &stack_size, &typestring_offset );
-- 
2.4.5

