From c7dc4a7cb6ce194725d5f2eeaac279f6a27c0d40 Mon Sep 17 00:00:00 2001
From: Jarkko Korpi <jarkko_korpi@hotmail.com>
Date: Sun, 18 Dec 2016 00:53:33 +0200
Subject: ntdll: Silence FIXME in RtlCaptureStackBackTrace stub function

---
 dlls/ntdll/signal_x86_64.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ntdll/signal_x86_64.c b/dlls/ntdll/signal_x86_64.c
index dbab1c5..be4d35d 100644
--- a/dlls/ntdll/signal_x86_64.c
+++ b/dlls/ntdll/signal_x86_64.c
@@ -3942,7 +3942,7 @@ __ASM_GLOBAL_FUNC( RtlRaiseException,
  */
 USHORT WINAPI RtlCaptureStackBackTrace( ULONG skip, ULONG count, PVOID *buffer, ULONG *hash )
 {
-    FIXME( "(%d, %d, %p, %p) stub!\n", skip, count, buffer, hash );
+    TRACE( "(%d, %d, %p, %p) stub!\n", skip, count, buffer, hash );
     return 0;
 }
 
-- 
2.9.0

