From d8452914d07d7a24aa2647bf7c98383d962693f9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sat, 27 Jun 2015 22:26:14 +0200
Subject: wined3d: Load dxtn dylib library on Mac OS X.

---
 dlls/wined3d/dxtn.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/wined3d/dxtn.c b/dlls/wined3d/dxtn.c
index 2639788..03e4b26 100644
--- a/dlls/wined3d/dxtn.c
+++ b/dlls/wined3d/dxtn.c
@@ -465,6 +465,10 @@ BOOL wined3d_dxtn_init(void)
 #ifdef SONAME_LIBTXC_DXTN
         SONAME_LIBTXC_DXTN,
 #endif
+#ifdef __APPLE__
+        "libtxc_dxtn.dylib",
+        "libtxc_dxtn_s2tc.dylib",
+#endif
         "libtxc_dxtn.so",
         "libtxc_dxtn_s2tc.so.0"
     };
-- 
2.4.3

