From 376bc1c96b3fe8aa79516198aa59763a762392ee Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 27 Sep 2015 19:35:05 +0200
Subject: user32: Refresh MDI menus when DefMDIChildProc(WM_SETTEXT) is
 called.

---
 dlls/user32/mdi.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/user32/mdi.c b/dlls/user32/mdi.c
index 41b5077..635c8b2 100644
--- a/dlls/user32/mdi.c
+++ b/dlls/user32/mdi.c
@@ -1425,6 +1425,7 @@ LRESULT WINAPI DefMDIChildProcA( HWND hwnd, UINT message,
 	DefWindowProcA(hwnd, message, wParam, lParam);
 	if( ci->hwndChildMaximized == hwnd )
 	    MDI_UpdateFrameText( GetParent(client), client, TRUE, NULL );
+        MDI_RefreshMenu( ci );
         return 1; /* success. FIXME: check text length */
 
     case WM_GETMINMAXINFO:
@@ -1465,6 +1466,7 @@ LRESULT WINAPI DefMDIChildProcW( HWND hwnd, UINT message,
         DefWindowProcW(hwnd, message, wParam, lParam);
         if( ci->hwndChildMaximized == hwnd )
             MDI_UpdateFrameText( GetParent(client), client, TRUE, NULL );
+        MDI_RefreshMenu( ci );
         return 1; /* success. FIXME: check text length */
 
     case WM_GETMINMAXINFO:
-- 
2.5.1

