From b2fec3a86116deb45df5c8f3f6ac07434d3d34a7 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 23 Jan 2016 20:23:04 +0100
Subject: server: Add missing check for objattr variable in load_registry
 wineserver call (Coverity).

---
 server/registry.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/server/registry.c b/server/registry.c
index 5723824..a35765c 100644
--- a/server/registry.c
+++ b/server/registry.c
@@ -2187,6 +2187,8 @@ DECL_HANDLER(load_registry)
     const struct security_descriptor *sd;
     const struct object_attributes *objattr = get_req_object_attributes( &sd, &name );
 
+    if (!objattr) return;
+
     if (!thread_single_check_privilege( current, &SeRestorePrivilege ))
     {
         set_error( STATUS_PRIVILEGE_NOT_HELD );
-- 
2.6.4

