From 4e690407c338be1553e6c2c18977364152477339 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 23 Jan 2016 20:23:48 +0100
Subject: server: Avoid invalid memory access if creation of namespace fails in
 create_directory (Coverity).

---
 server/directory.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/server/directory.c b/server/directory.c
index de049ef..937ab89 100644
--- a/server/directory.c
+++ b/server/directory.c
@@ -192,7 +192,7 @@ static struct directory *create_directory( struct directory *root, const struct
         if (!(dir->entries = create_namespace( hash_size )))
         {
             release_object( dir );
-            dir = NULL;
+            return NULL;
         }
         if (sd) default_set_sd( &dir->obj, sd, OWNER_SECURITY_INFORMATION | GROUP_SECURITY_INFORMATION |
                                 DACL_SECURITY_INFORMATION | SACL_SECURITY_INFORMATION );
-- 
2.6.4

