From 012441a764762063b3037d060769fea4356a295c Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 18 Apr 2018 03:55:14 +0000
Subject: [PATCH] dxgi: Return S_OK in SetMaximumFrameLatency

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 dlls/dxgi/device.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/dxgi/device.c b/dlls/dxgi/device.c
index c266cae..ee5f3e6 100644
--- a/dlls/dxgi/device.c
+++ b/dlls/dxgi/device.c
@@ -262,7 +262,7 @@ static HRESULT STDMETHODCALLTYPE dxgi_device_SetMaximumFrameLatency(IWineDXGIDev
     if (max_latency > DXGI_FRAME_LATENCY_MAX)
         return DXGI_ERROR_INVALID_CALL;
 
-    return E_NOTIMPL;
+    return S_OK;
 }
 
 static HRESULT STDMETHODCALLTYPE dxgi_device_GetMaximumFrameLatency(IWineDXGIDevice *iface, UINT *max_latency)
-- 
1.9.1

