From addabfd06c4d78e92fa795aaba74c448fb892a1b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Andr=C3=A9=20Hentschel?= <nerv@dawncrow.de>
Date: Sat, 3 May 2014 19:01:06 +0200
Subject: libwine: Use try_mmap_fixed for wine64 on FreeBSD

---
 libs/wine/mmap.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/libs/wine/mmap.c b/libs/wine/mmap.c
index b219147..36bf1b1 100644
--- a/libs/wine/mmap.c
+++ b/libs/wine/mmap.c
@@ -82,7 +82,8 @@ static inline int get_fdzero(void)
     return fd;
 }
 
-#if (defined(__svr4__) || defined(__NetBSD__)) && !defined(MAP_TRYFIXED)
+#if (defined(__svr4__) || defined(__NetBSD__) || (defined(__FreeBSD__) && defined(__x86_64__))) && \
+    !defined(MAP_TRYFIXED)
 /***********************************************************************
  *             try_mmap_fixed
  *
@@ -213,7 +214,8 @@ void *wine_anon_mmap( void *start, size_t size, int prot, int flags )
 #ifdef MAP_TRYFIXED
         /* If available, this will attempt a fixed mapping in-kernel */
         flags |= MAP_TRYFIXED;
-#elif defined(__svr4__) || defined(__NetBSD__) || defined(__APPLE__)
+#elif defined(__svr4__) || defined(__NetBSD__) || defined(__APPLE__) || \
+      (defined(__FreeBSD__) && defined(__x86_64__))
         if ( try_mmap_fixed( start, size, prot, flags, get_fdzero(), 0 ) )
             return start;
 #endif
-- 
2.1.2

