From 62d8d5c8322be7bf07f95bca2c26057e708eff5a Mon Sep 17 00:00:00 2001
From: Alex Henrie <alexhenrie24@gmail.com>
Date: Mon, 27 Jan 2020 23:08:45 -0700
Subject: [PATCH 2/5] include: Add RtlIpv6StringToAddress(Ex)[AW]

Signed-off-by: Alex Henrie <alexhenrie24@gmail.com>
---
 include/ip2string.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/include/ip2string.h b/include/ip2string.h
index 6c96c8d76d..0753ee82a9 100644
--- a/include/ip2string.h
+++ b/include/ip2string.h
@@ -30,6 +30,13 @@ NTSTATUS WINAPI RtlIpv4StringToAddressExA(const char *str, BOOLEAN strict, IN_AD
 NTSTATUS WINAPI RtlIpv4StringToAddressExW(const WCHAR *str, BOOLEAN strict, IN_ADDR *address, USHORT *port);
 #define RtlIpv4StringToAddressEx WINELIB_NAME_AW(RtlIpv4StringToAddressEx)
 
+NTSTATUS WINAPI RtlIpv6StringToAddressA(const char *str, const char **terminator, IN6_ADDR *address);
+NTSTATUS WINAPI RtlIpv6StringToAddressW(const WCHAR *str, const WCHAR **terminator, IN6_ADDR *address);
+#define RtlIpv6StringToAddress WINELIB_NAME_AW(RtlIpv6StringToAddress)
+NTSTATUS WINAPI RtlIpv6StringToAddressExA(const char *str, IN6_ADDR *address, ULONG *scope, USHORT *port);
+NTSTATUS WINAPI RtlIpv6StringToAddressExW(const WCHAR *str, IN6_ADDR *address, ULONG *scope, USHORT *port);
+#define RtlIpv6StringToAddressEx WINELIB_NAME_AW(RtlIpv6StringToAddressEx)
+
 #ifdef __cplusplus
 }
 #endif
-- 
2.24.1

