From 7b59e0a7d0962899762b5b7a466d16b97d771df2 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 27 Apr 2019 12:53:06 +1000
Subject: [PATCH 3/7] dinput: Return E_NOINTERFACE from IDirectInputDevice2
 QueryInterface

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 dlls/dinput/device.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/dinput/device.c b/dlls/dinput/device.c
index 63eb868..1317e91 100644
--- a/dlls/dinput/device.c
+++ b/dlls/dinput/device.c
@@ -1257,7 +1257,7 @@ HRESULT WINAPI IDirectInputDevice2WImpl_QueryInterface(LPDIRECTINPUTDEVICE8W ifa
     }
 
     WARN("Unsupported interface!\n");
-    return E_FAIL;
+    return E_NOINTERFACE;
 }
 
 HRESULT WINAPI IDirectInputDevice2AImpl_QueryInterface(LPDIRECTINPUTDEVICE8A iface, REFIID riid, LPVOID *ppobj)
-- 
1.9.1

