From 1b1158f85ae61f348207460ce98b1977eee79b66 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Sat, 27 Apr 2019 13:10:01 +1000
Subject: [PATCH 4/7] dinput: Dont report we cannot open a device

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 dlls/dinput/joystick_linuxinput.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/dlls/dinput/joystick_linuxinput.c b/dlls/dinput/joystick_linuxinput.c
index fb6b305..7343a50 100644
--- a/dlls/dinput/joystick_linuxinput.c
+++ b/dlls/dinput/joystick_linuxinput.c
@@ -219,10 +219,7 @@ static void find_joydevs(void)
         }
 
         if (fd == -1)
-        {
-            WARN("Failed to open \"%s\": %d %s\n", buf, errno, strerror(errno));
             continue;
-        }
 
         if (ioctl(fd, EVIOCGBIT(0, sizeof(joydev.evbits)), joydev.evbits) == -1)
         {
-- 
1.9.1

