From a1747217a5bae6416a157f8bcf9286845364a5fe Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 29 Aug 2015 06:24:21 +0200
Subject: d3d11: Fix compile failure with recent version of gcc.

---
 dlls/d3d11/utils.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/dlls/d3d11/utils.c b/dlls/d3d11/utils.c
index 307ef23..91de43c 100644
--- a/dlls/d3d11/utils.c
+++ b/dlls/d3d11/utils.c
@@ -465,10 +465,10 @@ UINT d3d10_cpu_access_flags_from_d3d11_cpu_access_flags(UINT cpu_access_flags)
 
 UINT d3d11_resource_misc_flags_from_d3d10_resource_misc_flags(UINT resource_misc_flags)
 {
-    static const UINT bitwise_identical_flags = D3D10_RESOURCE_MISC_GENERATE_MIPS
+    const UINT bitwise_identical_flags = D3D10_RESOURCE_MISC_GENERATE_MIPS
             | D3D10_RESOURCE_MISC_SHARED
             | D3D10_RESOURCE_MISC_TEXTURECUBE;
-    static const UINT handled_flags = bitwise_identical_flags
+    const UINT handled_flags = bitwise_identical_flags
             | D3D10_RESOURCE_MISC_SHARED_KEYEDMUTEX
             | D3D10_RESOURCE_MISC_GDI_COMPATIBLE;
     UINT d3d11_resource_misc_flags = resource_misc_flags & bitwise_identical_flags;
@@ -486,7 +486,7 @@ UINT d3d11_resource_misc_flags_from_d3d10_resource_misc_flags(UINT resource_misc
 
 UINT d3d10_resource_misc_flags_from_d3d11_resource_misc_flags(UINT resource_misc_flags)
 {
-    static const UINT bitwise_identical_flags = D3D11_RESOURCE_MISC_GENERATE_MIPS
+    const UINT bitwise_identical_flags = D3D11_RESOURCE_MISC_GENERATE_MIPS
             | D3D11_RESOURCE_MISC_SHARED
             | D3D11_RESOURCE_MISC_TEXTURECUBE
             | D3D11_RESOURCE_MISC_BUFFER_STRUCTURED
@@ -496,7 +496,7 @@ UINT d3d10_resource_misc_flags_from_d3d11_resource_misc_flags(UINT resource_misc
             | D3D11_RESOURCE_MISC_RESTRICT_SHARED_RESOURCE
             | D3D11_RESOURCE_MISC_RESTRICT_SHARED_RESOURCE_DRIVER
             | D3D11_RESOURCE_MISC_GUARDED;
-    static const UINT handled_flags = bitwise_identical_flags
+    const UINT handled_flags = bitwise_identical_flags
             | D3D11_RESOURCE_MISC_SHARED_KEYEDMUTEX
             | D3D11_RESOURCE_MISC_GDI_COMPATIBLE;
     UINT d3d10_resource_misc_flags = resource_misc_flags & bitwise_identical_flags;
-- 
2.5.0

