From 62276b1f14a98c34737b195705c79f60a2561f59 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Sun, 27 Nov 2016 14:07:12 +0800
Subject: user32: Fix groupbox rectangle calculation in the button's WM_SETTEXT
 handler.

---
 dlls/user32/button.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/user32/button.c b/dlls/user32/button.c
index 623b425..86b3b1e 100644
--- a/dlls/user32/button.c
+++ b/dlls/user32/button.c
@@ -394,6 +394,9 @@ LRESULT ButtonWndProc_common(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam,
 
             GetClientRect(hWnd, &client);
             rc = client;
+            /* FIXME: check other BS_* handlers */
+            if (btn_type == BS_GROUPBOX)
+                InflateRect(&rc, -7, 1); /* GB_Paint does this */
             BUTTON_CalcLabelRect(hWnd, hdc, &rc);
             /* Clip by client rect bounds */
             if (rc.right > client.right) rc.right = client.right;
-- 
2.9.0

