From 351506a226133691c2dac3e83e964b304b5ddd90 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Mon, 21 Mar 2016 16:23:21 +0100
Subject: wined3d: Add warning that CSMT patchset is disabled.

---
 dlls/wined3d/wined3d_main.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/wined3d/wined3d_main.c b/dlls/wined3d/wined3d_main.c
index fdb260e..d5c307e 100644
--- a/dlls/wined3d/wined3d_main.c
+++ b/dlls/wined3d/wined3d_main.c
@@ -327,6 +327,11 @@ static BOOL wined3d_dll_init(HINSTANCE hInstDLL)
         }
     }
 
+#if defined(STAGING_CSMT)
+    FIXME_(winediag)("Please note that the experimental wined3d CSMT patchset has been disabled in this\n");
+    FIXME_(winediag)("version of Wine Staging. Check out the release notes for further information.\n");
+#endif /* STAGING_CSMT */
+
     if (appkey) RegCloseKey( appkey );
     if (hkey) RegCloseKey( hkey );
 
-- 
2.7.1

