From a9fb17a4b519b6ca814cd9402531a1ca860634f1 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Fri, 10 Feb 2017 00:34:37 +0800
Subject: comctl32: Add support for PSPCB_ADDREF/PSPCB_RELEASE callback
 notifications. (v2)

---
 dlls/comctl32/propsheet.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/dlls/comctl32/propsheet.c b/dlls/comctl32/propsheet.c
index a3a6a5c4df8..4c4ca82a572 100644
--- a/dlls/comctl32/propsheet.c
+++ b/dlls/comctl32/propsheet.c
@@ -3013,6 +3013,9 @@ HPROPSHEETPAGE WINAPI CreatePropertySheetPageA(
     else
         ppsp->pszHeaderSubTitle = NULL;
 
+    if ((ppsp->dwFlags & PSH_USECALLBACK) && ppsp->pfnCallback)
+        ppsp->pfnCallback(0, PSPCB_ADDREF, ppsp);
+
     return (HPROPSHEETPAGE)ppsp;
 }
 
@@ -3059,6 +3062,9 @@ HPROPSHEETPAGE WINAPI CreatePropertySheetPageW(LPCPROPSHEETPAGEW lpPropSheetPage
     else
         ppsp->pszHeaderSubTitle = NULL;
 
+    if ((ppsp->dwFlags & PSH_USECALLBACK) && ppsp->pfnCallback)
+        ppsp->pfnCallback(0, PSPCB_ADDREF, ppsp);
+
     return (HPROPSHEETPAGE)ppsp;
 }
 
@@ -3080,6 +3086,9 @@ BOOL WINAPI DestroyPropertySheetPage(HPROPSHEETPAGE hPropPage)
   if (!psp)
      return FALSE;
 
+  if ((psp->dwFlags & PSH_USECALLBACK) && psp->pfnCallback)
+     psp->pfnCallback(0, PSPCB_RELEASE, psp);
+
   if (!(psp->dwFlags & PSP_DLGINDIRECT) && !IS_INTRESOURCE( psp->u.pszTemplate ))
      Free ((LPVOID)psp->u.pszTemplate);
 
-- 
2.11.0

