From 50389a00b3f63757d8cab2b36e63d88f955d5a92 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sat, 25 Jun 2016 07:11:52 +0200
Subject: include: Remove duplicate definition of WS_XML_NODE_POSITION.

---
 include/webservices.h | 1 -
 1 file changed, 1 deletion(-)

diff --git a/include/webservices.h b/include/webservices.h
index a5ffd51..3c808f5 100644
--- a/include/webservices.h
+++ b/include/webservices.h
@@ -73,7 +73,6 @@ typedef struct _WS_CUSTOM_HTTP_PROXY WS_CUSTOM_HTTP_PROXY;
 typedef struct _WS_HTTP_MESSAGE_MAPPING WS_HTTP_MESSAGE_MAPPING;
 typedef struct _WS_HTTP_HEADER_MAPPING WS_HTTP_HEADER_MAPPING;
 typedef struct _WS_HTTP_REDIRECT_CALLBACK_CONTEXT WS_HTTP_REDIRECT_CALLBACK_CONTEXT;
-typedef struct _WS_XML_NODE_POSITION WS_XML_NODE_POSITION;
 
 struct _WS_STRUCT_DESCRIPTION;
 struct _WS_XML_STRING;
-- 
2.8.0

