From 60d8abcd74abb34f8922c7b33742d73c47773ddf Mon Sep 17 00:00:00 2001
From: Christian Costa <titan.costa@gmail.com>
Date: Mon, 22 Oct 2012 21:00:52 +0200
Subject: d3dxof: Use sizeof(GUID) instead of hardcoding the value.

---
 dlls/d3dxof/d3dxof.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/d3dxof/d3dxof.c b/dlls/d3dxof/d3dxof.c
index 88647cf..d612109 100644
--- a/dlls/d3dxof/d3dxof.c
+++ b/dlls/d3dxof/d3dxof.c
@@ -619,7 +619,7 @@ static HRESULT WINAPI IDirectXFileDataImpl_GetId(IDirectXFileData* iface, LPGUID
   if (!pGuid)
     return DXFILEERR_BADVALUE;
 
-  memcpy(pGuid, &This->pobj->class_id, 16);
+  memcpy(pGuid, &This->pobj->class_id, sizeof(GUID));
 
   return DXFILE_OK;
 }
@@ -667,7 +667,7 @@ static HRESULT WINAPI IDirectXFileDataImpl_GetType(IDirectXFileData* iface, cons
   if (!pguid)
     return DXFILEERR_BADVALUE;
 
-  memcpy(&guid, &This->pobj->type, 16);
+  memcpy(&guid, &This->pobj->type, sizeof(GUID));
   *pguid = &guid;
 
   return DXFILE_OK;
@@ -886,7 +886,7 @@ static HRESULT WINAPI IDirectXFileDataReferenceImpl_GetId(IDirectXFileDataRefere
   if (!pGuid)
     return DXFILEERR_BADVALUE;
 
-  memcpy(pGuid, &This->ptarget->class_id, 16);
+  memcpy(pGuid, &This->ptarget->class_id, sizeof(GUID));
 
   return DXFILE_OK;
 }
-- 
2.3.0

