From d18752b72603e91c7bbbbfa623c83f3d13112ca1 Mon Sep 17 00:00:00 2001
From: Christian Costa <titan.costa@gmail.com>
Date: Mon, 22 Oct 2012 21:00:52 +0200
Subject: d3dxof: Use straight assignement instead of using memcpy with
 hardcoded size.

---
 dlls/d3dxof/d3dxof.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/d3dxof/d3dxof.c b/dlls/d3dxof/d3dxof.c
index 88647cf..8888d8b 100644
--- a/dlls/d3dxof/d3dxof.c
+++ b/dlls/d3dxof/d3dxof.c
@@ -619,7 +619,7 @@ static HRESULT WINAPI IDirectXFileDataImpl_GetId(IDirectXFileData* iface, LPGUID
   if (!pGuid)
     return DXFILEERR_BADVALUE;
 
-  memcpy(pGuid, &This->pobj->class_id, 16);
+  *pGuid = This->pobj->class_id;
 
   return DXFILE_OK;
 }
@@ -667,7 +667,7 @@ static HRESULT WINAPI IDirectXFileDataImpl_GetType(IDirectXFileData* iface, cons
   if (!pguid)
     return DXFILEERR_BADVALUE;
 
-  memcpy(&guid, &This->pobj->type, 16);
+  guid = This->pobj->type;
   *pguid = &guid;
 
   return DXFILE_OK;
@@ -886,7 +886,7 @@ static HRESULT WINAPI IDirectXFileDataReferenceImpl_GetId(IDirectXFileDataRefere
   if (!pGuid)
     return DXFILEERR_BADVALUE;
 
-  memcpy(pGuid, &This->ptarget->class_id, 16);
+  *pGuid = This->ptarget->class_id;
 
   return DXFILE_OK;
 }
-- 
2.3.0

