From f5b87201251c3fd2d10cc9208ade78d0a100d26a Mon Sep 17 00:00:00 2001
From: Hugh McMaster <hugh.mcmaster@outlook.com>
Date: Thu, 28 May 2015 21:29:24 +1000
Subject: wineconsole: Rearrange user dialog to make space for InsertMode
 checkbox

---
 programs/wineconsole/wineconsole.rc | 28 ++++++++++++++--------------
 1 file changed, 14 insertions(+), 14 deletions(-)

diff --git a/programs/wineconsole/wineconsole.rc b/programs/wineconsole/wineconsole.rc
index e4e0624..4dd06b7 100644
--- a/programs/wineconsole/wineconsole.rc
+++ b/programs/wineconsole/wineconsole.rc
@@ -54,31 +54,31 @@ IDS_USAGE_FOOTER           "\nExample:\n  wineconsole cmd\nStarts the Wine comma
 
 END
 
-IDD_OPTION DIALOG 36, 24, 140, 105
+IDD_OPTION DIALOG 36, 24, 140, 153
 STYLE DS_MODALFRAME | WS_POPUP | WS_CAPTION | WS_SYSMENU
 CAPTION "Options"
 FONT 8, "MS Shell Dlg"
 {
-	GROUPBOX "Cursor size", -1, 5, 5, 65, 56, BS_GROUPBOX
+    GROUPBOX "Cursor size", -1, 5, 5, 70, 54, BS_GROUPBOX
 	AUTORADIOBUTTON "&Small", IDC_OPT_CURSOR_SMALL, 9, 18, 50, 10, WS_TABSTOP
 	AUTORADIOBUTTON "&Medium", IDC_OPT_CURSOR_MEDIUM, 9, 30, 50, 10, WS_TABSTOP
 	AUTORADIOBUTTON "&Large", IDC_OPT_CURSOR_LARGE, 9, 42, 50, 10, WS_TABSTOP
 
-	GROUPBOX "Control", -1, 75, 5, 125, 56, BS_GROUPBOX
-	LTEXT "Popup menu", -1, 79, 18, 50, 18
-	AUTOCHECKBOX "&Control", IDC_OPT_CONF_CTRL, 138, 18, 60, 10, WS_TABSTOP
-	AUTOCHECKBOX "S&hift", IDC_OPT_CONF_SHIFT, 138, 30, 60, 10, WS_TABSTOP
-	LTEXT "Quick edit", -1, 79, 42, 50, 18
-	AUTOCHECKBOX "&enable", IDC_OPT_QUICK_EDIT, 138, 42, 60, 10, WS_TABSTOP
+    GROUPBOX "Console", -1, 80, 5, 120, 54, BS_GROUPBOX
+    AUTOCHECKBOX "&Quick edit", IDC_OPT_QUICK_EDIT, 84, 18, 70, 10, WS_TABSTOP
 
-	GROUPBOX "Command history", -1, 5, 63, 195, 40, BS_GROUPBOX
-	LTEXT "&Number of recalled commands:", -1, 9, 75, 78, 18
-	EDITTEXT IDC_OPT_HIST_SIZE, 87, 77, 31, 12, WS_TABSTOP|WS_BORDER|ES_NUMBER
+    GROUPBOX "Popup menu", -1, 5, 61, 70, 42, BS_GROUPBOX
+    AUTOCHECKBOX "&Control", IDC_OPT_CONF_CTRL, 9, 74, 60, 10, WS_TABSTOP
+    AUTOCHECKBOX "S&hift", IDC_OPT_CONF_SHIFT, 9, 86, 60, 10, WS_TABSTOP
+
+    GROUPBOX "Command history", -1, 5, 105, 195, 42, BS_GROUPBOX
+    LTEXT "&Number of recalled commands:", -1, 9, 118, 110, 10
+    EDITTEXT IDC_OPT_HIST_SIZE, 123, 117, 31, 12, WS_TABSTOP|WS_BORDER|ES_NUMBER
 	CONTROL "", IDC_OPT_HIST_SIZE_UD, "msctls_updown32", UDS_SETBUDDYINT|UDS_ALIGNRIGHT|UDS_AUTOBUDDY|UDS_ARROWKEYS|UDS_NOTHOUSANDS, 0, 0, 0, 0
-	AUTOCHECKBOX "&Remove doubles", IDC_OPT_HIST_NODOUBLE, 125, 75, 60, 18, WS_TABSTOP|BS_MULTILINE
+    AUTOCHECKBOX "&Remove doubles", IDC_OPT_HIST_NODOUBLE, 9, 130, 80, 10, WS_TABSTOP|BS_MULTILINE
 }
 
-IDD_FONT DIALOG 36, 24, 140, 105
+IDD_FONT DIALOG 36, 24, 140, 153
 STYLE DS_MODALFRAME | WS_POPUP | WS_CAPTION | WS_SYSMENU
 CAPTION "Font"
 FONT 8, "MS Shell Dlg"
@@ -94,7 +94,7 @@ FONT 8, "MS Shell Dlg"
 	LTEXT "", IDC_FNT_FONT_INFO, 128, 73, 80, 27
 }
 
-IDD_CONFIG DIALOG 36, 24, 140, 105
+IDD_CONFIG DIALOG 36, 24, 140, 153
 STYLE DS_MODALFRAME | WS_POPUP | WS_CAPTION | WS_SYSMENU
 CAPTION "Configuration"
 FONT 8, "MS Shell Dlg"
-- 
2.4.1

