From c89f895ce2d7e5a29491bf658bdae0b53fbf0c08 Mon Sep 17 00:00:00 2001
From: Hugh McMaster <hugh.mcmaster@outlook.com>
Date: Thu, 28 May 2015 21:29:47 +1000
Subject: wineconsole: Improve semantics of some poorly-worded resource strings

---
 programs/wineconsole/wineconsole.rc | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/programs/wineconsole/wineconsole.rc b/programs/wineconsole/wineconsole.rc
index 4dd06b7..4226ba3 100644
--- a/programs/wineconsole/wineconsole.rc
+++ b/programs/wineconsole/wineconsole.rc
@@ -65,17 +65,17 @@ FONT 8, "MS Shell Dlg"
 	AUTORADIOBUTTON "&Large", IDC_OPT_CURSOR_LARGE, 9, 42, 50, 10, WS_TABSTOP
 
     GROUPBOX "Console", -1, 80, 5, 120, 54, BS_GROUPBOX
-    AUTOCHECKBOX "&Quick edit", IDC_OPT_QUICK_EDIT, 84, 18, 70, 10, WS_TABSTOP
+    AUTOCHECKBOX "&QuickEdit mode", IDC_OPT_QUICK_EDIT, 84, 18, 70, 10, WS_TABSTOP
 
     GROUPBOX "Popup menu", -1, 5, 61, 70, 42, BS_GROUPBOX
     AUTOCHECKBOX "&Control", IDC_OPT_CONF_CTRL, 9, 74, 60, 10, WS_TABSTOP
     AUTOCHECKBOX "S&hift", IDC_OPT_CONF_SHIFT, 9, 86, 60, 10, WS_TABSTOP
 
     GROUPBOX "Command history", -1, 5, 105, 195, 42, BS_GROUPBOX
-    LTEXT "&Number of recalled commands:", -1, 9, 118, 110, 10
+    LTEXT "&Number of stored commands:", -1, 9, 118, 110, 10
     EDITTEXT IDC_OPT_HIST_SIZE, 123, 117, 31, 12, WS_TABSTOP|WS_BORDER|ES_NUMBER
 	CONTROL "", IDC_OPT_HIST_SIZE_UD, "msctls_updown32", UDS_SETBUDDYINT|UDS_ALIGNRIGHT|UDS_AUTOBUDDY|UDS_ARROWKEYS|UDS_NOTHOUSANDS, 0, 0, 0, 0
-    AUTOCHECKBOX "&Remove doubles", IDC_OPT_HIST_NODOUBLE, 9, 130, 80, 10, WS_TABSTOP|BS_MULTILINE
+    AUTOCHECKBOX "&Remove duplicates", IDC_OPT_HIST_NODOUBLE, 9, 130, 80, 10, WS_TABSTOP|BS_MULTILINE
 }
 
 IDD_FONT DIALOG 36, 24, 140, 153
-- 
2.4.1

