From 3a98fa11972dec8bdbc5139097f5c64f2a4fb1fb Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Fri, 21 Jul 2017 01:31:54 +0200
Subject: rpcrt4: Hold CS while iterating through protseqs list.

---
 dlls/rpcrt4/rpc_server.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/rpcrt4/rpc_server.c b/dlls/rpcrt4/rpc_server.c
index c350c5970c2..d1b5d64ecea 100644
--- a/dlls/rpcrt4/rpc_server.c
+++ b/dlls/rpcrt4/rpc_server.c
@@ -806,8 +806,10 @@ static RPC_STATUS RPCRT4_stop_listen(BOOL auto_listen)
 
   if (stop_listen) {
     RpcServerProtseq *cps;
+    EnterCriticalSection(&server_cs);
     LIST_FOR_EACH_ENTRY(cps, &protseqs, RpcServerProtseq, entry)
       RPCRT4_sync_with_server_thread(cps);
+    LeaveCriticalSection(&server_cs);
   }
 
   if (!auto_listen)
-- 
2.13.1

