From 2a88e8c6a5dc0d9b530664eda478088e7511c9e2 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Thu, 31 Jan 2019 10:43:03 +1100
Subject: [PATCH] crypt32: Fix compile error on older gnutls.

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 dlls/crypt32/pfx.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/crypt32/pfx.c b/dlls/crypt32/pfx.c
index cde2f19..46bf356 100644
--- a/dlls/crypt32/pfx.c
+++ b/dlls/crypt32/pfx.c
@@ -45,6 +45,8 @@ int gnutls_pkcs12_simple_parse(gnutls_pkcs12_t p12, const char *password,
     gnutls_x509_crt_t **extra_certs, unsigned int *extra_certs_len,
     gnutls_x509_crl_t * crl, unsigned int flags);
 
+int gnutls_x509_privkey_get_pk_algorithm2(gnutls_x509_privkey_t, unsigned int*);
+
 static void *libgnutls_handle;
 #define MAKE_FUNCPTR(f) static typeof(f) * p##f
 MAKE_FUNCPTR(gnutls_global_deinit);
-- 
1.9.1

