From 778d6ea39b0c49f5000efd401f1ab05ec64c7abc Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Wed, 30 Jan 2019 10:20:49 +1100
Subject: [PATCH] crypt32: Fix compile error on older gnutls

Signed-off-by: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
---
 dlls/crypt32/pfx.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/dlls/crypt32/pfx.c b/dlls/crypt32/pfx.c
index d824410..c4fb34d 100644
--- a/dlls/crypt32/pfx.c
+++ b/dlls/crypt32/pfx.c
@@ -38,6 +38,12 @@ WINE_DEFAULT_DEBUG_CHANNEL(crypt);
 #ifdef SONAME_LIBGNUTLS
 WINE_DECLARE_DEBUG_CHANNEL(winediag);
 
+/* Not present in gnutls version < 3.0 */
+int gnutls_pkcs12_simple_parse(gnutls_pkcs12_t p12, const char *password,
+    gnutls_x509_privkey_t *key, gnutls_x509_crt_t **chain, unsigned int *chain_len,
+    gnutls_x509_crt_t **extra_certs, unsigned int *extra_certs_len,
+    gnutls_x509_crl_t * crl, unsigned int flags);
+
 static void *libgnutls_handle;
 #define MAKE_FUNCPTR(f) static typeof(f) * p##f
 MAKE_FUNCPTR(gnutls_global_deinit);
-- 
1.9.1

