From 6c46c1e7d4c3ac8fe3482dc7d948b6d4397b6ef7 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 15 Mar 2015 01:33:05 +0100
Subject: wined3d: Silence repeated 'Unhandled blend factor 0' messages.

---
 dlls/wined3d/state.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/dlls/wined3d/state.c b/dlls/wined3d/state.c
index 540a1b36174..d7f70545905 100644
--- a/dlls/wined3d/state.c
+++ b/dlls/wined3d/state.c
@@ -415,8 +415,11 @@ static GLenum gl_blend_factor(enum wined3d_blend factor, const struct wined3d_fo
         case WINED3D_BLEND_INVSRC1ALPHA:
             return GL_ONE_MINUS_SRC1_ALPHA;
         default:
-            FIXME("Unhandled blend factor %#x.\n", factor);
+        {
+            static int once;
+            if (factor || !once++) FIXME("Unhandled blend factor %#x.\n", factor);
             return GL_NONE;
+        }
     }
 }
 
-- 
2.11.0

