From 9627792f549907cfa050d89f464f7a19a1ef1c59 Mon Sep 17 00:00:00 2001
From: Christian Costa <titan.costa@gmail.com>
Date: Sun, 25 Oct 2015 12:50:31 +0100
Subject: wined3d: Print FIXME only once in surface_cpu_blt.

---
 dlls/wined3d/surface.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/wined3d/surface.c b/dlls/wined3d/surface.c
index 814d4ab..8192c18 100644
--- a/dlls/wined3d/surface.c
+++ b/dlls/wined3d/surface.c
@@ -4238,7 +4238,8 @@ static HRESULT surface_cpu_blt(struct wined3d_texture *dst_texture, unsigned int
                 && (src_width != dst_width || src_height != dst_height))
         {
             /* Can happen when d3d9 apps do a StretchRect() call which isn't handled in GL. */
-            FIXME("Filter %s not supported in software blit.\n", debug_d3dtexturefiltertype(filter));
+            static int once;
+            if (!once++) FIXME("Filter %s not supported in software blit.\n", debug_d3dtexturefiltertype(filter));
         }
 
         xinc = (src_width << 16) / dst_width;
-- 
2.7.1

