From dbce3be53546b6d2f9db58c53d8d5376092dc83b Mon Sep 17 00:00:00 2001
From: Christian Costa <titan.costa@gmail.com>
Date: Sun, 3 Apr 2016 22:30:59 +0200
Subject: wined3d: Display FIXME only once when blen op is 0.

---
 dlls/wined3d/state.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/dlls/wined3d/state.c b/dlls/wined3d/state.c
index e5e550e..6ca506b 100644
--- a/dlls/wined3d/state.c
+++ b/dlls/wined3d/state.c
@@ -291,8 +291,11 @@ static GLenum gl_blend_op(const struct wined3d_gl_info *gl_info, enum wined3d_bl
         case WINED3D_BLEND_OP_MAX:
             return gl_info->supported[EXT_BLEND_MINMAX] ? GL_MAX : GL_FUNC_ADD;
         default:
-            FIXME("Unhandled blend op %#x.\n", op);
+        {
+            static int once;
+            if (op || !once++) FIXME("Unhandled blend op %#x.\n", op);
             return GL_FUNC_ADD;
+        }
     }
 }
 
-- 
2.7.1

