From 7a7067db1ba6ab620bb7979ee081fb846d95016f Mon Sep 17 00:00:00 2001
From: Christian Costa <titan.costa@gmail.com>
Date: Sun, 8 Jan 2017 00:35:27 +0100
Subject: wined3d: Silence noisy fixme Unrecognized stencil op 0.

---
 dlls/wined3d/state.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/dlls/wined3d/state.c b/dlls/wined3d/state.c
index f7e36676a83..b990b31acb5 100644
--- a/dlls/wined3d/state.c
+++ b/dlls/wined3d/state.c
@@ -848,8 +848,11 @@ static GLenum gl_stencil_op(enum wined3d_stencil_op op)
         case WINED3D_STENCIL_OP_DECR:
             return GL_DECR_WRAP;
         default:
-            FIXME("Unrecognized stencil op %#x.\n", op);
+        {
+            static int once;
+            if (op || !once++) FIXME("Unrecognized stencil op %#x.\n", op);
             return GL_KEEP;
+        }
     }
 }
 
-- 
2.11.0

