From f0fe03443eddebbe3997e9ca9e4ed6d8954b159b Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 19 Jul 2017 01:20:01 +0200
Subject: ntdll: Clear upper WORD of segment registers in RtlCaptureContext.

---
 dlls/ntdll/signal_i386.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/dlls/ntdll/signal_i386.c b/dlls/ntdll/signal_i386.c
index 9c4c07dea4a..03c0f5ebc79 100644
--- a/dlls/ntdll/signal_i386.c
+++ b/dlls/ntdll/signal_i386.c
@@ -1228,9 +1228,13 @@ __ASM_STDCALL_FUNC( RtlCaptureContext, 4,
                     "movl 8(%esp),%eax\n\t"    /* context */
                     "movl $0x10007,(%eax)\n\t" /* context->ContextFlags */
                     "movw %gs,0x8c(%eax)\n\t"  /* context->SegGs */
+                    "movw $0,0x8e(%eax)\n\t"
                     "movw %fs,0x90(%eax)\n\t"  /* context->SegFs */
+                    "movw $0,0x92(%eax)\n\t"
                     "movw %es,0x94(%eax)\n\t"  /* context->SegEs */
+                    "movw $0,0x96(%eax)\n\t"
                     "movw %ds,0x98(%eax)\n\t"  /* context->SegDs */
+                    "movw $0,0x9a(%eax)\n\t"
                     "movl %edi,0x9c(%eax)\n\t" /* context->Edi */
                     "movl %esi,0xa0(%eax)\n\t" /* context->Esi */
                     "movl %ebx,0xa4(%eax)\n\t" /* context->Ebx */
@@ -1241,6 +1245,7 @@ __ASM_STDCALL_FUNC( RtlCaptureContext, 4,
                     "movl 4(%ebp),%edx\n\t"
                     "movl %edx,0xb8(%eax)\n\t" /* context->Eip */
                     "movw %cs,0xbc(%eax)\n\t"  /* context->SegCs */
+                    "movw $0,0xbe(%eax)\n\t"
                     "pushfl\n\t"
                     __ASM_CFI(".cfi_adjust_cfa_offset 4\n\t")
                     "popl 0xc0(%eax)\n\t"      /* context->EFlags */
@@ -1248,6 +1253,7 @@ __ASM_STDCALL_FUNC( RtlCaptureContext, 4,
                     "leal 8(%ebp),%edx\n\t"
                     "movl %edx,0xc4(%eax)\n\t" /* context->Esp */
                     "movw %ss,0xc8(%eax)\n\t"  /* context->SegSs */
+                    "movw $0,0xca(%eax)\n\t"
                     "popl 0xb0(%eax)\n\t"      /* context->Eax */
                     __ASM_CFI(".cfi_adjust_cfa_offset -4\n\t")
                     "ret $4" )
-- 
2.13.1

