From c437a2f59781727e9dd2109eafc9fa750404c260 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Thu, 28 May 2015 22:47:20 +0200
Subject: wininet: Use random names when caching very long urls.

---
 dlls/wininet/urlcache.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/wininet/urlcache.c b/dlls/wininet/urlcache.c
index 276132b..0e5fe3c 100644
--- a/dlls/wininet/urlcache.c
+++ b/dlls/wininet/urlcache.c
@@ -2630,7 +2630,7 @@ static BOOL urlcache_entry_create(const char *url, const char *ext, WCHAR *full_
     if(!InternetCrackUrlA(url, 0, 0, &uc))
         uc.dwUrlPathLength = 0;
 
-    if(!uc.dwUrlPathLength) {
+    if(!uc.dwUrlPathLength || uc.dwUrlPathLength >= sizeof(file_name)) {
         file_name[0] = 0;
     }else {
         char *p, *e;
-- 
2.4.1

