From 97e522d70d8f94fe8af1baa4a1c49f5956df3caf Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Tue, 18 Aug 2015 21:33:58 +0200
Subject: krnl386.exe16: Increase buffer size in GetTempDrive.

---
 dlls/krnl386.exe16/file.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/krnl386.exe16/file.c b/dlls/krnl386.exe16/file.c
index 6c3cef3..77908c2 100644
--- a/dlls/krnl386.exe16/file.c
+++ b/dlls/krnl386.exe16/file.c
@@ -461,10 +461,10 @@ LONG WINAPI _hwrite16( HFILE16 hFile, LPCSTR buffer, LONG count )
  */
 UINT WINAPI GetTempDrive( BYTE ignored )
 {
-    WCHAR buffer[8];
+    WCHAR buffer[MAX_PATH];
     BYTE ret;
 
-    if (GetTempPathW( 8, buffer )) ret = (BYTE)toupperW(buffer[0]);
+    if (GetTempPathW( MAX_PATH, buffer )) ret = (BYTE)toupperW(buffer[0]);
     else ret = 'C';
     return MAKELONG( ret | (':' << 8), 1 );
 }
-- 
2.5.0

