From d99e740d73f63cb26e638c6daa550eb2bb28d061 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Thu, 10 Dec 2015 05:57:34 +0100
Subject: server: Increase size of PID table to 512 to reduce risk of
 collisions.

---
 server/process.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/server/process.c b/server/process.c
index 1fdaaca..d7ddc21 100644
--- a/server/process.c
+++ b/server/process.c
@@ -361,7 +361,7 @@ unsigned int alloc_ptid( void *ptr )
     else  /* need to grow the array */
     {
         unsigned int count = alloc_ptid_entries + (alloc_ptid_entries / 2);
-        if (!count) count = 64;
+        if (!count) count = 512;
         if (!(entry = realloc( ptid_entries, count * sizeof(*entry) )))
         {
             set_error( STATUS_NO_MEMORY );
-- 
2.6.2

