From 43f660323b880ccda7da022b95f5bfed2a1c1fc3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Wed, 26 Oct 2016 03:43:16 +0200
Subject: msvfw32: Use case insensitive compare in ICInfo for fourcc codes.

---
 dlls/msvfw32/msvideo_main.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/dlls/msvfw32/msvideo_main.c b/dlls/msvfw32/msvideo_main.c
index 0ae33ed..1b4e9ca 100644
--- a/dlls/msvfw32/msvideo_main.c
+++ b/dlls/msvfw32/msvideo_main.c
@@ -299,9 +299,8 @@ static BOOL ICInfo_enum_handler(const char *drv, unsigned int nr, void *param)
     ICINFO *lpicinfo = param;
     DWORD fccHandler = mmioStringToFOURCCA(drv + 5, 0);
 
-    /* exact match of fccHandler or nth driver found */
-    if ((lpicinfo->fccHandler != nr) && (lpicinfo->fccHandler != fccHandler))
-	return FALSE;
+    if (lpicinfo->fccHandler != nr && compare_fourcc(lpicinfo->fccHandler, fccHandler))
+        return FALSE;
 
     lpicinfo->fccHandler = fccHandler;
     lpicinfo->dwFlags = 0;
-- 
2.9.0

