From 3c33f1f4e9c9e829226bb6bba9a42154f00a48ef Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Thu, 29 Oct 2015 15:41:02 +0800
Subject: dlls/ole32: Handle failure of reading directory entry in
 IEnumSTATSTG::Next.

Signed-off-by: Dmitry Timoshkov <dmitry@baikal.ru>
---
 dlls/ole32/storage32.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/ole32/storage32.c b/dlls/ole32/storage32.c
index bfef4a2..4d775b4 100644
--- a/dlls/ole32/storage32.c
+++ b/dlls/ole32/storage32.c
@@ -1036,9 +1036,10 @@ static HRESULT WINAPI IEnumSTATSTGImpl_Next(
     /*
      * Read the entry from the storage.
      */
-    StorageBaseImpl_ReadDirEntry(This->parentStorage,
+    hr = StorageBaseImpl_ReadDirEntry(This->parentStorage,
       currentSearchNode,
       &currentEntry);
+    if (FAILED(hr)) break;
 
     /*
      * Copy the information to the return buffer.
-- 
2.6.1

