From 9b77ae10308afc29ee408f42b4ae854b6ce0be92 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Thu, 29 Oct 2015 15:41:15 +0800
Subject: dlls/ole32: Zero out returned stats when IEnumSTATSTG::Next reaches
 end of enumeration.

Signed-off-by: Dmitry Timoshkov <dmitry@baikal.ru>
---
 dlls/ole32/storage32.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/ole32/storage32.c b/dlls/ole32/storage32.c
index 4d775b4..f75e097 100644
--- a/dlls/ole32/storage32.c
+++ b/dlls/ole32/storage32.c
@@ -1031,7 +1031,10 @@ static HRESULT WINAPI IEnumSTATSTGImpl_Next(
     hr = IEnumSTATSTGImpl_GetNextRef(This, &currentSearchNode);
 
     if (FAILED(hr) || currentSearchNode == DIRENTRY_NULL)
+    {
+      memset(currentReturnStruct, 0, sizeof(*currentReturnStruct));
       break;
+    }
 
     /*
      * Read the entry from the storage.
-- 
2.6.1

