From d6c4b5bde83cd1cf8a71084841342e27c94fd8ad Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sun, 27 Sep 2015 15:27:42 +0200
Subject: user32: Ignore sizes passed in WM_SIZE for edit control.
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Michael Müller <michael@fds-team.de>
---
 dlls/user32/edit.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/dlls/user32/edit.c b/dlls/user32/edit.c
index 084bbff..0f5d86d 100644
--- a/dlls/user32/edit.c
+++ b/dlls/user32/edit.c
@@ -3925,10 +3925,11 @@ static void EDIT_WM_SetText(EDITSTATE *es, LPCWSTR text, BOOL unicode)
  */
 static void EDIT_WM_Size(EDITSTATE *es, UINT action, INT width, INT height)
 {
-	if ((action == SIZE_MAXIMIZED) || (action == SIZE_RESTORED)) {
+	if ((action == SIZE_MAXIMIZED) || (action == SIZE_RESTORED))
+	{
 		RECT rc;
-		TRACE("width = %d, height = %d\n", width, height);
-		SetRect(&rc, 0, 0, width, height);
+		/* Approach passes zero as width/height, so don't trust these values */
+		GetClientRect(es->hwndSelf, &rc);
 		EDIT_SetRectNP(es, &rc);
 		EDIT_UpdateText(es, NULL, TRUE);
 	}
-- 
2.5.1

