From 6444421b8793005293d4c8af51a6f7f287719598 Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Wed, 5 Nov 2014 18:54:02 +0100
Subject: d3dx9_36/tests: Remove code to skip test which should never fail.

---
 dlls/d3dx9_36/tests/mesh.c | 9 +--------
 1 file changed, 1 insertion(+), 8 deletions(-)

diff --git a/dlls/d3dx9_36/tests/mesh.c b/dlls/d3dx9_36/tests/mesh.c
index 2a617f5..870f62a 100644
--- a/dlls/d3dx9_36/tests/mesh.c
+++ b/dlls/d3dx9_36/tests/mesh.c
@@ -2579,24 +2579,17 @@ static void D3DXCreateBoxTest(void)
     hr = D3DXCreateBox(device,10.9f,20.0f,4.9f,&box, &ppBuffer);
     ok(hr==D3D_OK, "Expected D3D_OK, received %#x\n", hr);
 
-    if (FAILED(hr))
-    {
-        skip("D3DXCreateBox failed\n");
-        goto end;
-    }
-
     buffer = ID3DXBuffer_GetBufferPointer(ppBuffer);
     for(i=0; i<36; i++)
         ok(adjacency[i]==buffer[i], "expected adjacency %d: %#x, received %#x\n",i,adjacency[i], buffer[i]);
 
     box->lpVtbl->Release(box);
+    ID3DXBuffer_Release(ppBuffer);
 
     test_box(device, 10.9f, 20.0f, 4.9f);
 
-end:
     IDirect3DDevice9_Release(device);
     IDirect3D9_Release(d3d);
-    if (ppBuffer) ID3DXBuffer_Release(ppBuffer);
     DestroyWindow(wnd);
 }
 
-- 
2.1.3

