From 7ce6926804895efc9abe5ebee6190fd47a79421a Mon Sep 17 00:00:00 2001
From: Sebastian Lackner <sebastian@fds-team.de>
Date: Sun, 10 Jul 2016 06:35:28 +0200
Subject: secur32: Define missing gnutls_kx_algorithm_t values for old gnutls
 versions.

---
 dlls/secur32/schannel_gnutls.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/secur32/schannel_gnutls.c b/dlls/secur32/schannel_gnutls.c
index 7f2dbd6..9a96634 100644
--- a/dlls/secur32/schannel_gnutls.c
+++ b/dlls/secur32/schannel_gnutls.c
@@ -82,6 +82,11 @@ MAKE_FUNCPTR(gnutls_transport_set_push_function);
 #define GNUTLS_CIPHER_AES_192_CBC 92
 #define GNUTLS_CIPHER_AES_128_GCM 93
 #define GNUTLS_CIPHER_AES_256_GCM 94
+
+#define GNUTLS_KX_ANON_ECDH     11
+#define GNUTLS_KX_ECDHE_RSA     12
+#define GNUTLS_KX_ECDHE_ECDSA   13
+#define GNUTLS_KX_ECDHE_PSK     14
 #endif
 
 static int compat_cipher_get_block_size(gnutls_cipher_algorithm_t cipher)
-- 
2.8.0

