From a35e75bfe3a6358885e756396c8597d7b60be6e5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sun, 19 Mar 2017 19:08:34 +0100
Subject: [PATCH] ntdll: Catch windows int 0x2e syscall on i386.

---
 dlls/ntdll/unix/signal_i386.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/ntdll/unix/signal_i386.c b/dlls/ntdll/unix/signal_i386.c
index 60a47d53ba8..5cc92ab0ced 100644
--- a/dlls/ntdll/unix/signal_i386.c
+++ b/dlls/ntdll/unix/signal_i386.c
@@ -1749,6 +1749,11 @@ static BOOL handle_interrupt( unsigned int interrupt, ucontext_t *sigcontext, vo
         rec->ExceptionInformation[2] = context->Edx;
         setup_raise_exception( sigcontext, stack, rec, xcontext );
         return TRUE;
+    case 0x2e:
+        FIXME("unimplemented syscall handler for %#x\n", context->Eax);
+        EAX_sig(sigcontext) = STATUS_INVALID_SYSTEM_SERVICE;
+        EIP_sig(sigcontext) += 2;
+        return TRUE;
     default:
         return FALSE;
     }
-- 
2.28.0

