From 17b12b07aae111f5dde4f4dca3520ab93487b2aa Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sun, 19 Mar 2017 19:08:34 +0100
Subject: [PATCH] ntdll: Catch windows int 0x2e syscall on i386.

---
 dlls/ntdll/unix/signal_i386.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/ntdll/unix/signal_i386.c b/dlls/ntdll/unix/signal_i386.c
index 7ae4cda33a6..80fef44d055 100644
--- a/dlls/ntdll/unix/signal_i386.c
+++ b/dlls/ntdll/unix/signal_i386.c
@@ -1638,6 +1638,11 @@ static BOOL handle_interrupt( unsigned int interrupt, ucontext_t *sigcontext, st
         stack->rec.ExceptionInformation[2] = stack->context.Edx;
         setup_raise_exception( sigcontext, stack );
         return TRUE;
+    case 0x2e:
+        FIXME("unimplemented syscall handler for %#x\n", stack->context.Eax);
+        EAX_sig(sigcontext) = STATUS_INVALID_SYSTEM_SERVICE;
+        EIP_sig(sigcontext) += 2;
+        return TRUE;
     default:
         return FALSE;
     }
-- 
2.26.2

