From 4b73ac9cedc1531c6ae441c38d0bad37ed29c13f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Michael=20M=C3=BCller?= <michael@fds-team.de>
Date: Sun, 19 Mar 2017 19:08:34 +0100
Subject: [PATCH] ntdll: Catch windows int 0x2e syscall on i386.

---
 dlls/ntdll/unix/signal_i386.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/dlls/ntdll/unix/signal_i386.c b/dlls/ntdll/unix/signal_i386.c
index 19ba551773a..e30dbf9398e 100644
--- a/dlls/ntdll/unix/signal_i386.c
+++ b/dlls/ntdll/unix/signal_i386.c
@@ -1584,6 +1584,11 @@ static BOOL handle_interrupt( unsigned int interrupt, ucontext_t *sigcontext, vo
         rec->ExceptionInformation[2] = context->Edx;
         setup_raise_exception( sigcontext, stack, rec, context );
         return TRUE;
+    case 0x2e:
+        FIXME("unimplemented syscall handler for %#x\n", context->Eax);
+        EAX_sig(sigcontext) = STATUS_INVALID_SYSTEM_SERVICE;
+        EIP_sig(sigcontext) += 2;
+        return TRUE;
     default:
         return FALSE;
     }
-- 
2.27.0

